---
title: "Assessment of multi-model BMAseq and other single-model approaches (DESeq2, edgeR, limma) on inferring replicable differentially expressed genes"
author: "Anni Liu"
date: "`r format(Sys.Date(), '%B %d, %Y')`"
output: rmarkdown::html_vignette
---

```{r,echo=FALSE}
knitr::opts_chunk$set(
  options(scipen = 999),
  cache = TRUE,
  error = FALSE,
  message = FALSE,
  warning = FALSE,
  tidy.opts = list(width.cutoff = 60),
  fig.width = 8,
  fig.height = 6
)
```


# Attach libraries and functions
```{r}
# devtools::install("/Library/Frameworks/R.framework/Versions/4.2/Resources/library/BMAseq") {BMAseq} (2023Apr1 version)  
easypackages::libraries("BMAseq", "limma", "qvalue", "parallel", "tidyverse", 
                        "edgeR", "DESeq2", "data.table", "ggVennDiagram", "gridExtra", 
                        "openxlsx", "grDevices", "viridis", "cowplot", "clusterProfiler",
                        "org.Hs.eg.db", "latex2exp", "enrichplot") |> suppressPackageStartupMessages()

Bayesfactor <- BMAseq:::Bayesfactor
```


# Process data
```{r}
dat.expr <- dget("../ApplicationData/derived/dat.expr.Subcutaneous") 
dat.pheno <- dget("../ApplicationData/derived/dat.pheno.Subcutaneous") 
dim(dat.expr) # 24660 genes and 404 subjects
dim(dat.pheno) # 404 subjects and 13 phenotypes
dat.pheno[1:5, ]
dat.expr[1:5, 1:5]
paste0("The column names of gene expression data", ifelse(all(colnames(dat.expr) == rownames(dat.pheno)), " MATCH ", "NOT MATCH"), "the row names of phenotype data.")

# Pre-filter the genes
# Here we perform the median absolute deviation with the threshold of 0.8 to select genes that are most likely to distinguish the samples
dat.expr.new <- dat.expr[apply(dat.expr, 1, function(x) mad(x) > 0.8), ] # We have 24455 genes

seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
for (seed.i in seed.vec) {
  
  # Divide the gene expression data and phenotype data into the 50% training set and 50% test set 
  test.ind <- sample(1:nrow(dat.pheno), ceiling(0.5*nrow(dat.pheno)))
  dat.pheno.train <- dat.pheno[-test.ind, ]
  dat.pheno.test <- dat.pheno[test.ind, ]
  dat.expr.train <- dat.expr.new[, rownames(dat.pheno.train)]
  dat.expr.test <- dat.expr.new[, rownames(dat.pheno.test)]
  
  # Check if the gene expression data matches the phenotype data  
  print(paste0("The column names of training gene expression data", ifelse(all(colnames(dat.expr.train) == rownames(dat.pheno.train)), " MATCH ", "NOT MATCH"), "the row names of training phenotype data."))
  print(paste0("The column names of test gene expression data", ifelse(all(colnames(dat.expr.test) == rownames(dat.pheno.test)), " MATCH ", "NOT MATCH"), "the row names of test phenotype data."))
  
  # Make the table one of patient characteristics per random seed trial
  dat.pheno.train$group <- "Training"
  dat.pheno.test$group <- "Test"
  dat.pheno.all <- rbind(dat.pheno.train, dat.pheno.test)
  tab1 <- tbl_summary(data = dat.pheno.all,
                      by = "group") %>% 
    add_p() %>% 
    as_gt() %>%
    gt::gtsave(filename = paste0("../ApplicationResult/Multi/RandomSeed/TMM_Top2000/", date.analysis, "_", "Table1", "_", seed.num, ".rtf"))
  
  # Save gene expression and phenotype data
  saveRDS(dat.expr.train, 
          file = sprintf("../ApplicationData/derived/RandomSeed/dat.expr.train%s.RDS", seed.i))
  saveRDS(dat.expr.test, 
          file = sprintf("../ApplicationData/derived/RandomSeed/dat.expr.test%s.RDS", seed.i))
  saveRDS(dat.pheno.train, 
          file = sprintf("../ApplicationData/derived/RandomSeed/dat.pheno.train%s.RDS", seed.i))
  saveRDS(dat.pheno.test, 
          file = sprintf("../ApplicationData/derived/RandomSeed/dat.pheno.test%s.RDS", seed.i))
}
```


# Modify `BMAseq.multi.postprob` to be compatible with the trimmed-mean of M-values (TMM)
```{r}
##------BMAseq.multi.postprob.norm------
BMAseq.multi.postprob.norm <- function(dat.expr.counts, dat.pheno, var.pool, max.nvar, interaction = NULL, cut.BF = 1) {


    if (sum(colnames(dat.expr.counts) != rownames(dat.pheno)) > 0)
        stop("Two datasets dat.expr.counts and dat.pheno must be matched by subjects.")
    if (is.null(var.pool) || length(var.pool) < 2)
        stop("Must provide at least two variables of interest.")
    if (sum(var.pool %in% colnames(dat.pheno)) != length(var.pool))
        stop("Variables of interest must be in datasets dat.pheno.")
    if (is.null(max.nvar))
        stop("Must provide max.nvar.")


    n.var <- length(var.pool)
    n.gene <- nrow(dat.expr.counts)


    # estimate voom weights
    y0 <- voom(counts = dat.expr.counts, 
               lib.size = colSums(dat.expr.counts)*calcNormFactors(dat.expr.counts))
    input.dat.expr <- y0$E
    input.weights <- y0$weights


    # build model space
    result.modelspace <- Modelspace(dat.pheno, var.pool, max.nvar, interaction = interaction)
    input.model.space <- result.modelspace$model.space
    input.dat.pheno <- result.modelspace$dat.pheno.new


    # calculate Bayes factor
    out.bf <- vapply(1:length(input.model.space), function(i) {
        print(paste0(i, ". ", input.model.space[i]))
        design <- model.matrix(formula(input.model.space[i]), data = input.dat.pheno)
        colnames(design)[1] <- "Int"
        vapply(1:n.gene, function(k) Bayesfactor(design, input.dat.expr[k, ], input.weights[k, ]), FUN.VALUE = as.double(1))
    }, FUN.VALUE = as.double(1:n.gene))


    # obtain prior model probability
    bf.max <- t(apply(out.bf, 1, function(x) ifelse(x == max(x) & x > cut.BF, 1, 0)))
    prior.modelprob <- apply(bf.max, 2, sum)/n.gene
    pm.est <- prior.modelprob
    n.model <- length(prior.modelprob)
    pm.prior0 <- pm.prior1 <- rep(0.5/(n.model - 1), n.model - 1)
    alpha <- 0.2
    for (j in 1:30) {
        pm.prior1 <- pm.est[-1]/apply(t(apply(out.bf[, -1], 1, function(x) x/(1 - sum(pm.prior0) + sum(x * pm.prior0)))), 2,
            mean)
        pm.prior0 <- pm.prior0 + (pm.prior1 - pm.prior0) * alpha
    }
    prior.modelprob <- c(1 - sum(pm.prior0), pm.prior0)


    # calculate posterior model probability
    post.modelprob <- t(apply(out.bf, 1, function(x) x * prior.modelprob/sum(x * prior.modelprob)))
    rownames(post.modelprob) <- rownames(dat.expr.counts)
    colnames(post.modelprob) <- input.model.space



    return(list(dat.expr.logcpm = input.dat.expr, weights = input.weights, model.space = input.model.space, dat.pheno.new = input.dat.pheno,
        post.modelprob = post.modelprob, var.pool = var.pool, interaction = interaction))
}
```


# Univariable analysis using the single-model approaches
## Create functions of building univariable models 
```{r}
uni_TMM_top <- function(seed.num = 999, threshold = 2000) {
  ##------Set variables of interest------
  vars <- c("BMI", "AGE", "SEX", "MHABNWBC", "MHARTHTS", "MHCVD", "MHASTHMA", "MHBCTINF", "MHBLDDND", "MHCOCAINE5", "MHCOPD", "MHDRNKSTS", "SMTSPAX") 
  
  
  ##------voom + limma------
  design.train <- mclapply(1:length(vars),
                           function(i) model.matrix(~dat.pheno.train[vars][[i]]), 
                           mc.cores = 10)
  
  design.test <- mclapply(1:length(vars),
                          function(i) model.matrix(~dat.pheno.test[vars][[i]]),
                          mc.cores = 10)
  
  voom.train <- mclapply(1:length(vars),
                         function(i) voom(counts = dat.expr.train,
                                          design = design.train[[i]],
                                          lib.size = colSums(dat.expr.train)*calcNormFactors(dat.expr.train)),
                         mc.cores = 10)
  
  voom.fit.train <- mclapply(1:length(vars),
                             function(i) lmFit(voom.train[[i]][["E"]],
                                               design = design.train[[i]],
                                               weights = voom.train[[i]][["weights"]]),
                             mc.cores = 10)
  
  voom.eFDR.train <- mclapply(1:length(vars),
                              function(i) {
                                t <- voom.fit.train[[i]][["coefficients"]][, 2]/voom.fit.train[[i]][["stdev.unscaled"]][, 2]/voom.fit.train[[i]][["sigma"]]
                                p <- 2*pt(-abs(t), df = voom.fit.train[[i]][["df.residual"]])
                                return(qvalue(p)) },
                              mc.cores = 10)
  
  voom.eFDR.train2 <- mclapply(1:length(vars),
                               function(i) {
                                 q.val <- voom.eFDR.train[[i]][["qvalues"]]
                                 return(q.val[order(q.val)[1:threshold]])
                               },
                               mc.cores = 10)
  
  names(design.train) = names(voom.train) = names(voom.fit.train) = names(voom.eFDR.train) = names(voom.eFDR.train2) = vars 
  
  voom.test <- mclapply(1:length(vars),
                        function(i) voom(counts = dat.expr.test,
                                         design = design.test[[i]],
                                         lib.size = colSums(dat.expr.test)*calcNormFactors(dat.expr.test)),
                        mc.cores = 10)
  
  voom.fit.test <- mclapply(1:length(vars),
                            function(i) lmFit(voom.test[[i]][["E"]],
                                              design = design.test[[i]],
                                              weights = voom.test[[i]][["weights"]]), 
                            mc.cores = 10)
  
  voom.eFDR.test <- mclapply(1:length(vars),
                             function(i) {
                               t <- voom.fit.test[[i]][["coefficients"]][, 2]/voom.fit.test[[i]][["stdev.unscaled"]][, 2]/voom.fit.test[[i]][["sigma"]]
                               p <- 2*pt(-abs(t), df = voom.fit.test[[i]][["df.residual"]])
                               return(qvalue(p)) },
                             mc.cores = 10)
  
  voom.eFDR.test2 <- mclapply(1:length(vars),
                              function(i) {
                                q.val <- voom.eFDR.test[[i]][["qvalues"]]
                                return(q.val[order(q.val)[1:threshold]])
                              },
                              mc.cores = 10)
  
  names(design.test) = names(voom.test) = names(voom.fit.test) = names(voom.eFDR.test) = names(voom.eFDR.test2) = vars 
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, 
       vars, design.train, design.test, 
       voom.train, voom.fit.train, voom.eFDR.train, voom.eFDR.train2, 
       voom.test, voom.fit.test, voom.eFDR.test, voom.eFDR.test2, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/VoomLimmaUni%s.RData", threshold, seed.num))

  
  ##------voom + limma + eBayes------
  eBayes.fit.train <- mclapply(1:length(vars),
                               function(i) lmFit(voom.train[[i]][["E"]],
                                                 design = design.train[[i]],
                                                 weights = voom.train[[i]][["weights"]]) %>% eBayes(), 
                               mc.cores = 10)
  
  eBayes.eFDR.train <- mclapply(1:length(vars),
                                function(i) eBayes.fit.train[[i]][["p.value"]][, 2] %>% qvalue(),
                                mc.cores = 10)
  
  eBayes.eFDR.train2 <- mclapply(1:length(vars),
                                 function(i) {
                                   q.val <- eBayes.eFDR.train[[i]][["qvalues"]]
                                   return(q.val[order(q.val)[1:threshold]])},
                                 mc.cores = 10)
  
  names(eBayes.fit.train) = names(eBayes.eFDR.train) = names(eBayes.eFDR.train2) = vars 
  

  eBayes.fit.test <- mclapply(1:length(vars),
                              function(i) lmFit(voom.test[[i]][["E"]],
                                                design = design.test[[i]],
                                                weights = voom.test[[i]][["weights"]]) %>% eBayes(), 
                              mc.cores = 10)
  
  eBayes.eFDR.test <- mclapply(1:length(vars),
                               function(i) eBayes.fit.test[[i]][["p.value"]][, 2] %>% qvalue(),
                               mc.cores = 10)
  
  eBayes.eFDR.test2 <- mclapply(1:length(vars),
                                function(i) {
                                  q.val <- eBayes.eFDR.test[[i]][["qvalues"]]
                                  return(q.val[order(q.val)[1:threshold]])},
                                mc.cores = 10)
  
  names(eBayes.fit.test) = names(eBayes.eFDR.test) = names(eBayes.eFDR.test2) = vars 

  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, 
       vars, design.train, design.test, 
       voom.train, eBayes.fit.train, eBayes.eFDR.train, eBayes.eFDR.train2,
       voom.test, eBayes.fit.test, eBayes.eFDR.test, eBayes.eFDR.test2, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/eBayesUni%s.RData", threshold, seed.num))
  
  
  ##------edgeR------
  y.train <- mclapply(1:length(vars),
                      function(i) DGEList(counts = dat.expr.train, 
                                          lib.size = colSums(dat.expr.train)) %>% 
                        calcNormFactors() %>% 
                        estimateGLMTrendedDisp(design.train[[i]]),
                      mc.cores = 10) 
  
  edgeR.fit.train <- mclapply(1:length(vars),
                              function(i) glmQLFit(y.train[[i]], design.train[[i]]),
                              mc.cores = 10)
  
  qlf.train <- mclapply(1:length(vars),
                        function(i) glmQLFTest(edgeR.fit.train[[i]], coef = 2),
                        mc.cores = 10)
  
  edgeR.eFDR.train <- mclapply(1:length(vars),
                               function(i) qlf.train[[i]][["table"]][["PValue"]] %>% qvalue(),
                               mc.cores = 10)
  
  edgeR.eFDR.train2 <- mclapply(1:length(vars),
                                function(i) {
                                  q.val <- edgeR.eFDR.train[[i]][["qvalues"]]
                                  return(q.val[order(q.val)[1:threshold]]) 
                                },
                                mc.cores = 10)
  
  edgeR.eFDR.GeneName.train <- mclapply(1:length(vars),
                                        function(i) {
                                          q.val <- edgeR.eFDR.train[[i]][["qvalues"]]
                                          return(rownames(dat.expr.train)[order(q.val)[1:threshold]])
                                        },
                                        mc.cores = 10)
  
  names(y.train) = names(edgeR.fit.train) = names(qlf.train) = names(edgeR.eFDR.train) = names(edgeR.eFDR.train2) = names(edgeR.eFDR.GeneName.train) = vars 
  
  y.test <- mclapply(1:length(vars),
                     function(i) DGEList(counts = dat.expr.test, 
                                         lib.size = colSums(dat.expr.test)) %>%
                       calcNormFactors() %>%
                       estimateGLMTrendedDisp(design.test[[i]]),
                     mc.cores = 10) 
  
  edgeR.fit.test <- mclapply(1:length(vars),
                             function(i) glmQLFit(y.test[[i]], design.test[[i]]),
                             mc.cores = 10)
  
  qlf.test <- mclapply(1:length(vars),
                       function(i) glmQLFTest(edgeR.fit.test[[i]], coef = 2),
                       mc.cores = 10)
  
  edgeR.eFDR.test <- mclapply(1:length(vars),
                              function(i) qlf.test[[i]][["table"]][["PValue"]] %>% qvalue(),
                              mc.cores = 10)
  
  edgeR.eFDR.test2 <- mclapply(1:length(vars),
                               function(i) {
                                 q.val <- edgeR.eFDR.test[[i]][["qvalues"]]
                                 return(q.val[order(q.val)[1:threshold]]) 
                               },
                               mc.cores = 10)
  
  edgeR.eFDR.GeneName.test <- mclapply(1:length(vars),
                                       function(i) {
                                         q.val <- edgeR.eFDR.test[[i]][["qvalues"]]
                                         return(rownames(dat.expr.test)[order(q.val)[1:threshold]])
                                       },
                                       mc.cores = 10)
  
  names(y.test) = names(edgeR.fit.test) = names(qlf.test) = names(edgeR.eFDR.test) = names(edgeR.eFDR.test2) = names(edgeR.eFDR.GeneName.test) = vars 
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, 
       vars, design.train, design.test, 
       y.train, edgeR.fit.train, qlf.train, edgeR.eFDR.train, edgeR.eFDR.train2, edgeR.eFDR.GeneName.train, 
       y.test, edgeR.fit.test, qlf.test, edgeR.eFDR.test, edgeR.eFDR.test2, edgeR.eFDR.GeneName.test, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/edgeRUni%s.RData", threshold, seed.num))
  
  
  ##------DESeq2------
  cts.train <- as.matrix(dat.expr.train)
  cts.test <- as.matrix(dat.expr.test)
  coldata.train <- dat.pheno.train
  coldata.test <- dat.pheno.test
  
  # Check the sample order of the count matrix and column data in the training and test sets 
  all(rownames(coldata.test) == colnames(cts.test))
  all(rownames(coldata.train) == colnames(cts.train))
  
  # Transform the TMM normalization factors to be used in DESeq2
  lib.size <- colSums(cts.train)
  norm.factor <- calcNormFactors(cts.train, method = "TMM")
  size.factor <- lib.size*norm.factor/exp(mean(log(lib.size*norm.factor))) 
  
  res.train <- mclapply(1:length(vars),
                        function(i) {
                          dds <- DESeqDataSetFromMatrix(countData = cts.train, 
                                                        colData = coldata.train, 
                                                        design = formula(paste0("~", vars[i])))
                          sizeFactors(dds) <- size.factor
                          return(results(DESeq(dds)))},
                        mc.cores = 10) 
  
  DESeq2.eFDR.train <- mclapply(1:length(vars),
                                function(i) qvalue(res.train[[i]][["pvalue"]])$qvalues,
                                mc.cores = 10)
  
  DESeq2.eFDR.train2 <- mclapply(1:length(vars),
                                 function(i) {
                                   q.val <- DESeq2.eFDR.train[[i]]
                                   return(q.val[order(q.val)[1:threshold]])
                                 },
                                 mc.cores = 10)
  
  DESeq2.eFDR.GeneName.train <- mclapply(1:length(vars),
                                         function(i) {
                                           q.val <- DESeq2.eFDR.train[[i]]
                                           return(rownames(cts.train)[order(q.val)[1:threshold]])
                                         },
                                         mc.cores = 10)
  
  names(res.train) = names(DESeq2.eFDR.train) = names(DESeq2.eFDR.train2) = names(DESeq2.eFDR.GeneName.train) = vars 
  
  # Transform the TMM normalization factors to be used in DESeq2
  lib.size <- colSums(cts.test)
  norm.factor <- calcNormFactors(cts.test, method = "TMM")
  size.factor <- lib.size*norm.factor/exp(mean(log(lib.size*norm.factor))) 
  
  res.test <- mclapply(1:length(vars),
                       function(i) {
                         dds <- DESeqDataSetFromMatrix(countData = cts.test, 
                                                       colData = coldata.test, 
                                                       design = formula(paste0("~", vars[i])))
                         sizeFactors(dds) <- size.factor
                         return(results(DESeq(dds)))},
                       mc.cores = 10) 
  
  DESeq2.eFDR.test <- mclapply(1:length(vars),
                               function(i) qvalue(res.test[[i]][["pvalue"]])$qvalues,
                               mc.cores = 10)
  
  DESeq2.eFDR.test2 <- mclapply(1:length(vars),
                                function(i) {
                                  q.val <- DESeq2.eFDR.test[[i]]
                                  return(q.val[order(q.val)[1:threshold]])
                                },
                                mc.cores = 10)
  
  DESeq2.eFDR.GeneName.test <- mclapply(1:length(vars),
                                        function(i) {
                                          q.val <- DESeq2.eFDR.test[[i]]
                                          return(rownames(cts.test)[order(q.val)[1:threshold]])
                                        },
                                        mc.cores = 10)
  
  names(res.test) = names(DESeq2.eFDR.test) = names(DESeq2.eFDR.test2) = names(DESeq2.eFDR.GeneName.test) = vars 
  
  save(cts.train, cts.test, coldata.train, coldata.test, vars,
       res.train, DESeq2.eFDR.train, DESeq2.eFDR.train2, DESeq2.eFDR.GeneName.train, 
       res.test, DESeq2.eFDR.test, DESeq2.eFDR.test2, DESeq2.eFDR.GeneName.test, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/DESeq2Uni%s.RData", threshold, seed.num))
}
```


## Run and save modeling results
```{r}
threshold.i <- 5000
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
for (seed.i in seed.vec) {
  dat.expr.train <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/dat.expr.train%s.RDS", seed.i))
  dat.expr.test <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/dat.expr.test%s.RDS", seed.i))
  dat.pheno.train <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/dat.pheno.train%s.RDS", seed.i))
  dat.pheno.test <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/dat.pheno.test%s.RDS", seed.i))
  uni_TMM_top(seed.num = seed.i, threshold = threshold.i)
}
```


## Screen eligible phenotypes to be included in the donwnstream multivariable analysis and interaction analysis 
We filter out those phenotypes showing less than 5% (`0.05 * threshold`) cDEGs within a particular ranking threshold identified in at least one of the approaches (BMAseq, DESeq2, edgeR, voom + limma + eBayes, voom + limma) in the univariable analysis. Hence, in the donwnstream multivariable analysis and interaction analysis, we only include the phenotypes showing greater than or equal to 5% cDEGs within a particular ranking threshold in all approaches under the univariable framework. 

```{r}
threshold.vec <- seq(1000, 5000, 1000)
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
method.vec <- c("DESeq2", "edgeR", "eBayes", "VoomLimma")
vars <- c("BMI", "AGE", "SEX", "MHABNWBC", "MHARTHTS", "MHCVD", "MHASTHMA", "MHBCTINF", "MHBLDDND", "MHCOCAINE5", "MHCOPD", "MHDRNKSTS", "SMTSPAX") 
date.analysis <- format(Sys.Date(), "%Y%b%d")

for (threshold.i in threshold.vec) {
  for (seed.i in seed.vec) {
    for (method.i in method.vec) {
      load(file = sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sUni%s.RData", method.i, seed.i))
      if (method.i == "DESeq2") {
        cDEGs.train <- DESeq2.eFDR.GeneName.train
        cDEGs.test <- DESeq2.eFDR.GeneName.test
      } else if (method.i == "edgeR") {
        cDEGs.train <- edgeR.eFDR.GeneName.train
        cDEGs.test <- edgeR.eFDR.GeneName.test
      } else if (method.i == "eBayes") {
        cDEGs.train <- eBayes.eFDR.train2
        cDEGs.test <- eBayes.eFDR.test2
      } else {
        cDEGs.train <- voom.eFDR.train2
        cDEGs.test <- voom.eFDR.test2
      }
      if (method.i %in% c("DESeq2", "edgeR")) {
        plot.array <- mclapply(1:length(vars), 
                               function(x) ggVennDiagram(
                                 list(Train = cDEGs.train[[x]][1:threshold.i], 
                                      Test = cDEGs.test[[x]][1:threshold.i]), 
                                 label_alpha = 0, label_color = "white") +
                                 theme(legend.position = "none", 
                                 plot.title = element_text(hjust = 0.5, colour = "red")) + 
                                 ggtitle(vars[x]),
                               mc.cores = 10)
        g <- grid.arrange(do.call("arrangeGrob", c(plot.array, ncol = 3)))
        ggsave(filename = sprintf("../ApplicationResult/Uni/Trial/Top%s/%s_%s_%s.eps", threshold.i, date.analysis, method.i, seed.i),
               plot = g,
               device = "eps",
               dpi = 600,
               width = 8,
               height = 9,
               units = "in")
        for (i in vars) {
          if (intersect(cDEGs.train[[i]][1:threshold.i], cDEGs.test[[i]][1:threshold.i]) |> length() / threshold.i < 0.05) {
            print(paste("This phenotype", i, "should not be used in the multivariable analysis!"))}
        }
      } else {
        plot.array <- mclapply(1:length(vars), 
                               function(x) ggVennDiagram(
                                 list(Train = cDEGs.train[[x]][1:threshold.i] |> names(), 
                                      Test = cDEGs.test[[x]][1:threshold.i] |> names()), 
                                 label_alpha = 0, label_color = "white") +
                                 theme(legend.position = "none", 
                                 plot.title = element_text(hjust = 0.5, colour = "red")) + 
                                 ggtitle(vars[x]),
                               mc.cores = 10)
        g <- grid.arrange(do.call("arrangeGrob", c(plot.array, ncol = 3)))
        ggsave(filename = sprintf("../ApplicationResult/Uni/Trial/Top%s/%s_%s_%s.eps", threshold.i, date.analysis, method.i, seed.i),
               plot = g,
               device = "eps",
               dpi = 600,
               width = 8,
               height = 9,
               units = "in")
        for (i in vars) {
          if (intersect(cDEGs.train[[i]][1:threshold.i] |> names(), cDEGs.test[[i]][1:threshold.i] |> names()) |> length() / threshold.i < 0.05) {
            print(paste("This phenotype", i, "should not be used in the multivariable analysis!"))}
        }
      }
    }
  }
}
```


## Create functions to generate data for making boxplots of number of cDEGs and rate of cDEGs, respectively
```{r not.include.multiBMAseq}
##------Number of cDEGs------
boxplot_data <- function (threshold = 1000, random.seed = 8809678) {
  data.table(threshold=rep(threshold,4), 
             random.seed=random.seed,
             method=c("DESeq2","edgeR","eBayes","voom.limma"),
             variable=rep(c("BMI","AGE","SEX","MHABNWBC"),each=4),
             num.cDEGs=c(
               ##------BMI------
               intersect(DESeq2.eFDR.GeneName.train$BMI[1:threshold], DESeq2.eFDR.GeneName.test$BMI[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$BMI[1:threshold], edgeR.eFDR.GeneName.test$BMI[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$BMI[1:threshold]), names(eBayes.eFDR.test2$BMI[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$BMI[1:threshold]), names(voom.eFDR.test2$BMI[1:threshold])) |> length(),
               
               ##------AGE------
               intersect(DESeq2.eFDR.GeneName.train$AGE[1:threshold], DESeq2.eFDR.GeneName.test$AGE[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$AGE[1:threshold], edgeR.eFDR.GeneName.test$AGE[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$AGE[1:threshold]), names(eBayes.eFDR.test2$AGE[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$AGE[1:threshold]), names(voom.eFDR.test2$AGE[1:threshold])) |> length(),
               
               ##------SEX------
               intersect(DESeq2.eFDR.GeneName.train$SEX[1:threshold], DESeq2.eFDR.GeneName.test$SEX[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$SEX[1:threshold], edgeR.eFDR.GeneName.test$SEX[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$SEX[1:threshold]), names(eBayes.eFDR.test2$SEX[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$SEX[1:threshold]), names(voom.eFDR.test2$SEX[1:threshold])) |> length(),
               
               ##------MHABNWBC------
               intersect(DESeq2.eFDR.GeneName.train$MHABNWBC[1:threshold], DESeq2.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$MHABNWBC[1:threshold], edgeR.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$MHABNWBC[1:threshold]), names(eBayes.eFDR.test2$MHABNWBC[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$MHABNWBC[1:threshold]), names(voom.eFDR.test2$MHABNWBC[1:threshold])) |> length()))
}

##------Rate of cDEGs------
boxplot_data2 <- function (threshold = 1000, random.seed = 8809678) {
  data.table(threshold=rep(threshold,4), 
             random.seed=random.seed,
             method=c("DESeq2","edgeR","eBayes","voom.limma"),
             variable=rep(c("BMI","AGE","SEX","MHABNWBC"),each=4),
             rate.cDEGs=c(
               ##------BMI------
               intersect(DESeq2.eFDR.GeneName.train$BMI[1:threshold], DESeq2.eFDR.GeneName.test$BMI[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$BMI[1:threshold], edgeR.eFDR.GeneName.test$BMI[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$BMI[1:threshold]), names(eBayes.eFDR.test2$BMI[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$BMI[1:threshold]), names(voom.eFDR.test2$BMI[1:threshold])) |> length() / threshold,
               
               ##------AGE------
               intersect(DESeq2.eFDR.GeneName.train$AGE[1:threshold], DESeq2.eFDR.GeneName.test$AGE[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$AGE[1:threshold], edgeR.eFDR.GeneName.test$AGE[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$AGE[1:threshold]), names(eBayes.eFDR.test2$AGE[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$AGE[1:threshold]), names(voom.eFDR.test2$AGE[1:threshold])) |> length() / threshold,
               
               ##------SEX------
               intersect(DESeq2.eFDR.GeneName.train$SEX[1:threshold], DESeq2.eFDR.GeneName.test$SEX[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$SEX[1:threshold], edgeR.eFDR.GeneName.test$SEX[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$SEX[1:threshold]), names(eBayes.eFDR.test2$SEX[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$SEX[1:threshold]), names(voom.eFDR.test2$SEX[1:threshold])) |> length() / threshold,
               
               ##------MHABNWBC------
               intersect(DESeq2.eFDR.GeneName.train$MHABNWBC[1:threshold], DESeq2.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$MHABNWBC[1:threshold], edgeR.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$MHABNWBC[1:threshold]), names(eBayes.eFDR.test2$MHABNWBC[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$MHABNWBC[1:threshold]), names(voom.eFDR.test2$MHABNWBC[1:threshold])) |> length() / threshold))
}
```


```{r include.multiBMAseq}
boxplot_data <- function (threshold = 1000, random.seed = 8809678, type = "Uni") {
  data.table(threshold=rep(threshold,5), 
             random.seed=random.seed,
             method=c("BMAseq","DESeq2","edgeR","eBayes","voom.limma"),
             variable=rep(c("BMI","AGE","SEX","MHABNWBC"),each=5),
             num.cDEGs=c(
               ##------BMI------
               intersect(names(BMAseq.eFDR.Main.train$BMI[1:threshold]), names(BMAseq.eFDR.Main.test$BMI[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$BMI[1:threshold], DESeq2.eFDR.GeneName.test$BMI[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$BMI[1:threshold], edgeR.eFDR.GeneName.test$BMI[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$BMI[1:threshold]), names(eBayes.eFDR.test2$BMI[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$BMI[1:threshold]), names(voom.eFDR.test2$BMI[1:threshold])) |> length(),
               
               ##------AGE------
               intersect(names(BMAseq.eFDR.Main.train$AGE[1:threshold]), names(BMAseq.eFDR.Main.test$AGE[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$AGE[1:threshold], DESeq2.eFDR.GeneName.test$AGE[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$AGE[1:threshold], edgeR.eFDR.GeneName.test$AGE[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$AGE[1:threshold]), names(eBayes.eFDR.test2$AGE[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$AGE[1:threshold]), names(voom.eFDR.test2$AGE[1:threshold])) |> length(),
               
               ##------SEX------
               intersect(names(BMAseq.eFDR.Main.train$SEX[1:threshold]), names(BMAseq.eFDR.Main.test$SEX[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$SEX[1:threshold], DESeq2.eFDR.GeneName.test$SEX[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$SEX[1:threshold], edgeR.eFDR.GeneName.test$SEX[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$SEX[1:threshold]), names(eBayes.eFDR.test2$SEX[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$SEX[1:threshold]), names(voom.eFDR.test2$SEX[1:threshold])) |> length(),
               
               ##------MHABNWBC------
               intersect(names(BMAseq.eFDR.Main.train$MHABNWBC[1:threshold]), names(BMAseq.eFDR.Main.test$MHABNWBC[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$MHABNWBC[1:threshold], DESeq2.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$MHABNWBC[1:threshold], edgeR.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$MHABNWBC[1:threshold]), names(eBayes.eFDR.test2$MHABNWBC[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$MHABNWBC[1:threshold]), names(voom.eFDR.test2$MHABNWBC[1:threshold])) |> length()))
}

##------Rate of cDEGs------
boxplot_data2 <- function (threshold = 1000, random.seed = 8809678) {
  data.table(threshold=rep(threshold,5), 
             random.seed=random.seed,
             method=c("BMAseq","DESeq2","edgeR","eBayes","voom.limma"),
             variable=rep(c("BMI","AGE","SEX","MHABNWBC"),each=5),
             rate.cDEGs=c(
               ##------BMI------
               intersect(names(BMAseq.eFDR.Main.train$BMI[1:threshold]), names(BMAseq.eFDR.Main.test$BMI[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$BMI[1:threshold], DESeq2.eFDR.GeneName.test$BMI[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$BMI[1:threshold], edgeR.eFDR.GeneName.test$BMI[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$BMI[1:threshold]), names(eBayes.eFDR.test2$BMI[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$BMI[1:threshold]), names(voom.eFDR.test2$BMI[1:threshold])) |> length() / threshold,
               
               ##------AGE------
               intersect(names(BMAseq.eFDR.Main.train$AGE[1:threshold]), names(BMAseq.eFDR.Main.test$AGE[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$AGE[1:threshold], DESeq2.eFDR.GeneName.test$AGE[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$AGE[1:threshold], edgeR.eFDR.GeneName.test$AGE[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$AGE[1:threshold]), names(eBayes.eFDR.test2$AGE[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$AGE[1:threshold]), names(voom.eFDR.test2$AGE[1:threshold])) |> length() / threshold,
               
               ##------SEX------
               intersect(names(BMAseq.eFDR.Main.train$SEX[1:threshold]), names(BMAseq.eFDR.Main.test$SEX[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$SEX[1:threshold], DESeq2.eFDR.GeneName.test$SEX[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$SEX[1:threshold], edgeR.eFDR.GeneName.test$SEX[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$SEX[1:threshold]), names(eBayes.eFDR.test2$SEX[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$SEX[1:threshold]), names(voom.eFDR.test2$SEX[1:threshold])) |> length() / threshold,
               
               ##------MHABNWBC------
               intersect(names(BMAseq.eFDR.Main.train$MHABNWBC[1:threshold]), names(BMAseq.eFDR.Main.test$MHABNWBC[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$MHABNWBC[1:threshold], DESeq2.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$MHABNWBC[1:threshold], edgeR.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$MHABNWBC[1:threshold]), names(eBayes.eFDR.test2$MHABNWBC[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$MHABNWBC[1:threshold]), names(voom.eFDR.test2$MHABNWBC[1:threshold])) |> length() / threshold))
}
```


## Organize data for boxplotting
```{r not.include.multiBMAseq}
##------Number of cDEGs------
threshold.vec <- seq(1000, 5000, 1000)
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
boxplot.data.num <- NULL
for (threshold.i in threshold.vec) {
  for (seed.i in seed.vec) {
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/DESeq2Uni%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/edgeRUni%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/eBayesUni%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/VoomLimmaUni%s.RData", seed.i))
    boxplot.data.num <- rbind(boxplot.data.num, boxplot_data(threshold.i, seed.i))
  }
}
saveRDS(boxplot.data.num, "../ApplicationResult/Uni/RandomSeed/numcDEGs/Boxplot/boxplot.data.num.noBMA.RDS")


##------Rate of cDEGs------
threshold.vec <- seq(1000, 5000, 1000)
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
boxplot.data.rate <- NULL
for (threshold.i in threshold.vec) {
  for (seed.i in seed.vec) {
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/DESeq2Uni%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/edgeRUni%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/eBayesUni%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/VoomLimmaUni%s.RData", seed.i))
    boxplot.data.rate <- rbind(boxplot.data.rate, boxplot_data2(threshold.i, seed.i))
  }
}
saveRDS(boxplot.data.rate, "../ApplicationResult/Uni/RandomSeed/ratecDEGs/Boxplot/boxplot.data.rate.noBMA.RDS")
```

```{r include.multiBMAseq}
##------Number of cDEGs------
threshold.vec <- seq(1000, 5000, 1000)
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
boxplot.data.num <- NULL
for (threshold.i in threshold.vec) {
  for (seed.i in seed.vec) {
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/BMAseqMulti%s.RData", seed.i)) # Multivariable BMAseq model without interaction terms
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/DESeq2Uni%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/edgeRUni%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/eBayesUni%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/VoomLimmaUni%s.RData", seed.i))
    boxplot.data.num <- rbind(boxplot.data.num, boxplot_data(threshold.i, seed.i))
  }
}
saveRDS(boxplot.data.num, "../ApplicationResult/Uni/RandomSeed/numcDEGs/Boxplot/boxplot.data.num.BMA.RDS")


##------Rate of cDEGs------
threshold.vec <- seq(1000, 5000, 1000)
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
boxplot.data.rate <- NULL
for (threshold.i in threshold.vec) {
  for (seed.i in seed.vec) {
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/BMAseqMulti%s.RData", seed.i)) # Multivariable BMAseq model without interaction terms
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/DESeq2Uni%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/edgeRUni%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/eBayesUni%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/VoomLimmaUni%s.RData", seed.i))
    boxplot.data.rate <- rbind(boxplot.data.rate, boxplot_data2(threshold.i, seed.i))
  }
}
saveRDS(boxplot.data.rate, "../ApplicationResult/Uni/RandomSeed/ratecDEGs/Boxplot/boxplot.data.rate.BMA.RDS")
```


## Make and save boxplots
```{r not.include.multiBMAseq}
##------Number of cDEGs------
date.analysis <- format(Sys.Date(), "%Y%b%d")
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC")

boxplot.data.num <- readRDS("../ApplicationResult/Uni/RandomSeed/numcDEGs/Boxplot/boxplot.data.num.BMA.RDS")
boxplot.data.num[,method:=factor(method,levels=c("BMAseq","DESeq2","edgeR","eBayes","voom.limma"))]

for (var.i in var.vec) {
  g <- boxplot.data.num[variable==var.i] |>
    ggplot(aes(y = num.cDEGs, x = method, group = method)) + 
    geom_boxplot(outlier.shape = NA) + 
    geom_jitter(aes(color = method), size = 0.8) + 
    stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
    facet_wrap(vars(threshold)) + 
    labs(y = "Number of common differentially expressed genes", x = "Method") + 
    theme_bw() + 
    theme(legend.position = "none",
          text = element_text(family = "Arial"), 
          axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
    ggsave(filename = sprintf("../ApplicationResult/Uni/RandomSeed/numcDEGs/Boxplot/%s_%s.eps", date.analysis, var.i),
           plot = g, device = "eps", dpi = 600, width = 9, height = 6, units = "in")
}


##------Rate of cDEGs------
boxplot.data.rate <- readRDS("../ApplicationResult/Uni/RandomSeed/ratecDEGs/Boxplot/boxplot.data.rate.BMA.RDS")
boxplot.data.rate[,method:=factor(method,levels=c("BMAseq","DESeq2","edgeR","eBayes","voom.limma"))]

for (var.i in var.vec) {
  if (var.i != "MHABNWBC") {
    g <- boxplot.data.rate[variable==var.i] |>
      ggplot(aes(y = rate.cDEGs, x = method, group = method)) + 
      geom_boxplot(outlier.shape = NA) + 
      geom_jitter(aes(color = method), size = 0.8) + 
      stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
      facet_wrap(vars(threshold)) + 
      labs(y = "Rate of common differentially expressed genes", x = "Method") + 
      theme_bw() + 
      theme(legend.position = "none",
            axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5)) + 
      scale_y_continuous(labels = scales::label_number(accuracy = 0.01), breaks = seq(0, 0.5, 0.1), limits = c(0, 0.5))
      ggsave(filename = sprintf("../ApplicationResult/Uni/RandomSeed/ratecDEGs/Boxplot/%s_%s.eps", date.analysis, var.i),
             plot = g, device = "eps", dpi = 600, width = 9, height = 6, units = "in")
  } else {
    g <- boxplot.data.rate[variable==var.i] |>
      ggplot(aes(y = rate.cDEGs, x = method, group = method)) + 
      geom_boxplot(outlier.shape = NA) + 
      geom_jitter(aes(color = method), size = 0.8) + 
      stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
      facet_wrap(vars(threshold)) + 
      labs(y = "Rate of common differentially expressed genes", x = "Method") + 
      theme_bw() + 
      theme(legend.position = "none",
            axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5)) + 
      scale_y_continuous(labels = scales::label_number(accuracy = 0.01), breaks = seq(0, 0.8, 0.1), limits = c(0, 0.8))
      ggsave(filename = sprintf("../ApplicationResult/Uni/RandomSeed/ratecDEGs/Boxplot/%s_%s.eps", date.analysis, var.i),
             plot = g, device = "eps", dpi = 600, width = 9, height = 6, units = "in")
  }
}
```

```{r include.multiBMAseq}
##------Number of cDEGs------
date.analysis <- format(Sys.Date(), "%Y%b%d")
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC")

for (var.i in var.vec) {
  g <- boxplot.data.num[variable==var.i] |>
    ggplot(aes(y = num.cDEGs, x = method, group = method)) + 
    geom_boxplot(outlier.shape = NA) + 
    geom_jitter(aes(color = method), size = 0.8) + 
    stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
    facet_wrap(vars(threshold)) + 
    labs(y = "Number of common differentially expressed genes", x = "Method") + 
    theme_bw() + 
    theme(legend.position = "none",
          axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
    ggsave(filename = sprintf("../ApplicationResult/Uni/RandomSeed/numcDEGs/Boxplot/%s_%s.BMA.eps", date.analysis, var.i),
           plot = g, device = "eps", dpi = 600, width = 9, height = 6, units = "in")
}


##------Rate of cDEGs------
for (var.i in var.vec) {
  g <- boxplot.data.rate[variable==var.i] |>
    ggplot(aes(y = rate.cDEGs, x = method, group = method)) + 
    geom_boxplot(outlier.shape = NA) + 
    geom_jitter(aes(color = method), size = 0.8) + 
    stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
    facet_wrap(vars(threshold)) + 
    labs(y = "Rate of common differentially expressed genes", x = "Method") + 
    theme_bw() + 
    theme(legend.position = "none",
          axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
    ggsave(filename = sprintf("../ApplicationResult/Uni/RandomSeed/ratecDEGs/Boxplot/%s_%s.BMA.eps", date.analysis, var.i),
           plot = g, device = "eps", dpi = 600, width = 9, height = 6, units = "in")
}
```


## Estimate the duplicate rate of a cDEG associated with a variable
```{r}
cDEGs = cDEGs.list = NULL
threshold.vec <- seq(1000, 5000, 1000)
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)

duplicate_rate_cDEGs_per_variable <- function(var.name = "BMI", method.name = "DESeq2") {
  for (threshold.i in threshold.vec) {
    for (seed.i in seed.vec) {
      if (method.name %in% c("DESeq2", "edgeR")) {
        load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sUni%s.RData", method.name, seed.i))
        cDEGs <- c(intersect(get(sprintf("%s.eFDR.GeneName.train", method.name))[[var.name]][1:threshold.i], 
                             get(sprintf("%s.eFDR.GeneName.test", method.name))[[var.name]][1:threshold.i]))
      } else if (method.name == "eBayes") {
        load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sUni%s.RData", method.name, seed.i))
        cDEGs <- c(intersect(names(get(sprintf("%s.eFDR.train2", method.name))[[var.name]][1:threshold.i]), 
                             names(get(sprintf("%s.eFDR.test2", method.name))[[var.name]][1:threshold.i])))
      } else {
        load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sUni%s.RData", method.name, seed.i))
        method.name2 <- "voom"
        cDEGs <- c(intersect(names(get(sprintf("%s.eFDR.train2", method.name2))[[var.name]][1:threshold.i]), 
                             names(get(sprintf("%s.eFDR.test2", method.name2))[[var.name]][1:threshold.i])))
      }
      cDEGs.list <- append(cDEGs.list, list(cDEGs))
    }
    cDEGs.unique <- unique(unlist(cDEGs.list))
    
    # Create a matrix that shows the occurrence of each unique gene among the 10 random seed trials
    cDEGs.matrix <- matrix(0, nrow = length(cDEGs.unique), ncol = 10)
    rownames(cDEGs.matrix) <- cDEGs.unique
    colnames(cDEGs.matrix) <- seed.vec   
    for (i in 1:10) {
      cDEGs.i.seed <- cDEGs.list[[i]]
      for (j in 1:length(cDEGs.unique)) {
        if (cDEGs.unique[j] %in% cDEGs.i.seed) {
          cDEGs.matrix[j, i] <- 1
        }
      }
    }
    Sum <- apply(cDEGs.matrix, 1, sum)
    Percent <- Sum/10
    cDEGs.matrix <- cbind(cDEGs.matrix, Sum, Percent) # Add two new columns to the matrix
    saveRDS(cDEGs.matrix, sprintf("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/%sUni%s_%s.RDS", method.name, var.name, threshold.i))
    
    print(paste("Complete the duplicated rate matrix of cDEGs associated with", var.name, "identified by", method.name, "among 10 random seeds for the threshold", threshold.i))
    cDEGs = cDEGs.list =  NULL
  }
}

mapply(duplicate_rate_cDEGs_per_variable,
       rep(c("BMI", "AGE", "SEX", "MHABNWBC"), each = 4), 
       rep(c("DESeq2", "edgeR", "eBayes", "VoomLimma"), times = 4))


# For illustration purpose, only consider the phenotype BMI and 2000 ranking threshold
method.vec <- c("DESeq2", "edgeR", "eBayes", "VoomLimma")
var.name <- "BMI"
threshold.i <- 2000
for (method.name in method.vec) {
  data <- readRDS(sprintf("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/%sUni%s_%s.RDS", method.name, var.name, threshold.i))
  write.csv(data, sprintf("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/%sUni%s_%s.csv", method.name, var.name, threshold.i))
}
```


## Prepare and organize the data for boxplots
```{r not.include.multiBMAseq}
boxplot_duplicate_rate_data <- function (var.name = NULL, method.name = NULL, threshold = NULL) {
  load.data <- readRDS(sprintf("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/UniModel/%sUni%s_%s.RDS", method.name, var.name, threshold))
  temp.data <- data.table(threshold=threshold, 
                          method=method.name,
                          variable=var.name,
                          cDEGs.duplicate.rate.median=median(load.data[,12]),
                          cDEGs.duplicate.rate.mean=mean(load.data[,12]))
  return(temp.data)
}

organized_duplicate_rate_data <- mapply(boxplot_duplicate_rate_data,
                                        var.name = rep(c("BMI", "AGE", "SEX", "MHABNWBC"), each = 20), 
                                        method.name = rep(rep(c("DESeq2", "edgeR", "eBayes", "VoomLimma"), each = 5), times = 4),
                                        threshold = rep(rep(seq(1000, 5000, 1000), times = 4), times = 4), SIMPLIFY = F) |> rbindlist()

# data.frame(var.name, method.name, threshold) |> View()

saveRDS(organized_duplicate_rate_data, "../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/UniModel/organized_duplicate_rate_data.noBMA.RDS")
```


```{r include.multiBMAseq}
boxplot_duplicate_rate_data <- function (var.name = NULL, method.name = NULL, threshold = NULL) {
  if (method.name == "BMAseq") {
    load.data <- readRDS(sprintf("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/MultiModel/%sMulti%s_%s.RDS", method.name, var.name, threshold)) 
  } else {
    load.data <- readRDS(sprintf("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/UniModel/%sUni%s_%s.RDS", method.name, var.name, threshold))
  } 
  temp.data <- data.table(threshold=threshold, 
                          method=method.name,
                          variable=var.name,
                          cDEGs.duplicate.rate.median=median(load.data[,12]),
                          cDEGs.duplicate.rate.mean=mean(load.data[,12]))
  return(temp.data)
}

organized_duplicate_rate_data <- mapply(boxplot_duplicate_rate_data,
                                        var.name = rep(c("BMI", "AGE", "SEX", "MHABNWBC"), each = 25), 
                                        method.name = rep(rep(c("BMAseq", "DESeq2", "edgeR", "eBayes", "VoomLimma"), each = 5), times = 4),
                                        threshold = rep(rep(seq(1000, 5000, 1000), times = 5), times = 4), SIMPLIFY = F) |> rbindlist()

# data.frame(var.name, method.name, threshold) |> View()

saveRDS(organized_duplicate_rate_data, "../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/UniModel/organized_duplicate_rate_data.RDS")
```


## Make and save boxplots
```{r}
organized_duplicate_rate_data[,variable:=factor(variable,levels=c("BMI","AGE","SEX","MHABNWBC"))]
organized_duplicate_rate_data[,method:=factor(method,levels=c("BMAseq","DESeq2","edgeR","eBayes","VoomLimma"),labels=c("BMAseq","DESeq2","edgeR","eBayes","voom.limma"))]


organized_duplicate_rate_data[,cDEGs.duplicate.rate.median:=as.numeric(cDEGs.duplicate.rate.median)]
organized_duplicate_rate_data[,cDEGs.duplicate.rate.mean:=as.numeric(cDEGs.duplicate.rate.mean)]

date.analysis <- format(Sys.Date(), "%Y%b%d")
g1 <- organized_duplicate_rate_data |>
  ggplot(mapping = aes(y = cDEGs.duplicate.rate.median, x = method, group = method)) + 
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(aes(color = method), size = 0.8) + 
  facet_wrap(vars(variable)) + 
  scale_y_continuous(labels = scales::label_number(accuracy = 0.01),
                     breaks = seq(0, 1, 0.1),
                     limits = c(0, 1)) + 
  labs(y = "Median duplicate rate of cDEGs", x = "Method") + 
  theme_bw() + 
  theme(legend.position = "none",
        axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
ggsave(filename = sprintf("../ApplicationResult/Uni/RandomSeed/DuplicatedRateMatrix/Boxplot/%s_median.eps", date.analysis),
       plot = g1, device = "eps", dpi = 600, width = 9, height = 6, units = "in")

g2 <- organized_duplicate_rate_data |>
  ggplot(mapping = aes(y = cDEGs.duplicate.rate.mean, x = method, group = method)) + 
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(aes(color = method), size = 0.8) + 
  facet_wrap(vars(variable)) + 
  scale_y_continuous(labels = scales::label_number(accuracy = 0.01),
                     breaks = seq(0, 1, 0.1),
                     limits = c(0, 1)) + 
  labs(y = "Mean duplicate rate of cDEGs", x = "Method") + 
  theme_bw() + 
  theme(legend.position = "none",
        axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
ggsave(filename = sprintf("../ApplicationResult/Uni/RandomSeed/DuplicatedRateMatrix/Boxplot/%s_mean.eps", date.analysis),
       plot = g2, device = "eps", dpi = 600, width = 9, height = 6, units = "in")
```


## Make Venn diagrams of cDEGs
```{r threshold.2000}
threshold <- 2000
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC")
date.analysis <- format(Sys.Date(), "%Y%b%d")

make_vennplot <- function(var.name = NULL, threshold = NULL) {

  BMAseq.cDEGs <- intersect(names(BMAseq.eFDR.Main.train[[var.name]][1:threshold]), names(BMAseq.eFDR.Main.test[[var.name]][1:threshold]))
  DESeq2.cDEGs <- intersect(DESeq2.eFDR.GeneName.train[[var.name]][1:threshold], DESeq2.eFDR.GeneName.test[[var.name]][1:threshold])
  edgeR.cDEGs <- intersect(edgeR.eFDR.GeneName.train[[var.name]][1:threshold], edgeR.eFDR.GeneName.test[[var.name]][1:threshold])
  eBayes.cDEGs <- intersect(names(eBayes.eFDR.train2[[var.name]][1:threshold]), names(eBayes.eFDR.test2[[var.name]][1:threshold]))
  voom.limma.cDEGs <- intersect(names(voom.eFDR.train2[[var.name]][1:threshold]), names(voom.eFDR.test2[[var.name]][1:threshold]))
  
  g <- ggVennDiagram(list(BMAseq = BMAseq.cDEGs, 
                          DESeq2 = DESeq2.cDEGs,
                          edgeR = edgeR.cDEGs,
                          eBayes = eBayes.cDEGs,
                          voom.limma = voom.limma.cDEGs),
                     label_alpha = 0, label_color = "white") +
    theme(legend.position = "none", 
          plot.title = element_text(hjust = 0.5, colour = "red", size = 18)) + 
    ggtitle(paste("The intersection of cDEGs associated with the", var.name, "\nunder the seed", seed.i, "and threshold", threshold)) + 
    scale_fill_distiller(palette = "Dark2", direction = -1)
  
  ggsave(filename = paste0("../ApplicationResult/Uni/RandomSeed/TMM_Top", threshold, "/Vennplot/", date.analysis, "_", var.name, "_", seed.i, ".eps"),
         plot = g,
         device = "eps",
         dpi = 600, 
         width = 10,
         height = 7,
         units = "in")
}

for (seed.i in seed.vec) {

load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/BMAseqMulti%s.RData", seed.i))
  names(BMAseq.eFDR.Main.train) = names(BMAseq.eFDR.Main.test) = var.vec
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/UniModel/DESeq2Uni%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/UniModel/edgeRUni%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/UniModel/eBayesUni%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/UniModel/VoomLimmaUni%s.RData", seed.i))
  for (var.i in var.vec) {
    make_vennplot(var.name = var.i, threshold = 2000)
  }
}
```


```{r threshold.5000}
threshold <- 5000
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC")
date.analysis <- format(Sys.Date(), "%Y%b%d")

for (seed.i in seed.vec) {
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/BMAseqMulti%s.RData", seed.i))
  names(BMAseq.eFDR.Main.train) = names(BMAseq.eFDR.Main.test) = var.vec
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/UniModel/DESeq2Uni%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/UniModel/edgeRUni%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/UniModel/eBayesUni%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/UniModel/VoomLimmaUni%s.RData", seed.i))
  for (var.i in var.vec) {
    make_vennplot(var.name = var.i, threshold = 5000)
  }
}
```


## Extract unique cDEGs identified by BMAseq per seed and all unique cDEGs identified by BMAseq across 10 seeds
```{r}
file.name <- "../ApplicationResult/Uni/RandomSeed/TMM_Top2000/Vennplot/uniq_cDEGs_BMI_BMAseq_per_seed.txt"
unique.cDEGs.list <- NULL
extract_unique_cDEGs_per_seed <-function(var.name = NULL, threshold = NULL) {
  BMAseq.cDEGs <- intersect(names(BMAseq.eFDR.Main.train[[var.name]][1:threshold]), names(BMAseq.eFDR.Main.test[[var.name]][1:threshold]))
  DESeq2.cDEGs <- intersect(DESeq2.eFDR.GeneName.train[[var.name]][1:threshold], DESeq2.eFDR.GeneName.test[[var.name]][1:threshold])
  edgeR.cDEGs <- intersect(edgeR.eFDR.GeneName.train[[var.name]][1:threshold], edgeR.eFDR.GeneName.test[[var.name]][1:threshold])
  eBayes.cDEGs <- intersect(names(eBayes.eFDR.train2[[var.name]][1:threshold]), names(eBayes.eFDR.test2[[var.name]][1:threshold]))
  voom.limma.cDEGs <- intersect(names(voom.eFDR.train2[[var.name]][1:threshold]), names(voom.eFDR.test2[[var.name]][1:threshold]))
  unique.cDEGs <- setdiff(BMAseq.cDEGs, union(DESeq2.cDEGs, edgeR.cDEGs) |> union(eBayes.cDEGs) |> union(voom.limma.cDEGs))
  return(unique.cDEGs)
}

for (seed.i in seed.vec) {
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/BMAseqMulti%s.RData", seed.i)) # Multivariable BMAseq model without interaction terms
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/DESeq2Uni%s.RData", seed.i)) # Univariable DESeq2 model
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/edgeRUni%s.RData", seed.i)) # Univariable edgeR model
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/eBayesUni%s.RData", seed.i)) # Univariable eBayes model
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/VoomLimmaUni%s.RData", seed.i)) # Univariable VoomLimma model
  unique.cDEGs <- extract_unique_cDEGs_per_seed(var.name = "BMI", threshold = 2000)
  unique.cDEGs.list <- append(unique.cDEGs.list, list(unique.cDEGs))
  write(sprintf("\nThe following gene IDs are unique gene IDs associated with the main effect of %s identified by BMAseq under the seed %s and threshold %s", var.name, seed.i, threshold), file = file.name, append = T)
  write.table(unique.cDEGs, file = file.name, quote = F, col.names = F, append = T)
}
names(unique.cDEGs.list) <- seed.vec

file.name <- "../ApplicationResult/Uni/RandomSeed/TMM_Top2000/Vennplot/uniq_cDEGs_BMI_BMAseq_all_seeds.txt"
write.table(unlist(unique.cDEGs.list) |> unique(), file = file.name, quote = F, col.names = F)
```


# Multi-variables without interaction terms analysis
## Create functions of building multivariable models without interaction terms
```{r cut.FDR0.05}
multi_TMM_top <- function(seed.num = 999, threshold = 2000) {
  ##------Set variables of interest------
  vars.pool <- c("BMI", "AGE", "SEX", "MHABNWBC")  
  
  
  ##------BMAseq------
  output.multi.train.s1 <- BMAseq.multi.postprob.norm(dat.expr.counts = dat.expr.train,  
                                                      dat.pheno = dat.pheno.train, 
                                                      var.pool = vars.pool, 
                                                      max.nvar = 4, 
                                                      cut.BF = 1)
  
  output.multi.train.s2 <- BMAseq.multi.DEG(postprob.output = output.multi.train.s1, 
                                            cut.FDR = 0.05) 
  
  BMAseq.eFDR.Main.train <- mclapply(1:length(vars.pool),
                                     function(i) output.multi.train.s2$eFDR[, i][order(output.multi.train.s2$eFDR[, i])[1:threshold]],
                                     mc.cores = 10) # cut.FDR = 1 does not effect eFDR.main
  
  output.multi.test.s1 <- BMAseq.multi.postprob.norm(dat.expr.counts = dat.expr.test, 
                                                     dat.pheno = dat.pheno.test, 
                                                     var.pool = vars.pool, 
                                                     max.nvar = 4, 
                                                     cut.BF = 1)
  
  output.multi.test.s2 <- BMAseq.multi.DEG(postprob.output = output.multi.test.s1, 
                                           cut.FDR = 0.05)
  
  BMAseq.eFDR.Main.test <- mclapply(1:length(vars.pool),
                                    function(i) output.multi.test.s2$eFDR[, i][order(output.multi.test.s2$eFDR[, i])[1:threshold]],
                                    mc.cores = 10)

  names(BMAseq.eFDR.Main.train) = names(BMAseq.eFDR.Main.test) = vars.pool
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, vars.pool,
       output.multi.train.s1, output.multi.train.s2, output.multi.test.s1, output.multi.test.s2, 
       BMAseq.eFDR.Main.train, BMAseq.eFDR.Main.test, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/MultiModel/BMAseqMulti%s.RData", threshold, seed.num)) 
  
  
  ##------voom + limma------
  design.train <- model.matrix(~BMI + AGE + SEX + MHABNWBC, data = dat.pheno.train)
  design.test <- model.matrix(~BMI + AGE + SEX + MHABNWBC, data = dat.pheno.test)
    
  voom.train <- voom(counts = dat.expr.train,
                     design = design.train,
                     lib.size = colSums(dat.expr.train)*calcNormFactors(dat.expr.train)) 
  
  voom.fit.train <- lmFit(voom.train[["E"]],
                          design = design.train,
                          weights = voom.train[["weights"]])
  
  voom.eFDR.train <- mclapply(1:length(vars.pool),
                              function(i) {
                                t <- voom.fit.train[["coefficients"]][, i + 1]/voom.fit.train[["stdev.unscaled"]][, i + 1]/voom.fit.train[["sigma"]]
                                p <- 2*pt(-abs(t), df = voom.fit.train[["df.residual"]])
                                return(qvalue(p)) },
                              mc.cores = 10)
  
  voom.eFDR.train2 <- mclapply(1:length(vars.pool),
                               function(i) {
                                 q.val <- voom.eFDR.train[[i]][["qvalues"]]
                                 return(q.val[order(q.val)[1:threshold]])
                               },
                               mc.cores = 10)
  
  names(voom.eFDR.train) = names(voom.eFDR.train2) = vars.pool 
  
  voom.test <- voom(counts = dat.expr.test,
                    design = design.test,
                    lib.size = colSums(dat.expr.test)*calcNormFactors(dat.expr.test))
  
  voom.fit.test <- lmFit(voom.test[["E"]],
                         design = design.test,
                         weights = voom.test[["weights"]])
  
  voom.eFDR.test <- mclapply(1:length(vars.pool),
                             function(i) {
                               t <- voom.fit.test[["coefficients"]][, i + 1]/voom.fit.test[["stdev.unscaled"]][, i + 1]/voom.fit.test[["sigma"]]
                               p <- 2*pt(-abs(t), df = voom.fit.test[["df.residual"]])
                               return(qvalue(p)) },
                             mc.cores = 10)
  
  voom.eFDR.test2 <- mclapply(1:length(vars.pool),
                              function(i) {
                                q.val <- voom.eFDR.test[[i]][["qvalues"]]
                                return(q.val[order(q.val)[1:threshold]])
                              },
                              mc.cores = 10)
  
  names(voom.eFDR.test) = names(voom.eFDR.test2) = vars.pool 

  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, vars.pool,
       design.train, design.test, voom.train, voom.fit.train, voom.eFDR.train, voom.eFDR.train2, 
       voom.test, voom.fit.test, voom.eFDR.test, voom.eFDR.test2, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/MultiModel/VoomLimmaMulti%s.RData", threshold, seed.num))
  
  
  ##------voom + limma + eBayes------
  voom.train <- voom(counts = dat.expr.train,
                     design = design.train,
                     lib.size = colSums(dat.expr.train)*calcNormFactors(dat.expr.train))
  
  eBayes.fit.train <- lmFit(voom.train[["E"]],
                            design = design.train,
                            weights = voom.train[["weights"]]) %>% 
                        eBayes()
  
  eBayes.eFDR.train <- mclapply(1:length(vars.pool),
                                function(i) eBayes.fit.train[["p.value"]][, i + 1] %>% 
                                  qvalue(),
                                mc.cores = 10)
  
  eBayes.eFDR.train2 <- mclapply(1:length(vars.pool),
                                 function(i) {
                                   q.val <- eBayes.eFDR.train[[i]][["qvalues"]]
                                   return(q.val[order(q.val)[1:threshold]])
                                 },
                                 mc.cores = 10)
  
  names(eBayes.eFDR.train) = names(eBayes.eFDR.train2) = vars.pool 
  
  voom.test <- voom(counts = dat.expr.test,
                    design = design.test,
                    lib.size = colSums(dat.expr.test)*calcNormFactors(dat.expr.test))
  

  eBayes.fit.test <- lmFit(voom.test[["E"]],
                           design = design.test,
                           weights = voom.test[["weights"]]) %>% 
                       eBayes()
  
  eBayes.eFDR.test <- mclapply(1:length(vars.pool),
                               function(i) eBayes.fit.test[["p.value"]][, i + 1] %>% 
                                 qvalue(),
                               mc.cores = 10)
  
  eBayes.eFDR.test2 <- mclapply(1:length(vars.pool),
                                function(i) {
                                  q.val <- eBayes.eFDR.test[[i]][["qvalues"]]
                                  return(q.val[order(q.val)[1:threshold]])
                                },
                                mc.cores = 10)
  
  names(eBayes.eFDR.test) = names(eBayes.eFDR.test2) = vars.pool 
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, vars.pool,
       design.train, design.test, voom.train, eBayes.fit.train, eBayes.eFDR.train, eBayes.eFDR.train2,
       voom.test, eBayes.fit.test, eBayes.eFDR.test, eBayes.eFDR.test2, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/MultiModel/eBayesMulti%s.RData", threshold, seed.num))

  
  ##------edgeR------
  y.train <- DGEList(counts = dat.expr.train, 
                     lib.size = colSums(dat.expr.train)) %>% 
               calcNormFactors() %>% 
               estimateGLMTrendedDisp(design.train)
  
  edgeR.fit.train <- glmQLFit(y.train, design.train)
  
  qlf.train <- mclapply(1:length(vars.pool),
                        function(i) glmQLFTest(edgeR.fit.train, coef = i + 1),
                        mc.cores = 10)
  
  edgeR.eFDR.train <- mclapply(1:length(vars.pool),
                               function(i) qlf.train[[i]][["table"]][["PValue"]] %>%
                                 qvalue(),
                               mc.cores = 10)
  
  edgeR.eFDR.train2 <- mclapply(1:length(vars.pool),
                                function(i) {
                                  q.val <- edgeR.eFDR.train[[i]][["qvalues"]]
                                  return(q.val[order(q.val)[1:threshold]]) 
                                },
                                mc.cores = 10)
  
  edgeR.eFDR.GeneName.train <- mclapply(1:length(vars.pool),
                                        function(i) {
                                          q.val <- edgeR.eFDR.train[[i]][["qvalues"]]
                                          return(rownames(dat.expr.train)[order(q.val)[1:threshold]])
                                        },
                                        mc.cores = 10)
  
  names(qlf.train) = names(edgeR.eFDR.train) = names(edgeR.eFDR.train2) = names(edgeR.eFDR.GeneName.train) = vars.pool 
  
  y.test <- DGEList(counts = dat.expr.test, 
                    lib.size = colSums(dat.expr.test)) %>% 
              calcNormFactors() %>% 
              estimateGLMTrendedDisp(design.test)
  
  edgeR.fit.test <- glmQLFit(y.test, design.test)
  
  qlf.test <- mclapply(1:length(vars.pool),
                       function(i) glmQLFTest(edgeR.fit.test, coef = i + 1),
                       mc.cores = 10)
  
  edgeR.eFDR.test <- mclapply(1:length(vars.pool),
                              function(i) qlf.test[[i]][["table"]][["PValue"]] %>%
                                qvalue(),
                              mc.cores = 10)
  
  edgeR.eFDR.test2 <- mclapply(1:length(vars.pool),
                               function(i) {
                                 q.val <- edgeR.eFDR.test[[i]][["qvalues"]]
                                 return(q.val[order(q.val)[1:threshold]]) 
                               },
                               mc.cores = 10)
  
  edgeR.eFDR.GeneName.test <- mclapply(1:length(vars.pool),
                                       function(i) {
                                         q.val <- edgeR.eFDR.test[[i]][["qvalues"]]
                                         return(rownames(dat.expr.test)[order(q.val)[1:threshold]])
                                       },
                                      mc.cores = 10)
  
  names(qlf.test) = names(edgeR.eFDR.test) = names(edgeR.eFDR.test2) = names(edgeR.eFDR.GeneName.test) = vars.pool 
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, vars.pool, design.train, design.test, 
       y.train, edgeR.fit.train, edgeR.eFDR.train, edgeR.eFDR.train2, edgeR.eFDR.GeneName.train, 
       y.test, edgeR.fit.test, edgeR.eFDR.test, edgeR.eFDR.test2, edgeR.eFDR.GeneName.test, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/MultiModel/edgeRMulti%s.RData", threshold, seed.num))
  
  
  ##------DESeq2------
  cts.train <- as.matrix(dat.expr.train)
  cts.test <- as.matrix(dat.expr.test)
  coldata.train <- dat.pheno.train
  coldata.test <- dat.pheno.test
  
  # Check the sample order of the count matrix and column data in the training and test sets 
  all(rownames(coldata.test) == colnames(cts.test))
  all(rownames(coldata.train) == colnames(cts.train))
  
  name.formula <- c("BMI_high_vs_low", "AGE_old_vs_young", "SEX_male_vs_female", "MHABNWBC_yes_vs_no") 
  
  # Transform the TMM normalization factors to be used in DESeq2 (training set)
  lib.size <- colSums(cts.train)
  norm.factor <- calcNormFactors(cts.train, method = "TMM")
  size.factor <- lib.size*norm.factor/exp(mean(log(lib.size*norm.factor))) 
  
  dds <- DESeqDataSetFromMatrix(countData = cts.train, 
                                colData = coldata.train, 
                                design = ~BMI + AGE + SEX + MHABNWBC)
  sizeFactors(dds) <- size.factor

  res.train <- mclapply(1:length(vars.pool), 
                        function(i) {
                          return(results(DESeq(dds), name = name.formula[i])) },
                        mc.cores = 10)
  
  DESeq2.eFDR.train <- mclapply(1:length(vars.pool),
                                function(i) qvalue(res.train[[i]][["pvalue"]])$qvalues,
                                mc.cores = 10)
  
  DESeq2.eFDR.train2 <- mclapply(1:length(vars.pool),
                                 function(i) {
                                   q.val <- DESeq2.eFDR.train[[i]]
                                   return(q.val[order(q.val)[1:threshold]])
                                 },
                                 mc.cores = 10)
  
  DESeq2.eFDR.GeneName.train <- mclapply(1:length(vars.pool),
                                         function(i) {
                                           q.val <- DESeq2.eFDR.train[[i]]
                                           return(rownames(cts.train)[order(q.val)[1:threshold]])
                                         },
                                         mc.cores = 10)
  
  names(res.train) = names(DESeq2.eFDR.train) = names(DESeq2.eFDR.train2) = names(DESeq2.eFDR.GeneName.train) = vars.pool 
  
  # Transform the TMM normalization factors to be used in DESeq2 (test set)
  lib.size <- colSums(cts.test)
  norm.factor <- calcNormFactors(cts.test, method = "TMM")
  size.factor <- lib.size*norm.factor/exp(mean(log(lib.size*norm.factor))) 
  
  dds <- DESeqDataSetFromMatrix(countData = cts.test, 
                                colData = coldata.test, 
                                design = ~BMI + AGE + SEX + MHABNWBC)
  sizeFactors(dds) <- size.factor
  
  res.test <- mclapply(1:length(vars.pool), 
                       function(i) {
                         return(results(DESeq(dds), name = name.formula[i])) }, 
                       mc.cores = 10)
  
  DESeq2.eFDR.test <- mclapply(1:length(vars.pool),
                               function(i) qvalue(res.test[[i]][["pvalue"]])$qvalues,
                               mc.cores = 10)
  
  DESeq2.eFDR.test2 <- mclapply(1:length(vars.pool),
                                function(i) {
                                  q.val <- DESeq2.eFDR.test[[i]]
                                  return(q.val[order(q.val)[1:threshold]])
                                },
                                mc.cores = 10)
  
  DESeq2.eFDR.GeneName.test <- mclapply(1:length(vars.pool),
                                        function(i) {
                                          q.val <- DESeq2.eFDR.test[[i]]
                                          return(rownames(cts.test)[order(q.val)[1:threshold]])
                                        },
                                        mc.cores = 10)
  
  names(res.test) = names(DESeq2.eFDR.test) = names(DESeq2.eFDR.test2) = names(DESeq2.eFDR.GeneName.test) = vars.pool
  
  save(cts.train, cts.test, coldata.train, coldata.test, vars.pool,
       res.train, DESeq2.eFDR.train, DESeq2.eFDR.train2, DESeq2.eFDR.GeneName.train, 
       res.test, DESeq2.eFDR.test, DESeq2.eFDR.test2, DESeq2.eFDR.GeneName.test, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/MultiModel/DESeq2Multi%s.RData", threshold, seed.num))
}
```

```{r cut.FDR1}
multi_TMM_top <- function(seed.num = 999, threshold = 2000) {
  ##------Set variables of interest------
  vars.pool <- c("BMI", "AGE", "SEX", "MHABNWBC")  
  
  
  ##------BMAseq------
  output.multi.train.s1 <- BMAseq.multi.postprob.norm(dat.expr.counts = dat.expr.train,  
                                                      dat.pheno = dat.pheno.train, 
                                                      var.pool = vars.pool, 
                                                      max.nvar = 4, 
                                                      cut.BF = 1)
  
  output.multi.train.s2 <- BMAseq.multi.DEG(postprob.output = output.multi.train.s1, 
                                            cut.FDR = 1) 
  
  
  BMAseq.eFDR.Main.train <- mclapply(1:length(vars.pool),
                                     function(i) output.multi.train.s2$eFDR[, i][order(output.multi.train.s2$eFDR[, i])[1:threshold]],
                                     mc.cores = 10) 
  
  output.multi.test.s1 <- BMAseq.multi.postprob.norm(dat.expr.counts = dat.expr.test, 
                                                     dat.pheno = dat.pheno.test, 
                                                     var.pool = vars.pool, 
                                                     max.nvar = 4, 
                                                     cut.BF = 1)
  
  output.multi.test.s2 <- BMAseq.multi.DEG(postprob.output = output.multi.test.s1, 
                                           cut.FDR = 1)
  
  BMAseq.eFDR.Main.test <- mclapply(1:length(vars.pool),
                                    function(i) output.multi.test.s2$eFDR[, i][order(output.multi.test.s2$eFDR[, i])[1:threshold]],
                                    mc.cores = 10)

  names(BMAseq.eFDR.Main.train) = names(BMAseq.eFDR.Main.test) = vars.pool
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, vars.pool,
       output.multi.train.s1, output.multi.train.s2, output.multi.test.s1, output.multi.test.s2, 
       BMAseq.eFDR.Main.train, BMAseq.eFDR.Main.test, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/MultiModel/BMAseqMulti%sFDR1.RData", threshold, seed.num)) 
}
```

## Run and save modeling results
```{r}
threshold.i <- 5000
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
for (seed.i in seed.vec) {
  dat.expr.train <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/dat.expr.train%s.RDS", seed.i))
  dat.expr.test <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/dat.expr.test%s.RDS", seed.i))
  dat.pheno.train <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/dat.pheno.train%s.RDS", seed.i))
  dat.pheno.test <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/dat.pheno.test%s.RDS", seed.i))
  multi_TMM_top(seed.num = seed.i, threshold = threshold.i)
}
```


## Create functions to generate data for making boxplots of number of cDEGs and rate of cDEGs, respectively
```{r}
##------Number of cDEGs------
boxplot_data <- function (threshold = 1000, random.seed = 8809678) {
  data.table(threshold=rep(threshold,5), 
             random.seed=random.seed,
             method=c("BMAseq","DESeq2","edgeR","eBayes","voom.limma"),
             variable=rep(c("BMI","AGE","SEX","MHABNWBC"),each=5),
             num.cDEGs=c(
               ##------BMI------
               intersect(names(BMAseq.eFDR.Main.train$BMI[1:threshold]), names(BMAseq.eFDR.Main.test$BMI[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$BMI[1:threshold], DESeq2.eFDR.GeneName.test$BMI[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$BMI[1:threshold], edgeR.eFDR.GeneName.test$BMI[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$BMI[1:threshold]), names(eBayes.eFDR.test2$BMI[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$BMI[1:threshold]), names(voom.eFDR.test2$BMI[1:threshold])) |> length(),
               
               ##------AGE------
               intersect(names(BMAseq.eFDR.Main.train$AGE[1:threshold]), names(BMAseq.eFDR.Main.test$AGE[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$AGE[1:threshold], DESeq2.eFDR.GeneName.test$AGE[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$AGE[1:threshold], edgeR.eFDR.GeneName.test$AGE[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$AGE[1:threshold]), names(eBayes.eFDR.test2$AGE[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$AGE[1:threshold]), names(voom.eFDR.test2$AGE[1:threshold])) |> length(),
               
               ##------SEX------
               intersect(names(BMAseq.eFDR.Main.train$SEX[1:threshold]), names(BMAseq.eFDR.Main.test$SEX[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$SEX[1:threshold], DESeq2.eFDR.GeneName.test$SEX[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$SEX[1:threshold], edgeR.eFDR.GeneName.test$SEX[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$SEX[1:threshold]), names(eBayes.eFDR.test2$SEX[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$SEX[1:threshold]), names(voom.eFDR.test2$SEX[1:threshold])) |> length(),
               
               ##------MHABNWBC------
               intersect(names(BMAseq.eFDR.Main.train$MHABNWBC[1:threshold]), names(BMAseq.eFDR.Main.test$MHABNWBC[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$MHABNWBC[1:threshold], DESeq2.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$MHABNWBC[1:threshold], edgeR.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$MHABNWBC[1:threshold]), names(eBayes.eFDR.test2$MHABNWBC[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$MHABNWBC[1:threshold]), names(voom.eFDR.test2$MHABNWBC[1:threshold])) |> length()))
}

##------Rate of cDEGs------
boxplot_data2 <- function (threshold = 1000, random.seed = 8809678) {
  data.table(threshold=rep(threshold,5), 
             random.seed=random.seed,
             method=c("BMAseq","DESeq2","edgeR","eBayes","voom.limma"),
             variable=rep(c("BMI","AGE","SEX","MHABNWBC"),each=5),
             rate.cDEGs=c(
               ##------BMI------
               intersect(names(BMAseq.eFDR.Main.train$BMI[1:threshold]), names(BMAseq.eFDR.Main.test$BMI[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$BMI[1:threshold], DESeq2.eFDR.GeneName.test$BMI[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$BMI[1:threshold], edgeR.eFDR.GeneName.test$BMI[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$BMI[1:threshold]), names(eBayes.eFDR.test2$BMI[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$BMI[1:threshold]), names(voom.eFDR.test2$BMI[1:threshold])) |> length() / threshold,
               
               ##------AGE------
               intersect(names(BMAseq.eFDR.Main.train$AGE[1:threshold]), names(BMAseq.eFDR.Main.test$AGE[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$AGE[1:threshold], DESeq2.eFDR.GeneName.test$AGE[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$AGE[1:threshold], edgeR.eFDR.GeneName.test$AGE[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$AGE[1:threshold]), names(eBayes.eFDR.test2$AGE[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$AGE[1:threshold]), names(voom.eFDR.test2$AGE[1:threshold])) |> length() / threshold,
               
               ##------SEX------
               intersect(names(BMAseq.eFDR.Main.train$SEX[1:threshold]), names(BMAseq.eFDR.Main.test$SEX[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$SEX[1:threshold], DESeq2.eFDR.GeneName.test$SEX[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$SEX[1:threshold], edgeR.eFDR.GeneName.test$SEX[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$SEX[1:threshold]), names(eBayes.eFDR.test2$SEX[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$SEX[1:threshold]), names(voom.eFDR.test2$SEX[1:threshold])) |> length() / threshold,
               
               ##------MHABNWBC------
               intersect(names(BMAseq.eFDR.Main.train$MHABNWBC[1:threshold]), names(BMAseq.eFDR.Main.test$MHABNWBC[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$MHABNWBC[1:threshold], DESeq2.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$MHABNWBC[1:threshold], edgeR.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$MHABNWBC[1:threshold]), names(eBayes.eFDR.test2$MHABNWBC[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$MHABNWBC[1:threshold]), names(voom.eFDR.test2$MHABNWBC[1:threshold])) |> length() / threshold))
}
```


## Organize data for boxplotting
```{r}
##------Number of cDEGs------
threshold.vec <- seq(1000, 5000, 1000)
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
boxplot.data.num <- NULL
for (threshold.i in threshold.vec) {
  for (seed.i in seed.vec) {
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/BMAseqMulti%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/DESeq2Multi%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/edgeRMulti%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/eBayesMulti%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/VoomLimmaMulti%s.RData", seed.i))
    boxplot.data.num <- rbind(boxplot.data.num, boxplot_data(threshold.i, seed.i))
  }
}
saveRDS(boxplot.data.num, "../ApplicationResult/Multi/RandomSeed/numcDEGs/Boxplot/boxplot.data.num.RDS")


##------Rate of cDEGs------
threshold.vec <- seq(1000, 5000, 1000)
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
boxplot.data.rate <- NULL
for (threshold.i in threshold.vec) {
  for (seed.i in seed.vec) {
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/BMAseqMulti%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/DESeq2Multi%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/edgeRMulti%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/eBayesMulti%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/VoomLimmaMulti%s.RData", seed.i))
    boxplot.data.rate <- rbind(boxplot.data.rate, boxplot_data2(threshold.i, seed.i))
  }
}
saveRDS(boxplot.data.rate, "../ApplicationResult/Multi/RandomSeed/ratecDEGs/Boxplot/boxplot.data.rate.RDS")
```


## Make and save boxplots
```{r}
##------Number of cDEGs------
date.analysis <- format(Sys.Date(), "%Y%b%d")
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC")

## Read organized boxplot data
boxplot.data.num <- readRDS("../ApplicationResult/Multi/RandomSeed/numcDEGs/Boxplot/boxplot.data.num.RDS")
boxplot.data.num[,method:=factor(method,levels=c("BMAseq","DESeq2","edgeR","eBayes","voom.limma"))]

for (var.i in var.vec) {
  g <- boxplot.data.num[variable==var.i] |>
    ggplot(aes(y = num.cDEGs, x = method, group = method)) + 
    geom_boxplot(outlier.shape = NA) + 
    geom_jitter(aes(color = method), size = 0.8) + 
    stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
    facet_wrap(vars(threshold)) + 
    labs(y = "Number of common differentially expressed genes", x = "Method") + 
    theme_bw() + 
    theme(legend.position = "none",
          axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
    ggsave(filename = sprintf("../ApplicationResult/Multi/RandomSeed/numcDEGs/Boxplot/%s_%s.eps", date.analysis, var.i),
           plot = g, device = "eps", dpi = 600, width = 9, height = 6, units = "in")
}


##------Rate of cDEGs------
## Read organized boxplot data
boxplot.data.rate <- readRDS("../ApplicationResult/Multi/RandomSeed/ratecDEGs/Boxplot/boxplot.data.rate.RDS")
boxplot.data.rate[,method:=factor(method,levels=c("BMAseq","DESeq2","edgeR","eBayes","voom.limma"))]

for (var.i in var.vec) {
  if (var.i != "MHABNWBC") {
    g <- boxplot.data.rate[variable==var.i] |>
      ggplot(aes(y = rate.cDEGs, x = method, group = method)) + 
      geom_boxplot(outlier.shape = NA) + 
      geom_jitter(aes(color = method), size = 0.8) + 
      stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
      facet_wrap(vars(threshold)) + 
      labs(y = "Rate of common differentially expressed genes", x = "Method") + 
      theme_bw() + 
      theme(legend.position = "none",
            axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5)) +
      scale_y_continuous(labels = scales::label_number(accuracy = 0.01), breaks = seq(0, 0.5, 0.1), limits = c(0, 0.5))
    ggsave(filename = sprintf("../ApplicationResult/Multi/RandomSeed/ratecDEGs/Boxplot/%s_%s.eps", date.analysis, var.i),
             plot = g, device = "eps", dpi = 600, width = 9, height = 6, units = "in")
  } else {
    g <- boxplot.data.rate[variable==var.i] |>
      ggplot(aes(y = rate.cDEGs, x = method, group = method)) + 
      geom_boxplot(outlier.shape = NA) + 
      geom_jitter(aes(color = method), size = 0.8) + 
      stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
      facet_wrap(vars(threshold)) + 
      labs(y = "Rate of common differentially expressed genes", x = "Method") + 
      theme_bw() + 
      theme(legend.position = "none",
            axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5)) +
      scale_y_continuous(labels = scales::label_number(accuracy = 0.01), breaks = seq(0, 0.8, 0.1), limits = c(0, 0.8))
      ggsave(filename = sprintf("../ApplicationResult/Multi/RandomSeed/ratecDEGs/Boxplot/%s_%s.eps", date.analysis, var.i),
             plot = g, device = "eps", dpi = 600, width = 9, height = 6, units = "in")
  }
}
```


## Estimate the duplicate rate of a cDEG associated with a variable
```{r}
cDEGs = cDEGs.list = NULL
threshold.vec <- seq(1000, 5000, 1000)
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)

duplicate_rate_cDEGs_per_variable <- function(var.name = "BMI", method.name = "BMAseq") {
  for (threshold.i in threshold.vec) {
    for (seed.i in seed.vec) {
      if (method.name == "BMAseq") {
        load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sMulti%s.RData", method.name, seed.i))
        cDEGs <- c(intersect(names(get(sprintf("%s.eFDR.Main.train", method.name))[[var.name]][1:threshold.i]), 
                             names(get(sprintf("%s.eFDR.Main.test", method.name))[[var.name]][1:threshold.i])))
      } else if (method.name %in% c("DESeq2", "edgeR")) {
        load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sMulti%s.RData", method.name, seed.i))
        cDEGs <- c(intersect(get(sprintf("%s.eFDR.GeneName.train", method.name))[[var.name]][1:threshold.i], 
                             get(sprintf("%s.eFDR.GeneName.test", method.name))[[var.name]][1:threshold.i]))
      } else if (method.name == "eBayes") {
        load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sMulti%s.RData", method.name, seed.i))
        cDEGs <- c(intersect(names(get(sprintf("%s.eFDR.train2", method.name))[[var.name]][1:threshold.i]), 
                             names(get(sprintf("%s.eFDR.test2", method.name))[[var.name]][1:threshold.i])))
      } else {
        load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sMulti%s.RData", method.name, seed.i))
        method.name2 <- "voom"
        cDEGs <- c(intersect(names(get(sprintf("%s.eFDR.train2", method.name2))[[var.name]][1:threshold.i]), 
                             names(get(sprintf("%s.eFDR.test2", method.name2))[[var.name]][1:threshold.i])))
      }
      cDEGs.list <- append(cDEGs.list, list(cDEGs))
    }
    cDEGs.unique <- unique(unlist(cDEGs.list))
    
    # Create a matrix that shows the occurrence of each unique gene among the 10 random seed trials
    cDEGs.matrix <- matrix(0, nrow = length(cDEGs.unique), ncol = 10)
    rownames(cDEGs.matrix) <- cDEGs.unique
    colnames(cDEGs.matrix) <- seed.vec   
    for (i in 1:10) {
      cDEGs.i.seed <- cDEGs.list[[i]]
      for (j in 1:length(cDEGs.unique)) {
        if (cDEGs.unique[j] %in% cDEGs.i.seed) {
          cDEGs.matrix[j, i] <- 1
        }
      }
    }
    Sum <- apply(cDEGs.matrix, 1, sum)
    Percent <- Sum/10
    cDEGs.matrix <- cbind(cDEGs.matrix, Sum, Percent) # Add two new columns to the matrix
    saveRDS(cDEGs.matrix, sprintf("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/%sMulti%s_%s.RDS", method.name, var.name, threshold.i))
    
    print(paste("Complete the duplicated rate matrix of cDEGs associated with", var.name, "identified by", method.name, "among 10 random seeds for the threshold", threshold.i))
    cDEGs = cDEGs.list =  NULL
  }
}

mapply(duplicate_rate_cDEGs_per_variable,
       rep(c("BMI", "AGE", "SEX", "MHABNWBC"), each = 5), 
       rep(c("BMAseq", "DESeq2", "edgeR", "eBayes", "VoomLimma"), times = 4))

method.vec <- c("BMAseq", "DESeq2", "edgeR", "eBayes", "VoomLimma")
var.name <- "BMI"
threshold.i <- 2000
for (method.name in method.vec) {
  data <- readRDS(sprintf("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/MultiModel/%sMulti%s_%s.RDS", method.name, var.name, threshold.i))
  write.csv(data, sprintf("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/MultiModel/%sMulti%s_%s.csv", method.name, var.name, threshold.i))
}
```


## Prepare and organize the data for boxplots
```{r}
boxplot_duplicate_rate_data <- function (var.name = NULL, method.name = NULL, threshold = NULL) {
  load.data <- readRDS(sprintf("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/MultiModel/%sMulti%s_%s.RDS", method.name, var.name, threshold))
  temp.data <- data.table(threshold=threshold, 
                          method=method.name,
                          variable=var.name,
                          cDEGs.duplicate.rate.median=median(load.data[,12]),
                          cDEGs.duplicate.rate.mean=mean(load.data[,12]))
  return(temp.data)
}

organized_duplicate_rate_data <- mapply(boxplot_duplicate_rate_data,
                                        var.name = rep(c("BMI", "AGE","SEX", "MHABNWBC"), each = 25), 
                                        method.name = rep(rep(c("BMAseq", "DESeq2", "edgeR", "eBayes", "VoomLimma"), each = 5), times = 4),
                                        threshold = rep(rep(seq(1000, 5000, 1000), times = 5), times = 4), SIMPLIFY = F) |> rbindlist()

saveRDS(organized_duplicate_rate_data, "../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/MultiModel/organized_duplicate_rate_data.RDS")

# data.frame(var.name, method.name, threshold)
```


## Make and save the duplicate rate boxplots
```{r}
organized_duplicate_rate_data <- readRDS("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/MultiModel/organized_duplicate_rate_data.RDS")
organized_duplicate_rate_data[,variable:=factor(variable,levels=c("BMI","AGE","SEX","MHABNWBC"))]
organized_duplicate_rate_data[,method:=factor(method,levels=c("BMAseq","DESeq2","edgeR","eBayes","VoomLimma"),labels=c("BMAseq","DESeq2","edgeR","eBayes","voom.limma"))]
organized_duplicate_rate_data[,cDEGs.duplicate.rate.median:=as.numeric(cDEGs.duplicate.rate.median)]
organized_duplicate_rate_data[,cDEGs.duplicate.rate.mean:=as.numeric(cDEGs.duplicate.rate.mean)]
date.analysis <- format(Sys.Date(), "%Y%b%d")
g1 <- organized_duplicate_rate_data |>
  ggplot(mapping = aes(y = cDEGs.duplicate.rate.median, x = method, group = method)) + 
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(aes(color = method), size = 0.8) + 
  facet_wrap(vars(variable)) + 
  scale_y_continuous(labels = scales::label_number(accuracy = 0.01),
                     breaks = seq(0, 1, 0.1), 
                     limits = c(0, 1)) + 
  labs(y = "Median duplicate rate of cDEGs", x = "Method") + 
  theme_bw() + 
  theme(legend.position = "none",
        axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
ggsave(filename = sprintf("../ApplicationResult/Multi/RandomSeed/DuplicatedRateMatrix/Boxplot/%s_median.eps", date.analysis),
       plot = g1, device = "eps", dpi = 600, width = 9, height = 6, units = "in")

g2 <- organized_duplicate_rate_data |>
  ggplot(mapping = aes(y = cDEGs.duplicate.rate.mean, x = method, group = method)) + 
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(aes(color = method), size = 0.8) + 
  facet_wrap(vars(variable)) + 
  scale_y_continuous(labels = scales::label_number(accuracy = 0.01),
                     breaks = seq(0, 1, 0.1), 
                     limits = c(0, 1)) + 
  labs(y = "Mean duplicate rate of cDEGs", x = "Method") + 
  theme_bw() + 
  theme(legend.position = "none",
        axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
ggsave(filename = sprintf("../ApplicationResult/Multi/RandomSeed/DuplicatedRateMatrix/Boxplot/%s_mean.eps", date.analysis),
       plot = g2, device = "eps", dpi = 600, width = 9, height = 6, units = "in")
```


## Make Venn diagrams of cDEGs
```{r threshold.2000}
threshold <- 2000
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC")
date.analysis <- format(Sys.Date(), "%Y%b%d")

make_vennplot <- function(var.name = NULL, threshold = NULL) {

  BMAseq.cDEGs <- intersect(names(BMAseq.eFDR.Main.train[[var.name]][1:threshold]), names(BMAseq.eFDR.Main.test[[var.name]][1:threshold]))
  DESeq2.cDEGs <- intersect(DESeq2.eFDR.GeneName.train[[var.name]][1:threshold], DESeq2.eFDR.GeneName.test[[var.name]][1:threshold])
  edgeR.cDEGs <- intersect(edgeR.eFDR.GeneName.train[[var.name]][1:threshold], edgeR.eFDR.GeneName.test[[var.name]][1:threshold])
  eBayes.cDEGs <- intersect(names(eBayes.eFDR.train2[[var.name]][1:threshold]), names(eBayes.eFDR.test2[[var.name]][1:threshold]))
  voom.limma.cDEGs <- intersect(names(voom.eFDR.train2[[var.name]][1:threshold]), names(voom.eFDR.test2[[var.name]][1:threshold]))
  
  g <- ggVennDiagram(list(BMAseq = BMAseq.cDEGs, 
                          DESeq2 = DESeq2.cDEGs,
                          edgeR = edgeR.cDEGs,
                          eBayes = eBayes.cDEGs,
                          voom.limma = voom.limma.cDEGs),
                     label_alpha = 0, label_color = "white") +
    theme(legend.position = "none", 
          plot.title = element_text(hjust = 0.5, colour = "red", size = 18)) + 
    ggtitle(paste("The intersection of cDEGs associated with the", var.name, "\nunder the seed", seed.i, "and threshold", threshold)) + 
    scale_fill_distiller(palette = "Dark2", direction = -1)
  
  ggsave(filename = paste0("../ApplicationResult/Multi/RandomSeed/TMM_Top", threshold, "/Vennplot/", date.analysis, "_", var.name, "_", seed.i, ".eps"),
         plot = g,
         device = "eps",
         dpi = 600, 
         width = 10,
         height = 7,
         units = "in")
}

for (seed.i in seed.vec) {
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/BMAseqMulti%s.RData", seed.i))
  names(BMAseq.eFDR.Main.train) = names(BMAseq.eFDR.Main.test) = var.vec
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/DESeq2Multi%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/edgeRMulti%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/eBayesMulti%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/VoomLimmaMulti%s.RData", seed.i))
  for (var.i in var.vec) {
    make_vennplot(var.name = var.i, threshold = 2000)
  }
}
```


```{r threshold.5000}
threshold <- 5000
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC")
date.analysis <- format(Sys.Date(), "%Y%b%d")

for (seed.i in seed.vec) {
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/BMAseqMulti%s.RData", seed.i))
  names(BMAseq.eFDR.Main.train) = names(BMAseq.eFDR.Main.test) = var.vec
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/DESeq2Multi%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/edgeRMulti%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/eBayesMulti%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/VoomLimmaMulti%s.RData", seed.i))
  for (var.i in var.vec) {
    make_vennplot(var.name = var.i, threshold = 5000)
  }
}
```


## Extract unique cDEGs identified by BMAseq per seed and all unique cDEGs identified by BMAseq across 10 seeds
```{r}
file.name <- "../ApplicationResult/Multi/RandomSeed/TMM_Top2000/Vennplot/uniq_cDEGs_BMI_BMAseq_per_seed.txt"
unique.cDEGs.list <- NULL
extract_unique_cDEGs_per_seed <- function(var.name = NULL, threshold = NULL) {
  BMAseq.cDEGs <- intersect(names(BMAseq.eFDR.Main.train[[var.name]][1:threshold]), names(BMAseq.eFDR.Main.test[[var.name]][1:threshold]))
  DESeq2.cDEGs <- intersect(DESeq2.eFDR.GeneName.train[[var.name]][1:threshold], DESeq2.eFDR.GeneName.test[[var.name]][1:threshold])
  edgeR.cDEGs <- intersect(edgeR.eFDR.GeneName.train[[var.name]][1:threshold], edgeR.eFDR.GeneName.test[[var.name]][1:threshold])
  eBayes.cDEGs <- intersect(names(eBayes.eFDR.train2[[var.name]][1:threshold]), names(eBayes.eFDR.test2[[var.name]][1:threshold]))
  voom.limma.cDEGs <- intersect(names(voom.eFDR.train2[[var.name]][1:threshold]), names(voom.eFDR.test2[[var.name]][1:threshold]))
  unique.cDEGs <- setdiff(BMAseq.cDEGs, union(DESeq2.cDEGs, edgeR.cDEGs) |> union(eBayes.cDEGs) |> union(voom.limma.cDEGs))
  return(unique.cDEGs)
}

for (seed.i in seed.vec) {
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/BMAseqMulti%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/DESeq2Multi%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/edgeRMulti%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/eBayesMulti%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/VoomLimmaMulti%s.RData", seed.i))
  unique.cDEGs <- extract_unique_cDEGs_per_seed(var.name = "BMI", threshold = 2000)
  unique.cDEGs.list <- append(unique.cDEGs.list, list(unique.cDEGs))
  write(sprintf("\nThe following gene IDs are unique gene IDs associated with the main effect of %s identified by BMAseq under the seed %s and threshold %s", var.name, seed.i, threshold), file = file.name, append = T)
  write.table(unique.cDEGs, file = file.name, quote = F, col.names = F, append = T)
}
names(unique.cDEGs.list) <- seed.vec

file.name <- "../ApplicationResult/Multi/RandomSeed/TMM_Top2000/Vennplot/uniq_cDEGs_BMI_BMAseq_all_seeds.txt"
write.table(unlist(unique.cDEGs.list) |> unique(), file = file.name, quote = F, col.names = F)
```


## Estimate the proportion of cDEGs explained by univariate or multivariate best models
```{r}
# Build a BMAseq best model list associated with cDEGs/non-cDEGs per variable per random seed per ranking threshold
collect_bestmodel <- function (var.name = NULL, threshold = NULL, seed = NULL) {
  temp.train <- output.multi.train.s2[["DEG.bestmodel"]][[var.name]]
  temp.test <- output.multi.test.s2[["DEG.bestmodel"]][[var.name]]
  DEG.train <- BMAseq.eFDR.Main.train[[var.name]][1:threshold] |> names()
  DEG.test <- BMAseq.eFDR.Main.test[[var.name]][1:threshold] |> names()
  cDEG.all <- intersect(DEG.train, DEG.test)
  noncDEG.train <- setdiff(DEG.train, cDEG.all)
  noncDEG.test <- setdiff(DEG.test, cDEG.all)
  cDEG.bestmodel.train <- temp.train[temp.train[, "name.DEG"] %in% cDEG.all, ] 
  cDEG.bestmodel.test <- temp.test[temp.test[, "name.DEG"] %in% cDEG.all, ] 
  noncDEG.bestmodel.train <- temp.train[temp.train[, "name.DEG"] %in% noncDEG.train, ] 
  noncDEG.bestmodel.test <- temp.test[temp.test[, "name.DEG"] %in% noncDEG.test, ] 
  colnames(cDEG.bestmodel.train)[3] <- sprintf("bestmodel.train.%s", seed)
  colnames(cDEG.bestmodel.test)[3] <- sprintf("bestmodel.test.%s", seed)
  colnames(noncDEG.bestmodel.train)[3] <- sprintf("bestmodel.train.%s", seed)
  colnames(noncDEG.bestmodel.test)[3] <- sprintf("bestmodel.test.%s", seed)
  cDEG.bestmodel.all <- merge(cDEG.bestmodel.train, cDEG.bestmodel.test, by = c("index.DEG", "name.DEG"))
  noncDEG.bestmodel.all <- merge(noncDEG.bestmodel.train, noncDEG.bestmodel.test, all = T)
  return(list(cDEG.model.result = cDEG.bestmodel.all, noncDEG.model.result = noncDEG.bestmodel.all))
}

# Run the best model collector for each variable, each threshold, each seed
cDEG.bestmodel.list = noncDEG.bestmodel.list = NULL
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC")
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
threshold.vec <- seq(1000, 5000, 1000)
for (threshold.i in threshold.vec) { 
  for (seed.i in seed.vec) { 
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/BMAseqMulti%sFDR1.RData", seed.i))
    for (var.i in var.vec) { 
      cDEG.bestmodel.i <- collect_bestmodel(threshold = threshold.i, seed = seed.i, var.name = var.i)[["cDEG.model.result"]] |> list()
      noncDEG.bestmodel.i <- collect_bestmodel(threshold = threshold.i, seed = seed.i, var.name = var.i)[["noncDEG.model.result"]] |> list()
      names(cDEG.bestmodel.i) = names(noncDEG.bestmodel.i) = paste0(var.i, "+", threshold.i, "+", seed.i) 
      cDEG.bestmodel.list <- c(cDEG.bestmodel.list, cDEG.bestmodel.i)
      noncDEG.bestmodel.list <- c(noncDEG.bestmodel.list, noncDEG.bestmodel.i)
    }
  }
}

# Save all best models of cDEGs and non-cDEGs among each variable, each threshold, each seed
date.analysis <- format(Sys.Date(), "%Y%b%d")
save(cDEG.bestmodel.list, noncDEG.bestmodel.list, file = sprintf("../ApplicationResult/Multi/RandomSeed/BestModelProportion/%s_BMAseq.bestmodel.list.RData", date.analysis))
load("../ApplicationResult/Multi/RandomSeed/BestModelProportion/2023May23_BMAseq.bestmodel.list.RData")


# Assign each BMAseq best model to each cDEG per variable per random seed per ranking threshold
cDEG = cDEG.list = bestmodel.list = NULL
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC")
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
threshold.vec <- seq(1000, 5000, 1000)
most_fr_bestmodel_cDEG_per_var <- function(var.name = "BMI") {
  for (threshold.i in threshold.vec) { 
    for (seed.i in seed.vec) { 
      file.name <- sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/BMAseqMulti%sFDR1.RData", seed.i)
      load(file.name)
      object.list <- ls(pattern = "dat.*|.*s1")
      rm(list = object.list)
      temp.train <- output.multi.train.s2[["DEG.bestmodel"]][[var.name]]
      temp.test <- output.multi.test.s2[["DEG.bestmodel"]][[var.name]]
      cDEG <- intersect(BMAseq.eFDR.Main.train[[var.name]][1:threshold.i] |> names(), BMAseq.eFDR.Main.test[[var.name]][1:threshold.i] |> names())
      cDEG.list <- c(cDEG.list, list(cDEG))
    }
    cDEG.unique <- unique(unlist(cDEG.list)) 
    cDEG.bestmodel.matrix <- matrix(0, nrow = length(cDEG.unique), ncol = 20) 
    rownames(cDEG.bestmodel.matrix) <- cDEG.unique
    colnames(cDEG.bestmodel.matrix) <- paste(c("bestmodel.train", "bestmodel.test"), rep(seed.vec, each = 2), sep = ".")  
    for (i in 1:10) { 
      seed.value <- seed.vec[i]
      temp.cDEG.bestmodel <- cDEG.bestmodel.list[[sprintf("%s+%s+%s", var.name, threshold.i, seed.value)]]
      temp.noncDEG.bestmodel <- noncDEG.bestmodel.list[[sprintf("%s+%s+%s", var.name, threshold.i, seed.value)]]
      cDEG.per.seed <- temp.cDEG.bestmodel$name.DEG 
      noncDEG.per.seed <- temp.noncDEG.bestmodel$name.DEG 
      for (j in 1:length(cDEG.unique)) { 
        if (cDEG.unique[j] %in% cDEG.per.seed) { 
          cDEG.bestmodel.matrix[j, 2*i-1] <- temp.cDEG.bestmodel[which(temp.cDEG.bestmodel$name.DEG == cDEG.unique[j]), sprintf("bestmodel.train.%s", seed.value)] 
          cDEG.bestmodel.matrix[j, 2*i] <- temp.cDEG.bestmodel[which(temp.cDEG.bestmodel$name.DEG == cDEG.unique[j]), sprintf("bestmodel.test.%s", seed.value)]
        } else if (cDEG.unique[j] %in% noncDEG.per.seed) { 
          cDEG.bestmodel.matrix[j, 2*i-1] <- temp.noncDEG.bestmodel[which(temp.noncDEG.bestmodel$name.DEG == cDEG.unique[j]), sprintf("bestmodel.train.%s", seed.value)]
          cDEG.bestmodel.matrix[j, 2*i] <- temp.noncDEG.bestmodel[which(temp.noncDEG.bestmodel$name.DEG == cDEG.unique[j]), sprintf("bestmodel.test.%s", seed.value)]
        } else { 
          cDEG.bestmodel.matrix[j, 2*i-1] <- NA
          cDEG.bestmodel.matrix[j, 2*i] <- NA
        }
      }
    }
    most.fr.bestmodel <- apply(cDEG.bestmodel.matrix, 1, function(x) names(table(x))[which(table(x) == max(table(x)))]) 
    most.fr.bestmodel.onecol <- lapply(most.fr.bestmodel, function(x) paste0(x, collapse = "/")) |> unlist()
    sum.uni <- lapply(most.fr.bestmodel, function (x) any(sprintf("~1+%s", var.name) %in% x)) |> unlist() |> sum()
    bestmodel.multi <- lapply(most.fr.bestmodel, function (x) { (x != sprintf("~1+%s", var.name)) & grepl(var.name, x) }) 
    sum.multi <- lapply(bestmodel.multi, sum)
    sum.multi <- ifelse(sum.multi > 1, 1, sum.multi) |> unlist() |> sum() 
    percent.uni <- sum.uni / sum(sum.uni + sum.multi)
    percent.multi <- sum.multi / sum(sum.uni + sum.multi)
    cDEG.bestmodel.matrix <- cbind(cDEG.bestmodel.matrix, most.fr.bestmodel.onecol, sum.uni, sum.multi, percent.uni, percent.multi)
    file.name <- sprintf("../ApplicationData/derived/RandomSeed/BestModelMatrix/%s_%s.RDS", var.name, threshold.i)
    saveRDS(cDEG.bestmodel.matrix, file.name)
    print(paste("Complete the BMAseq best model matrix of cDEG associated with", var.name, "among 10 random seeds for the threshold", threshold.i))
    cDEG = cDEG.list =  NULL
    object.list <- ls(pattern = "BMAseq.*|.*s2")
    rm(list = object.list)
  }
}  

mapply(FUN = most_fr_bestmodel_cDEG_per_var,
       c("BMI", "AGE", "SEX", "MHABNWBC"))

# Spot check the data 
BMI_1000 <- readRDS("../ApplicationData/derived/RandomSeed/BestModelMatrix/BMI_1000.RDS")

# Export the matrix into xslx
library(openxlsx)
threshold.vec <- seq(1000, 5000, 1000)
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC")
for (threshold.i in threshold.vec) {
  for (var.i in var.vec) {
    file.name <- sprintf("../ApplicationData/derived/RandomSeed/BestModelMatrix/%s_%s.RDS", var.i, threshold.i)
    data <- readRDS(file.name) |> data.frame()
    file.name <- sprintf("../ApplicationData/derived/RandomSeed/BestModelMatrix/%s_%s.xlsx", var.i, threshold.i)
    write.xlsx(x = data, file = file.name, rowNames = T, colWidths = 40, firstActiveRow = 2, firstActiveCol = 2)
  }
}
```


## Collect unique genes identified by each approach
```{r}
threshold <- 5000
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC")
date.analysis <- format(Sys.Date(), "%Y%b%d")
var.name <- "BMI"


cDEGs.type.vec <- c("BMAseq.cDEGs", 
                    "DESeq2.UVM.cDEGs",
                    "edgeR.UVM.cDEGs",
                    "eBayes.UVM.cDEGs",
                    "voom.limma.UVM.cDEGs",
                    "DESeq2.MVM.cDEGs",
                    "edgeR.MVM.cDEGs",
                    "eBayes.MVM.cDEGs",
                    "voom.limma.MVM.cDEGs")

for (var.name in var.vec) {
  for (target.type in cDEGs.type.vec) {
    unique.cDEGs.all.seed <- NULL
    target.type.idx <- which(cDEGs.type.vec == target.type)
    nontarget.type.idx <- which(cDEGs.type.vec != target.type)
    
    for (seed.i in seed.vec) {
      ##------BMAseq------
      file.name <- sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/Trim/BMAseqMulti%s.RData", seed.i) # Multivariable BMAseq model without interaction terms
      data.env <- local({load(file.name); environment()})
      names(data.env$BMAseq.eFDR.Main.train) = names(data.env$BMAseq.eFDR.Main.test) = var.vec
      BMAseq.cDEGs <- intersect(names(data.env$BMAseq.eFDR.Main.train[[var.name]][1:threshold]), 
                                names(data.env$BMAseq.eFDR.Main.test[[var.name]][1:threshold]))
      rm(list = "data.env")
      
      ##------DESeq2_UVM------
      file.name <- sprintf("../ApplicationData/derived/RandomSeed/Top5000/UniModel/Trim/DESeq2Uni%s.RData", seed.i) # Univariable DESeq2 model
      data.env <- local({load(file.name); environment()})
      DESeq2.UVM.cDEGs <- intersect(data.env$DESeq2.eFDR.GeneName.train[[var.name]][1:threshold], 
                                    data.env$DESeq2.eFDR.GeneName.test[[var.name]][1:threshold])
      rm(list = "data.env")
      
      ##------edgeR_UVM------
      file.name <- sprintf("../ApplicationData/derived/RandomSeed/Top5000/UniModel/Trim/edgeRUni%s.RData", seed.i) # Univariable edgeR model
      data.env <- local({load(file.name); environment()})
      edgeR.UVM.cDEGs <- intersect(data.env$edgeR.eFDR.GeneName.train[[var.name]][1:threshold],
                                   data.env$edgeR.eFDR.GeneName.test[[var.name]][1:threshold])
      rm(list = "data.env")
      
      ##------eBayes_UVM------
      file.name <- sprintf("../ApplicationData/derived/RandomSeed/Top5000/UniModel/Trim/eBayesUni%s.RData", seed.i) # Univariable eBayes model
      data.env <- local({load(file.name); environment()})
      eBayes.UVM.cDEGs <- intersect(names(data.env$eBayes.eFDR.train2[[var.name]][1:threshold]), 
                                    names(data.env$eBayes.eFDR.test2[[var.name]][1:threshold]))
      rm(list = "data.env")
      
      ##------voom.limma_UVM------
      file.name <- sprintf("../ApplicationData/derived/RandomSeed/Top5000/UniModel/Trim/VoomLimmaUni%s.RData", seed.i) # Univariable VoomLimma model
      data.env <- local({load(file.name); environment()})
      voom.limma.UVM.cDEGs <- intersect(names(data.env$voom.eFDR.train2[[var.name]][1:threshold]), 
                                        names(data.env$voom.eFDR.test2[[var.name]][1:threshold]))
      rm(list = "data.env")
      
      ##------DESeq2_MVM------
      file.name <- sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/Trim/DESeq2Multi%s.RData", seed.i) # Multivariable DESeq2 model
      data.env <- local({load(file.name); environment()})
      DESeq2.MVM.cDEGs <- intersect(data.env$DESeq2.eFDR.GeneName.train[[var.name]][1:threshold], 
                                    data.env$DESeq2.eFDR.GeneName.test[[var.name]][1:threshold])
      rm(list = "data.env")
      
      ##------edgeR_MVM------
      file.name <- sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/Trim/edgeRMulti%s.RData", seed.i) # Multivariable edgeR model
      data.env <- local({load(file.name); environment()})
      edgeR.MVM.cDEGs <- intersect(data.env$edgeR.eFDR.GeneName.train[[var.name]][1:threshold],
                                   data.env$edgeR.eFDR.GeneName.test[[var.name]][1:threshold])
      rm(list = "data.env")
      
      ##------eBayes_MVM------
      file.name <- sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/Trim/eBayesMulti%s.RData", seed.i) # Multivariable eBayes model
      data.env <- local({load(file.name); environment()})
      eBayes.MVM.cDEGs <- intersect(names(data.env$eBayes.eFDR.train2[[var.name]][1:threshold]), 
                                    names(data.env$eBayes.eFDR.test2[[var.name]][1:threshold]))
      rm(list = "data.env")
      
      ##------voom.limma_MVM------
      file.name <- sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/Trim/VoomLimmaMulti%s.RData", seed.i) # Multivariable VoomLimma model
      data.env <- local({load(file.name); environment()})
      voom.limma.MVM.cDEGs <- intersect(names(data.env$voom.eFDR.train2[[var.name]][1:threshold]), 
                                        names(data.env$voom.eFDR.test2[[var.name]][1:threshold]))
      rm(list = "data.env")
      
      unique.cDEGs.per.seed <-
        setdiff(
          get(target.type),
          union(get(cDEGs.type.vec[nontarget.type.idx[1]]), get(cDEGs.type.vec[nontarget.type.idx[2]])) |>
            union(get(cDEGs.type.vec[nontarget.type.idx[3]])) |>
            union(get(cDEGs.type.vec[nontarget.type.idx[4]])) |>
            union(get(cDEGs.type.vec[nontarget.type.idx[5]])) |>
            union(get(cDEGs.type.vec[nontarget.type.idx[6]])) |>
            union(get(cDEGs.type.vec[nontarget.type.idx[7]])) |>
            union(get(cDEGs.type.vec[nontarget.type.idx[8]]))
        )
      
      method.name <- gsub("\\.", "_", 
                          gsub(".cDEGs", "", target.type))
      txt.file.name <- sprintf("../ApplicationResult/UniqueGene/TMM_Top5000/%s_uniq_cDEGs_%s_per_seed.txt", method.name, var.name)
      write(sprintf("\nThe following gene IDs are unique gene IDs associated with the main effect of %s identified by %s under the seed %s and threshold %s", 
                    var.name, method.name, seed.i, threshold), file = txt.file.name, append = T)
      write.table(unique.cDEGs.per.seed, 
                  file = txt.file.name, 
                  quote = F, col.names = F, append = T)
      
      unique.cDEGs.all.seed <- append(unique.cDEGs.all.seed, list(unique.cDEGs.per.seed))
    }
    names(unique.cDEGs.all.seed) <- seed.vec
    saveRDS(unique.cDEGs.all.seed, file = sprintf("../ApplicationResult/UniqueGene/TMM_Top5000/%s.uniq.cDEGs.%s.all.seed.RDS", method.name, var.name))
  }
}
```


## Construct GO enrichment analysis
```{r table}
keytypes(org.Hs.eg.db)

##------Load cDEGs data------
seed.i = 8809678
var.name = "BMI"
threshold = 5000

##------BMAseq-----
method.name = "BMAseq"
model.type = "Multi"
file.name <- sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sModel/Trim/%s%s%s.RData", 
                     model.type, method.name, model.type, seed.i) 
data.env <- local({load(file.name); environment()})
BMAseq.cDEGs <- intersect(names(data.env$BMAseq.eFDR.Main.train[[var.name]][1:threshold]), 
                          names(data.env$BMAseq.eFDR.Main.test[[var.name]][1:threshold]))
BMAseq.unique.cDEGs.all.seed <- readRDS(sprintf("../ApplicationResult/UniqueGene/TMM_Top5000/%s.uniq.cDEGs.%s.all.seed.RDS",
                                                method.name, var.name))
BMAseq.unique.cDEGs <- BMAseq.unique.cDEGs.all.seed[[as.character(seed.i)]]

# Delete any character after the comma
BMAseq.cDEGs.8809678 <- sub("\\..*", "", BMAseq.cDEGs) 
BMAseq.unique.cDEGs.8809678 <- sub("\\..*", "", BMAseq.unique.cDEGs) 

# GO over-representation analysis
BMAseq.go.ora <-
  enrichGO(
    gene          = BMAseq.unique.cDEGs.8809678,
    # universe      = BMAseq.cDEGs.8809678,
    OrgDb         = "org.Hs.eg.db",
    keyType       = "ENSEMBL",
    ont           = "BP",
    pAdjustMethod = "BH",
    pvalueCutoff  = 0.05,
    qvalueCutoff  = 0.2,
    readable      = TRUE
  )
BMAseq.go.ora@result |> View()  
BMAseq.go.ora@result[order(BMAseq.go.ora@result$Count, decreasing = T), ] |> View()
BMAseq.go.ora@result[order(BMAseq.go.ora@result$Count, decreasing = T), "Description"][1:10]
# [1] "small molecule catabolic process"               
# [2] "amino acid metabolic process"                   
# [3] "organic acid catabolic process"                 
# [4] "carboxylic acid catabolic process"              
# [5] "axonogenesis"                                   
# [6] "purine nucleotide biosynthetic process"         
# [7] "purine-containing compound biosynthetic process"
# [8] "nucleotide biosynthetic process"                
# [9] "nucleoside phosphate biosynthetic process"      
# [10] "lipid catabolic process"          
BMAseq.go.ora@result[order(BMAseq.go.ora@result$Count, decreasing = T), "geneID"][1:10]
# [1] "MTHFS/APOBEC3H/ECHS1/RIDA/PPARA/APOBEC3G/DBT/ALDH5A1/AUH/DECR1/GCSH"
# [2] "MTHFS/ECHS1/RIDA/NARS1/MARS2/DBT/ALDH5A1/APIP/AUH/GCSH"             
# [3] "MTHFS/ECHS1/RIDA/PPARA/DBT/ALDH5A1/AUH/DECR1/GCSH"                  
# [4] "MTHFS/ECHS1/RIDA/PPARA/DBT/ALDH5A1/AUH/DECR1/GCSH"                  
# [5] "SPP1/PLPPR4/AFG3L2/CDH11/FN1/GLI2/NRDC/EFNB1/NTRK2"                 
# [6] "AMPD2/SLC25A1/PDHX/PRPS2/PPARA/ATP6/PARP10/ND2"                     
# [7] "AMPD2/SLC25A1/PDHX/PRPS2/PPARA/ATP6/PARP10/ND2"                     
# [8] "AMPD2/SLC25A1/PDHX/PRPS2/PPARA/ATP6/PARP10/ND2"                     
# [9] "AMPD2/SLC25A1/PDHX/PRPS2/PPARA/ATP6/PARP10/ND2"                     
# [10] "SPP1/ECHS1/PLAAT4/GDPD3/RAB7A/PPARA/AUH/DECR1"  

x <- strsplit("MTHFS/APOBEC3H/ECHS1/RIDA/PPARA/APOBEC3G/DBT/ALDH5A1/AUH/DECR1/GCSH", split = "/") |> unlist()
suppressMessages({
  ids <- bitr(x, fromType = "SYMBOL", 
              toType = c("ENSEMBL"), 
              OrgDb = "org.Hs.eg.db")
})
ids
# SYMBOL         ENSEMBL
# 1     MTHFS ENSG00000136371
# 2  APOBEC3H ENSG00000100298
# 3     ECHS1 ENSG00000127884
# 4      RIDA ENSG00000132541
# 5     PPARA ENSG00000186951
# 6  APOBEC3G ENSG00000239713
# 7       DBT ENSG00000137992
# 8   ALDH5A1 ENSG00000112294
# 9       AUH ENSG00000148090
# 10    DECR1 ENSG00000104325
# 11     GCSH ENSG00000140905


##------DESeq2_UVM-----
method.name = "DESeq2"
model.type = "Uni"
model.type.2 = "UVM"
file.name <- sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sModel/Trim/%s%s%s.RData", 
                     model.type, method.name, model.type, seed.i) 
data.env <- local({load(file.name); environment()})
DESeq2_UVM.cDEGs <- intersect(data.env$DESeq2.eFDR.GeneName.train[[var.name]][1:threshold], 
                              data.env$DESeq2.eFDR.GeneName.test[[var.name]][1:threshold])
DESeq2_UVM.unique.cDEGs.all.seed <- readRDS(sprintf("../ApplicationResult/UniqueGene/TMM_Top5000/%s_%s.uniq.cDEGs.%s.all.seed.RDS",
                                                    method.name, model.type.2, var.name))
DESeq2_UVM.unique.cDEGs <- DESeq2_UVM.unique.cDEGs.all.seed[[as.character(seed.i)]]

# Delete any character after the comma
DESeq2_UVM.cDEGs.8809678 <- sub("\\..*", "", DESeq2_UVM.cDEGs) 
DESeq2_UVM.unique.cDEGs.8809678 <- sub("\\..*", "", DESeq2_UVM.unique.cDEGs) 

# GO over-representation analysis
DESeq2_UVM.go.ora <-
  enrichGO(
    gene          = DESeq2_UVM.unique.cDEGs.8809678,
    # universe      = DESeq2_UVM.cDEGs.8809678,
    OrgDb         = "org.Hs.eg.db",
    keyType       = "ENSEMBL",
    ont           = "BP",
    pAdjustMethod = "BH",
    pvalueCutoff  = 0.05,
    qvalueCutoff  = 0.2,
    readable      = TRUE
  )
DESeq2_UVM.go.ora@result |> View()  
# DESeq2_UVM.go.ora.dt <- as.data.table(DESeq2_UVM.go.ora)
# clusterProfiler::dotplot(object = DESeq2_UVM.go.ora, 
#                          color = "pvalue",
#                          showCategory = 10) +
#   theme(axis.text.y = element_text(size = 7))


DESeq2_UVM.go.ora@result[order(DESeq2_UVM.go.ora@result$Count, decreasing = T), ] |> View()
DESeq2_UVM.go.ora@result[order(DESeq2_UVM.go.ora@result$Count, decreasing = T), "Description"][1:10]
# [1] "epidermis development"                          
# [2] "keratinocyte differentiation"                   
# [3] "epidermal cell differentiation"                 
# [4] "skin development"                               
# [5] "intermediate filament organization"             
# [6] "intermediate filament cytoskeleton organization"
# [7] "intermediate filament-based process"            
# [8] "keratinization"                                 
# [9] "positive regulation of mitotic cell cycle"      
# [10] "connective tissue development"      
DESeq2_UVM.go.ora@result[order(DESeq2_UVM.go.ora@result$Count, decreasing = T), "geneID"][1:10]
# [1] "KRT6B/KRT6A/KRT14/KRT5/COL17A1/KRT77/ZBED2" "KRT6B/KRT6A/KRT14/KRT5/KRT77/ZBED2"        
# [3] "KRT6B/KRT6A/KRT14/KRT5/KRT77/ZBED2"         "KRT6B/KRT6A/KRT14/KRT5/KRT77/ZBED2"        
# [5] "KRT6B/KRT6A/KRT14/KRT5/KRT77"               "KRT6B/KRT6A/KRT14/KRT5/KRT77"              
# [7] "KRT6B/KRT6A/KRT14/KRT5/KRT77"               "KRT6B/KRT6A/KRT5/KRT77"                    
# [9] "CDC25C/UBE2C/CDC7/FOXA1"                    "BARX2/HMGCS2/FOXA1/OSR2" 
x <- strsplit("KRT6B/KRT6A/KRT14/KRT5/COL17A1/KRT77/ZBED2", split = "/") |> unlist()
suppressMessages({
  ids <- bitr(x, fromType = "SYMBOL", 
              toType = c("ENSEMBL"), 
              OrgDb = "org.Hs.eg.db")
})
ids
# SYMBOL         ENSEMBL
# 1   KRT6B ENSG00000185479
# 2   KRT6A ENSG00000205420
# 3   KRT14 ENSG00000186847
# 4    KRT5 ENSG00000186081
# 5 COL17A1 ENSG00000065618
# 6   KRT77 ENSG00000189182
# 7   ZBED2 ENSG00000177494


##------DESeq2_MVM-----
method.name = "DESeq2"
model.type = "Multi"
model.type.2 = "MVM"
file.name <- sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sModel/Trim/%s%s%s.RData", 
                     model.type, method.name, model.type, seed.i) 
data.env <- local({load(file.name); environment()})
DESeq2_MVM.cDEGs <- intersect(data.env$DESeq2.eFDR.GeneName.train[[var.name]][1:threshold], 
                              data.env$DESeq2.eFDR.GeneName.test[[var.name]][1:threshold])
DESeq2_MVM.unique.cDEGs.all.seed <- readRDS(sprintf("../ApplicationResult/UniqueGene/TMM_Top5000/%s_%s.uniq.cDEGs.%s.all.seed.RDS",
                                                    method.name, model.type.2, var.name))
DESeq2_MVM.unique.cDEGs <- DESeq2_MVM.unique.cDEGs.all.seed[[as.character(seed.i)]]

# Delete any character after the comma
DESeq2_MVM.cDEGs.8809678 <- sub("\\..*", "", DESeq2_MVM.cDEGs) 
DESeq2_MVM.unique.cDEGs.8809678 <- sub("\\..*", "", DESeq2_MVM.unique.cDEGs) 

# GO over-representation analysis
DESeq2_MVM.go.ora <-
  enrichGO(
    gene          = DESeq2_MVM.unique.cDEGs.8809678,
    # universe      = DESeq2_MVM.cDEGs.8809678,
    OrgDb         = "org.Hs.eg.db",
    keyType       = "ENSEMBL",
    ont           = "BP",
    pAdjustMethod = "BH",
    pvalueCutoff  = 0.05,
    qvalueCutoff  = 0.2,
    readable      = TRUE
  )
DESeq2_MVM.go.ora@result |> View()  

DESeq2_MVM.go.ora@result[order(DESeq2_MVM.go.ora@result$Count, decreasing = T), ] |> View()
DESeq2_MVM.go.ora@result[order(DESeq2_MVM.go.ora@result$Count, decreasing = T), "Description"][1:10]
# [1] "response to alcohol"                    "cellular response to alcohol"          
# [3] "cellular response to salt"              "response to salt"                      
# [5] "cAMP biosynthetic process"              "meiotic sister chromatid segregation"  
# [7] "meiotic sister chromatid cohesion"      "meiosis II"                            
# [9] "meiosis II cell cycle process"          "cyclic nucleotide biosynthetic process"
DESeq2_MVM.go.ora@result[order(DESeq2_MVM.go.ora@result$Count, decreasing = T), "geneID"][1:10]
# [1] "ADCY8/CES1/CSF3/ADCY7" "ADCY8/CES1/ADCY7"      "ADCY8/ACHE/ADCY7"     
# [4] "ADCY8/ACHE/ADCY7"      "ADCY8/ADCY7"           "BUB1/HORMAD2"         
# [7] "BUB1/HORMAD2"          "BUB1/HORMAD2"          "BUB1/HORMAD2"         
# [10] "ADCY8/ADCY7"          
x <- strsplit("ADCY8/CES1/CSF3/ADCY7", split = "/") |> unlist()
suppressMessages({
  ids <- bitr(x, fromType = "SYMBOL", 
              toType = c("ENSEMBL"), 
              OrgDb = "org.Hs.eg.db")
})
ids
# SYMBOL         ENSEMBL
# 1  ADCY8 ENSG00000155897
# 2   CES1 ENSG00000198848
# 3   CES1 ENSG00000262243
# 4   CSF3 ENSG00000108342
# 5  ADCY7 ENSG00000121281


##------edgeR_UVM-----
method.name = "edgeR"
model.type = "Uni"
model.type.2 = "UVM"
file.name <- sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sModel/Trim/%s%s%s.RData", 
                     model.type, method.name, model.type, seed.i) 
data.env <- local({load(file.name); environment()})
edgeR_UVM.cDEGs <- intersect(data.env$edgeR.eFDR.GeneName.train[[var.name]][1:threshold], 
                              data.env$edgeR.eFDR.GeneName.test[[var.name]][1:threshold])
edgeR_UVM.unique.cDEGs.all.seed <- readRDS(sprintf("../ApplicationResult/UniqueGene/TMM_Top5000/%s_%s.uniq.cDEGs.%s.all.seed.RDS",
                                                    method.name, model.type.2, var.name))
edgeR_UVM.unique.cDEGs <- edgeR_UVM.unique.cDEGs.all.seed[[as.character(seed.i)]]

# Delete any character after the comma
edgeR_UVM.cDEGs.8809678 <- sub("\\..*", "", edgeR_UVM.cDEGs) 
edgeR_UVM.unique.cDEGs.8809678 <- sub("\\..*", "", edgeR_UVM.unique.cDEGs) 

# GO over-representation analysis
edgeR_UVM.go.ora <-
  enrichGO(
    gene          = edgeR_UVM.unique.cDEGs.8809678,
    # universe      = edgeR_UVM.cDEGs.8809678,
    OrgDb         = "org.Hs.eg.db",
    keyType       = "ENSEMBL",
    ont           = "BP",
    pAdjustMethod = "BH",
    pvalueCutoff  = 0.05,
    qvalueCutoff  = 0.2,
    readable      = TRUE
  )
edgeR_UVM.go.ora@result |> View()  

edgeR_UVM.go.ora@result[order(edgeR_UVM.go.ora@result$Count, decreasing = T), ] |> View()
edgeR_UVM.go.ora@result[order(edgeR_UVM.go.ora@result$Count, decreasing = T), "Description"][1:10]
# [1] "epidermis development"              "skin development"                  
# [3] "hormone secretion"                  "hormone transport"                 
# [5] "lipid catabolic process"            "lipid transport"                   
# [7] "skin epidermis development"         "regulation of hormone secretion"   
# [9] "organic hydroxy compound transport" "amide transport"     
edgeR_UVM.go.ora@result[order(edgeR_UVM.go.ora@result$Count, decreasing = T), "geneID"][1:10]
# [1] "KRT25/KRT71/SPRR1A/WNT10A/FLG2/FOXQ1/KRTDAP/ALOXE3/LIPN/KLK7/ABCA12/C1orf68/LCE1A/LCE1C"
# [2] "KRT25/KRT71/SPRR1A/WNT10A/FLG2/FOXQ1/ALOXE3/LIPN/ABCA12/LCE1A/LCE1C"                    
# [3] "PTPRN/CHGA/POMC/CELA2A/PLA2G3/ABCA12/FAM3D/F2"                                          
# [4] "PTPRN/CHGA/POMC/CELA2A/PLA2G3/ABCA12/FAM3D/F2"                                          
# [5] "CYP2W1/CEL/PNLIP/CLPS/APOC3/PLA2G1B/LIPH/LIPN"                                          
# [6] "CEL/POMC/PNLIP/APOC3/PLA2G1B/PLA2G3/ABCA12/STAR"                                        
# [7] "KRT25/KRT71/WNT10A/FLG2/FOXQ1/ALOXE3/ABCA12"                                            
# [8] "CHGA/POMC/CELA2A/PLA2G3/ABCA12/FAM3D/F2"                                                
# [9] "CHGA/CEL/POMC/PNLIP/APOC3/ABCA12/STAR"                                                  
# [10] "PTPRN/CHGA/CELA2A/KMO/ABCA12/FAM3D/F2"       
x <- strsplit("KRT25/KRT71/SPRR1A/WNT10A/FLG2/FOXQ1/KRTDAP/ALOXE3/LIPN/KLK7/ABCA12/C1orf68/LCE1A/LCE1C", split = "/") |> unlist()
suppressMessages({
  ids <- bitr(x, fromType = "SYMBOL", 
              toType = c("ENSEMBL"), 
              OrgDb = "org.Hs.eg.db")
})
ids
# SYMBOL         ENSEMBL
# 1    KRT25 ENSG00000204897
# 2    KRT71 ENSG00000139648
# 3   SPRR1A ENSG00000169474
# 4   WNT10A ENSG00000135925
# 5     FLG2 ENSG00000143520
# 6    FOXQ1 ENSG00000164379
# 7   KRTDAP ENSG00000188508
# 8   ALOXE3 ENSG00000179148
# 9     LIPN ENSG00000204020
# 10    KLK7 ENSG00000169035
# 11  ABCA12 ENSG00000144452
# 12 C1orf68 ENSG00000198854
# 13   LCE1A ENSG00000186844
# 14   LCE1C ENSG00000197084


##------edgeR_MVM-----
method.name = "edgeR"
model.type = "Multi"
model.type.2 = "MVM"
file.name <- sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sModel/Trim/%s%s%s.RData", 
                     model.type, method.name, model.type, seed.i) 
data.env <- local({load(file.name); environment()})
edgeR_MVM.cDEGs <- intersect(data.env$edgeR.eFDR.GeneName.train[[var.name]][1:threshold], 
                              data.env$edgeR.eFDR.GeneName.test[[var.name]][1:threshold])
edgeR_MVM.unique.cDEGs.all.seed <- readRDS(sprintf("../ApplicationResult/UniqueGene/TMM_Top5000/%s_%s.uniq.cDEGs.%s.all.seed.RDS",
                                                    method.name, model.type.2, var.name))
edgeR_MVM.unique.cDEGs <- edgeR_MVM.unique.cDEGs.all.seed[[as.character(seed.i)]]

# Delete any character after the comma
edgeR_MVM.cDEGs.8809678 <- sub("\\..*", "", edgeR_MVM.cDEGs) 
edgeR_MVM.unique.cDEGs.8809678 <- sub("\\..*", "", edgeR_MVM.unique.cDEGs) 

# GO over-representation analysis
edgeR_MVM.go.ora <-
  enrichGO(
    gene          = edgeR_MVM.unique.cDEGs.8809678,
    # universe      = edgeR_MVM.cDEGs.8809678,
    OrgDb         = "org.Hs.eg.db",
    keyType       = "ENSEMBL",
    ont           = "BP",
    pAdjustMethod = "BH",
    pvalueCutoff  = 0.05,
    qvalueCutoff  = 0.2,
    readable      = TRUE
  )
edgeR_MVM.go.ora@result |> View()  

edgeR_MVM.go.ora@result[order(edgeR_MVM.go.ora@result$Count, decreasing = T), ] |> View()
edgeR_MVM.go.ora@result[order(edgeR_MVM.go.ora@result$Count, decreasing = T), "Description"][1:10]
# [1] "maturation of SSU-rRNA"                        
# [2] "ribosomal small subunit biogenesis"            
# [3] "rRNA processing"                               
# [4] "glial cell differentiation"                    
# [5] "ribonucleoprotein complex assembly"            
# [6] "ribonucleoprotein complex subunit organization"
# [7] "rRNA metabolic process"                        
# [8] "positive regulation of growth"                 
# [9] "ribosome biogenesis"                           
# [10] "gliogenesis"       
edgeR_MVM.go.ora@result[order(edgeR_MVM.go.ora@result$Count, decreasing = T), "geneID"][1:10]
# [1] "RPS16/SNU13"  "RPS16/SNU13"  "RPS16/SNU13"  "POU3F2/FGF5" 
# [5] "SNU13/SF3B6"  "SNU13/SF3B6"  "RPS16/SNU13"  "KRT17/POU3F2"
# [9] "RPS16/SNU13"  "POU3F2/FGF5"          
x <- strsplit("RPS16/SNU13", split = "/") |> unlist()
suppressMessages({
  ids <- bitr(x, fromType = "SYMBOL", 
              toType = c("ENSEMBL"), 
              OrgDb = "org.Hs.eg.db")
})
ids
# SYMBOL         ENSEMBL
# 1  RPS16 ENSG00000105193
# 2  SNU13 ENSG00000100138


##------eBayes_UVM-----
method.name = "eBayes"
model.type = "Uni"
model.type.2 = "UVM"
file.name <- sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sModel/Trim/%s%s%s.RData", 
                     model.type, method.name, model.type, seed.i) 
data.env <- local({load(file.name); environment()})
eBayes_UVM.cDEGs <- intersect(names(data.env$eBayes.eFDR.train2[[var.name]][1:threshold]), 
                              names(data.env$eBayes.eFDR.test2[[var.name]][1:threshold]))
eBayes_UVM.unique.cDEGs.all.seed <- readRDS(sprintf("../ApplicationResult/UniqueGene/TMM_Top5000/%s_%s.uniq.cDEGs.%s.all.seed.RDS",
                                                   method.name, model.type.2, var.name))
eBayes_UVM.unique.cDEGs <- eBayes_UVM.unique.cDEGs.all.seed[[as.character(seed.i)]]

# Delete any character after the comma
eBayes_UVM.cDEGs.8809678 <- sub("\\..*", "", eBayes_UVM.cDEGs) 
eBayes_UVM.unique.cDEGs.8809678 <- sub("\\..*", "", eBayes_UVM.unique.cDEGs) 

# GO over-representation analysis
eBayes_UVM.go.ora <-
  enrichGO(
    gene          = eBayes_UVM.unique.cDEGs.8809678,
    # universe      = eBayes_UVM.cDEGs.8809678,
    OrgDb         = "org.Hs.eg.db",
    keyType       = "ENSEMBL",
    ont           = "BP",
    pAdjustMethod = "BH",
    pvalueCutoff  = 0.05,
    qvalueCutoff  = 0.2,
    readable      = TRUE
  )
eBayes_UVM.go.ora@result |> View()  

eBayes_UVM.go.ora@result[order(eBayes_UVM.go.ora@result$Count, decreasing = T), ] |> View()
eBayes_UVM.go.ora@result[order(eBayes_UVM.go.ora@result$Count, decreasing = T), "Description"][1:10]
# [1] "regulation of fear response"                
# [2] "response to redox state"                    
# [3] "positive regulation of behavior"            
# [4] "behavioral fear response"                   
# [5] "behavioral defense response"                
# [6] "fear response"                              
# [7] "circadian regulation of gene expression"    
# [8] "regulation of behavior"                     
# [9] "multicellular organismal response to stress"
# [10] "positive regulation of DNA repair"    
eBayes_UVM.go.ora@result[order(eBayes_UVM.go.ora@result$Count, decreasing = T), "geneID"][1:10]
# [1] "NPAS2" "NPAS2" "NPAS2" "NPAS2" "NPAS2" "NPAS2" "NPAS2" "NPAS2" "NPAS2"
# [10] "NPAS2"       
x <- strsplit("NPAS2", split = "/") |> unlist()
suppressMessages({
  ids <- bitr(x, fromType = "SYMBOL", 
              toType = c("ENSEMBL"), 
              OrgDb = "org.Hs.eg.db")
})
ids
# SYMBOL         ENSEMBL
# 1  NPAS2 ENSG00000170485


##------eBayes_MVM-----
method.name = "eBayes"
model.type = "Multi"
model.type.2 = "MVM"
file.name <- sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sModel/Trim/%s%s%s.RData", 
                     model.type, method.name, model.type, seed.i) 
data.env <- local({load(file.name); environment()})
eBayes_MVM.cDEGs <- intersect(names(data.env$eBayes.eFDR.train2[[var.name]][1:threshold]), 
                              names(data.env$eBayes.eFDR.test2[[var.name]][1:threshold]))
eBayes_MVM.unique.cDEGs.all.seed <- readRDS(sprintf("../ApplicationResult/UniqueGene/TMM_Top5000/%s_%s.uniq.cDEGs.%s.all.seed.RDS",
                                                    method.name, model.type.2, var.name))
eBayes_MVM.unique.cDEGs <- eBayes_MVM.unique.cDEGs.all.seed[[as.character(seed.i)]]

# Delete any character after the comma
eBayes_MVM.cDEGs.8809678 <- sub("\\..*", "", eBayes_MVM.cDEGs) 
eBayes_MVM.unique.cDEGs.8809678 <- sub("\\..*", "", eBayes_MVM.unique.cDEGs) 

# GO over-representation analysis
eBayes_MVM.go.ora <-
  enrichGO(
    gene          = eBayes_MVM.unique.cDEGs.8809678,
    # universe      = eBayes_MVM.cDEGs.8809678,
    OrgDb         = "org.Hs.eg.db",
    keyType       = "ENSEMBL",
    ont           = "BP",
    pAdjustMethod = "BH",
    pvalueCutoff  = 0.05,
    qvalueCutoff  = 0.2,
    readable      = TRUE
  )
eBayes_MVM.go.ora@result |> View()  

eBayes_MVM.go.ora@result[order(eBayes_MVM.go.ora@result$Count, decreasing = T), ] |> View()
eBayes_MVM.go.ora@result[order(eBayes_MVM.go.ora@result$Count, decreasing = T), "Description"][1:10]
# [1] "positive regulation of protein kinase activity"                                  
# [2] "positive regulation of kinase activity"                                          
# [3] "positive regulation of cell development"                                         
# [4] "positive regulation of synapse maturation"                                       
# [5] "regulation of T-helper 2 cell differentiation"                                   
# [6] "motor neuron migration"                                                          
# [7] "antigen processing and presentation of exogenous peptide antigen via MHC class I"
# [8] "interneuron migration"                                                           
# [9] "CD40 signaling pathway"                                                          
# [10] "neurotransmitter-gated ion channel clustering"   
eBayes_MVM.go.ora@result[order(eBayes_MVM.go.ora@result$Count, decreasing = T), "geneID"][1:10]
# [1] "CD86/RELN" "CD86/RELN" "CD86/RELN" "RELN"      "CD86"      "RELN"     
# [7] "IFI30"     "RELN"      "CD86"      "RELN" 
x <- strsplit("CD86/RELN", split = "/") |> unlist()
suppressMessages({
  ids <- bitr(x, fromType = "SYMBOL", 
              toType = c("ENSEMBL"), 
              OrgDb = "org.Hs.eg.db")
})
ids
# SYMBOL         ENSEMBL
# 1   CD86 ENSG00000114013
# 2   RELN ENSG00000189056


##------voom.limma_UVM-----
method.name = "VoomLimma"
method.name.2 = "voom_limma"
model.type = "Uni"
model.type.2 = "UVM"
file.name <- sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sModel/Trim/%s%s%s.RData", 
                     model.type, method.name, model.type, seed.i) 
data.env <- local({load(file.name); environment()})
voom.limma_UVM.cDEGs <- intersect(names(data.env$voom.eFDR.train2[[var.name]][1:threshold]), 
                                  names(data.env$voom.eFDR.test2[[var.name]][1:threshold]))
voom.limma_UVM.unique.cDEGs.all.seed <- readRDS(sprintf("../ApplicationResult/UniqueGene/TMM_Top5000/%s_%s.uniq.cDEGs.%s.all.seed.RDS",
                                                        method.name.2, model.type.2, var.name))
voom.limma_UVM.unique.cDEGs <- voom.limma_UVM.unique.cDEGs.all.seed[[as.character(seed.i)]]

# Delete any character after the comma
voom.limma_UVM.cDEGs.8809678 <- sub("\\..*", "", voom.limma_UVM.cDEGs) 
voom.limma_UVM.unique.cDEGs.8809678 <- sub("\\..*", "", voom.limma_UVM.unique.cDEGs) 

# GO over-representation analysis
voom.limma_UVM.go.ora <-
  enrichGO(
    gene          = voom.limma_UVM.unique.cDEGs.8809678,
    # universe      = voom.limma_UVM.cDEGs.8809678,
    OrgDb         = "org.Hs.eg.db",
    keyType       = "ENSEMBL",
    ont           = "BP",
    pAdjustMethod = "BH",
    pvalueCutoff  = 0.05,
    qvalueCutoff  = 0.2,
    readable      = TRUE
  )
voom.limma_UVM.go.ora@result |> View()  

voom.limma_UVM.go.ora@result[order(voom.limma_UVM.go.ora@result$Count, decreasing = T), ] |> View()
voom.limma_UVM.go.ora@result[order(voom.limma_UVM.go.ora@result$Count, decreasing = T), "Description"][1:10]
# [1] "establishment of protein localization to organelle"             
# [2] "protein insertion into mitochondrial inner membrane"            
# [3] "3'-UTR-mediated mRNA destabilization"                           
# [4] "negative regulation by host of viral transcription"             
# [5] "3'-UTR-mediated mRNA stabilization"                             
# [6] "protein insertion into mitochondrial membrane"                  
# [7] "positive regulation of protein import into nucleus"             
# [8] "establishment of protein localization to mitochondrial membrane"
# [9] "inner mitochondrial membrane organization"                      
# [10] "nuclear membrane organization"     
voom.limma_UVM.go.ora@result[order(voom.limma_UVM.go.ora@result$Count, decreasing = T), "geneID"][1:10]
# [1] "TARDBP/TIMM8A" "TIMM8A"        "TARDBP"        "TARDBP"       
# [5] "TARDBP"        "TIMM8A"        "TARDBP"        "TIMM8A"       
# [9] "TIMM8A"        "TARDBP"    
x <- strsplit("TARDBP/TIMM8A", split = "/") |> unlist()
suppressMessages({
  ids <- bitr(x, fromType = "SYMBOL", 
              toType = c("ENSEMBL"), 
              OrgDb = "org.Hs.eg.db")
})
ids
# SYMBOL         ENSEMBL
# 1 TARDBP ENSG00000120948
# 2 TIMM8A ENSG00000126953


##------voom.limma_MVM-----
method.name = "VoomLimma"
method.name.2 = "voom_limma"
model.type = "Multi"
model.type.2 = "MVM"
file.name <- sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sModel/Trim/%s%s%s.RData", 
                     model.type, method.name, model.type, seed.i) 
data.env <- local({load(file.name); environment()})
voom.limma_MVM.cDEGs <- intersect(names(data.env$voom.eFDR.train2[[var.name]][1:threshold]), 
                                  names(data.env$voom.eFDR.test2[[var.name]][1:threshold]))
voom.limma_MVM.unique.cDEGs.all.seed <- readRDS(sprintf("../ApplicationResult/UniqueGene/TMM_Top5000/%s_%s.uniq.cDEGs.%s.all.seed.RDS",
                                                        method.name.2, model.type.2, var.name))
voom.limma_MVM.unique.cDEGs <- voom.limma_MVM.unique.cDEGs.all.seed[[as.character(seed.i)]]

# Delete any character after the comma
voom.limma_MVM.cDEGs.8809678 <- sub("\\..*", "", voom.limma_MVM.cDEGs) 
voom.limma_MVM.unique.cDEGs.8809678 <- sub("\\..*", "", voom.limma_MVM.unique.cDEGs) 

# GO over-representation analysis
voom.limma_MVM.go.ora <-
  enrichGO(
    gene          = voom.limma_MVM.unique.cDEGs.8809678,
    # universe      = voom.limma_MVM.cDEGs.8809678,
    OrgDb         = "org.Hs.eg.db",
    keyType       = "ENSEMBL",
    ont           = "BP",
    pAdjustMethod = "BH",
    pvalueCutoff  = 0.05,
    qvalueCutoff  = 0.2,
    readable      = TRUE
  )
voom.limma_MVM.go.ora@result |> View()  

voom.limma_MVM.go.ora@result[order(voom.limma_MVM.go.ora@result$Count, decreasing = T), ] |> View()
voom.limma_MVM.go.ora@result[order(voom.limma_MVM.go.ora@result$Count, decreasing = T), "Description"][1:10]
# [1] "ribonucleoprotein complex assembly"            
# [2] "ribonucleoprotein complex subunit organization"
# [3] "pre-mRNA cleavage required for polyadenylation"
# [4] "mRNA cleavage involved in mRNA processing"     
# [5] "regulation of mRNA polyadenylation"            
# [6] "protein heterotetramerization"                 
# [7] "toll-like receptor 9 signaling pathway"        
# [8] "U2-type prespliceosome assembly"               
# [9] "mRNA cleavage"                                 
# [10] "regulation of mRNA 3'-end processing"   
voom.limma_MVM.go.ora@result[order(voom.limma_MVM.go.ora@result$Count, decreasing = T), "geneID"][1:10]
# [1] "CPSF7/BUD13" "CPSF7/BUD13" "CPSF7"       "CPSF7"       "CPSF7"      
# [6] "CPSF7"       "EPG5"        "BUD13"       "CPSF7"       "CPSF7"  
x <- strsplit("CPSF7/BUD13", split = "/") |> unlist()
suppressMessages({
  ids <- bitr(x, fromType = "SYMBOL", 
              toType = c("ENSEMBL"), 
              OrgDb = "org.Hs.eg.db")
})
ids
# SYMBOL         ENSEMBL
# 1  CPSF7 ENSG00000149532
# 2  BUD13 ENSG00000137656
```


```{r dotplot}
make_dotplot <- function(data.set = ora.obj.ordered, 
                         top.n = 20, 
                         method.name = "BMAseq", 
                         color.lower = NULL, 
                         color.upper = NULL, 
                         color.type = c("pvalue", "qscore")) {
  if (nrow(data.set) < top.n) {
    data.set <- data.set
  } else {
    data.set <- data.set[1:top.n, ]
  }
  term.ordered <- rev(data.set$Description)
  color.type <- match.arg(color.type, c("pvalue", "qscore"))
  min.count <- min(data.set$Count)
  max.count <- max(data.set$Count)
  
  if (is.null(color.lower)) {
    color.lower <- 0
  }
  if (is.null(color.upper)) {
    if (color.type == "pvalue") {
      color.upper <- range(data.set$pvalue)[2] |> round(2)
    } else if (color.type == "qscore") {
      data.set <- mutate(data.set, qscore = -log(p.adjust, base = 10))
      color.upper <- range(data.set$qscore)[2] |> round(2)
    }
  } 
  
  p <- ggplot(data = data.set, 
              mapping = aes(x = Count, y = Description)) + 
    geom_point(aes(size = Count, color = get(color.type))) +
    scale_colour_gradient(limits = c(color.lower, color.upper), low = "blue", high = "red") + 
    scale_size_continuous(breaks = seq(min.count, max.count, 1)) + # Count, not continuous scale
    scale_x_continuous(breaks = seq(min.count, max.count, 1)) + # Count, not continuous scale
    scale_y_discrete(limits = term.ordered, labels = scales::label_wrap(70)) + 
    labs(y = NULL,
         color = ifelse(color.type == "pvalue", "p-value", "q score")) +
    ggtitle(paste0("Unique cDEGs of ", var.name)) + 
    theme_bw(base_size = 14, base_family = "Arial") +
    theme(plot.title = element_text(hjust = 0.5, face = "bold"),
          axis.text.y = element_text(size = 13, color = "black")) + 
    guides(size = guide_legend(order = 1)) # Order the discrete legend
  
  return(p)
}


# Prepare dotplot data
# var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC") 
date.analysis <- format(Sys.Date(), "%Y%b%d")
seed.i <- 8809678
threshold <- 5000
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC") 
var.vec <- "BMI"
method.type.vec <- c("BMAseq", 
                     "DESeq2_UVM",
                     "edgeR_UVM",
                     "eBayes_UVM",
                     "voom.limma_UVM",
                     "DESeq2_MVM",
                     "edgeR_MVM",
                     "eBayes_MVM",
                     "voom.limma_MVM")

for (var.name in var.vec) {
  for (method.name in method.type.vec) {
    if (grepl("voom.limma", method.name)) {
      method.name <- paste0("voom_limma", sub(".*_", "_", method.name))
    }
    unique.cDEGs.all.seed <- readRDS(sprintf("../ApplicationResult/UniqueGene/TMM_Top5000/%s.uniq.cDEGs.%s.all.seed.RDS",
                                             method.name, var.name))
    unique.cDEGs <- unique.cDEGs.all.seed[[as.character(seed.i)]]
    unique.cDEGs.8809678 <- sub("\\..*", "", unique.cDEGs)      # Delete any character after the comma

    ora.obj <-
      enrichGO(
        gene          = unique.cDEGs.8809678,
        OrgDb         = "org.Hs.eg.db",
        keyType       = "ENSEMBL",
        ont           = "BP",
        pAdjustMethod = "BH",
        pvalueCutoff  = 0.05,
        qvalueCutoff  = 0.2,
        readable      = T
      )
    
    ora.obj.df <- ora.obj@result
    ora.obj.ordered <- ora.obj.df[order(ora.obj.df$Count, decreasing = T), ]
    
    p <- make_dotplot(data.set = ora.obj.ordered, 
                      top.n = 20, 
                      method.name = method.i, 
                      color.lower = NULL, 
                      color.upper = NULL, 
                      color.type = "qscore")
    
    ggsave(filename = sprintf("../ApplicationResult/AddViz/DotPlot/%s_%s_%s_%s.eps", date.analysis, method.name, var.name, seed.i),
           plot = p, device = cairo_ps, dpi = 600, width = 10, height = 6, units = "in")
    
  }
}
```


Here, we show biological insights of replicable DEGs that are uniquely and consistently detected with BMAseq or univariable edgeR.
```{r dotplot_cnetplot}
##------BMAseq-----
# Show all cDEGs uniquely identified by BMAseq that appear in at least one seed
var.name <- "BMI"
threshold.i <- 5000
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
class_freq_all_seed_list <- vector(mode = "list", length = 10L)
names(class_freq_all_seed_list) <- seed.vec 
for (seed.i in seed.vec) {
  file.name <- paste0("../ApplicationData/derived/RandomSeed/HeatmapBoxplotData/", var.name, "_", threshold.i, "_", seed.i, ".RDS")
  class_freq_per_seed <- readRDS(file.name)
  class_freq_per_seed_BMAseq <- filter(class_freq_per_seed, Class == "100000000") |>
    dplyr::select(Class) |>
    mutate(Seed = seed.i) |>
    rownames_to_column("cDEG")
  class_freq_all_seed_list[[as.character(seed.i)]] <- class_freq_per_seed_BMAseq 
}

class_freq_all_seed_df <- do.call(rbind, class_freq_all_seed_list)

# Show all cDEGs uniquely identified by BMAseq that appear in at least 5 seeds
class_freq_all_seed_df_new <- summarize(.data = class_freq_all_seed_df, cDEG.freq = n(), .by = "cDEG")
class_freq_all_seed_df_select <- dplyr::filter(class_freq_all_seed_df_new, cDEG.freq >= 5)
unique_cDEG_all_seed <- sub("\\..*", "", class_freq_all_seed_df_select$cDEG)
unique_cDEG_all_seed_gene_symbol <- bitr(unique_cDEG_all_seed, fromType = "ENSEMBL", 
                                         toType = c("SYMBOL"), 
                                         OrgDb = "org.Hs.eg.db")

# Perform ORA analysis on cDEGs uniquely identified by BMAseq that appear in at least 5 seeds
ora.obj <-
  enrichGO(
    gene          = unique_cDEG_all_seed,
    OrgDb         = "org.Hs.eg.db",
    keyType       = "ENSEMBL",
    ont           = "BP",
    pAdjustMethod = "BH",
    pvalueCutoff  = 0.05,
    qvalueCutoff  = 0.2,
    readable      = T
  )

ora.obj.df <- ora.obj@result
ora.obj.ordered <- ora.obj.df[order(ora.obj.df$Count, decreasing = T), ]

p <- cnetplot(x = ora.obj,
         showCategory = ora.obj.ordered$Description[1:6],
         circular = T, 
         color.params = list(foldChange = NULL, edge = TRUE)) + 
  theme(legend.title = element_text(size = 14),
        legend.text = element_text(size = 12))

date.analysis <- format(Sys.Date(), "%Y%b%d")
ggsave(filename = sprintf("../ApplicationResult/AddViz/cnetplot/%s_%s_%s_%s.eps", date.analysis, "BMAseq", var.name, "all.seed"),
       plot = p, device = cairo_ps, dpi = 600, width = 11, height = 7, units = "in")

make_dotplot <- function(data.set = ora.obj.ordered, 
                         top.n = 20, 
                         method.name = "BMAseq", 
                         color.lower = NULL, 
                         color.upper = NULL, 
                         color.type = c("pvalue", "qscore")) {
  if (nrow(data.set) < top.n) {
    data.set <- data.set
  } else {
    data.set <- data.set[1:top.n, ]
  }
  term.ordered <- rev(data.set$Description)
  color.type <- match.arg(color.type, c("pvalue", "qscore"))
  min.count <- min(data.set$Count)
  max.count <- max(data.set$Count)
  
  if (is.null(color.lower)) {
    color.lower <- 0
  }
  if (is.null(color.upper)) {
    if (color.type == "pvalue") {
      color.upper <- range(data.set$pvalue)[2] |> round(2)
    } else if (color.type == "qscore") {
      data.set <- mutate(data.set, qscore = -log(pvalue, base = 10)) # Use the original p value to calculate the q score
      color.upper <- range(data.set$qscore)[2] |> round(2) + 0.5 # Increase the upper bound of q score
    }
  } 
  
  p <- ggplot(data = data.set, 
              mapping = aes(x = Count, y = Description)) + 
    geom_point(aes(size = Count, color = get(color.type))) +
    scale_colour_gradient(limits = c(color.lower, color.upper), low = "blue", high = "red") + 
    scale_size_continuous(breaks = seq(min.count, max.count, 1)) + # Count, not continuous scale
    scale_x_continuous(breaks = seq(min.count, max.count, 1)) + # Count, not continuous scale
    scale_y_discrete(limits = term.ordered, labels = scales::label_wrap(70)) + 
    labs(y = NULL,
         color = ifelse(color.type == "pvalue", "p-value", 
                        TeX("-$log_{10}$ (p-value)"))
         ) +
    ggtitle(paste0("Unique cDEGs of ", var.name)) + 
    theme_bw(base_size = 14, base_family = "Arial") +
    theme(plot.title = element_text(hjust = 0.5, face = "bold"),
          axis.text.y = element_text(size = 13, color = "black")) + 
    guides(size = guide_legend(order = 1)) # Order the discrete legend
  
  return(p)
}


p <- make_dotplot(data.set = ora.obj.ordered, 
                  top.n = 20, 
                  method.name = "BMAseq", 
                  color.lower = NULL, 
                  color.upper = NULL, 
                  color.type = "qscore")

date.analysis <- format(Sys.Date(), "%Y%b%d")
ggsave(filename = sprintf("../ApplicationResult/AddViz/DotPlot/%s_%s_%s_%s.eps", date.analysis, "BMAseq", var.name, "all.seed"),
       plot = p, device = cairo_ps, dpi = 600, width = 10, height = 6, units = "in")

##------edgeR_UVM-----
# Show all cDEGs uniquely identified by edgeR_UVM that appear in at least one seed
var.name <- "BMI"
threshold.i <- 5000
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
class_freq_all_seed_list <- vector(mode = "list", length = 10L)
names(class_freq_all_seed_list) <- seed.vec 
for (seed.i in seed.vec) {
  ## Import the matrix that records the class of each cDEGs per seed
  file.name <- paste0("../ApplicationData/derived/RandomSeed/HeatmapBoxplotData/", var.name, "_", threshold.i, "_", seed.i, ".RDS")
  class_freq_per_seed <- readRDS(file.name)
  class_freq_per_seed_BMAseq <- filter(class_freq_per_seed, Class == "000100000") |>
    dplyr::select(Class) |>
    mutate(Seed = seed.i) |>
    rownames_to_column("cDEG")
  class_freq_all_seed_list[[as.character(seed.i)]] <- class_freq_per_seed_BMAseq 
}

class_freq_all_seed_df <- do.call(rbind, class_freq_all_seed_list)

# Show all cDEGs uniquely identified by edgeR_UVM that appear in at least 5 seeds 
class_freq_all_seed_df_new <- summarize(.data = class_freq_all_seed_df, cDEG.freq = n(), .by = "cDEG")
class_freq_all_seed_df_select <- dplyr::filter(class_freq_all_seed_df_new, cDEG.freq >= 5)
unique_cDEG_all_seed <- sub("\\..*", "", class_freq_all_seed_df_select$cDEG)
unique_cDEG_all_seed_gene_symbol <- bitr(unique_cDEG_all_seed, fromType = "ENSEMBL", 
                                         toType = c("SYMBOL"), 
                                         OrgDb = "org.Hs.eg.db")

# Perform ORA analysis on cDEGs uniquely identified by edgeR_UVM that appear in at least 5 seeds
ora.obj <-
  enrichGO(
    gene          = unique_cDEG_all_seed,
    OrgDb         = "org.Hs.eg.db",
    keyType       = "ENSEMBL",
    ont           = "BP",
    pAdjustMethod = "BH",
    pvalueCutoff  = 0.05,
    qvalueCutoff  = 0.2,
    readable      = T
  )

ora.obj.df <- ora.obj@result
ora.obj.ordered <- ora.obj.df[order(ora.obj.df$Count, decreasing = T), ]

p <- cnetplot(x = ora.obj,
         showCategory = ora.obj.ordered$Description[1:6],
         circular = T, 
         color.params = list(foldChange = NULL, edge = TRUE)) + 
  theme(legend.title = element_text(size = 14),
        legend.text = element_text(size = 12))

date.analysis <- format(Sys.Date(), "%Y%b%d")
ggsave(filename = sprintf("../ApplicationResult/AddViz/cnetplot/%s_%s_%s_%s.eps", date.analysis, "edgeR_UVM", var.name, "all.seed"),
       plot = p, device = cairo_ps, dpi = 600, width = 11, height = 7, units = "in")

p <- make_dotplot(data.set = ora.obj.ordered, 
                  top.n = 20, 
                  method.name = "edgeR_UVM", 
                  color.lower = NULL, 
                  color.upper = NULL, 
                  color.type = "qscore")

date.analysis <- format(Sys.Date(), "%Y%b%d")
ggsave(filename = sprintf("../ApplicationResult/AddViz/DotPlot/%s_%s_%s_%s.eps", date.analysis, "edgeR_UVM", var.name, "all.seed"),
       plot = p, device = cairo_ps, dpi = 600, width = 10, height = 6, units = "in")
```


## Make boxplots of unique cDEGs in one seed 8809678
```{r MTHFS}
# Set the plot theme
theme_BMA <- function(base_size = 12,
                      base_family = "Arial") {
  theme_classic(base_size = base_size, base_family = base_family) +
    theme(
      legend.position = "none",
      panel.background = element_rect(fill = "white", colour = "white"),
      axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5),
      strip.background = element_rect(fill = "white", color = "black", linewidth = 0.6), # Customize the box border in strip
      strip.placement = "inside"
    )
}

seed.i <- c(8809678)
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC")
dat.expr.train <- readRDS(sprintf("../ApplicationData/derived/RandomSeed/dat.expr.train%s.RDS", seed.i))
dat.pheno.train <- readRDS(sprintf("../ApplicationData/derived/RandomSeed/dat.pheno.train%s.RDS", seed.i))
dat.expr.test <- readRDS(sprintf("../ApplicationData/derived/RandomSeed/dat.expr.test%s.RDS", seed.i))
dat.pheno.test <- readRDS(sprintf("../ApplicationData/derived/RandomSeed/dat.pheno.test%s.RDS", seed.i))

MTHFS.expr.train <- dat.expr.train[rownames(dat.expr.train) == "ENSG00000136371.5", ]
MTHFS.expr.train.t <- t(MTHFS.expr.train)
MTHFS.train <- data.frame(expr = MTHFS.expr.train.t[, 1],
                          BMI = dat.pheno.train$BMI, 
                          AGE = dat.pheno.train$AGE,
                          SEX = dat.pheno.train$SEX,
                          MHABNWBC = dat.pheno.train$MHABNWBC)
MTHFS.train$BMI_MHABNWBC <- paste(paste0("BMI", MTHFS.train$BMI), paste0("WBC", MTHFS.train$MHABNWBC), sep = " + ")
MTHFS.train$BMI_AGE_SEX_MHABNWBC <- paste(paste0("BMI", MTHFS.train$BMI), 
                                          paste0("AGE", MTHFS.train$AGE), 
                                          paste0("SEX", MTHFS.train$SEX), 
                                          paste0("WBC", MTHFS.train$MHABNWBC), sep = " + ")
MTHFS.train$BMI_AGE_MHABNWBC <- paste(paste0("BMI", MTHFS.train$BMI), 
                                      paste0("AGE", MTHFS.train$AGE), 
                                      paste0("WBC", MTHFS.train$MHABNWBC), sep = " + ")
MTHFS.expr.test <- dat.expr.test[rownames(dat.expr.test) == "ENSG00000136371.5", ]
MTHFS.expr.test.t <- t(MTHFS.expr.test)
MTHFS.test <- data.frame(expr = MTHFS.expr.test.t[, 1],
                         BMI = dat.pheno.test$BMI, 
                         AGE = dat.pheno.test$AGE,
                         SEX = dat.pheno.test$SEX,
                         MHABNWBC = dat.pheno.test$MHABNWBC)
MTHFS.test$BMI_MHABNWBC <- paste(paste0("BMI", MTHFS.test$BMI), paste0("WBC", MTHFS.test$MHABNWBC), sep = " + ")
MTHFS.test$BMI_AGE_SEX_MHABNWBC <- paste(paste0("BMI", MTHFS.test$BMI), 
                                         paste0("AGE", MTHFS.test$AGE), 
                                         paste0("SEX", MTHFS.test$SEX), 
                                         paste0("WBC", MTHFS.test$MHABNWBC), sep = " + ")
MTHFS.test$BMI_AGE_MHABNWBC <- paste(paste0("BMI", MTHFS.test$BMI), 
                                     paste0("AGE", MTHFS.test$AGE), 
                                     paste0("WBC", MTHFS.test$MHABNWBC), sep = " + ")

make_boxplot <- function(gene.symbol = "MTHFS", data.type = "train", var.x = "BMI", test.type = "wilcox.test") {
  p <- ggplot(data = get(paste0(gene.symbol, ".", data.type)),
         mapping = aes(x = get(var.x), y = expr)) +
    geom_boxplot(aes(group = get(var.x), color = get(var.x)),
                 outlier.shape = NA) + 
    geom_jitter(aes(color = get(var.x)), size = 0.8) + 
    stat_compare_means(method = test.type, 
                       label.x.npc = "middle",
                       # aes(label = ifelse(p < 1.e-3, "p < 0.001", sprintf("p = %4.3f", as.numeric(..p.format..))))
                       ) + 
    theme_BMA() + 
    scale_color_viridis(discrete = T) + 
    ylab(paste0(gene.symbol, " expression level")) + 
    xlab(var.x)
  return(p)
}
p1 <- make_boxplot(gene.symbol = "MTHFS", data.type = "train", var.x = "BMI", test.type = "wilcox.test")
p2 <- make_boxplot(gene.symbol = "MTHFS", data.type = "train", var.x = "BMI_MHABNWBC", test.type = "kruskal.test")
p3 <- make_boxplot(gene.symbol = "MTHFS", data.type = "train", var.x = "BMI_AGE_SEX_MHABNWBC", test.type = "kruskal.test")
p4 <- make_boxplot(gene.symbol = "MTHFS", data.type = "train", var.x = "BMI_AGE_MHABNWBC", test.type = "kruskal.test")
plot_grid(p1, p2, p3, p4, ncol = 4, labels = LETTERS[1:4])

p5 <- make_boxplot(gene.symbol = "MTHFS", data.type = "test", var.x = "BMI", test.type = "wilcox.test")
p6 <- make_boxplot(gene.symbol = "MTHFS", data.type = "test", var.x = "BMI_MHABNWBC", test.type = "kruskal.test")
p7 <- make_boxplot(gene.symbol = "MTHFS", data.type = "test", var.x = "BMI_AGE_SEX_MHABNWBC", test.type = "kruskal.test")
p8 <- make_boxplot(gene.symbol = "MTHFS", data.type = "test", var.x = "MHABNWBC", test.type = "kruskal.test")
plot_grid(p5, p6, p7, p8, ncol = 4, labels = LETTERS[1:4])
```


## Make boxplots of unique cDEGs across all seeds
### BMAseq
```{r IRF3}
##------Map gene symbols to ENSEMBL IDs------
ensembl.id <- bitr("IRF3", 
                   fromType = "SYMBOL", 
                   toType = "ENSEMBL", 
                   OrgDb = "org.Hs.eg.db")
ensembl.id
# SYMBOL         ENSEMBL
# 1   IRF3 ENSG00000126456

##------Check the most frequent BMAseq best model across 10 seeds------
best.model <- read_excel("../ApplicationData/derived/RandomSeed/BestModelMatrix/BMI_5000.xlsx")
# ~1+MHABNWBC

##------Use the whole data to make the boxplot------
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC")
dat.expr.train <- readRDS(sprintf("../ApplicationData/derived/RandomSeed/dat.expr.train%s.RDS", seed.i))
dat.pheno.train <- readRDS(sprintf("../ApplicationData/derived/RandomSeed/dat.pheno.train%s.RDS", seed.i))
dat.expr.test <- readRDS(sprintf("../ApplicationData/derived/RandomSeed/dat.expr.test%s.RDS", seed.i))
dat.pheno.test <- readRDS(sprintf("../ApplicationData/derived/RandomSeed/dat.pheno.test%s.RDS", seed.i))
dat.expr.all <- cbind(dat.expr.train, dat.expr.test)
dat.pheno.all <- rbind(dat.pheno.train, dat.pheno.test)
IRF3.expr.all <- dat.expr.all[rownames(dat.expr.all) == "ENSG00000126456.11", ]
IRF3.expr.all.t <- t(IRF3.expr.all)
IRF3.all <- data.frame(expr = IRF3.expr.all.t[, 1],
                       BMI = dat.pheno.all$BMI, 
                       AGE = dat.pheno.all$AGE,
                       SEX = dat.pheno.all$SEX,
                       MHABNWBC = dat.pheno.all$MHABNWBC)

##------Set the factor levels of phenotypes------
IRF3.all$BMI_AGE_SEX_MHABNWBC <- paste(paste0("AGE", IRF3.all$AGE), 
                                       paste0("SEX", IRF3.all$SEX), 
                                       paste0("WBC", IRF3.all$MHABNWBC),
                                       paste0("BMI", IRF3.all$BMI), sep = " + ")
combs <- table(IRF3.all$BMI_AGE_SEX_MHABNWBC) |> names()
IRF3.all$BMI_AGE_SEX_MHABNWBC <- factor(IRF3.all$BMI_AGE_SEX_MHABNWBC,
                                        levels = c(combs[grep(pattern = "BMIlow", x = combs)],
                                                   combs[grep(pattern = "BMIhigh", x = combs)]))


IRF3.all$BMI_MHABNWBC <- paste(paste0("WBC", IRF3.all$MHABNWBC), 
                               paste0("BMI", IRF3.all$BMI), sep = " + ")
combs <- table(IRF3.all$BMI_MHABNWBC) |> names()
IRF3.all$BMI_MHABNWBC <- factor(IRF3.all$BMI_MHABNWBC,
                                levels = c(combs[grep(pattern = "BMIlow", x = combs)],
                                           combs[grep(pattern = "BMIhigh", x = combs)]))


IRF3.all$AGE_MHABNWBC <- paste(paste0("AGE", IRF3.all$AGE), 
                               paste0("WBC", IRF3.all$MHABNWBC), sep = " + ")
combs <- table(IRF3.all$AGE_MHABNWBC) |> names()
IRF3.all$AGE_MHABNWBC <- factor(IRF3.all$AGE_MHABNWBC,
                                levels = c(combs[grep(pattern = "WBCno", x = combs)],
                                           combs[grep(pattern = "WBCyes", x = combs)]))

##------Customize the plot theme------
theme_BMA <- function(base_size = 14,
                      base_family = "Arial") {
  theme_classic(base_size = base_size, base_family = base_family) +
    theme(
      legend.position = "none",
      panel.background = element_rect(fill = "white", colour = "white"),
      axis.text.x = element_text(size = 12, angle = 90, hjust = 1, color = "black", vjust = 0.5),
      strip.background = element_rect(fill = "white", color = "black", linewidth = 0.6), # Customize the box border in strip
      strip.placement = "inside"
    )
}

##------Make a boxplot function------
make_boxplot <- function(gene.symbol = "MTHFS", 
                         data.type = "train", 
                         var.x = "BMI", 
                         test.type = "wilcox.test",
                         paired_comparisons_list = NULL,
                         y_transform = NULL) {
  p <- ggplot(data = get(paste0(gene.symbol, ".", data.type)),
              mapping = aes(x = get(var.x), y = expr)) +
    geom_boxplot(aes(group = get(var.x), color = get(var.x)),
                 lwd = 1.2,
                 outlier.shape = NA) + 
    geom_jitter(aes(color = get(var.x)), size = 0.8) + 
    stat_compare_means(method = test.type, # This is the overall comparison
                       label.x.npc = "middle",
                       size = 5) + 
    theme_BMA() + 
    scale_color_viridis(discrete = T) + 
    xlab(gsub(pattern = "_", replacement = " + ", x = var.x))
  
  if (is.null(paired_comparisons_list)) {
    p
  } else {
    p <- p + stat_compare_means(label.x.npc = "middle",
                                comparisons = paired_comparisons_list,
                                method = "wilcox.test")
    # aes(label = ifelse(p < 1.e-3, "p < 0.001", sprintf("p = %4.3f", as.numeric(..p.format..))))
  }
  if (is.null(y_transform)) {
    p <- p + ylab(paste0(gene.symbol, " expression level"))
    return(p)
  } else {
    p <- p + ylab(bquote(paste("-log"[10]*" ", .(gene.symbol), " expression level"))) + 
      scale_y_continuous(trans = y_transform)
    return(p)
  }
}

##------Make boxplots------
p1 <- make_boxplot(gene.symbol = "IRF3", data.type = "all", var.x = "BMI", test.type = "wilcox.test")
p2 <- make_boxplot(gene.symbol = "IRF3", data.type = "all", var.x = "MHABNWBC", test.type = "wilcox.test")
p3 <- make_boxplot(gene.symbol = "IRF3", data.type = "all", var.x = "BMI_MHABNWBC", 
                   test.type = "kruskal.test")
p4 <- make_boxplot(gene.symbol = "IRF3", data.type = "all", var.x = "AGE_MHABNWBC", test.type = "kruskal.test")
p5 <- make_boxplot(gene.symbol = "IRF3", data.type = "all", var.x = "BMI_AGE_SEX_MHABNWBC", test.type = "kruskal.test")
p6 <- make_boxplot(gene.symbol = "IRF3", data.type = "all", var.x = "AGE", test.type = "wilcox.test")

##------Put plots together in a panel------
g1 <- plot_grid(p1, p5, ncol = 2, labels = LETTERS[1:2])
g2 <- plot_grid(p2, p3, p4, ncol = 3, labels = LETTERS[3:5])
g3 <- plot_grid(p2, p3, ncol = 2, labels = LETTERS[3:4])
g4 <- plot_grid(p6, p4, ncol = 2, labels = LETTERS[1:2])

##------Save plots------
date.analysis <- format(Sys.Date(), "%Y%b%d")
ggsave(filename = sprintf("../ApplicationResult/AddViz/IRF3_BoxPlot/%s_%s_%s_%s.1.eps", date.analysis, "BMAseq", var.name, "all.seed"),
       plot = g1, device = cairo_ps, dpi = 600, width = 12, height = 7, units = "in")

ggsave(filename = sprintf("../ApplicationResult/AddViz/IRF3_BoxPlot/%s_%s_%s_%s.2.eps", date.analysis, "BMAseq", var.name, "all.seed"),
       plot = g2, device = cairo_ps, dpi = 600, width = 12, height = 6, units = "in")

ggsave(filename = sprintf("../ApplicationResult/AddViz/IRF3_BoxPlot/%s_%s_%s_%s.3.eps", date.analysis, "BMAseq", var.name, "all.seed"),
       plot = g3, device = cairo_ps, dpi = 600, width = 12, height = 6, units = "in")

ggsave(filename = sprintf("../ApplicationResult/AddViz/IRF3_BoxPlot/%s_%s_%s_%s.4.eps", date.analysis, "BMAseq", var.name, "all.seed"),
       plot = g4, device = cairo_ps, dpi = 600, width = 12, height = 6, units = "in")
```


### edgeR_UVM
```{r KRT71}
##------Map gene symbols to ENSEMBL IDs------
ensembl.id <- bitr("KRT71", 
                   fromType = "SYMBOL", 
                   toType = "ENSEMBL", 
                   OrgDb = "org.Hs.eg.db")
ensembl.id
# SYMBOL         ENSEMBL
# 1  KRT71 ENSG00000139648

##------Use the whole data to make the boxplot------
KRT71.expr.all <- dat.expr.all[grepl(pattern = "ENSG00000139648", x = rownames(dat.expr.all)), ]
KRT71.expr.all.t <- t(KRT71.expr.all)
KRT71.all <- data.frame(expr = KRT71.expr.all.t[, 1] + 1,
                        BMI = dat.pheno.all$BMI, 
                        AGE = dat.pheno.all$AGE,
                        SEX = dat.pheno.all$SEX,
                        MHABNWBC = dat.pheno.all$MHABNWBC)

##------Set the factor levels of phenotypes------
KRT71.all$BMI_AGE_SEX_MHABNWBC <- paste(paste0("AGE", KRT71.all$AGE), 
                                        paste0("SEX", KRT71.all$SEX), 
                                        paste0("WBC", KRT71.all$MHABNWBC),
                                        paste0("BMI", KRT71.all$BMI), sep = " + ")
combs <- table(KRT71.all$BMI_AGE_SEX_MHABNWBC) |> names()
KRT71.all$BMI_AGE_SEX_MHABNWBC <- factor(KRT71.all$BMI_AGE_SEX_MHABNWBC,
                                         levels = c(combs[grep(pattern = "BMIlow", x = combs)],
                                                    combs[grep(pattern = "BMIhigh", x = combs)]))

##------Make boxplots------
p1 <- make_boxplot(gene.symbol = "KRT71", data.type = "all", var.x = "BMI", test.type = "wilcox.test", y_transform = "log10")
p2 <- make_boxplot(gene.symbol = "KRT71", data.type = "all", var.x = "BMI_AGE_SEX_MHABNWBC", test.type = "kruskal.test", y_transform = "log10")
g1 <- plot_grid(p1, p2, ncol = 2, labels = LETTERS[1:2])
ggsave(filename = sprintf("../ApplicationResult/AddViz/KRT71_BoxPlot/%s_%s_%s_%s.1.eps", date.analysis, "BMAseq", var.name, "all.seed"),
       plot = g1, device = cairo_ps, dpi = 600, width = 12, height = 7, units = "in")
```


```{r LCE1A}
##------Map gene symbols to ENSEMBL IDs------
ensembl.id <- bitr("LCE1A", 
                   fromType = "SYMBOL", 
                   toType = "ENSEMBL", 
                   OrgDb = "org.Hs.eg.db")
ensembl.id
# SYMBOL         ENSEMBL
# 1  LCE1A ENSG00000186844

##------Use the whole data to make the boxplot------
LCE1A.expr.all <- dat.expr.all[grepl(pattern = "ENSG00000186844", x = rownames(dat.expr.all)), ]
LCE1A.expr.all.t <- t(LCE1A.expr.all)
LCE1A.all <- data.frame(expr = LCE1A.expr.all.t[, 1] + 1,
                        BMI = dat.pheno.all$BMI, 
                        AGE = dat.pheno.all$AGE,
                        SEX = dat.pheno.all$SEX,
                        MHABNWBC = dat.pheno.all$MHABNWBC)

##------Set the factor levels of phenotypes------
LCE1A.all$BMI_AGE_SEX_MHABNWBC <- paste(paste0("AGE", LCE1A.all$AGE), 
                                        paste0("SEX", LCE1A.all$SEX), 
                                        paste0("WBC", LCE1A.all$MHABNWBC),
                                        paste0("BMI", LCE1A.all$BMI), sep = " + ")
combs <- table(LCE1A.all$BMI_AGE_SEX_MHABNWBC) |> names()
LCE1A.all$BMI_AGE_SEX_MHABNWBC <- factor(LCE1A.all$BMI_AGE_SEX_MHABNWBC,
                                         levels = c(combs[grep(pattern = "BMIlow", x = combs)],
                                                    combs[grep(pattern = "BMIhigh", x = combs)]))

##------Make boxplots------
p1 <- make_boxplot(gene.symbol = "LCE1A", data.type = "all", var.x = "BMI", test.type = "wilcox.test", y_transform = "log10")
p2 <- make_boxplot(gene.symbol = "LCE1A", data.type = "all", var.x = "BMI_AGE_SEX_MHABNWBC", test.type = "kruskal.test", y_transform = "log10")
g1 <- plot_grid(p1, p2, ncol = 2, labels = LETTERS[1:2])
ggsave(filename = sprintf("../ApplicationResult/AddViz/LCE1A_BoxPlot/%s_%s_%s_%s.1.eps", date.analysis, "BMAseq", var.name, "all.seed"),
       plot = g1, device = cairo_ps, dpi = 600, width = 12, height = 7, units = "in")
```


```{r LCE2B}
##------Map gene symbols to ENSEMBL IDs------
ensembl.id <- bitr("LCE2B", 
                   fromType = "SYMBOL", 
                   toType = "ENSEMBL", 
                   OrgDb = "org.Hs.eg.db")
ensembl.id
# SYMBOL         ENSEMBL
# 1  LCE2B ENSG00000159455

##------Use the whole data to make the boxplot------
LCE2B.expr.all <- dat.expr.all[grepl(pattern = "ENSG00000159455", x = rownames(dat.expr.all)), ]
LCE2B.expr.all.t <- t(LCE2B.expr.all)
LCE2B.all <- data.frame(expr = LCE2B.expr.all.t[, 1] + 1,
                        BMI = dat.pheno.all$BMI, 
                        AGE = dat.pheno.all$AGE,
                        SEX = dat.pheno.all$SEX,
                        MHABNWBC = dat.pheno.all$MHABNWBC)

##------Set the factor levels of phenotypes------
LCE2B.all$BMI_AGE_SEX_MHABNWBC <- paste(paste0("AGE", LCE2B.all$AGE), 
                                        paste0("SEX", LCE2B.all$SEX), 
                                        paste0("WBC", LCE2B.all$MHABNWBC),
                                        paste0("BMI", LCE2B.all$BMI), sep = " + ")
combs <- table(LCE2B.all$BMI_AGE_SEX_MHABNWBC) |> names()
LCE2B.all$BMI_AGE_SEX_MHABNWBC <- factor(LCE2B.all$BMI_AGE_SEX_MHABNWBC,
                                         levels = c(combs[grep(pattern = "BMIlow", x = combs)],
                                                    combs[grep(pattern = "BMIhigh", x = combs)]))

##------Make boxplots------
p1 <- make_boxplot(gene.symbol = "LCE2B", data.type = "all", var.x = "BMI", test.type = "wilcox.test", y_transform = "log10")
p2 <- make_boxplot(gene.symbol = "LCE2B", data.type = "all", var.x = "BMI_AGE_SEX_MHABNWBC", test.type = "kruskal.test", y_transform = "log10")
g1 <- plot_grid(p1, p2, ncol = 2, labels = LETTERS[1:2])
ggsave(filename = sprintf("../ApplicationResult/AddViz/LCE2B_BoxPlot/%s_%s_%s_%s.1.eps", date.analysis, "BMAseq", var.name, "all.seed"),
       plot = g1, device = cairo_ps, dpi = 600, width = 12, height = 7, units = "in")
```


```{r LORICRIN}
##------Map gene symbols to ENSEMBL IDs------
ensembl.id <- bitr("LORICRIN", 
                   fromType = "SYMBOL", 
                   toType = "ENSEMBL", 
                   OrgDb = "org.Hs.eg.db")
ensembl.id
# SYMBOL         ENSEMBL
# 1  LORICRIN ENSG00000203782

##------Use the whole data to make the boxplot------
LORICRIN.expr.all <- dat.expr.all[grepl(pattern = "ENSG00000203782", x = rownames(dat.expr.all)), ]
LORICRIN.expr.all.t <- t(LORICRIN.expr.all)
LORICRIN.all <- data.frame(expr = LORICRIN.expr.all.t[, 1] + 1,
                        BMI = dat.pheno.all$BMI, 
                        AGE = dat.pheno.all$AGE,
                        SEX = dat.pheno.all$SEX,
                        MHABNWBC = dat.pheno.all$MHABNWBC)

##------Set the factor levels of phenotypes------
LORICRIN.all$BMI_AGE_SEX_MHABNWBC <- paste(paste0("AGE", LORICRIN.all$AGE), 
                                           paste0("SEX", LORICRIN.all$SEX), 
                                           paste0("WBC", LORICRIN.all$MHABNWBC),
                                           paste0("BMI", LORICRIN.all$BMI), sep = " + ")
combs <- table(LORICRIN.all$BMI_AGE_SEX_MHABNWBC) |> names()
LORICRIN.all$BMI_AGE_SEX_MHABNWBC <- factor(LORICRIN.all$BMI_AGE_SEX_MHABNWBC,
                                            levels = c(combs[grep(pattern = "BMIlow", x = combs)],
                                                       combs[grep(pattern = "BMIhigh", x = combs)]))

##------Make boxplots------
p1 <- make_boxplot(gene.symbol = "LORICRIN", data.type = "all", var.x = "BMI", test.type = "wilcox.test", y_transform = "log10")
p2 <- make_boxplot(gene.symbol = "LORICRIN", data.type = "all", var.x = "BMI_AGE_SEX_MHABNWBC", test.type = "kruskal.test", y_transform = "log10")
g1 <- plot_grid(p1, p2, ncol = 2, labels = LETTERS[1:2])
ggsave(filename = sprintf("../ApplicationResult/AddViz/LORICRIN_BoxPlot/%s_%s_%s_%s.1.eps", date.analysis, "BMAseq", var.name, "all.seed"),
       plot = g1, device = cairo_ps, dpi = 600, width = 12, height = 7, units = "in")
```


# Interaction analysis
## Create functions of building multivariable models with interaction terms
```{r}
multiInt_TMM_top <- function(seed.num = 999, threshold = 2000) {
  
  ##------Set variables of interest------
  var.pool <- c("BMI", "AGE", "SEX", "MHABNWBC")  
  interact.pool <- c("BMIxSEX", "BMI.BMI&SEX", "SEX.BMI&SEX")
  var.pool.add <- c(var.pool, "BMIxSEX")
  
  
  ##------BMAseq------
  output.multi.int1.train <- BMAseq.multi.postprob.norm(
    dat.expr.counts = dat.expr.train,  
    dat.pheno = dat.pheno.train, 
    var.pool = var.pool, 
    max.nvar = 4, 
    interaction = "BMI&SEX", 
    cut.BF = 1)
  
  output.multi.int2.train <- BMAseq.multi.DEG(
    postprob.output = output.multi.int1.train, 
    cut.FDR = 0.05)
  
  output.multi.int1.test <- BMAseq.multi.postprob.norm(
    dat.expr.counts = dat.expr.test, 
    dat.pheno = dat.pheno.test, 
    var.pool = var.pool, 
    max.nvar = 4, 
    interaction = "BMI&SEX", 
    cut.BF = 1)
  
  output.multi.int2.test <- BMAseq.multi.DEG(
    postprob.output = output.multi.int1.test, 
    cut.FDR = 0.05)
  
  BMAseq.eFDR.Main.train <- mclapply(1:length(var.pool),
                                     function(i) output.multi.int2.train$eFDR[, i][order(output.multi.int2.train$eFDR[, i])[1:threshold]],
                                     mc.cores = 10)
  
  BMAseq.eFDR.Main.test <- mclapply(1:length(var.pool),
                                    function(i) output.multi.int2.test$eFDR[, i][order(output.multi.int2.test$eFDR[, i])[1:threshold]],
                                     mc.cores = 10)
  
  names(BMAseq.eFDR.Main.train) = names(BMAseq.eFDR.Main.test) = var.pool
  
  BMAseq.eFDR.Interaction.train <- mclapply(5:7,
                                            function(i) output.multi.int2.train$eFDR[, i][order(output.multi.int2.train$eFDR[, i])[1:threshold]],
                                            mc.cores = 10)
  
  BMAseq.eFDR.Interaction.test <- mclapply(5:7,
                                           function(i) output.multi.int2.test$eFDR[, i][order(output.multi.int2.test$eFDR[, i])[1:threshold]],
                                           mc.cores = 10)
  
  names(BMAseq.eFDR.Interaction.train) = names(BMAseq.eFDR.Interaction.test) = interact.pool
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, var.pool, interact.pool, var.pool.add, 
       output.multi.int1.train, output.multi.int2.train, output.multi.int1.test, output.multi.int2.test, 
       BMAseq.eFDR.Main.train, BMAseq.eFDR.Main.test, BMAseq.eFDR.Interaction.train, BMAseq.eFDR.Interaction.test, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/MultiIntModel/BMAseqMultiInt%s.RData", threshold, seed.num)) 
  
  
  ##------voom + limma------
  design.train <- model.matrix(~BMI + AGE + SEX + MHABNWBC + BMI*SEX, data = dat.pheno.train)
  design.test <- model.matrix(~BMI + AGE + SEX + MHABNWBC + BMI*SEX, data = dat.pheno.test)
  
  voom.train <- voom(counts = dat.expr.train,
                     design = design.train,
                     lib.size = colSums(dat.expr.train)*calcNormFactors(dat.expr.train)) 
  
  voom.fit.train <- lmFit(voom.train[["E"]],
                          design = design.train,
                          weights = voom.train[["weights"]])
  
  voom.eFDR.train <- mclapply(1:length(var.pool.add),
                              function(i) {
                                t <- voom.fit.train[["coefficients"]][, i + 1]/voom.fit.train[["stdev.unscaled"]][, i + 1]/voom.fit.train[["sigma"]]
                                p <- 2*pt(-abs(t), df = voom.fit.train[["df.residual"]])
                                return(qvalue(p)) },
                              mc.cores = 10)
  
  voom.eFDR.train2 <- mclapply(1:length(var.pool.add),
                               function(i) {
                                 q.val <- voom.eFDR.train[[i]][["qvalues"]]
                                 return(q.val[order(q.val)[1:threshold]])
                               },
                               mc.cores = 10)
  
  names(voom.eFDR.train) = names(voom.eFDR.train2) = var.pool.add 
  
  voom.test <- voom(counts = dat.expr.test,
                    design = design.test,
                    lib.size = colSums(dat.expr.test)*calcNormFactors(dat.expr.test))
  
  voom.fit.test <- lmFit(voom.test[["E"]],
                         design = design.test,
                         weights = voom.test[["weights"]])
  
  voom.eFDR.test <- mclapply(1:length(var.pool.add),
                             function(i) {
                               t <- voom.fit.test[["coefficients"]][, i + 1]/voom.fit.test[["stdev.unscaled"]][, i + 1]/voom.fit.test[["sigma"]]
                               p <- 2*pt(-abs(t), df = voom.fit.test[["df.residual"]])
                               return(qvalue(p)) },
                             mc.cores = 10)
  
  voom.eFDR.test2 <- mclapply(1:length(var.pool.add),
                              function(i) {
                                q.val <- voom.eFDR.test[[i]][["qvalues"]]
                                return(q.val[order(q.val)[1:threshold]])
                              },
                                mc.cores = 10)
  
  names(voom.eFDR.test) = names(voom.eFDR.test2) = var.pool.add 
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, var.pool, var.pool.add, 
       design.train, design.test, voom.train, voom.fit.train, voom.eFDR.train, voom.eFDR.train2, 
       voom.test, voom.fit.test, voom.eFDR.test, voom.eFDR.test2, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/MultiIntModel/VoomLimmaMultiInt%s.RData", threshold, seed.num))
  
  
  ##------voom + limma + eBayes------
  voom.train <- voom(counts = dat.expr.train,
                     design = design.train,
                     lib.size = colSums(dat.expr.train)*calcNormFactors(dat.expr.train))
  
  eBayes.fit.train <- lmFit(voom.train[["E"]],
                            design = design.train,
                            weights = voom.train[["weights"]]) %>% 
    eBayes()
  
  eBayes.eFDR.train <- mclapply(1:length(var.pool.add),
                                function(i) eBayes.fit.train[["p.value"]][, i + 1] %>% 
                                  qvalue(),
                                mc.cores = 10)
  
  eBayes.eFDR.train2 <- mclapply(1:length(var.pool.add),
                                 function(i) {
                                   q.val <- eBayes.eFDR.train[[i]][["qvalues"]]
                                   return(q.val[order(q.val)[1:threshold]])
                                 },
                                 mc.cores = 10)
  
  names(eBayes.eFDR.train) = names(eBayes.eFDR.train2) = var.pool.add 
  
  voom.test <- voom(counts = dat.expr.test,
                    design = design.test,
                    lib.size = colSums(dat.expr.test)*calcNormFactors(dat.expr.test))
  
  eBayes.fit.test <- lmFit(voom.test[["E"]],
                           design = design.test,
                           weights = voom.test[["weights"]]) %>% 
    eBayes()
  
  eBayes.eFDR.test <- mclapply(1:length(var.pool.add),
                               function(i) eBayes.fit.test[["p.value"]][, i + 1] %>% 
                                 qvalue(),
                               mc.cores = 10)
  
  eBayes.eFDR.test2 <- mclapply(1:length(var.pool.add),
                                function(i) {
                                  q.val <- eBayes.eFDR.test[[i]][["qvalues"]]
                                  return(q.val[order(q.val)[1:threshold]]) 
                                },
                                mc.cores = 10)
  
  names(eBayes.eFDR.test) = names(eBayes.eFDR.test2) = var.pool.add 
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, var.pool, var.pool.add, 
       design.train, design.test, voom.train, eBayes.fit.train, eBayes.eFDR.train, eBayes.eFDR.train2, 
       voom.test, eBayes.fit.test, eBayes.eFDR.test, eBayes.eFDR.test2, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/MultiIntModel/eBayesMultiInt%s.RData", threshold, seed.num))
  
  
  ##------edgeR------
  y.train <- DGEList(counts = dat.expr.train, 
                     lib.size = colSums(dat.expr.train)) %>% 
    calcNormFactors() %>% 
    estimateGLMTrendedDisp(design.train)
  
  edgeR.fit.train <- glmQLFit(y.train, design.train)
  
  qlf.train <- mclapply(1:length(var.pool.add),
                        function(i) glmQLFTest(edgeR.fit.train, coef = i + 1),
                        mc.cores = 10)
  
  edgeR.eFDR.train <- mclapply(1:length(var.pool.add),
                               function(i) qlf.train[[i]][["table"]][["PValue"]] %>%
                                 qvalue(),
                               mc.cores = 10)
  
  edgeR.eFDR.train2 <- mclapply(1:length(var.pool.add),
                                function(i) {
                                  q.val <- edgeR.eFDR.train[[i]][["qvalues"]]
                                  return(q.val[order(q.val)[1:threshold]]) 
                                },
                                mc.cores = 10)
  
  edgeR.eFDR.GeneName.train <- mclapply(1:length(var.pool.add),
                                        function(i) {
                                          q.val <- edgeR.eFDR.train[[i]][["qvalues"]]
                                          return(rownames(dat.expr.train)[order(q.val)[1:threshold]])
                                        },
                                        mc.cores = 10)
  
  names(qlf.train) = names(edgeR.eFDR.train) = names(edgeR.eFDR.train2) = names(edgeR.eFDR.GeneName.train) = var.pool.add 
  
  y.test <- DGEList(counts = dat.expr.test, 
                    lib.size = colSums(dat.expr.test)) %>% 
    calcNormFactors() %>% 
    estimateGLMTrendedDisp(design.test)
  
  edgeR.fit.test <- glmQLFit(y.test, design.test)
  
  qlf.test <- mclapply(1:length(var.pool.add),
                       function(i) glmQLFTest(edgeR.fit.test, coef = i + 1),
                       mc.cores = 10)
  
  edgeR.eFDR.test <- mclapply(1:length(var.pool.add),
                              function(i) qlf.test[[i]][["table"]][["PValue"]] %>% 
                                qvalue(),
                              mc.cores = 10)
  
  edgeR.eFDR.test2 <- mclapply(1:length(var.pool.add),
                               function(i) {
                                 q.val <- edgeR.eFDR.test[[i]][["qvalues"]]
                                 return(q.val[order(q.val)[1:threshold]]) 
                               },
                               mc.cores = 10)
  
  edgeR.eFDR.GeneName.test <- mclapply(1:length(var.pool.add),
                                       function(i) {
                                         q.val <- edgeR.eFDR.test[[i]][["qvalues"]]
                                         return(rownames(dat.expr.test)[order(q.val)[1:threshold]])
                                       },
                                       mc.cores = 10)
  
  names(qlf.test) = names(edgeR.eFDR.test) = names(edgeR.eFDR.test2) = names(edgeR.eFDR.GeneName.test) = var.pool.add 
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, var.pool, var.pool.add, design.train, design.test, 
       y.train, edgeR.fit.train, edgeR.eFDR.train, edgeR.eFDR.train2, edgeR.eFDR.GeneName.train, 
       y.test, edgeR.fit.test, edgeR.eFDR.test, edgeR.eFDR.test2, edgeR.eFDR.GeneName.test, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/MultiIntModel/edgeRMultiInt%s.RData", threshold, seed.num))
  
  
  ##------DESeq2------
  cts.train <- as.matrix(dat.expr.train)
  cts.test <- as.matrix(dat.expr.test)
  coldata.train <- dat.pheno.train
  coldata.test <- dat.pheno.test
  
  name.formula <- c("BMI_high_vs_low", "AGE_old_vs_young", "SEX_male_vs_female", 
                    "MHABNWBC_yes_vs_no", "BMIhigh.SEXmale")  
  
  lib.size <- colSums(cts.train)
  norm.factor <- calcNormFactors(cts.train, method = "TMM")
  size.factor <- lib.size*norm.factor/exp(mean(log(lib.size*norm.factor))) 
  
  dds <- DESeqDataSetFromMatrix(countData = cts.train, 
                                colData = coldata.train, 
                                design = ~BMI + AGE + SEX + MHABNWBC + BMI*SEX)
  sizeFactors(dds) <- size.factor
  
  res.train <- mclapply(1:length(var.pool.add), 
                        function(i) {
                          return(results(DESeq(dds), name = name.formula[i])) },
                        mc.cores = 10)
  
  DESeq2.eFDR.train <- mclapply(1:length(var.pool.add),
                                function(i) qvalue(res.train[[i]][["pvalue"]])$qvalues,
                                mc.cores = 10)
  
  DESeq2.eFDR.train2 <- mclapply(1:length(var.pool.add),
                                 function(i) {
                                   q.val <- DESeq2.eFDR.train[[i]]
                                   return(q.val[order(q.val)[1:threshold]])
                                 },
                                 mc.cores = 10)
  
  DESeq2.eFDR.GeneName.train <- mclapply(1:length(var.pool.add),
                                         function(i) {
                                           q.val <- DESeq2.eFDR.train[[i]]
                                           return(rownames(cts.train)[order(q.val)[1:threshold]])
                                         },
                                         mc.cores = 10)
  
  names(res.train) = names(DESeq2.eFDR.train) = names(DESeq2.eFDR.train2) = names(DESeq2.eFDR.GeneName.train) = var.pool.add 
  
  lib.size <- colSums(cts.test)
  norm.factor <- calcNormFactors(cts.test, method = "TMM")
  size.factor <- lib.size*norm.factor/exp(mean(log(lib.size*norm.factor))) 
  
  dds <- DESeqDataSetFromMatrix(countData = cts.test, 
                                colData = coldata.test, 
                                design = ~BMI + AGE + SEX + MHABNWBC + BMI*SEX)
  sizeFactors(dds) <- size.factor
  
  res.test <- mclapply(1:length(var.pool.add), 
                       function(i) {
                         return(results(DESeq(dds), name = name.formula[i])) }, 
                       mc.cores = 10)
  
  DESeq2.eFDR.test <- mclapply(1:length(var.pool.add),
                               function(i) qvalue(res.test[[i]][["pvalue"]])$qvalues,
                               mc.cores = 10)
  
  DESeq2.eFDR.test2 <- mclapply(1:length(var.pool.add),
                                function(i) {
                                  q.val <- DESeq2.eFDR.test[[i]]
                                  return(q.val[order(q.val)[1:threshold]])
                                },
                                mc.cores = 10)
  
  DESeq2.eFDR.GeneName.test <- mclapply(1:length(var.pool.add),
                                        function(i) {
                                          q.val <- DESeq2.eFDR.test[[i]]
                                          return(rownames(cts.test)[order(q.val)[1:threshold]])
                                        },
                                        mc.cores = 10)
  
  names(res.test) = names(DESeq2.eFDR.test) = names(DESeq2.eFDR.test2) = names(DESeq2.eFDR.GeneName.test) = var.pool.add
  
  save(cts.train, cts.test, coldata.train, coldata.test, var.pool, var.pool.add, 
       res.train, DESeq2.eFDR.train, DESeq2.eFDR.train2, DESeq2.eFDR.GeneName.train, 
       res.test, DESeq2.eFDR.test, DESeq2.eFDR.test2, DESeq2.eFDR.GeneName.test, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/MultiIntModel/DESeq2MultiInt%s.RData", threshold, seed.num))
}
```


## Run and save modeling results
```{r}
threshold.i <- 5000
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
for (seed.i in seed.vec) {
  dat.expr.train <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/dat.expr.train%s.RDS", seed.i))
  dat.expr.test <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/dat.expr.test%s.RDS", seed.i))
  dat.pheno.train <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/dat.pheno.train%s.RDS", seed.i))
  dat.pheno.test <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/dat.pheno.test%s.RDS", seed.i))
  multiInt_TMM_top(seed.num = seed.i, threshold = threshold.i)
}
```


## Create functions to generate data for making boxplots of number of cDEGs and rate of cDEGs, respectively
```{r}
##------Number of cDEGs------
boxplot_data <- function (threshold = 1000, random.seed = 8809678) {
  data.table(threshold=rep(threshold,5), 
             random.seed=random.seed,
             method=c("BMAseq","DESeq2","edgeR","eBayes","voom.limma"),
             variable=rep(c("BMI","AGE","SEX","WBC","BMIxSEX"),each=5),
             num.cDEGs=c(
               ##------BMI------
               intersect(names(BMAseq.eFDR.Main.train$BMI[1:threshold]), names(BMAseq.eFDR.Main.test$BMI[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$BMI[1:threshold], DESeq2.eFDR.GeneName.test$BMI[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$BMI[1:threshold], edgeR.eFDR.GeneName.test$BMI[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$BMI[1:threshold]), names(eBayes.eFDR.test2$BMI[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$BMI[1:threshold]), names(voom.eFDR.test2$BMI[1:threshold])) |> length(),
               
               
               ##------AGE------
               intersect(names(BMAseq.eFDR.Main.train$AGE[1:threshold]), names(BMAseq.eFDR.Main.test$AGE[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$AGE[1:threshold], DESeq2.eFDR.GeneName.test$AGE[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$AGE[1:threshold], edgeR.eFDR.GeneName.test$AGE[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$AGE[1:threshold]), names(eBayes.eFDR.test2$AGE[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$AGE[1:threshold]), names(voom.eFDR.test2$AGE[1:threshold])) |> length(),
               
               
               ##------SEX------
               intersect(names(BMAseq.eFDR.Main.train$SEX[1:threshold]), names(BMAseq.eFDR.Main.test$SEX[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$SEX[1:threshold], DESeq2.eFDR.GeneName.test$SEX[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$SEX[1:threshold], edgeR.eFDR.GeneName.test$SEX[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$SEX[1:threshold]), names(eBayes.eFDR.test2$SEX[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$SEX[1:threshold]), names(voom.eFDR.test2$SEX[1:threshold])) |> length(),
               
               
               ##------MHABNWBC------
               intersect(names(BMAseq.eFDR.Main.train$MHABNWBC[1:threshold]), names(BMAseq.eFDR.Main.test$MHABNWBC[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$MHABNWBC[1:threshold], DESeq2.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$MHABNWBC[1:threshold], edgeR.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$MHABNWBC[1:threshold]), names(eBayes.eFDR.test2$MHABNWBC[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$MHABNWBC[1:threshold]), names(voom.eFDR.test2$MHABNWBC[1:threshold])) |> length(),
               
               ##------BMIxSEX------
               intersect(names(BMAseq.eFDR.Interaction.train$BMIxSEX[1:threshold]),  names(BMAseq.eFDR.Interaction.test$BMIxSEX[1:threshold])) |> length(), # ALiu: fix
               intersect(DESeq2.eFDR.GeneName.train$BMIxSEX[1:threshold], DESeq2.eFDR.GeneName.test$BMIxSEX[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$BMIxSEX[1:threshold], edgeR.eFDR.GeneName.test$BMIxSEX[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$BMIxSEX[1:threshold]), names(eBayes.eFDR.test2$BMIxSEX[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$BMIxSEX[1:threshold]), names(voom.eFDR.test2$BMIxSEX[1:threshold])) |> length()))
}


##------Rate of cDEGs------
boxplot_data2 <- function (threshold = 1000, random.seed = 8809678) {
  data.table(threshold=rep(threshold,5), 
             random.seed=random.seed,
             method=c("BMAseq","DESeq2","edgeR","eBayes","voom.limma"),
             variable=rep(c("BMI","AGE","SEX","WBC","BMIxSEX"),each=5),
             rate.cDEGs=c(
               ##------BMI------
               intersect(names(BMAseq.eFDR.Main.train$BMI[1:threshold]), names(BMAseq.eFDR.Main.test$BMI[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$BMI[1:threshold], DESeq2.eFDR.GeneName.test$BMI[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$BMI[1:threshold], edgeR.eFDR.GeneName.test$BMI[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$BMI[1:threshold]), names(eBayes.eFDR.test2$BMI[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$BMI[1:threshold]), names(voom.eFDR.test2$BMI[1:threshold])) |> length() / threshold,
               
               
               ##------AGE------
               intersect(names(BMAseq.eFDR.Main.train$AGE[1:threshold]), names(BMAseq.eFDR.Main.test$AGE[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$AGE[1:threshold], DESeq2.eFDR.GeneName.test$AGE[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$AGE[1:threshold], edgeR.eFDR.GeneName.test$AGE[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$AGE[1:threshold]), names(eBayes.eFDR.test2$AGE[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$AGE[1:threshold]), names(voom.eFDR.test2$AGE[1:threshold])) |> length() / threshold,
               
               
               ##------SEX------
               intersect(names(BMAseq.eFDR.Main.train$SEX[1:threshold]), names(BMAseq.eFDR.Main.test$SEX[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$SEX[1:threshold], DESeq2.eFDR.GeneName.test$SEX[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$SEX[1:threshold], edgeR.eFDR.GeneName.test$SEX[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$SEX[1:threshold]), names(eBayes.eFDR.test2$SEX[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$SEX[1:threshold]), names(voom.eFDR.test2$SEX[1:threshold])) |> length() / threshold,
               
               
               ##------MHABNWBC------
               intersect(names(BMAseq.eFDR.Main.train$MHABNWBC[1:threshold]), names(BMAseq.eFDR.Main.test$MHABNWBC[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$MHABNWBC[1:threshold], DESeq2.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$MHABNWBC[1:threshold], edgeR.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$MHABNWBC[1:threshold]), names(eBayes.eFDR.test2$MHABNWBC[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$MHABNWBC[1:threshold]), names(voom.eFDR.test2$MHABNWBC[1:threshold])) |> length() / threshold,
               
               
               ##------BMIxSEX------
               intersect(names(BMAseq.eFDR.Interaction.train$BMIxSEX[1:threshold]), names(BMAseq.eFDR.Interaction.test$BMIxSEX[1:threshold])) |> length() / threshold, # ALiu: fix
               intersect(DESeq2.eFDR.GeneName.train$BMIxSEX[1:threshold], DESeq2.eFDR.GeneName.test$BMIxSEX[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$BMIxSEX[1:threshold], edgeR.eFDR.GeneName.test$BMIxSEX[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$BMIxSEX[1:threshold]), names(eBayes.eFDR.test2$BMIxSEX[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$BMIxSEX[1:threshold]), names(voom.eFDR.test2$BMIxSEX[1:threshold])) |> length() / threshold))
}
```


## Organize data for boxplotting
```{r}
##------Number of cDEGs------
threshold.vec <- seq(1000, 5000, 1000)
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
boxplot.data.num <- NULL
for (threshold.i in threshold.vec) {
  for (seed.i in seed.vec) {
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiIntModel/BMAseqMultiInt%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiIntModel/DESeq2MultiInt%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiIntModel/edgeRMultiInt%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiIntModel/eBayesMultiInt%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiIntModel/VoomLimmaMultiInt%s.RData", seed.i))
    boxplot.data.num <- rbind(boxplot.data.num, boxplot_data(threshold.i, seed.i))
  }
}
saveRDS(boxplot.data.num, "../ApplicationResult/Multi_Interaction/RandomSeed/numcDEGs/Boxplot/boxplot.data.num.RDS")


##------Rate of cDEGs------
threshold.vec <- seq(1000, 5000, 1000)
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
boxplot.data.rate <- NULL
for (threshold.i in threshold.vec) {
  for (seed.i in seed.vec) {
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiIntModel/BMAseqMultiInt%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiIntModel/DESeq2MultiInt%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiIntModel/edgeRMultiInt%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiIntModel/eBayesMultiInt%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiIntModel/VoomLimmaMultiInt%s.RData", seed.i))
    boxplot.data.rate <- rbind(boxplot.data.rate, boxplot_data2(threshold.i, seed.i))
  }
}
saveRDS(boxplot.data.rate, "../ApplicationResult/Multi_Interaction/RandomSeed/ratecDEGs/Boxplot/boxplot.data.rate.RDS")
```


## Make and save boxplots
```{r}
##------Number of cDEGs------
date.analysis <- format(Sys.Date(), "%Y%b%d")
var.vec <- c("BMI", "AGE", "SEX", "WBC", "BMIxSEX")

boxplot.data.num <- readRDS("../ApplicationResult/Multi_Interaction/RandomSeed/numcDEGs/Boxplot/boxplot.data.num.RDS")
boxplot.data.num[,method:=factor(method,levels=c("BMAseq","DESeq2","edgeR","eBayes","voom.limma"))]


for (var.i in var.vec) {
  g <- boxplot.data.num[variable==var.i] |>
    ggplot(aes(y = num.cDEGs, x = method, group = method)) + 
    geom_boxplot(outlier.shape = NA) + 
    geom_jitter(aes(color = method), size = 0.8) + 
    stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
    facet_wrap(vars(threshold)) + 
    labs(y = "Number of common differentially expressed genes", x = "Method") + 
    theme_bw() + 
    theme(legend.position = "none",
          axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
    ggsave(filename = sprintf("../ApplicationResult/Multi_Interaction/RandomSeed/numcDEGs/Boxplot/%s_%s.eps", date.analysis, var.i),
           plot = g, device = "eps", dpi = 600, width = 9, height = 6, units = "in")
}


##------Rate of cDEGs------
boxplot.data.rate <- readRDS("../ApplicationResult/Multi_Interaction/RandomSeed/ratecDEGs/Boxplot/boxplot.data.rate.RDS")
boxplot.data.rate[,method:=factor(method,levels=c("BMAseq","DESeq2","edgeR","eBayes","voom.limma"))]

for (var.i in var.vec) {
  if (var.i != "WBC") {
    g <- boxplot.data.rate[variable==var.i] |>
      ggplot(aes(y = rate.cDEGs, x = method, group = method)) + 
      geom_boxplot(outlier.shape = NA) + 
      geom_jitter(aes(color = method), size = 0.8) + 
      stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
      facet_wrap(vars(threshold)) + 
      labs(y = "Rate of common differentially expressed genes", x = "Method") + 
      theme_bw() + 
      theme(legend.position = "none",
            axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5)) + 
      scale_y_continuous(labels = scales::label_number(accuracy = 0.01), breaks = seq(0, 0.5, 0.1), limits = c(0, 0.5))
      ggsave(filename = sprintf("../ApplicationResult/Multi_Interaction/RandomSeed/ratecDEGs/Boxplot/%s_%s.eps", date.analysis, var.i),
             plot = g, device = "eps", dpi = 600, width = 9, height = 6, units = "in")
  } else {
    g <- boxplot.data.rate[variable==var.i] |>
      ggplot(aes(y = rate.cDEGs, x = method, group = method)) + 
      geom_boxplot(outlier.shape = NA) + 
      geom_jitter(aes(color = method), size = 0.8) + 
      stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
      facet_wrap(vars(threshold)) + 
      labs(y = "Rate of common differentially expressed genes", x = "Method") + 
      theme_bw() + 
      theme(legend.position = "none",
            axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5)) + 
      scale_y_continuous(labels = scales::label_number(accuracy = 0.01), breaks = seq(0, 0.8, 0.1), limits = c(0, 0.8))
      ggsave(filename = sprintf("../ApplicationResult/Multi_Interaction/RandomSeed/ratecDEGs/Boxplot/%s_%s.eps", date.analysis, var.i),
             plot = g, device = "eps", dpi = 600, width = 9, height = 6, units = "in")
  }
}
```


## Estimate the duplicate rate of a common differentially expressed gene associated with a variable
The duplicate rate of a common differentially expressed gene associated with a variable is the occurrences of this common differentially expressed gene associated with a variable, which is identified by a particular model (e.g., multivariable BMAseq with interaction terms), among 10 random seed trials, divided by 10, within the particular ranking threshold (e.g., top 2000)
```{r}
cDEGs = cDEGs.list = NULL
threshold.vec <- seq(1000, 5000, 1000)
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)

duplicate_rate_cDEGs_per_variable <- function(var.name = "BMI", method.name = "BMAseq") {
  for (threshold.i in threshold.vec) {
    for (seed.i in seed.vec) {
      if (method.name == "BMAseq" & var.name != "BMIxSEX") {
        load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sMultiInt%s.RData", method.name, seed.i))
        cDEGs <- c(intersect(names(get(sprintf("%s.eFDR.Main.train", method.name))[[var.name]][1:threshold.i]), 
                             names(get(sprintf("%s.eFDR.Main.test", method.name))[[var.name]][1:threshold.i])))
        } else if (method.name == "BMAseq" & var.name == "BMIxSEX") {
        load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sMultiInt%s.RData", method.name, seed.i))
        var.name2 <- "BMI"
        cDEGs <- c(intersect(names(get(sprintf("%s.eFDR.Interaction.train", method.name))[[var.name2]][1:threshold.i]), 
                             names(get(sprintf("%s.eFDR.Interaction.test", method.name))[[var.name2]][1:threshold.i])))
      } else if (method.name %in% c("DESeq2", "edgeR")) {
        load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sMultiInt%s.RData", method.name, seed.i))
        cDEGs <- c(intersect(get(sprintf("%s.eFDR.GeneName.train", method.name))[[var.name]][1:threshold.i], 
                             get(sprintf("%s.eFDR.GeneName.test", method.name))[[var.name]][1:threshold.i]))
      } else if (method.name == "VoomLimma") {
        load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sMultiInt%s.RData", method.name, seed.i))
        method.name2 <- "voom"
        cDEGs <- c(intersect(names(get(sprintf("%s.eFDR.train2", method.name2))[[var.name]][1:threshold.i]), 
                             names(get(sprintf("%s.eFDR.test2", method.name2))[[var.name]][1:threshold.i])))
      } else {
        load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sMultiInt%s.RData", method.name, seed.i))
        cDEGs <- c(intersect(names(get(sprintf("%s.eFDR.train2", method.name))[[var.name]][1:threshold.i]), 
                             names(get(sprintf("%s.eFDR.test2", method.name))[[var.name]][1:threshold.i])))
      }
      cDEGs.list <- append(cDEGs.list, list(cDEGs))
    }
    cDEGs.unique <- unique(unlist(cDEGs.list))
    
    # Create a matrix that shows the occurrence of each unique gene among the 10 random seed trials
    cDEGs.matrix <- matrix(0, nrow = length(cDEGs.unique), ncol = 10)
    rownames(cDEGs.matrix) <- cDEGs.unique
    colnames(cDEGs.matrix) <- seed.vec   
    for (i in 1:10) {
      cDEGs.i.seed <- cDEGs.list[[i]]
      for (j in 1:length(cDEGs.unique)) {
        if (cDEGs.unique[j] %in% cDEGs.i.seed) {
          cDEGs.matrix[j, i] <- 1
        }
      }
    }
    Sum <- apply(cDEGs.matrix, 1, sum)
    Percent <- Sum/10
    cDEGs.matrix <- cbind(cDEGs.matrix, Sum, Percent) # Add two new columns to the matrix
    saveRDS(cDEGs.matrix, sprintf("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/%sMultiInt%s_%s.RDS", method.name, var.name, threshold.i))
    
    print(paste("Complete the duplicated rate matrix of cDEGs associated with", var.name, "identified by", method.name, "among 10 random seeds for the threshold", threshold.i))
    cDEGs = cDEGs.list =  NULL
  }
}

mapply(duplicate_rate_cDEGs_per_variable,
       rep(c("BMI", "AGE","SEX", "MHABNWBC", "BMIxSEX"), each = 5), 
       rep(c("BMAseq", "DESeq2", "edgeR", "eBayes", "VoomLimma"), times = 5))
```


## Prepare and organize the data for boxplots
```{r}
boxplot_duplicate_rate_data <- function (var.name = NULL, method.name = NULL, threshold = NULL) {
  load.data <- readRDS(sprintf("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/MultiIntModel/%sMultiInt%s_%s.RDS", method.name, var.name, threshold))
  temp.data <- data.table(threshold=threshold, 
                          method=method.name,
                          variable=var.name,
                          cDEGs.duplicate.rate.median=median(load.data[,12]),
                          cDEGs.duplicate.rate.mean=mean(load.data[,12]))
  return(temp.data)
}

organized_duplicate_rate_data <- mapply(boxplot_duplicate_rate_data,
                                        var.name = rep(c("BMI", "AGE","SEX", "MHABNWBC", "BMIxSEX"), each = 25), 
                                        method.name = rep(rep(c("BMAseq", "DESeq2", "edgeR", "eBayes", "VoomLimma"), each = 5), times = 5),
                                        threshold = rep(rep(seq(1000, 5000, 1000), times = 5), times = 5), SIMPLIFY = F) |> rbindlist()

saveRDS(organized_duplicate_rate_data, "../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/MultiIntModel/organized_duplicate_rate_data.RDS")
```


## Make and save boxplots
```{r}
organized_duplicate_rate_data <- readRDS("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/MultiIntModel/organized_duplicate_rate_data.RDS")

organized_duplicate_rate_data[,variable:=factor(variable,levels=c("BMI","AGE","SEX","MHABNWBC","BMIxSEX"))]
organized_duplicate_rate_data[,method:=factor(method,levels=c("BMAseq","DESeq2","edgeR","eBayes","VoomLimma"),labels=c("BMAseq","DESeq2","edgeR","eBayes","voom.limma"))]

organized_duplicate_rate_data[,cDEGs.duplicate.rate.median:=as.numeric(cDEGs.duplicate.rate.median)]
organized_duplicate_rate_data[,cDEGs.duplicate.rate.mean:=as.numeric(cDEGs.duplicate.rate.mean)]
date.analysis <- format(Sys.Date(), "%Y%b%d")
g1 <- organized_duplicate_rate_data[variable=="BMIxSEX"] |> # Focus on BMIxSEX
  ggplot(mapping = aes(y = cDEGs.duplicate.rate.median, x = method, group = method)) + 
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(aes(color = method)) + 
  facet_wrap(vars(variable)) + 
  scale_y_continuous(labels = scales::label_number(accuracy = 0.01),
                     breaks = seq(0, 1, 0.1),
                     limits = c(0, 1)) + 
  labs(y = "Median duplicate rate of cDEGs", x = "Method") + 
  theme_bw() + 
  theme(legend.position = "none",
        axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
ggsave(filename = sprintf("../ApplicationResult/Multi_Interaction/RandomSeed/DuplicatedRateMatrix/Boxplot/%s_median.eps", date.analysis),
       plot = g1, device = "eps", dpi = 600, width = 5, height = 4, units = "in")

g2 <- organized_duplicate_rate_data[variable=="BMIxSEX"] |> # Focus on BMIxSEX
  ggplot(mapping = aes(y = cDEGs.duplicate.rate.mean, x = method, group = method)) + 
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(aes(color = method)) + 
  facet_wrap(vars(variable)) + 
  scale_y_continuous(labels = scales::label_number(accuracy = 0.01),
                     breaks = seq(0, 1, 0.1),
                     limits = c(0, 1)) + 
  labs(y = "Mean duplicate rate of cDEGs", x = "Method") + 
  theme_bw() + 
  theme(legend.position = "none",
        axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
ggsave(filename = sprintf("../ApplicationResult/Multi_Interaction/RandomSeed/DuplicatedRateMatrix/Boxplot/%s_mean.eps", date.analysis),
       plot = g2, device = "eps", dpi = 600, width = 5, height = 4, units = "in")
```


## Make Venn diagrams of cDEGs
```{r}
threshold <- 2000
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC", "BMIxSEX")

date.analysis <- format(Sys.Date(), "%Y%b%d")
make_vennplot <- function(var.name = NULL, threshold = NULL) {
  if (var.name == "BMIxSEX") {
    var.name2 <- "BMI"
    BMAseq.cDEGs <- intersect(names(BMAseq.eFDR.Interaction.train[[var.name2]][1:threshold]), names(BMAseq.eFDR.Interaction.test[[var.name2]][1:threshold]))
  } else {
    BMAseq.cDEGs <- intersect(names(BMAseq.eFDR.Main.train[[var.name]][1:threshold]), names(BMAseq.eFDR.Main.test[[var.name]][1:threshold]))
  }
  DESeq2.cDEGs <- intersect(DESeq2.eFDR.GeneName.train[[var.name]][1:threshold], DESeq2.eFDR.GeneName.test[[var.name]][1:threshold])
  edgeR.cDEGs <- intersect(edgeR.eFDR.GeneName.train[[var.name]][1:threshold], edgeR.eFDR.GeneName.test[[var.name]][1:threshold])
  eBayes.cDEGs <- intersect(names(eBayes.eFDR.train2[[var.name]][1:threshold]), names(eBayes.eFDR.test2[[var.name]][1:threshold]))
  voom.limma.cDEGs <- intersect(names(voom.eFDR.train2[[var.name]][1:threshold]), names(voom.eFDR.test2[[var.name]][1:threshold]))
  g <- ggVennDiagram(list(BMAseq = BMAseq.cDEGs, 
                          DESeq2 = DESeq2.cDEGs,
                          edgeR = edgeR.cDEGs,
                          eBayes = eBayes.cDEGs,
                          voom.limma = voom.limma.cDEGs),
                     label_alpha = 0, label_color = "white") +
    theme(legend.position = "none", 
          plot.title = element_text(hjust = 0.5, colour = "red", size = 18)) + 
    ggtitle(paste("The intersection of common differentially expressed genes \nassociated with the", var.name)) + 
    scale_fill_distiller(palette = "Dark2", direction = -1)
  
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Vennplot/", date.analysis, "_", var.name, "_", seed.i, ".eps"),
         plot = g,
         device = "eps",
         dpi = 600, 
         width = 10,
         height = 7,
         units = "in")
}

for (seed.i in seed.vec) {
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/BMAseqMultiInt%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/DESeq2MultiInt%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/edgeRMultiInt%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/eBayesMultiInt%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/VoomLimmaMultiInt%s.RData", seed.i))
  for (var.i in var.vec) {
    make_vennplot(var.name = var.i, threshold = 2000)
  }
}
```


## Extract unique BMIxSEX - related cDEGs identified by BMAseq per seed and all unique cDEGs identified by BMAseq across 10 seeds
```{r}
file.name <- "../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Vennplot/uniq_cDEGs_BMIxSEX_BMAseq_per_seed.txt"
unique.cDEGs.list <- NULL
extract_unique_cDEGs_per_seed <-function(var.name = NULL, var.name.BMA = NULL, threshold = NULL) {
  BMAseq.cDEGs <- intersect(names(BMAseq.eFDR.Interaction.train[[var.name.BMA]][1:threshold]), names(BMAseq.eFDR.Interaction.test[[var.name.BMA]][1:threshold]))
  DESeq2.cDEGs <- intersect(DESeq2.eFDR.GeneName.train[[var.name]][1:threshold], DESeq2.eFDR.GeneName.test[[var.name]][1:threshold])
  edgeR.cDEGs <- intersect(edgeR.eFDR.GeneName.train[[var.name]][1:threshold], edgeR.eFDR.GeneName.test[[var.name]][1:threshold])
  eBayes.cDEGs <- intersect(names(eBayes.eFDR.train2[[var.name]][1:threshold]), names(eBayes.eFDR.test2[[var.name]][1:threshold]))
  voom.limma.cDEGs <- intersect(names(voom.eFDR.train2[[var.name]][1:threshold]), names(voom.eFDR.test2[[var.name]][1:threshold]))
  unique.cDEGs <- setdiff(BMAseq.cDEGs, union(DESeq2.cDEGs, edgeR.cDEGs) |> union(eBayes.cDEGs) |> union(voom.limma.cDEGs))
  return(unique.cDEGs)
}

for (seed.i in seed.vec) {
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiIntModel/BMAseqMultiInt%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiIntModel/DESeq2MultiInt%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiIntModel/edgeRMultiInt%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiIntModel/eBayesMultiInt%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiIntModel/VoomLimmaMultiInt%s.RData", seed.i))
  unique.cDEGs <- extract_unique_cDEGs_per_seed(var.name = "BMIxSEX", var.name.BMA = "BMI", threshold = 2000)
  unique.cDEGs.list <- append(unique.cDEGs.list, list(unique.cDEGs))
  write(sprintf("\nThe following gene IDs are unique gene IDs associated with the interaction effect of %s identified by BMAseq under the seed %s and threshold %s", var.name, seed.i, threshold), file = file.name, append = T)
  write.table(unique.cDEGs, file = file.name, quote = F, col.names = F, append = T)
}
names(unique.cDEGs.list) <- seed.vec

file.name <- "../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Vennplot/uniq_cDEGs_BMIxSEX_BMAseq_all_seeds.txt"
write.table(unlist(unique.cDEGs.list) |> unique(), file = file.name, quote = F, col.names = F)
```


# Additional visualization
## Rearrange cDEGs number plots
### All variables
```{r}
##------Load the cDEGs number data modeled under the univariable/multivariable framework------
cDEGs_num_uni <- readRDS("../ApplicationResult/Uni/RandomSeed/numcDEGs/Boxplot/boxplot.data.num.BMA.RDS")
cDEGs_num_uni <- cDEGs_num_uni[cDEGs_num_uni$method!="BMAseq", ]
cDEGs_num_multi <- readRDS("../ApplicationResult/Multi/RandomSeed/numcDEGs/Boxplot/boxplot.data.num.RDS")

##------Refine the cDEGs number data------
cDEGs_num_uni$method <- paste0(cDEGs_num_uni$method, "_UVM")
cDEGs_num_multi$method <- ifelse(cDEGs_num_multi$method != "BMAseq", paste0(cDEGs_num_multi$method, "_MVM"), cDEGs_num_multi$method)
cDEGs_num_all <- rbind(cDEGs_num_uni, cDEGs_num_multi)
cDEGs_num_all$method <- factor(cDEGs_num_all$method,
                               levels = c("BMAseq", "DESeq2_UVM", "DESeq2_MVM", "edgeR_UVM", "edgeR_MVM", "eBayes_UVM", "eBayes_MVM", "voom.limma_UVM", "voom.limma_MVM"))
cDEGs_num_all$variable <- factor(cDEGs_num_all$variable,levels=c("BMI","AGE","SEX","MHABNWBC"))

##------Set the plot theme------
theme_BMA <- function(
    base_size = 12,
    base_family = "Arial"
) {
  theme_classic(base_size = base_size, base_family = base_family) +
    theme(
      legend.position = "none",
      panel.background = element_rect(fill = "white", colour = "white"),
      axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5),
      strip.background = element_rect(fill = "white", color = "black", linewidth = 0.6), # Customize the box border in strip
      strip.placement = "inside"
    )
}

##------Make the boxplot------
date.analysis <- format(Sys.Date(), "%Y%b%d")
var.vec <- c("BMI","AGE","SEX","MHABNWBC")
for (var.i in var.vec) {
  g <- cDEGs_num_all[variable==var.i] |>
    ggplot(aes(y = num.cDEGs, x = method, group = method)) + 
    geom_boxplot(outlier.shape = NA, mapping = aes(color = method)) + 
    geom_jitter(aes(color = method), size = 0.8) + 
    # stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
    scale_color_manual(values = c("#A73030FF", "#3B3B3BFF", "#0073C2FF", "#EFC000FF", "#868686FF", "#8F7700FF", "#7AA6DCFF", "#003C67FF", "#CD534CFF")) + # Avoid red-green combinations difficult for colour-blind readers 
    facet_wrap(vars(threshold)) +
    geom_hline(aes(yintercept = -Inf)) + 
    geom_vline(aes(xintercept = -Inf)) + 
    coord_cartesian(clip = "off") + 
    labs(y = paste0("Number of ", var.i, "-related cDEGs"), x = "Method") + 
    theme_BMA()
    ggsave(filename = sprintf("../ApplicationResult/AddViz/numcDEGs/%s_%s.eps", date.analysis, var.i),
           plot = g, device = "eps", dpi = 600, width = 8, height = 6, units = "in")
}
```


### BMI
```{r}
var.i <- "BMI"
g <- cDEGs_num_all[variable == var.i] |>
  ggplot(aes(y = num.cDEGs, x = method, group = method)) +
  geom_boxplot(outlier.shape = NA, mapping = aes(color = method), lwd = 0.8) +
  geom_jitter(aes(color = method), size = 1) +
  # stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +
  scale_color_manual(
    values = c(
      "#A73030FF",
      "#3B3B3BFF",
      "#0073C2FF",
      "#EFC000FF",
      "#868686FF",
      "#8F7700FF",
      "#7AA6DCFF",
      "#003C67FF",
      "#CD534CFF"
    )
  ) + # Avoid red-green combinations difficult for colour-blind readers
  facet_wrap(vars(threshold)) +
  geom_hline(aes(yintercept = -Inf)) +
  geom_vline(aes(xintercept = -Inf)) +
  coord_cartesian(clip = "off") +
  labs(y = paste0("Number of cDEGs"),
       x = "Method") +
  theme_BMA()
ggsave(
  filename = sprintf(
    "../ApplicationResult/AddViz/numcDEGs/%s_%s.eps",
    date.analysis,
    var.i
  ),
  plot = g,
  device = "eps",
  dpi = 600,
  width = 8,
  height = 6,
  units = "in"
)
```


### Each threshold in BMI
```{r}
var.i <- "BMI"
threshold.vec <- seq(1000, 5000, 1000)
for (threshold.i in threshold.vec) { 
  cDEGs_num_threshold <- filter(cDEGs_num_all, threshold == threshold.i)
  g <- cDEGs_num_threshold[variable == var.i] |>
    ggplot(aes(y = num.cDEGs, x = method, group = method)) +
    geom_boxplot(outlier.shape = NA, mapping = aes(color = method), lwd = 0.8) +
    geom_jitter(aes(color = method), size = 1) +
    scale_color_manual(
      values = c(
        "#A73030FF",
        "#3B3B3BFF",
        "#0073C2FF",
        "#EFC000FF",
        "#868686FF",
        "#8F7700FF",
        "#7AA6DCFF",
        "#003C67FF",
        "#CD534CFF"
      )
    ) + # Avoid red-green combinations difficult for color-blind readers
    facet_wrap(vars(threshold)) +
    geom_hline(aes(yintercept = -Inf)) +
    geom_vline(aes(xintercept = -Inf)) +
    coord_cartesian(clip = "off") +
    labs(y = paste0("Number of cDEGs"),
         x = "Method") +
    theme_BMA()
  ggsave(
    filename = sprintf(
      "../ApplicationResult/AddViz/numcDEGs/%s_%s_%s.eps",
      date.analysis,
      var.i,
      threshold.i
    ),
    plot = g,
    device = "eps",
    dpi = 600,
    width = 4,
    height = 4,
    units = "in"
  )
}
```


## Rearrange cDEGs rate plots
### All variables
```{r}
##------Load the cDEGs rate data modeled under the univariable/multivariable framework------
cDEGs_rate_uni <- readRDS("../ApplicationResult/Uni/RandomSeed/ratecDEGs/Boxplot/boxplot.data.rate.BMA.RDS")
cDEGs_rate_uni <- cDEGs_rate_uni[cDEGs_rate_uni$method!="BMAseq", ]
cDEGs_rate_multi <- readRDS("../ApplicationResult/Multi/RandomSeed/ratecDEGs/Boxplot/boxplot.data.rate.RDS")

##------Refine the cDEGs rate data------
cDEGs_rate_uni$method <- paste0(cDEGs_rate_uni$method, "_UVM")
cDEGs_rate_multi$method <- ifelse(cDEGs_rate_multi$method != "BMAseq", paste0(cDEGs_rate_multi$method, "_MVM"), cDEGs_rate_multi$method)

cDEGs_rate_all <- rbind(cDEGs_rate_uni, cDEGs_rate_multi)
cDEGs_rate_all$method <- factor(cDEGs_rate_all$method,
                               levels = c("BMAseq", "DESeq2_UVM", "DESeq2_MVM", "edgeR_UVM", "edgeR_MVM", "eBayes_UVM", "eBayes_MVM", "voom.limma_UVM", "voom.limma_MVM"))
cDEGs_rate_all$variable <- factor(cDEGs_rate_all$variable,levels=c("BMI","AGE","SEX","MHABNWBC"))

##------Make the boxplot panel------
for (var.i in var.vec) {
  if (var.i != "MHABNWBC") {
    g <- cDEGs_rate_all[variable==var.i] |>
      ggplot(aes(y = rate.cDEGs, x = method, group = method)) + 
      geom_boxplot(outlier.shape = NA, mapping = aes(color = method)) + 
      geom_jitter(aes(color = method), size = 0.8) + 
      # stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
      scale_color_manual(values = c("#A73030FF", "#3B3B3BFF", "#0073C2FF", "#EFC000FF", "#868686FF", "#8F7700FF", "#7AA6DCFF", "#003C67FF", "#CD534CFF")) + # Avoid red-green combinations difficult for color-blind readers 
      scale_y_continuous(labels = scales::label_number(accuracy = 0.01), breaks = seq(0, 0.5, 0.1), limits = c(0, 0.5)) + 
      facet_wrap(vars(threshold)) +
      geom_hline(aes(yintercept = -Inf)) + 
      geom_vline(aes(xintercept = -Inf)) + 
      coord_cartesian(clip = "off") + 
      labs(y = paste0("Rate of ", var.i, "-related cDEGs"), 
           x = "Method") + 
      theme_BMA()
    ggsave(filename = sprintf("../ApplicationResult/AddViz/ratecDEGs/%s_%s.eps", date.analysis, var.i),
           plot = g, device = "eps", dpi = 600, width = 8, height = 6, units = "in")
  } else {
    g <- cDEGs_rate_all[variable==var.i] |>
      ggplot(aes(y = rate.cDEGs, x = method, group = method)) + 
      geom_boxplot(outlier.shape = NA, mapping = aes(color = method)) + 
      geom_jitter(aes(color = method), size = 0.8) + 
      # stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
      scale_color_manual(values = c("#A73030FF", "#3B3B3BFF", "#0073C2FF", "#EFC000FF", "#868686FF", "#8F7700FF", "#7AA6DCFF", "#003C67FF", "#CD534CFF")) + # Avoid red-green combinations difficult for color-blind readers 
      scale_y_continuous(labels = scales::label_number(accuracy = 0.01), breaks = seq(0, 0.8, 0.1), limits = c(0, 0.8)) + 
      facet_wrap(vars(threshold)) + 
      geom_hline(aes(yintercept = -Inf)) + 
      geom_vline(aes(xintercept = -Inf)) + 
      coord_cartesian(clip = "off") + 
      labs(y = paste0("Rate of ", var.i, "-related cDEGs"), 
           x = "Method") + 
      theme_BMA()
      ggsave(filename = sprintf("../ApplicationResult/AddViz/ratecDEGs/%s_%s.eps", date.analysis, var.i),
             plot = g, device = "eps", dpi = 600, width = 8, height = 6, units = "in")
  }
}
```


### BMI
```{r}
var.i <- "BMI"
g <- cDEGs_rate_all[variable == var.i] |>
  ggplot(aes(y = rate.cDEGs, x = method, group = method)) +
  geom_boxplot(outlier.shape = NA, mapping = aes(color = method), lwd = 0.8) +
  geom_jitter(aes(color = method), size = 1) +
  # stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +
  scale_color_manual(
    values = c(
      "#A73030FF",
      "#3B3B3BFF",
      "#0073C2FF",
      "#EFC000FF",
      "#868686FF",
      "#8F7700FF",
      "#7AA6DCFF",
      "#003C67FF",
      "#CD534CFF"
    )
  ) + # Avoid red-green combinations difficult for color-blind readers
  scale_y_continuous(
    labels = scales::label_number(accuracy = 0.01),
    breaks = seq(0, 0.8, 0.1),
    limits = c(0, 0.8)
  ) +
  facet_wrap(vars(threshold)) +
  geom_hline(aes(yintercept = -Inf)) +
  geom_vline(aes(xintercept = -Inf)) +
  coord_cartesian(clip = "off") +
  labs(y = paste0("Rate of cDEGs"),
       x = "Method") +
  theme_BMA()
ggsave(
  filename = sprintf(
    "../ApplicationResult/AddViz/ratecDEGs/%s_%s.eps",
    date.analysis,
    var.i
  ),
  plot = g,
  device = "eps",
  dpi = 600,
  width = 8,
  height = 6,
  units = "in"
)
```


### Each threshold in BMI
```{r}
var.i <- "BMI"
threshold.vec <- seq(1000, 5000, 1000)
for (threshold.i in threshold.vec) { 
  cDEGs_rate_threshold <- filter(cDEGs_rate_all, threshold == threshold.i)
  g <- cDEGs_rate_threshold[variable == var.i] |>
    ggplot(aes(y = rate.cDEGs, x = method, group = method)) +
    geom_boxplot(outlier.shape = NA, mapping = aes(color = method), lwd = 0.8) +
    geom_jitter(aes(color = method), size = 1) +
    scale_color_manual(
      values = c(
        "#A73030FF",
        "#3B3B3BFF",
        "#0073C2FF",
        "#EFC000FF",
        "#868686FF",
        "#8F7700FF",
        "#7AA6DCFF",
        "#003C67FF",
        "#CD534CFF"
      )
    ) + # Avoid red-green combinations difficult for color-blind readers
    scale_y_continuous(
      labels = scales::label_number(accuracy = 0.01)
    ) +
    facet_wrap(vars(threshold)) +
    geom_hline(aes(yintercept = -Inf)) +
    geom_vline(aes(xintercept = -Inf)) +
    coord_cartesian(clip = "off") +
    labs(y = paste0("Rate of cDEGs"),
         x = "Method") +
    theme_BMA()
  ggsave(
    filename = sprintf(
      "../ApplicationResult/AddViz/ratecDEGs/%s_%s_%s.eps",
      date.analysis,
      var.i,
      threshold.i
    ),
    plot = g,
    device = "eps",
    dpi = 600,
    width = 4,
    height = 4,
    units = "in"
  )
}
```


## Make cDEGs number line plots grouped by random seeds
### Each threshold in BMI
```{r}
##------Preprocess data------
cDEGs_num_uni <- readRDS("../ApplicationResult/Uni/RandomSeed/numcDEGs/Boxplot/boxplot.data.num.BMA.RDS")
cDEGs_num_uni <- cDEGs_num_uni[cDEGs_num_uni$method!="BMAseq", ]
cDEGs_num_multi <- readRDS("../ApplicationResult/Multi/RandomSeed/numcDEGs/Boxplot/boxplot.data.num.RDS")

cDEGs_num_uni$method <- paste0(cDEGs_num_uni$method, "_UVM")
cDEGs_num_multi$method <- ifelse(cDEGs_num_multi$method != "BMAseq", paste0(cDEGs_num_multi$method, "_MVM"), cDEGs_num_multi$method)
cDEGs_num_all <- rbind(cDEGs_num_uni, cDEGs_num_multi)
cDEGs_num_all$method <- factor(cDEGs_num_all$method,
                               levels = c("BMAseq", "DESeq2_UVM", "DESeq2_MVM", "edgeR_UVM", "edgeR_MVM", "eBayes_UVM", "eBayes_MVM", "voom.limma_UVM", "voom.limma_MVM"))
cDEGs_num_all$variable <- factor(cDEGs_num_all$variable,levels=c("BMI","AGE","SEX","MHABNWBC"))
cDEGs_num_all$random.seed <- factor(cDEGs_rate_all$random.seed,
                                    levels = c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999))

var.i <- "BMI"
threshold.vec <- seq(1000, 5000, 1000)
for (threshold.i in threshold.vec) {
  cDEGs_num_threshold <- filter(cDEGs_num_all, threshold == threshold.i, variable == var.i)
  g <- cDEGs_num_threshold |>
    ggplot(aes(y = num.cDEGs, x = method, group = random.seed)) +
    geom_line(aes(color = random.seed)) +
    geom_point(aes(color = random.seed), size = 0.8) + 
    # Annotate the max value of cDEGs num per seed by enlarging its point size
    geom_point(data = cDEGs_num_threshold |>
                 group_by(random.seed) |>
                 mutate(max.num = max(num.cDEGs)) |>
                 dplyr::filter(num.cDEGs == max.num),
               aes(y = max.num, x = method, color = random.seed), 
               size = 2) +
    facet_wrap(vars(threshold)) +
    labs(
      y = paste0("Number of cDEGs"),
      x = "Method",
      color = "Seed"
    ) +
    theme_BMA() +
    theme(# legend.position = "none",
      axis.text.x = element_text(
        size = 9,
        angle = 90,
        hjust = 1,
        color = "black",
        vjust = 0.5
      ))
  
    if (threshold.i != 5000) {
      ggsave(
        filename = sprintf(
          "../ApplicationResult/AddViz/numcDEGs/%s_%s_line_%s.eps",
          date.analysis,
          var.i,
          threshold.i
        ),
        plot = g,
        device = "eps",
        dpi = 600,
        width = 4,
        height = 4,
        units = "in")
    } else {
      g <- g + theme(legend.position = "right")
      ggsave(
        filename = sprintf(
          "../ApplicationResult/AddViz/numcDEGs/%s_%s_line_%s.eps",
          date.analysis,
          var.i,
          threshold.i
        ),
        plot = g,
        device = "eps",
        dpi = 600,
        width = 5.1,
        height = 4,
        units = "in")
    }
}
```


## Make cDEGs rate line plots grouped by random seeds
### All variables
```{r}
##------Preprocess data------
cDEGs_rate_uni <- readRDS("../ApplicationResult/Uni/RandomSeed/ratecDEGs/Boxplot/boxplot.data.rate.BMA.RDS")
cDEGs_rate_uni <- cDEGs_rate_uni[cDEGs_rate_uni$method!="BMAseq", ]
cDEGs_rate_multi <- readRDS("../ApplicationResult/Multi/RandomSeed/ratecDEGs/Boxplot/boxplot.data.rate.RDS")
cDEGs_rate_uni$method <- paste0(cDEGs_rate_uni$method, "_UVM")
cDEGs_rate_multi$method <- ifelse(cDEGs_rate_multi$method != "BMAseq", paste0(cDEGs_rate_multi$method, "_MVM"), cDEGs_rate_multi$method)

cDEGs_rate_all <- rbind(cDEGs_rate_uni, cDEGs_rate_multi)
cDEGs_rate_all$method <- factor(cDEGs_rate_all$method,
                                levels = c("BMAseq", "DESeq2_UVM", "DESeq2_MVM", "edgeR_UVM", "edgeR_MVM", "eBayes_UVM", "eBayes_MVM", "voom.limma_UVM", "voom.limma_MVM"))
cDEGs_rate_all$variable <- factor(cDEGs_rate_all$variable,levels=c("BMI","AGE","SEX","MHABNWBC"))
cDEGs_rate_all$random.seed <- factor(cDEGs_rate_all$random.seed,
                                     levels = c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999))

##------Make the lineplot panel------
for (var.i in var.vec) {
  if (var.i != "MHABNWBC") {
    g <- cDEGs_rate_all[variable==var.i] |>
      ggplot(aes(y = rate.cDEGs, x = method, group = random.seed)) + 
      geom_line(aes(color = random.seed)) + 
      geom_point(aes(color = random.seed), size = 0.8) + 
      # stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
      facet_wrap(vars(threshold)) + 
      labs(y = paste0("Rate of ", var.i, "-related cDEGs"), 
           x = "Method",
           color = "Seed") + 
      theme_bw() + 
      theme(# legend.position = "none",
            axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5)) +
      scale_y_continuous(labels = scales::label_number(accuracy = 0.01), breaks = seq(0, 0.5, 0.1), limits = c(0, 0.5))
    ggsave(filename = sprintf("../ApplicationResult/AddViz/ratecDEGs/%s_%s_line.eps", date.analysis, var.i),
             plot = g, device = "eps", dpi = 600, width = 9, height = 6, units = "in")
  } else {
    g <- cDEGs_rate_all[variable==var.i] |>
      ggplot(aes(y = rate.cDEGs, x = method, group = random.seed)) + 
      geom_line(aes(color = random.seed)) + 
      geom_point(aes(color = random.seed), size = 0.8) + 
      # stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
      facet_wrap(vars(threshold)) + 
      labs(y = paste0("Rate of ", var.i, "-related cDEGs"), 
           x = "Method",
           color = "Seed") + 
      theme_bw() + 
      theme(# legend.position = "none",
            axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5)) +
      scale_y_continuous(labels = scales::label_number(accuracy = 0.01), breaks = seq(0, 0.8, 0.1), limits = c(0, 0.8))
      ggsave(filename = sprintf("../ApplicationResult/AddViz/ratecDEGs/%s_%s_line.eps", date.analysis, var.i),
             plot = g, device = "eps", dpi = 600, width = 9, height = 6, units = "in")
  }
}
```


### Each threshold in BMI
```{r}
var.i <- "BMI"
threshold.vec <- seq(1000, 5000, 1000)
for (threshold.i in threshold.vec) {
  cDEGs_rate_threshold <- filter(cDEGs_rate_all, threshold == threshold.i, variable == var.i)
  g <- cDEGs_rate_threshold |>
    ggplot(aes(y = rate.cDEGs, x = method, group = random.seed)) +
    geom_line(aes(color = random.seed)) +
    geom_point(aes(color = random.seed), size = 0.8) + 
    # Annotate the max value of cDEGs rate per seed by enlarging its point size
    geom_point(data = cDEGs_rate_threshold |>
                 group_by(random.seed) |>
                 mutate(max.rate = max(rate.cDEGs)) |>
                 dplyr::filter(rate.cDEGs == max.rate),
               aes(y = max.rate, x = method, color = random.seed), 
               size = 2) +
    facet_wrap(vars(threshold)) +
    labs(
      y = paste0("Rate of cDEGs"),
      x = "Method",
      color = "Seed"
    ) +
    theme_BMA() +
    theme(# legend.position = "none",
      axis.text.x = element_text(
        size = 9,
        angle = 90,
        hjust = 1,
        color = "black",
        vjust = 0.5
      )) +
    scale_y_continuous(
      labels = scales::label_number(accuracy = 0.01)
    )
  
  if (threshold.i != 5000) {
    ggsave(
      filename = sprintf(
        "../ApplicationResult/AddViz/ratecDEGs/%s_%s_line_%s.eps",
        date.analysis,
        var.i,
        threshold.i
      ),
      plot = g,
      device = "eps",
      dpi = 600,
      width = 4,
      height = 4,
      units = "in")
  } else {
    g <- g + theme(legend.position = "right")
    ggsave(
      filename = sprintf(
        "../ApplicationResult/AddViz/ratecDEGs/%s_%s_line_%s.eps",
        date.analysis,
        var.i,
        threshold.i
      ),
      plot = g,
      device = "eps",
      dpi = 600,
      width = 5.1,
      height = 4,
      units = "in"
    )
  }
}
```


## [Optional] Make cDEGs rate line and box plots grouped by random seeds
```{r}
##------Preprocess data------
cDEGs_rate_uni <- readRDS("../ApplicationResult/Uni/RandomSeed/ratecDEGs/Boxplot/boxplot.data.rate.BMA.RDS")
cDEGs_rate_uni <- cDEGs_rate_uni[cDEGs_rate_uni$method!="BMAseq", ]
cDEGs_rate_multi <- readRDS("../ApplicationResult/Multi/RandomSeed/ratecDEGs/Boxplot/boxplot.data.rate.RDS")
cDEGs_rate_uni$method <- paste0(cDEGs_rate_uni$method, "_UVM")
cDEGs_rate_multi$method <- ifelse(cDEGs_rate_multi$method != "BMAseq", paste0(cDEGs_rate_multi$method, "_MVM"), cDEGs_rate_multi$method)

cDEGs_rate_all <- rbind(cDEGs_rate_uni, cDEGs_rate_multi)
cDEGs_rate_all$method <- factor(cDEGs_rate_all$method,
                                levels = c("BMAseq", "DESeq2_UVM", "DESeq2_MVM", "edgeR_UVM", "edgeR_MVM", "eBayes_UVM", "eBayes_MVM", "voom.limma_UVM", "voom.limma_MVM"))
cDEGs_rate_all$variable <- factor(cDEGs_rate_all$variable,levels=c("BMI","AGE","SEX","MHABNWBC"))
cDEGs_rate_all$random.seed <- factor(cDEGs_rate_all$random.seed,
                                     levels = c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999))

##------Make the line and box plot panel------
for (var.i in var.vec) {
  if (var.i != "MHABNWBC") {
    g <- cDEGs_rate_all[variable==var.i] |>
      ggplot(aes(y = rate.cDEGs, x = method, group = random.seed)) + 
      geom_line(aes(color = random.seed)) + 
      geom_boxplot(aes(group = method), outlier.shape = NA, alpha = 0.5) + 
      geom_point(aes(color = random.seed), size = 0.8, alpha = 0.6) + 
      # stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
      facet_wrap(vars(threshold)) + 
      labs(y = paste0("Rate of ", var.i, "-related cDEGs"), 
           x = "Method",
           color = "Seed") + 
      theme_bw() + 
      theme(# legend.position = "none",
            axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5)) +
      scale_y_continuous(labels = scales::label_number(accuracy = 0.01), breaks = seq(0, 0.5, 0.1), limits = c(0, 0.5))
    ggsave(filename = sprintf("../ApplicationResult/AddViz/ratecDEGs/%s_%s_line_box.eps", date.analysis, var.i),
             plot = g, device = "eps", dpi = 600, width = 9, height = 6, units = "in")
  } else {
    g <- cDEGs_rate_all[variable==var.i] |>
      ggplot(aes(y = rate.cDEGs, x = method, group = random.seed)) + 
      geom_line(aes(color = random.seed)) + 
      geom_boxplot(aes(group = method), outlier.shape = NA, alpha = 0.5) + 
      geom_point(aes(color = random.seed), size = 0.8, alpha = 0.6) + 
      # stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
      facet_wrap(vars(threshold)) + 
      labs(y = paste0("Rate of ", var.i, "-related cDEGs"), 
           x = "Method",
           color = "Seed") + 
      theme_bw() + 
      theme(# legend.position = "none",
            axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5)) +
      scale_y_continuous(labels = scales::label_number(accuracy = 0.01), breaks = seq(0, 0.8, 0.1), limits = c(0, 0.8))
      ggsave(filename = sprintf("../ApplicationResult/AddViz/ratecDEGs/%s_%s_line_box.eps", date.analysis, var.i),
             plot = g, device = "eps", dpi = 600, width = 9, height = 6, units = "in")
  }
}
```


## Rearrange duplicated rate plots
### All variables
```{r}
##------Load the duplicated rate data modeled under the univariable/multivariable framework------
duplicate_rate_uni <- readRDS("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/UniModel/organized_duplicate_rate_data.noBMA.RDS")

duplicate_rate_multi <- readRDS("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/MultiModel/organized_duplicate_rate_data.RDS")

##------Refine the duplicated rate data------
duplicate_rate_uni$method <- paste0(duplicate_rate_uni$method, "_UVM")
duplicate_rate_multi$method <- ifelse(duplicate_rate_multi$method != "BMAseq", paste0(duplicate_rate_multi$method, "_MVM"), duplicate_rate_multi$method)
duplicate_rate_all <- rbind(duplicate_rate_uni, duplicate_rate_multi)
duplicate_rate_all$method <- factor(duplicate_rate_all$method,
                                    levels = c("BMAseq", "DESeq2_UVM", "DESeq2_MVM", "edgeR_UVM", "edgeR_MVM", "eBayes_UVM", "eBayes_MVM", "VoomLimma_UVM", "VoomLimma_MVM"),
                                    labels = c("BMAseq", "DESeq2_UVM", "DESeq2_MVM", "edgeR_UVM", "edgeR_MVM", "eBayes_UVM", "eBayes_MVM", "voom.limma_UVM", "voom.limma_MVM"))
duplicate_rate_all$variable <- factor(duplicate_rate_all$variable,levels=c("BMI","AGE","SEX","MHABNWBC"))

##------Make the boxplot panel------
date.analysis <- format(Sys.Date(), "%Y%b%d")
g1 <- duplicate_rate_all |>
  ggplot(mapping = aes(y = cDEGs.duplicate.rate.median, x = method, group = method)) + 
  geom_boxplot(outlier.shape = NA, mapping = aes(color = method)) + 
  geom_jitter(aes(color = method), size = 0.8) + 
  facet_wrap(vars(variable)) + 
  geom_hline(aes(yintercept = -Inf)) + 
  geom_vline(aes(xintercept = -Inf)) + 
  coord_cartesian(clip = "off") + 
  scale_y_continuous(labels = scales::label_number(accuracy = 0.01),
                     breaks = seq(0, 1, 0.1),
                     limits = c(0, 1)) + 
  scale_color_manual(values = c("#A73030FF", "#3B3B3BFF", "#0073C2FF", "#EFC000FF", "#868686FF", "#8F7700FF", "#7AA6DCFF", "#003C67FF", "#CD534CFF")) + # Avoid red-green combinations difficult for color-blind readers 
  labs(y = "Median duplicate rate of cDEGs", x = "Method") + 
  theme_BMA()
ggsave(filename = sprintf("../ApplicationResult/AddViz/DuplicatedRate/%s_median.eps", date.analysis),
       plot = g1, device = "eps", dpi = 600, width = 5, height = 5, units = "in")

g2 <- duplicate_rate_all |>
  ggplot(mapping = aes(y = cDEGs.duplicate.rate.mean, x = method, group = method)) + 
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(aes(color = method), size = 0.8) + 
  facet_wrap(vars(variable)) + 
  geom_hline(aes(yintercept = -Inf)) + 
  geom_vline(aes(xintercept = -Inf)) + 
  coord_cartesian(clip = "off") + 
  scale_y_continuous(labels = scales::label_number(accuracy = 0.01),
                     breaks = seq(0, 1, 0.1),
                     limits = c(0, 0.8)) + 
  scale_color_manual(values = c("#A73030FF", "#3B3B3BFF", "#0073C2FF", "#EFC000FF", "#868686FF", "#8F7700FF", "#7AA6DCFF", "#003C67FF", "#CD534CFF")) + # Avoid red-green combinations difficult for color-blind readers 
  labs(y = "Mean duplicate rate of cDEGs", x = "Method") + 
  theme_BMA()
ggsave(filename = sprintf("../ApplicationResult/AddViz/DuplicatedRate/%s_mean.eps", date.analysis),
       plot = g2, device = "eps", dpi = 600, width = 5, height = 5, units = "in")
```


### BMI
```{r}
duplicate_rate_bmi <- filter(duplicate_rate_all, variable == "BMI")
g2 <- duplicate_rate_bmi |>
  ggplot(mapping = aes(y = cDEGs.duplicate.rate.mean, x = method, group = method)) + 
  geom_boxplot(outlier.shape = NA, lwd = 1) + 
  geom_jitter(aes(color = method), size = 1.5) + 
  geom_hline(aes(yintercept = -Inf)) + 
  geom_vline(aes(xintercept = -Inf)) + 
  coord_cartesian(clip = "off") + 
  scale_y_continuous(labels = scales::label_number(accuracy = 0.01)) + 
  scale_color_manual(values = c("#A73030FF", "#3B3B3BFF", "#0073C2FF", "#EFC000FF", "#868686FF", "#8F7700FF", "#7AA6DCFF", "#003C67FF", "#CD534CFF")) + # Avoid red-green combinations difficult for color-blind readers 
  labs(y = "Mean duplicate rate of cDEGs", x = "Method") + 
  theme_BMA()
ggsave(filename = sprintf("../ApplicationResult/AddViz/DuplicatedRate/%s_mean.eps", date.analysis),
       plot = g2, device = "eps", dpi = 600, width = 4, height = 5, units = "in")
```


## Make Venn diagrams in cDEGs between BMAseq and univariable or multivariable single model approaches
```{r layout.1}
threshold.vec <- 5000
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC")
model.method.vec <- c("DESeq2", "edgeR", "eBayes", "VoomLimma")

get_venndata <- function(seed.i = 8809678, 
                         var.name = "BMI", 
                         model.type.1 = "Multi", 
                         model.type.2 = "Uni",
                         model.type.3 = "Multi",
                         model.method.1 = "BMAseq",
                         model.method.2 = "DESeq2",
                         model.method.3 = "DESeq2",
                         threshold = 5000,
                         ...) {
  file.name.1 <- sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sModel/Trim/%s%s%s.RData", 
                         model.type.1, model.method.1, model.type.1, seed.i)
  load(file.name.1)
  if (model.method.1 == "BMAseq") {
    names(BMAseq.eFDR.Main.train) = names(BMAseq.eFDR.Main.test) = var.vec
    cDEGs.1 <- intersect(names(BMAseq.eFDR.Main.train[[var.name]][1:threshold]), 
                         names(BMAseq.eFDR.Main.test[[var.name]][1:threshold]))
  } else if (model.method.1 == "DESeq2") {
    cDEGs.1 <- intersect(DESeq2.eFDR.GeneName.train[[var.name]][1:threshold], 
                         DESeq2.eFDR.GeneName.test[[var.name]][1:threshold])
  } else if (model.method.1 == "edgeR") {
    cDEGs.1 <- intersect(edgeR.eFDR.GeneName.train[[var.name]][1:threshold],
                         edgeR.eFDR.GeneName.test[[var.name]][1:threshold])
  } else if (model.method.1 == "eBayes") {
    cDEGs.1 <- intersect(names(eBayes.eFDR.train2[[var.name]][1:threshold]),
                         names(eBayes.eFDR.test2[[var.name]][1:threshold]))
  } else if (model.method.1 == "VoomLimma") {
    cDEGs.1 <- intersect(names(voom.eFDR.train2[[var.name]][1:threshold]),
                         names(voom.eFDR.test2[[var.name]][1:threshold]))
  }
  
  file.name.2 <- sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sModel/Trim/%s%s%s.RData", 
                         model.type.2, model.method.2, model.type.2, seed.i)
  load(file.name.2)
  if (model.method.2 == "DESeq2") {
    cDEGs.2 <- intersect(DESeq2.eFDR.GeneName.train[[var.name]][1:threshold], 
                         DESeq2.eFDR.GeneName.test[[var.name]][1:threshold])
  } else if (model.method.2 == "edgeR") {
    cDEGs.2 <- intersect(edgeR.eFDR.GeneName.train[[var.name]][1:threshold],
                         edgeR.eFDR.GeneName.test[[var.name]][1:threshold])
  } else if (model.method.2 == "eBayes") {
    cDEGs.2 <- intersect(names(eBayes.eFDR.train2[[var.name]][1:threshold]),
                         names(eBayes.eFDR.test2[[var.name]][1:threshold]))
  } else if (model.method.2 == "VoomLimma") {
    cDEGs.2 <- intersect(names(voom.eFDR.train2[[var.name]][1:threshold]),
                         names(voom.eFDR.test2[[var.name]][1:threshold]))
  }
  
  file.name.3 <- sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sModel/Trim/%s%s%s.RData", 
                         model.type.3, model.method.3, model.type.3, seed.i)
  load(file.name.3)
  if (model.method.3 == "DESeq2") {
    cDEGs.3 <- intersect(DESeq2.eFDR.GeneName.train[[var.name]][1:threshold], 
                         DESeq2.eFDR.GeneName.test[[var.name]][1:threshold])
  } else if (model.method.3 == "edgeR") {
    cDEGs.3 <- intersect(edgeR.eFDR.GeneName.train[[var.name]][1:threshold],
                         edgeR.eFDR.GeneName.test[[var.name]][1:threshold])
  } else if (model.method.3 == "eBayes") {
    cDEGs.3 <- intersect(names(eBayes.eFDR.train2[[var.name]][1:threshold]),
                         names(eBayes.eFDR.test2[[var.name]][1:threshold]))
  } else if (model.method.3 == "VoomLimma") {
    cDEGs.3 <- intersect(names(voom.eFDR.train2[[var.name]][1:threshold]),
                         names(voom.eFDR.test2[[var.name]][1:threshold]))
  }
  
  return(list(cDEGs.1, cDEGs.2, cDEGs.3))
}


plot_venn <- function(seed.i = 8809678, 
                      var.name = "BMI", 
                      model.method.1 = "BMAseq",
                      model.method.2 = "DESeq2",
                      model.method.3 = "DESeq2",
                      model.type.1 = "Multi",
                      model.type.2 = "Uni",
                      model.type.3 = "Multi",
                      threshold = 5000,
                      ...) {

  tri_venndata <- get_venndata(seed.i = seed.i, 
                               var.name = var.name, 
                               model.method.1 = model.method.1,
                               model.method.2 = model.method.2,
                               model.method.3 = model.method.3,
                               model.type.1 = model.type.1,
                               model.type.2 = model.type.2,
                               model.type.3 = model.type.3,
                               threshold = threshold)
  
  # print(tri_venndata)
  
  label_names <- c(paste0(ifelse(
    model.method.1 == "BMAseq",
    "BMAseq",
    ifelse(
      model.type.1 == "Multi",
      paste0(model.method.1, "_MVM"),
      paste0(model.method.1, "_UVM")
    )
  )),
  paste0(ifelse(
    model.method.2 == "BMAseq",
    "BMAseq",
    ifelse(
      model.type.2 == "Multi",
      paste0(model.method.2, "_MVM"),
      paste0(model.method.2, "_UVM")
    )
  )),
  paste0(ifelse(
    model.method.3 == "BMAseq",
    "BMAseq",
    ifelse(
      model.type.3 == "Multi",
      paste0(model.method.3, "_MVM"),
      paste0(model.method.3, "_UVM")
    )
  )))
  label_names <- ifelse(grepl("VoomLimma", label_names) == T, gsub("VoomLimma", "voom.limma", label_names), label_names)
  
  venn_data <- list(tri_venndata[[1]], tri_venndata[[2]], tri_venndata[[3]])
  names(venn_data) <- label_names
  
  make_venn <- ggVennDiagram(
    x = venn_data,
    label = "count",
    label_alpha = 0,
    label_color = "black",
    label_size = 10,
    set_size = 6
  )
  make_venn$layers[[1]]$mapping <- aes(fill = name) # Set the discrete scale
  make_venn <- make_venn + 
    scale_x_continuous(expand = expansion(mult = 0.6)) + # Expand axis to show long set labels
    scale_color_manual(values = c("black", "black", "black")) + # Black circle border
    scale_fill_manual(values = c("#3B499299", "#BB002199", "#00828099", "#A2005699", "#80818099", "#63197999", "#008B4599")) + 
    theme(legend.position = "none",
          plot.title = element_text(size = 20, hjust = 0.5, colour = "red")) + 
  ggtitle(paste0(var.name, " (seed ", seed.i, ", ranking threshold ", threshold, ")"))
  return(make_venn)
}

# Loop every seed, every var.name, every ranking threshold
for (threshold.i in threshold.vec) {
  for (var.i in var.vec) {
    for (seed.i in seed.vec) {
      for (model.method.i in model.method.vec) {
        p1 <- plot_venn(seed.i = seed.i, 
                        var.name = var.i, 
                        model.method.1 = "BMAseq",
                        model.method.2 = model.method.i,
                        model.method.3 = model.method.i,
                        model.type.1 = "Multi",
                        model.type.2 = "Uni",
                        model.type.3 = "Multi",
                        threshold = threshold.i)
        date.analysis <- format(Sys.Date(), "%Y%b%d")
        file.name <- sprintf("../ApplicationResult/AddViz/VennDigram/Top5000/%s/%s_%s_%s.png", model.method.i, date.analysis, var.i, seed.i) 
        ggsave(filename = file.name, plot = p1, device = "png", dpi = 600, width = 12, height = 9, units = "in")        
      }
    }
  }
}
```


```{r layout.2}
threshold.vec <- 5000
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC")
model.method.vec <- c("DESeq2", "edgeR", "eBayes", "VoomLimma")

count_list_3 <- NULL
for (threshold.i in threshold.vec) {
  count_list_2 <- NULL
  for (var.i in var.vec) {
    count_list_1 <- NULL
    for (model.method.i in model.method.vec) {
      V1 = V2 = V3 = V4 = V5 = V6 = V7 = V8 = V9 = V10 = V11 = rep(NA, 7)
      count_df <- data.frame(V1, V2, V3, V4, V5, V6, V7, V8, V9, V10, V11)
      colnames(count_df) <- c("set.name", seed.vec)
      for (seed.i in seed.vec) { 
        tri_venndata <- get_venndata_call(seed.i = seed.i, 
                                     var.name = var.i, 
                                     model.method.1 = "BMAseq",
                                     model.method.2 = model.method.i,
                                     model.method.3 = model.method.i,
                                     model.type.1 = "Multi",
                                     model.type.2 = "Uni",
                                     model.type.3 = "Multi",
                                     threshold = threshold.i)
        model.method.1 = tri_venndata[[4]][["model.method.1"]]
        model.method.2 = model.method.i
        model.method.3 = model.method.i
        model.type.1 = tri_venndata[[4]][["model.type.1"]]
        model.type.2 = tri_venndata[[4]][["model.type.2"]]
        model.type.3 = tri_venndata[[4]][["model.type.3"]]
        
        label_names <- c(paste0(ifelse(
          model.method.1 == "BMAseq", "BMAseq",
          ifelse(model.type.1 == "Multi", 
                 paste0(model.method.1, "_MVM"),
                 paste0(model.method.1, "_UVM")
          )
        )),
        paste0(ifelse(
          model.method.2 == "BMAseq", "BMAseq",
          ifelse(model.type.2 == "Multi", 
                 paste0(model.method.2, "_MVM"),
                 paste0(model.method.2, "_UVM")
          )
        )),
        paste0(ifelse(
          model.method.3 == "BMAseq", "BMAseq",
          ifelse(model.type.3 == "Multi",
                 paste0(model.method.3, "_MVM"),
                 paste0(model.method.3, "_UVM")
          )))
        )
        label_names <- ifelse(grepl("VoomLimma", label_names) == T, 
                              gsub("VoomLimma", "voom.limma", label_names), 
                              label_names)
        
        tri_venndata <- list(tri_venndata[[1]], tri_venndata[[2]], tri_venndata[[3]])
        names(tri_venndata) <- label_names
        processed_data <- Venn(tri_venndata) |> process_data()
        count_data <- processed_data@region |> as.data.frame() |> select(name, count)
        col_idx <- Position(function(x) x == seed.i, seed.vec) # Locate the position of a seed in a seed vector
        if (any(is.na(count_df[, 1]))) { # Assign the set name only once to save the computational steps
          count_df[, 1] <- count_data[["name"]]
        }
        count_df[, col_idx + 1] <- count_data[["count"]]
        # merge(count_df, count_data, by = "name")
      }
      count_list_per_method <- list(count_df)
      names(count_list_per_method) <- paste(threshold.i, var.i, model.method.i, sep = "+")
      count_list_1 <- c(count_list_1, count_list_per_method)
    }
    count_list_per_var <- list(count_list_1)
    names(count_list_per_var) <- var.i
    count_list_2 <- c(count_list_2, count_list_per_var)
  }
  count_list_per_threshold <- list(count_list_2)
  names(count_list_per_threshold) <- threshold.i
  count_list_3 <- c(count_list_3, count_list_per_threshold)
  return(count_list_3)
}

analysis.date <- format(Sys.Date(), "%Y%b%d")
saveRDS(object = count_list_3, file = paste(analysis.date, "count_list_3.RDS", sep = "_"))

##------Make a boxplot data organizing function------
get_boxplotdata <- function(var.name = var.i) {
  df <- do.call(rbind, count_list_3[["5000"]][[var.name]])
  df$model.method <- c(rep(c("DESeq2", "edgeR", "eBayes", "voom.limma"), each = 7))
  df.long <- pivot_longer(data = df,
                          cols = !c(set.name, model.method),
                          names_to = "seed",
                          values_to = "count")
  df.long$set.name <- ifelse(df.long$set.name == "BMAseq", 
                             paste0(df.long$set.name, "_vs_", df.long$model.method),
                             df.long$set.name)
  df.long$set.name <- gsub("^BMAseq_vs_DESeq2$", "BMAseq_vs_DESeq2_UVM_vs_DESeq2_MVM", df.long$set.name)
  df.long$set.name <- gsub("^BMAseq_vs_edgeR$", "BMAseq_vs_edgeR_UVM_vs_edgeR_MVM", df.long$set.name)
  df.long$set.name <- gsub("^BMAseq_vs_eBayes$", "BMAseq_vs_eBayes_UVM_vs_eBayes_MVM", df.long$set.name)
  df.long$set.name <- gsub("^BMAseq_vs_voom.limma$", "BMAseq_vs_voom.limma_UVM_vs_voom.limma_MVM", df.long$set.name)
  df.long$set.name <- gsub("^DESeq2_UVM$", "DESeq2_UVM_vs_BMAseq_vs_DESeq2_MVM", df.long$set.name)
  df.long$set.name <- gsub("^DESeq2_MVM$", "DESeq2_MVM_vs_BMAseq_vs_DESeq2_UVM", df.long$set.name)
  df.long$set.name <- gsub("^edgeR_UVM$", "edgeR_UVM_vs_BMAseq_vs_edgeR_MVM", df.long$set.name)
  df.long$set.name <- gsub("^edgeR_MVM$", "edgeR_MVM_vs_BMAseq_vs_edgeR_UVM", df.long$set.name)
  df.long$set.name <- gsub("^eBayes_UVM$", "eBayes_UVM_vs_BMAseq_vs_eBayes_MVM", df.long$set.name)
  df.long$set.name <- gsub("^eBayes_MVM$", "eBayes_MVM_vs_BMAseq_vs_eBayes_UVM", df.long$set.name)
  df.long$set.name <- gsub("^voom.limma_UVM$", "voom.limma_UVM_vs_BMAseq_vs_voom.limma_MVM", df.long$set.name)
  df.long$set.name <- gsub("^voom.limma_MVM$", "voom.limma_MVM_vs_BMAseq_vs_voom.limma_UVM", df.long$set.name)
  
  df.long$set.name <- factor(df.long$set.name, 
                             levels = c(
                               "BMAseq..DESeq2_UVM..DESeq2_MVM",
                               "BMAseq..edgeR_UVM..edgeR_MVM",
                               "BMAseq..eBayes_UVM..eBayes_MVM",
                               "BMAseq..voom.limma_UVM..voom.limma_MVM",
                               
                               "BMAseq..DESeq2_UVM",
                               "BMAseq..DESeq2_MVM",
                               "BMAseq..edgeR_UVM",
                               "BMAseq..edgeR_MVM",
                               "BMAseq..eBayes_UVM",
                               "BMAseq..eBayes_MVM",
                               "BMAseq..voom.limma_UVM",
                               "BMAseq..voom.limma_MVM",
                               "DESeq2_UVM..DESeq2_MVM",
                               "edgeR_UVM..edgeR_MVM",
                               "eBayes_UVM..eBayes_MVM",
                               "voom.limma_UVM..voom.limma_MVM",
                               
                               "BMAseq_vs_DESeq2_UVM_vs_DESeq2_MVM",
                               "BMAseq_vs_edgeR_UVM_vs_edgeR_MVM",
                               "BMAseq_vs_eBayes_UVM_vs_eBayes_MVM",
                               "BMAseq_vs_voom.limma_UVM_vs_voom.limma_MVM",
                               "DESeq2_UVM_vs_BMAseq_vs_DESeq2_MVM",
                               "DESeq2_MVM_vs_BMAseq_vs_DESeq2_UVM",
                               "edgeR_UVM_vs_BMAseq_vs_edgeR_MVM",
                               "edgeR_MVM_vs_BMAseq_vs_edgeR_UVM",
                               "eBayes_UVM_vs_BMAseq_vs_eBayes_MVM",
                               "eBayes_MVM_vs_BMAseq_vs_eBayes_UVM",
                               "voom.limma_UVM_vs_BMAseq_vs_voom.limma_MVM",
                               "voom.limma_MVM_vs_BMAseq_vs_voom.limma_UVM"
                             ))
  return(df.long)
}

##------Set the plot theme------
theme_BMA <- function(base_size = 12,
                      base_family = "Arial") {
  theme_classic(base_size = base_size, base_family = base_family) +
    theme(
      legend.position = "none",
      panel.background = element_rect(fill = "white", colour = "white"),
      axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5),
      strip.background = element_rect(fill = "white", color = "black", linewidth = 0.6), # Customize the box border in strip
      strip.placement = "inside"
    )
}

##------Make a boxplot plotting function------
make_boxplot <- function(var.name = var.name, data.set = data.set) {
  p <- ggplot(data = data.set,
              aes(y = count, x = set.name)) + 
    geom_boxplot(aes(group = set.name, color = set.name),
                 outlier.shape = NA) + 
    geom_jitter(aes(color = set.name), size = 0.8) + 
    geom_vline(xintercept = 16.5, linetype = "dashed") + 
    annotate(geom = "text", x = 8.5, y = 1200, label = "Shared discovery", size = 5) + 
    annotate(geom = "text", x = 22.5, y = 1200, label = "Unique discovery", size = 5) + 
    xlab("Set name") + 
    ylab(paste0("Number of ", var.name, "-related cDEGs")) + 
    theme_BMA() + 
    scale_color_manual(values = (c(rep("#DF8F44FF", 4), rep("#B24745FF", 12), rep("#6A6599FF", 12)))) + 
    coord_flip() 
  date.analysis <- format(Sys.Date(), "%Y%b%d")
  ggsave(filename = sprintf("../ApplicationResult/AddViz/VennDigram/Top5000/All/%s_%s.png", date.analysis, var.name),
         plot = p, device = "png", dpi = 600, width = 11, height = 7, units = "in")
}

var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC")
for (var.i in var.vec) {
  boxplot.data <- get_boxplotdata(var.name = var.i)
  make_boxplot(var.name = var.i, data.set = boxplot.data)
}
```


## Make complicated heatmap and boxplot in unique and shared cDEGs between BMAseq and univariable or multivariable single model approaches
```{r}
##------Estimate the number of cDEGs of each class per seed------
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC")
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
threshold.vec <- 5000

for (threshold.i in threshold.vec) {
  for (var.name in var.vec) {
    class.freq.list <- vector(mode = "list", length = length(seed.vec))
    names(class.freq.list) <- seed.vec
    for (seed.i in seed.vec) {
      target.files <- list.files(path = "../ApplicationData/derived/RandomSeed/Top5000/MultiModel/", pattern = sprintf("*Multi%s.RData", seed.i), full.names = T)
      temp <- new.env()
      lapply(target.files, load, temp)
      names(temp$BMAseq.eFDR.Main.train) = names(temp$BMAseq.eFDR.Main.test) = var.vec
      BMAseq.cDEGs <- with(temp, intersect(names(BMAseq.eFDR.Main.train[[var.name]][1:threshold.i]), names(BMAseq.eFDR.Main.test[[var.name]][1:threshold.i])))
      DESeq2.MVM.cDEGs <- with(temp, intersect(DESeq2.eFDR.GeneName.train[[var.name]][1:threshold.i], DESeq2.eFDR.GeneName.test[[var.name]][1:threshold.i]))
      edgeR.MVM.cDEGs <- with(temp, intersect(edgeR.eFDR.GeneName.train[[var.name]][1:threshold.i], edgeR.eFDR.GeneName.test[[var.name]][1:threshold.i]))
      eBayes.MVM.cDEGs <- with(temp, intersect(names(eBayes.eFDR.train2[[var.name]][1:threshold.i]), names(eBayes.eFDR.test2[[var.name]][1:threshold.i])))
      voom.limma.MVM.cDEGs <- with(temp, intersect(names(voom.eFDR.train2[[var.name]][1:threshold.i]), names(voom.eFDR.test2[[var.name]][1:threshold.i])))
      rm(temp)
      
      target.files <- list.files(path = "../ApplicationData/derived/RandomSeed/Top5000/UniModel/", pattern = sprintf("*Uni%s.RData", seed.i), full.names = T)
      temp <- new.env()
      lapply(target.files, load, temp)
      DESeq2.UVM.cDEGs <- with(temp, intersect(DESeq2.eFDR.GeneName.train[[var.name]][1:threshold.i], DESeq2.eFDR.GeneName.test[[var.name]][1:threshold.i]))
      edgeR.UVM.cDEGs <- with(temp, intersect(edgeR.eFDR.GeneName.train[[var.name]][1:threshold.i], edgeR.eFDR.GeneName.test[[var.name]][1:threshold.i]))
      eBayes.UVM.cDEGs <- with(temp, intersect(names(eBayes.eFDR.train2[[var.name]][1:threshold.i]), names(eBayes.eFDR.test2[[var.name]][1:threshold.i])))
      voom.limma.UVM.cDEGs <- with(temp, intersect(names(voom.eFDR.train2[[var.name]][1:threshold.i]), names(voom.eFDR.test2[[var.name]][1:threshold.i])))
      rm(temp)
      cDEGs.all <- unique(c(BMAseq.cDEGs, DESeq2.MVM.cDEGs, edgeR.MVM.cDEGs, eBayes.MVM.cDEGs, voom.limma.MVM.cDEGs,
                            DESeq2.UVM.cDEGs, edgeR.UVM.cDEGs, eBayes.UVM.cDEGs, voom.limma.UVM.cDEGs))
      no.cDEGs <- length(cDEGs.all)
      cDEGs.mat <- matrix(0, nrow = no.cDEGs, ncol = 10) 
      rownames(cDEGs.mat) <- cDEGs.all
      colnames(cDEGs.mat) <- c("BMAseq", "DESeq2_UVM", "DESeq2_MVM", "edgeR_UVM", "edgeR_MVM",
                               "eBayes_UVM", "eBayes_MVM", "voom.limma_UVM", "voom.limma_MVM", "Class")
      cDEGs.mat[, 1] <- as.integer(cDEGs.all %in% BMAseq.cDEGs)
      cDEGs.mat[, 2] <- as.integer(cDEGs.all %in% DESeq2.UVM.cDEGs)
      cDEGs.mat[, 3] <- as.integer(cDEGs.all %in% DESeq2.MVM.cDEGs)
      cDEGs.mat[, 4] <- as.integer(cDEGs.all %in% edgeR.UVM.cDEGs)
      cDEGs.mat[, 5] <- as.integer(cDEGs.all %in% edgeR.MVM.cDEGs)
      cDEGs.mat[, 6] <- as.integer(cDEGs.all %in% eBayes.UVM.cDEGs)
      cDEGs.mat[, 7] <- as.integer(cDEGs.all %in% eBayes.MVM.cDEGs)
      cDEGs.mat[, 8] <- as.integer(cDEGs.all %in% voom.limma.UVM.cDEGs)
      cDEGs.mat[, 9] <- as.integer(cDEGs.all %in% voom.limma.MVM.cDEGs)
      cDEGs.mat[, 10] <- apply(cDEGs.mat[, 1:9], 1, function(x) paste(x, collapse = ""))
      cDEGs.df <- cDEGs.mat |> as.data.frame()
      
      file.name <- paste0("../ApplicationData/derived/RandomSeed/HeatmapBoxplotData/", var.name, "_", threshold.i, "_", seed.i)
      saveRDS(cDEGs.df, paste0(file.name, ".RDS"))
      write.xlsx(x = cDEGs.df, file = paste0(file.name, ".xlsx"), rowNames = T, colWidths = 40, firstActiveRow = 2, firstActiveCol = 2)
      
      class.freq <- cDEGs.df |>
        dplyr::count(Class, sort = T, name = "Frequency") |>
        mutate(Seed = seed.i) |>
        select(Seed, Class, Frequency)
      
      class.freq.list[[seed.i]] <- class.freq
    }
    class.freq.df <- do.call(rbind, class.freq.list)
    saveRDS(class.freq.df, file = paste0("../ApplicationData/derived/RandomSeed/HeatmapBoxplotData/", var.name, "_", threshold.i, "_ClassFreq"))
  }
}


##------Plotting------
# Set the plot theme
theme_BMA <- function(base_size = 12,
                      base_family = "Arial") {
  theme_classic(base_size = base_size, base_family = base_family) +
    theme(
      legend.position = "none",
      panel.background = element_rect(fill = "white", colour = "white"),
      axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5),
      strip.background = element_rect(fill = "white", color = "black", linewidth = 0.6), # Customize the box border in strip
      strip.placement = "inside"
    )
}

# Make a function to count the number of 1's in a string
count_ones <- function(string) {
  sum(as.integer(strsplit(string, "")[[1]]))
}

# Make a function to calculate the number of ones and their positions in a string
get_ones_info <- function(str) {
  ones_count <- sum(strsplit(str, "")[[1]] == "1")
  ones_positions <- which(strsplit(str, "")[[1]] == "1")
  return(list(string = str, count = ones_count, positions = ones_positions))
}

# Make the visual for each phenotype
threshold.i <- 5000
date.analysis <- format(Sys.Date(), "%Y%b%d")
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC")

for (var.name in var.vec) {
  class.freq <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/HeatmapBoxplotData/%s_5000_ClassFreq", var.name))
  class.freq.top10.per.seed <- class.freq |> 
    group_by(Seed) |> 
    slice_max(order_by = Frequency, n = 10, with_ties = F) |>
    ungroup() 
  
  # Filter out the top 10 most frequent class that appears less than 5 times across 10 seeds
  uniq_class <- unique(class.freq.top10.per.seed$Class)
  uniq_class_freq <- table(factor(class.freq.top10.per.seed$Class, levels = uniq_class))
  uniq_class_eligible <- uniq_class[uniq_class_freq >= 5]
  class.freq.top10.per.seed <- class.freq.top10.per.seed |> filter(Class %in% uniq_class_eligible)
  
  # Factorize and order the levels based on count of 1's
  uniq_class_2 <- unique(class.freq.top10.per.seed$Class)
  
  ones_info <- lapply(uniq_class_2, get_ones_info)
  
  df <- do.call(rbind, ones_info) |> data.frame()
  df <- df[order(df$count |> unlist(), sapply(df$positions, function(x) min(x))), ]
  new_class_level <- df$string |> unlist()
  
  # Reorder the class levels for plotting
  class.freq.top10.per.seed$Class <- factor(class.freq.top10.per.seed$Class, levels = rev(new_class_level))
  
  p1 <- ggplot(data = class.freq.top10.per.seed,
               aes(y = Frequency,
                   x = Class)) + 
    geom_boxplot(aes(group = Class, color = Class),
                 outlier.shape = NA) + 
    geom_jitter(aes(color = Class), size = 0.8) + 
    ylab(paste0("Number of ", var.name, "-related cDEGs")) + 
    xlab("") + 
    theme_BMA() + 
    coord_flip() + 
    theme(axis.text.x = element_text(angle = 0),
          axis.text.y = element_blank(),
          axis.ticks.y = element_blank()) + 
    scale_color_viridis(discrete = T,
                        direction = -1)
  
  aux <- ggplot_build(p1)
  y_ori <- aux[["layout"]][["panel_scales_y"]][[1]][["range"]][["range"]][2]
  if (var.name == "BMI") {
    y_pos <- y_ori - 300
  } else if (var.name == "AGE") {
    y_pos <- y_ori - 400
  } else if (var.name == "SEX") {
    y_pos <- y_ori - 400
  } else if (var.name == "MHABNWBC") {
    y_pos <- y_ori - 900
  }
  
  p1 <- p1 + 
    annotate("text", x = 9, y = y_pos, label = "a1: BMAseq", hjust = 0) +
    annotate("text", x = 8.5, y = y_pos, label = "a2: DESeq2_UVM", hjust = 0) +
    annotate("text", x = 8, y = y_pos, label = "a3: DESeq2_MVM", hjust = 0) +
    annotate("text", x = 7.5, y = y_pos, label = "a4: edgeR_UVM", hjust = 0) +
    annotate("text", x = 7, y = y_pos, label = "a5: edgeR_MVM", hjust = 0) +
    annotate("text", x = 6.5, y = y_pos, label = "a6: eBayes_UVM", hjust = 0) +
    annotate("text", x = 6, y = y_pos, label = "a7: eBayes_MVM", hjust = 0) +
    annotate("text", x = 5.5, y = y_pos, label = "a8: voom.limma_UVM", hjust = 0) +
    annotate("text", x = 5, y = y_pos, label = "a9: voom.limma_MVM", hjust = 0)
  
  # Set the input vector
  input_vector <- uniq_class_2
  
  # Set the approach vector
  approach_positions <- c("BMAseq", "DESeq2_UVM", "DESeq2_MVM", "edgeR_UVM", "edgeR_MVM", "eBayes_UVM", "eBayes_MVM", "voom.limma_UVM", "voom.limma_MVM")
  
  # Set the empty data frame to store the results
  result_df <- data.frame(Y = character(), X = character(), Z = integer(), stringsAsFactors = FALSE)
  
  # Process each string in the input vector
  for (i in seq_along(input_vector)) {
    string <- input_vector[i]
    
    # Find the positions of 1s in the string
    ones_positions <- which(strsplit(string, "")[[1]] == "1")
    
    # Create rows for the data frame based on the positions
    for (pos in ones_positions) {
      # Map the position of 1 to the approach
      row <- data.frame(Y = string, X = approach_positions[pos], Z = 1)
      result_df <- rbind(result_df, row)
    }
  }
  
  result_df$X <- factor(result_df$X, levels = approach_positions)
  result_df$Y <- factor(result_df$Y, levels = rev(new_class_level))
  p2 <- ggplot(result_df, aes(X, Y, fill = Z)) + 
    geom_tile(color = "white",
              lwd = 1.5,
              linetype = 1,
              aes(fill = Y)) + 
    labs(x = "Approach",
         y = "Class") + 
    theme_BMA() + 
    theme(axis.text.x = element_text(angle = 0, hjust = 1, color = "black", vjust = 0.5),
          axis.text.y = element_text(color = "black"),
          axis.line = element_blank(),
          axis.ticks = element_blank(),
          panel.border = element_rect(colour = "black", fill = NA, linewidth = 2)) + 
    scale_x_discrete(labels = paste0("a", 1:9)) + 
    scale_fill_viridis(discrete = T,
                       direction = -1)
  
  g <- cowplot::plot_grid(p2, p1)
  ggsave(filename = sprintf("../ApplicationResult/AddViz/HeatmapBoxplot/%s_%s_%s.eps", date.analysis, var.name, threshold.i),
         plot = g,
         device = cairo_ps,
         dpi = 600,
         width = 11,
         height = 5,
         units = "in")
}
```


# Session information
```{r}
sessionInfo()
```

