---
title: "Assessment of BMAseq and other single-model approaches on inferring replicable differentially expressed genes"
author: "Anni Liu"
date: "`r format(Sys.Date(), '%B %d, %Y')`"
output: rmarkdown::html_vignette 
---

```{r,echo=FALSE}
knitr::opts_chunk$set(
  # options(scipen = 999, digits = 3),
  cache = TRUE,
  error = FALSE,
  message = FALSE,
  warning = FALSE,
  tidy.opts = list(width.cutoff = 60),
  fig.width = 8,
  fig.height = 6
)
```


# Attach libraries and functions
```{r}
easypackages::libraries("BMAseq", "limma", "qvalue", "parallel", "tidyverse", "edgeR", "DESeq2", "data.table", "ggVennDiagram", "gridExtra") |> suppressPackageStartupMessages()

# Attach functions in {BMAseq} (2022Dec5 version)  
walk(c("Bayesfactor.R", "BMAseq.multi.DEG.old.R", "Post.incl.modelprob.old.R", "eFDR.DEG.old.R", "DEG.bestmodel.old.R"), source)
```


# Process data
```{r}
dat.expr <- dget("../ApplicationData/derived/dat.expr.Subcutaneous") 
dat.pheno <- dget("../ApplicationData/derived/dat.pheno.Subcutaneous") 
dim(dat.expr) # 24660 genes and 404 subjects
dim(dat.pheno) # 404 subjects and 13 phenotypes
dat.pheno[1:5, ]
dat.expr[1:5, 1:5]
paste0("The column names of gene expression data", ifelse(all(colnames(dat.expr) == rownames(dat.pheno)), " MATCH ", "NOT MATCH"), "the row names of phenotype data.")

# Pre-filter the genes
# Here we perform the median absolute deviation with the threshold of 0.8 to select genes that are most likely to distinguish the samples
dat.expr.new <- dat.expr[apply(dat.expr, 1, function(x) mad(x) > 0.8), ] # We have 24455 genes

seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
for (seed.i in seed.vec) {
  
  # Divide the gene expression data and phenotype data into the 50% training set and 50% test set 
  test.ind <- sample(1:nrow(dat.pheno), ceiling(0.5*nrow(dat.pheno)))
  dat.pheno.train <- dat.pheno[-test.ind, ]
  dat.pheno.test <- dat.pheno[test.ind, ]
  dat.expr.train <- dat.expr.new[, rownames(dat.pheno.train)]
  dat.expr.test <- dat.expr.new[, rownames(dat.pheno.test)]
  
  # Check if the gene expression data matches the phenotype data  
  print(paste0("The column names of training gene expression data", ifelse(all(colnames(dat.expr.train) == rownames(dat.pheno.train)), " MATCH ", "NOT MATCH"), "the row names of training phenotype data."))
  print(paste0("The column names of test gene expression data", ifelse(all(colnames(dat.expr.test) == rownames(dat.pheno.test)), " MATCH ", "NOT MATCH"), "the row names of test phenotype data."))
  
  # Make the table one of patient characteristics per random seed trial
  dat.pheno.train$group <- "Training"
  dat.pheno.test$group <- "Test"
  dat.pheno.all <- rbind(dat.pheno.train, dat.pheno.test)
  tab1 <- tbl_summary(data = dat.pheno.all,
                      by = "group") %>% 
    add_p() %>% 
    as_gt() %>%
    gt::gtsave(filename = paste0("../ApplicationResult/Multi/RandomSeed/TMM_Top2000/", date.analysis, "_", "Table1", "_", seed.num, ".rtf"))
  
  # Save gene expression and phenotype data
  saveRDS(dat.expr.train, 
          file = sprintf("../ApplicationData/derived/RandomSeed/dat.expr.train%s.RDS", seed.i))
  saveRDS(dat.expr.test, 
          file = sprintf("../ApplicationData/derived/RandomSeed/dat.expr.test%s.RDS", seed.i))
  saveRDS(dat.pheno.train, 
          file = sprintf("../ApplicationData/derived/RandomSeed/dat.pheno.train%s.RDS", seed.i))
  saveRDS(dat.pheno.test, 
          file = sprintf("../ApplicationData/derived/RandomSeed/dat.pheno.test%s.RDS", seed.i))
}
```


# Modify BMAseq functions to be compatible with the trimmed-mean of M-values (TMM)
```{r}
##------BMAseq.multi.postprob.norm------
BMAseq.multi.postprob.norm <- function(dat.expr.counts, dat.pheno, var.pool, max.nvar, interaction = NULL, cut.BF = 1) {


    if (sum(colnames(dat.expr.counts) != rownames(dat.pheno)) > 0)
        stop("Two datasets dat.expr.counts and dat.pheno must be matched by subjects.")
    if (is.null(var.pool) || length(var.pool) < 2)
        stop("Must provide at least two variables of interest.")
    if (sum(var.pool %in% colnames(dat.pheno)) != length(var.pool))
        stop("Variables of interest must be in datasets dat.pheno.")
    if (is.null(max.nvar))
        stop("Must provide max.nvar.")


    n.var <- length(var.pool)
    n.gene <- nrow(dat.expr.counts)


    # Estimate voom weights
    y0 <- voom(counts = dat.expr.counts, 
               lib.size = colSums(dat.expr.counts)*calcNormFactors(dat.expr.counts))
    input.dat.expr <- y0$E
    input.weights <- y0$weights


    # Build model space
    result.modelspace <- Modelspace(dat.pheno, var.pool, max.nvar, interaction = interaction)
    input.model.space <- result.modelspace$model.space
    input.dat.pheno <- result.modelspace$dat.pheno.new


    # Calculate Bayes factor
    out.bf <- vapply(1:length(input.model.space), function(i) {
        print(paste0(i, ". ", input.model.space[i]))
        design <- model.matrix(formula(input.model.space[i]), data = input.dat.pheno)
        colnames(design)[1] <- "Int"
        vapply(1:n.gene, function(k) Bayesfactor(design, input.dat.expr[k, ], input.weights[k, ]), FUN.VALUE = as.double(1))
    }, FUN.VALUE = as.double(1:n.gene))


    # Obtain prior model probability
    bf.max <- t(apply(out.bf, 1, function(x) ifelse(x == max(x) & x > cut.BF, 1, 0)))
    prior.modelprob <- apply(bf.max, 2, sum)/n.gene
    pm.est <- prior.modelprob
    n.model <- length(prior.modelprob)
    pm.prior0 <- pm.prior1 <- rep(0.5/(n.model - 1), n.model - 1)
    alpha <- 0.2
    for (j in 1:30) {
        pm.prior1 <- pm.est[-1]/apply(t(apply(out.bf[, -1], 1, function(x) x/(1 - sum(pm.prior0) + sum(x * pm.prior0)))), 2,
            mean)
        pm.prior0 <- pm.prior0 + (pm.prior1 - pm.prior0) * alpha
    }
    prior.modelprob <- c(1 - sum(pm.prior0), pm.prior0)


    # Calculate posterior model probability
    post.modelprob <- t(apply(out.bf, 1, function(x) x * prior.modelprob/sum(x * prior.modelprob)))
    rownames(post.modelprob) <- rownames(dat.expr.counts)
    colnames(post.modelprob) <- input.model.space


    return(list(dat.expr.logcpm = input.dat.expr, weights = input.weights, model.space = input.model.space, dat.pheno.new = input.dat.pheno, post.modelprob = post.modelprob))
}
```


# Univariable analysis using the single-model approaches
## Create functions of building univariable models 
```{r}
uni_TMM_top <- function(seed.num = 999, threshold = 2000) {
  ##------Set variables of interest------
  vars <- c("BMI", "AGE", "SEX", "MHABNWBC", "MHARTHTS", "MHCVD", "MHASTHMA", "MHBCTINF", "MHBLDDND", "MHCOCAINE5", "MHCOPD", "MHDRNKSTS", "SMTSPAX") 
  
  
  ##------voom + limma------
  design.train <- mclapply(1:length(vars),
                           function(i) model.matrix(~dat.pheno.train[vars][[i]]), 
                           mc.cores = 10)
  
  design.test <- mclapply(1:length(vars),
                          function(i) model.matrix(~dat.pheno.test[vars][[i]]),
                          mc.cores = 10)
  
  voom.train <- mclapply(1:length(vars),
                         function(i) voom(counts = dat.expr.train,
                                          design = design.train[[i]],
                                          lib.size = colSums(dat.expr.train)*calcNormFactors(dat.expr.train)),
                         mc.cores = 10)
  
  voom.fit.train <- mclapply(1:length(vars),
                             function(i) lmFit(voom.train[[i]][["E"]],
                                               design = design.train[[i]],
                                               weights = voom.train[[i]][["weights"]]),
                             mc.cores = 10)
  
  voom.eFDR.train <- mclapply(1:length(vars),
                              function(i) {
                                t <- voom.fit.train[[i]][["coefficients"]][, 2]/voom.fit.train[[i]][["stdev.unscaled"]][, 2]/voom.fit.train[[i]][["sigma"]]
                                p <- 2*pt(-abs(t), df = voom.fit.train[[i]][["df.residual"]])
                                return(qvalue(p)) },
                              mc.cores = 10)
  
  voom.eFDR.train2 <- mclapply(1:length(vars),
                             function(i) {
                               q.val <- voom.eFDR.train[[i]][["qvalues"]]
                               return(q.val[order(q.val)[1:threshold]])
                             },
                             mc.cores = 10)
  
  names(design.train) = names(voom.train) = names(voom.fit.train) = names(voom.eFDR.train) = names(voom.eFDR.train2) = vars 
  
  voom.test <- mclapply(1:length(vars),
                        function(i) voom(counts = dat.expr.test,
                                         design = design.test[[i]],
                                         lib.size = colSums(dat.expr.test)*calcNormFactors(dat.expr.test)),
                        mc.cores = 10)
  
  voom.fit.test <- mclapply(1:length(vars),
                            function(i) lmFit(voom.test[[i]][["E"]],
                                              design = design.test[[i]],
                                              weights = voom.test[[i]][["weights"]]), 
                            mc.cores = 10)
  
  voom.eFDR.test <- mclapply(1:length(vars),
                             function(i) {
                               t <- voom.fit.test[[i]][["coefficients"]][, 2]/voom.fit.test[[i]][["stdev.unscaled"]][, 2]/voom.fit.test[[i]][["sigma"]]
                               p <- 2*pt(-abs(t), df = voom.fit.test[[i]][["df.residual"]])
                               return(qvalue(p)) },
                             mc.cores = 10)
  
  voom.eFDR.test2 <- mclapply(1:length(vars),
                              function(i) {
                                q.val <- voom.eFDR.test[[i]][["qvalues"]]
                                return(q.val[order(q.val)[1:threshold]])
                              },
                              mc.cores = 10)
  
  names(design.test) = names(voom.test) = names(voom.fit.test) = names(voom.eFDR.test) = names(voom.eFDR.test2) = vars 
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, 
       vars, design.train, design.test, 
       voom.train, voom.fit.train, voom.eFDR.train, voom.eFDR.train2, 
       voom.test, voom.fit.test, voom.eFDR.test, voom.eFDR.test2, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/VoomLimmaUni%s.RData", threshold, seed.num))

  
  ##------voom + limma + eBayes------
  eBayes.fit.train <- mclapply(1:length(vars),
                               function(i) lmFit(voom.train[[i]][["E"]],
                                                 design = design.train[[i]],
                                                 weights = voom.train[[i]][["weights"]]) %>% eBayes(), 
                               mc.cores = 10)
  
  eBayes.eFDR.train <- mclapply(1:length(vars),
                                function(i) eBayes.fit.train[[i]][["p.value"]][, 2] %>% qvalue(),
                                mc.cores = 10)
  
  eBayes.eFDR.train2 <- mclapply(1:length(vars),
                                 function(i) {
                                   q.val <- eBayes.eFDR.train[[i]][["qvalues"]]
                                   return(q.val[order(q.val)[1:threshold]])},
                                 mc.cores = 10)
  
  names(eBayes.fit.train) = names(eBayes.eFDR.train) = names(eBayes.eFDR.train2) = vars 
  

  eBayes.fit.test <- mclapply(1:length(vars),
                              function(i) lmFit(voom.test[[i]][["E"]],
                                                design = design.test[[i]],
                                                weights = voom.test[[i]][["weights"]]) %>% eBayes(), 
                              mc.cores = 10)
  
  eBayes.eFDR.test <- mclapply(1:length(vars),
                               function(i) eBayes.fit.test[[i]][["p.value"]][, 2] %>% qvalue(),
                               mc.cores = 10)
  
  eBayes.eFDR.test2 <- mclapply(1:length(vars),
                                function(i) {
                                  q.val <- eBayes.eFDR.test[[i]][["qvalues"]]
                                  return(q.val[order(q.val)[1:threshold]])},
                                mc.cores = 10)
  
  names(eBayes.fit.test) = names(eBayes.eFDR.test) = names(eBayes.eFDR.test2) = vars 

  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, 
       vars, design.train, design.test, 
       voom.train, eBayes.fit.train, eBayes.eFDR.train, eBayes.eFDR.train2,
       voom.test, eBayes.fit.test, eBayes.eFDR.test, eBayes.eFDR.test2, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/eBayesUni%s.RData", threshold, seed.num))
  
  
  ##------edgeR------
  y.train <- mclapply(1:length(vars),
                      function(i) DGEList(counts = dat.expr.train, 
                                          lib.size = colSums(dat.expr.train)) %>% 
                        calcNormFactors() %>% 
                        estimateGLMTrendedDisp(design.train[[i]]),
                      mc.cores = 10) 
  
  edgeR.fit.train <- mclapply(1:length(vars),
                              function(i) glmQLFit(y.train[[i]], design.train[[i]]),
                              mc.cores = 10)
  
  qlf.train <- mclapply(1:length(vars),
                        function(i) glmQLFTest(edgeR.fit.train[[i]], coef = 2),
                        mc.cores = 10)
  
  edgeR.eFDR.train <- mclapply(1:length(vars),
                               function(i) qlf.train[[i]][["table"]][["PValue"]] %>% qvalue(),
                               mc.cores = 10)
  
  edgeR.eFDR.train2 <- mclapply(1:length(vars),
                                function(i) {
                                  q.val <- edgeR.eFDR.train[[i]][["qvalues"]]
                                  return(q.val[order(q.val)[1:threshold]]) 
                                },
                                mc.cores = 10)
  
  edgeR.eFDR.GeneName.train <- mclapply(1:length(vars),
                                        function(i) {
                                          q.val <- edgeR.eFDR.train[[i]][["qvalues"]]
                                          return(rownames(dat.expr.train)[order(q.val)[1:threshold]])
                                        },
                                        mc.cores = 10)
  
  names(y.train) = names(edgeR.fit.train) = names(qlf.train) = names(edgeR.eFDR.train) = names(edgeR.eFDR.train2) = names(edgeR.eFDR.GeneName.train) = vars 
  
  y.test <- mclapply(1:length(vars),
                     function(i) DGEList(counts = dat.expr.test, 
                                         lib.size = colSums(dat.expr.test)) %>%
                       calcNormFactors() %>%
                       estimateGLMTrendedDisp(design.test[[i]]),
                     mc.cores = 10) 
  
  edgeR.fit.test <- mclapply(1:length(vars),
                             function(i) glmQLFit(y.test[[i]], design.test[[i]]),
                             mc.cores = 10)
  
  qlf.test <- mclapply(1:length(vars),
                       function(i) glmQLFTest(edgeR.fit.test[[i]], coef = 2),
                       mc.cores = 10)
  
  edgeR.eFDR.test <- mclapply(1:length(vars),
                              function(i) qlf.test[[i]][["table"]][["PValue"]] %>% qvalue(),
                              mc.cores = 10)
  
  edgeR.eFDR.test2 <- mclapply(1:length(vars),
                               function(i) {
                                 q.val <- edgeR.eFDR.test[[i]][["qvalues"]]
                                 return(q.val[order(q.val)[1:threshold]]) 
                               },
                               mc.cores = 10)
  
  edgeR.eFDR.GeneName.test <- mclapply(1:length(vars),
                                       function(i) {
                                         q.val <- edgeR.eFDR.test[[i]][["qvalues"]]
                                         return(rownames(dat.expr.test)[order(q.val)[1:threshold]])
                                       },
                                       mc.cores = 10)
  
  names(y.test) = names(edgeR.fit.test) = names(qlf.test) = names(edgeR.eFDR.test) = names(edgeR.eFDR.test2) = names(edgeR.eFDR.GeneName.test) = vars 
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, 
       vars, design.train, design.test, 
       y.train, edgeR.fit.train, qlf.train, edgeR.eFDR.train, edgeR.eFDR.train2, edgeR.eFDR.GeneName.train, 
       y.test, edgeR.fit.test, qlf.test, edgeR.eFDR.test, edgeR.eFDR.test2, edgeR.eFDR.GeneName.test, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/edgeRUni%s.RData", threshold, seed.num))
  
  
  ##------DESeq2------
  cts.train <- as.matrix(dat.expr.train)
  cts.test <- as.matrix(dat.expr.test)
  coldata.train <- dat.pheno.train
  coldata.test <- dat.pheno.test
  
  # Check the sample order of the count matrix and column data in the training and test sets 
  all(rownames(coldata.test) == colnames(cts.test))
  all(rownames(coldata.train) == colnames(cts.train))
  
  # Transform the TMM normalization factors to be used in DESeq2
  lib.size <- colSums(cts.train)
  norm.factor <- calcNormFactors(cts.train, method = "TMM")
  size.factor <- lib.size*norm.factor/exp(mean(log(lib.size*norm.factor))) 
  
  res.train <- mclapply(1:length(vars),
                        function(i) {
                          dds <- DESeqDataSetFromMatrix(countData = cts.train, 
                                                        colData = coldata.train, 
                                                        design = formula(paste0("~", vars[i])))
                          sizeFactors(dds) <- size.factor
                          return(results(DESeq(dds)))},
                        mc.cores = 10) 
  
  DESeq2.eFDR.train <- mclapply(1:length(vars),
                                function(i) qvalue(res.train[[i]][["pvalue"]])$qvalues,
                                mc.cores = 10)
  
  DESeq2.eFDR.train2 <- mclapply(1:length(vars),
                                 function(i) {
                                   q.val <- DESeq2.eFDR.train[[i]]
                                   return(q.val[order(q.val)[1:threshold]])
                                 },
                                 mc.cores = 10)
  
  DESeq2.eFDR.GeneName.train <- mclapply(1:length(vars),
                                         function(i) {
                                           q.val <- DESeq2.eFDR.train[[i]]
                                           return(rownames(cts.train)[order(q.val)[1:threshold]])
                                         },
                                         mc.cores = 10)
  
  names(res.train) = names(DESeq2.eFDR.train) = names(DESeq2.eFDR.train2) = names(DESeq2.eFDR.GeneName.train) = vars 
  
  # Transform the TMM normalization factors to be used in DESeq2
  lib.size <- colSums(cts.test)
  norm.factor <- calcNormFactors(cts.test, method = "TMM")
  size.factor <- lib.size*norm.factor/exp(mean(log(lib.size*norm.factor))) 
  
  res.test <- mclapply(1:length(vars),
                       function(i) {
                         dds <- DESeqDataSetFromMatrix(countData = cts.test, 
                                                       colData = coldata.test, 
                                                       design = formula(paste0("~", vars[i])))
                         sizeFactors(dds) <- size.factor
                         return(results(DESeq(dds)))},
                       mc.cores = 10) 
  
  DESeq2.eFDR.test <- mclapply(1:length(vars),
                               function(i) qvalue(res.test[[i]][["pvalue"]])$qvalues,
                               mc.cores = 10)
  
  DESeq2.eFDR.test2 <- mclapply(1:length(vars),
                                function(i) {
                                  q.val <- DESeq2.eFDR.test[[i]]
                                  return(q.val[order(q.val)[1:threshold]])
                                },
                                mc.cores = 10)
  
  DESeq2.eFDR.GeneName.test <- mclapply(1:length(vars),
                                        function(i) {
                                          q.val <- DESeq2.eFDR.test[[i]]
                                          return(rownames(cts.test)[order(q.val)[1:threshold]])
                                        },
                                        mc.cores = 10)
  
  names(res.test) = names(DESeq2.eFDR.test) = names(DESeq2.eFDR.test2) = names(DESeq2.eFDR.GeneName.test) = vars 
  
  save(cts.train, cts.test, coldata.train, coldata.test, vars,
       res.train, DESeq2.eFDR.train, DESeq2.eFDR.train2, DESeq2.eFDR.GeneName.train, 
       res.test, DESeq2.eFDR.test, DESeq2.eFDR.test2, DESeq2.eFDR.GeneName.test, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/DESeq2Uni%s.RData", threshold, seed.num))
}
```


## Run and save modeling results
```{r}
threshold.i <- 5000
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
for (seed.i in seed.vec) {
  dat.expr.train <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/dat.expr.train%s.RDS", seed.i))
  dat.expr.test <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/dat.expr.test%s.RDS", seed.i))
  dat.pheno.train <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/dat.pheno.train%s.RDS", seed.i))
  dat.pheno.test <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/dat.pheno.test%s.RDS", seed.i))
  uni_TMM_top(seed.num = seed.i, threshold = threshold.i)
}
```


## Screen eligible phenotypes to be included in the donwnstream multivariable analysis and interaction analysis 
We filter out those phenotypes showing less than 5% (`0.05 * threshold`) cDEGs within a particular ranking threshold identified in at least one of the approaches (BMAseq, DESeq2, edgeR, voom + limma + eBayes, voom + limma) in the univariable analysis. Hence, in the donwnstream multivariable analysis and interaction analysis, we only include the phenotypes showing greater than or equal to 5% cDEGs within a particular ranking threshold in all approaches under the univariable framework. 

```{r}
threshold.vec <- seq(1000, 5000, 1000)
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
method.vec <- c("DESeq2", "edgeR", "eBayes", "VoomLimma")
vars <- c("BMI", "AGE", "SEX", "MHABNWBC", "MHARTHTS", "MHCVD", "MHASTHMA", "MHBCTINF", "MHBLDDND", "MHCOCAINE5", "MHCOPD", "MHDRNKSTS", "SMTSPAX") 
date.analysis <- format(Sys.Date(), "%Y%b%d")

for (threshold.i in threshold.vec) {
  for (seed.i in seed.vec) {
    for (method.i in method.vec) {
      load(file = sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sUni%s.RData", method.i, seed.i))
      if (method.i == "DESeq2") {
        cDEGs.train <- DESeq2.eFDR.GeneName.train
        cDEGs.test <- DESeq2.eFDR.GeneName.test
      } else if (method.i == "edgeR") {
        cDEGs.train <- edgeR.eFDR.GeneName.train
        cDEGs.test <- edgeR.eFDR.GeneName.test
      } else if (method.i == "eBayes") {
        cDEGs.train <- eBayes.eFDR.train2
        cDEGs.test <- eBayes.eFDR.test2
      } else {
        cDEGs.train <- voom.eFDR.train2
        cDEGs.test <- voom.eFDR.test2
      }
      if (method.i %in% c("DESeq2", "edgeR")) {
        plot.array <- mclapply(1:length(vars), 
                               function(x) ggVennDiagram(
                                 list(Train = cDEGs.train[[x]][1:threshold.i], 
                                      Test = cDEGs.test[[x]][1:threshold.i]), 
                                 label_alpha = 0, label_color = "white") +
                                 theme(legend.position = "none", 
                                 plot.title = element_text(hjust = 0.5, colour = "red")) + 
                                 ggtitle(vars[x]),
                               mc.cores = 10)
        g <- grid.arrange(do.call("arrangeGrob", c(plot.array, ncol = 3)))
        ggsave(filename = sprintf("../ApplicationResult/Uni/Trial/Top%s/%s_%s_%s.png", threshold.i, date.analysis, method.i, seed.i),
               plot = g,
               device = "png",
               width = 8,
               height = 9,
               units = "in")
        for (i in vars) {
          if (intersect(cDEGs.train[[i]][1:threshold.i], cDEGs.test[[i]][1:threshold.i]) |> length() / threshold.i < 0.05) {
            print(paste("This phenotype", i, "should not be used in the multivariable analysis!"))}
        }
      } else {
        plot.array <- mclapply(1:length(vars), 
                               function(x) ggVennDiagram(
                                 list(Train = cDEGs.train[[x]][1:threshold.i] |> names(), 
                                      Test = cDEGs.test[[x]][1:threshold.i] |> names()), 
                                 label_alpha = 0, label_color = "white") +
                                 theme(legend.position = "none", 
                                 plot.title = element_text(hjust = 0.5, colour = "red")) + 
                                 ggtitle(vars[x]),
                               mc.cores = 10)
        g <- grid.arrange(do.call("arrangeGrob", c(plot.array, ncol = 3)))
        ggsave(filename = sprintf("../ApplicationResult/Uni/Trial/Top%s/%s_%s_%s.png", threshold.i, date.analysis, method.i, seed.i),
               plot = g,
               device = "png",
               width = 8,
               height = 9,
               units = "in")
        for (i in vars) {
          if (intersect(cDEGs.train[[i]][1:threshold.i] |> names(), cDEGs.test[[i]][1:threshold.i] |> names()) |> length() / threshold.i < 0.05) {
            print(paste("This phenotype", i, "should not be used in the multivariable analysis!"))}
        }
      }
    }
  }
}
```


## Create functions to generate data for making boxplots of number of cDEGs and rate of cDEGs, respectively
```{r not.include.BMAseq}
##------Number of cDEGs------
boxplot_data <- function (threshold = 1000, random.seed = 8809678) {
  data.table(threshold=rep(threshold,4), 
             random.seed=random.seed,
             method=c("DESeq2","edgeR","eBayes","voom.limma"),
             variable=rep(c("BMI","AGE","SEX","MHABNWBC"),each=4),
             num.cDEGs=c(
               ##------BMI------
               intersect(DESeq2.eFDR.GeneName.train$BMI[1:threshold], DESeq2.eFDR.GeneName.test$BMI[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$BMI[1:threshold], edgeR.eFDR.GeneName.test$BMI[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$BMI[1:threshold]), names(eBayes.eFDR.test2$BMI[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$BMI[1:threshold]), names(voom.eFDR.test2$BMI[1:threshold])) |> length(),
               
               ##------AGE------
               intersect(DESeq2.eFDR.GeneName.train$AGE[1:threshold], DESeq2.eFDR.GeneName.test$AGE[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$AGE[1:threshold], edgeR.eFDR.GeneName.test$AGE[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$AGE[1:threshold]), names(eBayes.eFDR.test2$AGE[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$AGE[1:threshold]), names(voom.eFDR.test2$AGE[1:threshold])) |> length(),
               
               ##------SEX------
               intersect(DESeq2.eFDR.GeneName.train$SEX[1:threshold], DESeq2.eFDR.GeneName.test$SEX[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$SEX[1:threshold], edgeR.eFDR.GeneName.test$SEX[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$SEX[1:threshold]), names(eBayes.eFDR.test2$SEX[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$SEX[1:threshold]), names(voom.eFDR.test2$SEX[1:threshold])) |> length(),
               
               ##------MHABNWBC------
               intersect(DESeq2.eFDR.GeneName.train$MHABNWBC[1:threshold], DESeq2.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$MHABNWBC[1:threshold], edgeR.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$MHABNWBC[1:threshold]), names(eBayes.eFDR.test2$MHABNWBC[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$MHABNWBC[1:threshold]), names(voom.eFDR.test2$MHABNWBC[1:threshold])) |> length()))
}

##------Rate of cDEGs------
boxplot_data2 <- function (threshold = 1000, random.seed = 8809678) {
  data.table(threshold=rep(threshold,4), 
             random.seed=random.seed,
             method=c("DESeq2","edgeR","eBayes","voom.limma"),
             variable=rep(c("BMI","AGE","SEX","MHABNWBC"),each=4),
             rate.cDEGs=c(
               ##------BMI------
               intersect(DESeq2.eFDR.GeneName.train$BMI[1:threshold], DESeq2.eFDR.GeneName.test$BMI[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$BMI[1:threshold], edgeR.eFDR.GeneName.test$BMI[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$BMI[1:threshold]), names(eBayes.eFDR.test2$BMI[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$BMI[1:threshold]), names(voom.eFDR.test2$BMI[1:threshold])) |> length() / threshold,
               
               ##------AGE------
               intersect(DESeq2.eFDR.GeneName.train$AGE[1:threshold], DESeq2.eFDR.GeneName.test$AGE[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$AGE[1:threshold], edgeR.eFDR.GeneName.test$AGE[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$AGE[1:threshold]), names(eBayes.eFDR.test2$AGE[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$AGE[1:threshold]), names(voom.eFDR.test2$AGE[1:threshold])) |> length() / threshold,
               
               ##------SEX------
               intersect(DESeq2.eFDR.GeneName.train$SEX[1:threshold], DESeq2.eFDR.GeneName.test$SEX[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$SEX[1:threshold], edgeR.eFDR.GeneName.test$SEX[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$SEX[1:threshold]), names(eBayes.eFDR.test2$SEX[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$SEX[1:threshold]), names(voom.eFDR.test2$SEX[1:threshold])) |> length() / threshold,
               
               ##------MHABNWBC------
               intersect(DESeq2.eFDR.GeneName.train$MHABNWBC[1:threshold], DESeq2.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$MHABNWBC[1:threshold], edgeR.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$MHABNWBC[1:threshold]), names(eBayes.eFDR.test2$MHABNWBC[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$MHABNWBC[1:threshold]), names(voom.eFDR.test2$MHABNWBC[1:threshold])) |> length() / threshold))
}
```

```{r include.BMAseq}
boxplot_data <- function (threshold = 1000, random.seed = 8809678, type = "Uni") {
  data.table(threshold=rep(threshold,5), 
             random.seed=random.seed,
             method=c("BMAseq","DESeq2","edgeR","eBayes","voom.limma"),
             variable=rep(c("BMI","AGE","SEX","MHABNWBC"),each=5),
             num.cDEGs=c(
               ##------BMI------
               intersect(names(BMAseq.eFDR.Main.train$BMI[1:threshold]), names(BMAseq.eFDR.Main.test$BMI[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$BMI[1:threshold], DESeq2.eFDR.GeneName.test$BMI[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$BMI[1:threshold], edgeR.eFDR.GeneName.test$BMI[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$BMI[1:threshold]), names(eBayes.eFDR.test2$BMI[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$BMI[1:threshold]), names(voom.eFDR.test2$BMI[1:threshold])) |> length(),
               
               ##------AGE------
               intersect(names(BMAseq.eFDR.Main.train$AGE[1:threshold]), names(BMAseq.eFDR.Main.test$AGE[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$AGE[1:threshold], DESeq2.eFDR.GeneName.test$AGE[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$AGE[1:threshold], edgeR.eFDR.GeneName.test$AGE[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$AGE[1:threshold]), names(eBayes.eFDR.test2$AGE[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$AGE[1:threshold]), names(voom.eFDR.test2$AGE[1:threshold])) |> length(),
               
               ##------SEX------
               intersect(names(BMAseq.eFDR.Main.train$SEX[1:threshold]), names(BMAseq.eFDR.Main.test$SEX[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$SEX[1:threshold], DESeq2.eFDR.GeneName.test$SEX[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$SEX[1:threshold], edgeR.eFDR.GeneName.test$SEX[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$SEX[1:threshold]), names(eBayes.eFDR.test2$SEX[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$SEX[1:threshold]), names(voom.eFDR.test2$SEX[1:threshold])) |> length(),
               
               ##------MHABNWBC------
               intersect(names(BMAseq.eFDR.Main.train$MHABNWBC[1:threshold]), names(BMAseq.eFDR.Main.test$MHABNWBC[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$MHABNWBC[1:threshold], DESeq2.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$MHABNWBC[1:threshold], edgeR.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$MHABNWBC[1:threshold]), names(eBayes.eFDR.test2$MHABNWBC[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$MHABNWBC[1:threshold]), names(voom.eFDR.test2$MHABNWBC[1:threshold])) |> length()))
}

##------Rate of cDEGs------
boxplot_data2 <- function (threshold = 1000, random.seed = 8809678) {
  data.table(threshold=rep(threshold,5), 
             random.seed=random.seed,
             method=c("BMAseq","DESeq2","edgeR","eBayes","voom.limma"),
             variable=rep(c("BMI","AGE","SEX","MHABNWBC"),each=5),
             rate.cDEGs=c(
               ##------BMI------
               intersect(names(BMAseq.eFDR.Main.train$BMI[1:threshold]), names(BMAseq.eFDR.Main.test$BMI[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$BMI[1:threshold], DESeq2.eFDR.GeneName.test$BMI[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$BMI[1:threshold], edgeR.eFDR.GeneName.test$BMI[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$BMI[1:threshold]), names(eBayes.eFDR.test2$BMI[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$BMI[1:threshold]), names(voom.eFDR.test2$BMI[1:threshold])) |> length() / threshold,
               
               ##------AGE------
               intersect(names(BMAseq.eFDR.Main.train$AGE[1:threshold]), names(BMAseq.eFDR.Main.test$AGE[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$AGE[1:threshold], DESeq2.eFDR.GeneName.test$AGE[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$AGE[1:threshold], edgeR.eFDR.GeneName.test$AGE[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$AGE[1:threshold]), names(eBayes.eFDR.test2$AGE[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$AGE[1:threshold]), names(voom.eFDR.test2$AGE[1:threshold])) |> length() / threshold,
               
               ##------SEX------
               intersect(names(BMAseq.eFDR.Main.train$SEX[1:threshold]), names(BMAseq.eFDR.Main.test$SEX[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$SEX[1:threshold], DESeq2.eFDR.GeneName.test$SEX[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$SEX[1:threshold], edgeR.eFDR.GeneName.test$SEX[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$SEX[1:threshold]), names(eBayes.eFDR.test2$SEX[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$SEX[1:threshold]), names(voom.eFDR.test2$SEX[1:threshold])) |> length() / threshold,
               
               ##------MHABNWBC------
               intersect(names(BMAseq.eFDR.Main.train$MHABNWBC[1:threshold]), names(BMAseq.eFDR.Main.test$MHABNWBC[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$MHABNWBC[1:threshold], DESeq2.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$MHABNWBC[1:threshold], edgeR.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$MHABNWBC[1:threshold]), names(eBayes.eFDR.test2$MHABNWBC[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$MHABNWBC[1:threshold]), names(voom.eFDR.test2$MHABNWBC[1:threshold])) |> length() / threshold))
}
```


## Organize data for boxplotting
```{r not.include.BMAseq}
##------Number of cDEGs------
threshold.vec <- seq(1000, 5000, 1000)
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
boxplot.data.num <- NULL
for (threshold.i in threshold.vec) {
  for (seed.i in seed.vec) {
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/DESeq2Uni%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/edgeRUni%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/eBayesUni%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/VoomLimmaUni%s.RData", seed.i))
    boxplot.data.num <- rbind(boxplot.data.num, boxplot_data(threshold.i, seed.i))
  }
}
saveRDS(boxplot.data.num, "../ApplicationResult/Uni/RandomSeed/numcDEGs/Boxplot/boxplot.data.num.RDS")


##------Rate of cDEGs------
threshold.vec <- seq(1000, 5000, 1000)
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
boxplot.data.rate <- NULL
for (threshold.i in threshold.vec) {
  for (seed.i in seed.vec) {
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/DESeq2Uni%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/edgeRUni%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/eBayesUni%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/VoomLimmaUni%s.RData", seed.i))
    boxplot.data.rate <- rbind(boxplot.data.rate, boxplot_data2(threshold.i, seed.i))
  }
}
saveRDS(boxplot.data.rate, "../ApplicationResult/Uni/RandomSeed/ratecDEGs/Boxplot/boxplot.data.rate.RDS")
```

```{r include.multiBMAseq}
##------Number of cDEGs------
threshold.vec <- seq(1000, 5000, 1000)
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
boxplot.data.num <- NULL
for (threshold.i in threshold.vec) {
  for (seed.i in seed.vec) {
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/BMAseqMulti%s.RData", seed.i)) # Multivariable BMAseq model without interaction terms
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/DESeq2Uni%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/edgeRUni%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/eBayesUni%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/VoomLimmaUni%s.RData", seed.i))
    boxplot.data.num <- rbind(boxplot.data.num, boxplot_data(threshold.i, seed.i))
  }
}
saveRDS(boxplot.data.num, "../ApplicationResult/Uni/RandomSeed/numcDEGs/Boxplot/boxplot.data.num.BMA.RDS")


##------Rate of cDEGs------
threshold.vec <- seq(1000, 5000, 1000)
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
boxplot.data.rate <- NULL
for (threshold.i in threshold.vec) {
  for (seed.i in seed.vec) {
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/BMAseqMulti%s.RData", seed.i)) # Multivariable BMAseq model without interaction terms
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/DESeq2Uni%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/edgeRUni%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/eBayesUni%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/VoomLimmaUni%s.RData", seed.i))
    boxplot.data.rate <- rbind(boxplot.data.rate, boxplot_data2(threshold.i, seed.i))
  }
}
saveRDS(boxplot.data.rate, "../ApplicationResult/Uni/RandomSeed/ratecDEGs/Boxplot/boxplot.data.rate.BMA.RDS")
```


## Make and save boxplots
```{r not.include.multiBMAseq}
##------Number of cDEGs------
date.analysis <- format(Sys.Date(), "%Y%b%d")
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC")

for (var.i in var.vec) {
  g <- boxplot.data.num[variable==var.i] |>
    ggplot(aes(y = num.cDEGs, x = method, group = method)) + 
    geom_boxplot(outlier.shape = NA) + 
    geom_jitter(aes(color = method), size = 0.8) + 
    stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
    facet_wrap(vars(threshold)) + 
    labs(y = "Number of common differentially expressed genes", x = "Method") + 
    theme_bw() + 
    theme(legend.position = "none",
          axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
    ggsave(filename = sprintf("../ApplicationResult/Uni/RandomSeed/numcDEGs/Boxplot/%s_%s.png", date.analysis, var.i),
           plot = g, device = "png", width = 9, height = 6, units = "in")
}


##------Rate of cDEGs------
for (var.i in var.vec) {
  g <- boxplot.data.rate[variable==var.i] |>
    ggplot(aes(y = rate.cDEGs, x = method, group = method)) + 
    geom_boxplot(outlier.shape = NA) + 
    geom_jitter(aes(color = method), size = 0.8) + 
    stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
    facet_wrap(vars(threshold)) + 
    labs(y = "Rate of common differentially expressed genes", x = "Method") + 
    theme_bw() + 
    theme(legend.position = "none",
          axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
    ggsave(filename = sprintf("../ApplicationResult/Uni/RandomSeed/ratecDEGs/Boxplot/%s_%s.png", date.analysis, var.i),
           plot = g, device = "png", width = 9, height = 6, units = "in")
}
```

```{r include.multiBMAseq}
##------Number of cDEGs------
date.analysis <- format(Sys.Date(), "%Y%b%d")
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC")

for (var.i in var.vec) {
  g <- boxplot.data.num[variable==var.i] |>
    ggplot(aes(y = num.cDEGs, x = method, group = method)) + 
    geom_boxplot(outlier.shape = NA) + 
    geom_jitter(aes(color = method), size = 0.8) + 
    stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
    facet_wrap(vars(threshold)) + 
    labs(y = "Number of common differentially expressed genes", x = "Method") + 
    theme_bw() + 
    theme(legend.position = "none",
          axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
    ggsave(filename = sprintf("../ApplicationResult/Uni/RandomSeed/numcDEGs/Boxplot/%s_%s.BMA.png", date.analysis, var.i),
           plot = g, device = "png", width = 9, height = 6, units = "in")
}


##------Rate of cDEGs------
for (var.i in var.vec) {
  g <- boxplot.data.rate[variable==var.i] |>
    ggplot(aes(y = rate.cDEGs, x = method, group = method)) + 
    geom_boxplot(outlier.shape = NA) + 
    geom_jitter(aes(color = method), size = 0.8) + 
    stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
    facet_wrap(vars(threshold)) + 
    labs(y = "Rate of common differentially expressed genes", x = "Method") + 
    theme_bw() + 
    theme(legend.position = "none",
          axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
    ggsave(filename = sprintf("../ApplicationResult/Uni/RandomSeed/ratecDEGs/Boxplot/%s_%s.BMA.png", date.analysis, var.i),
           plot = g, device = "png", width = 9, height = 6, units = "in")
}
```


## Estimate the duplicate rate of a common differentially expressed gene associated with a variable
```{r}
cDEGs = cDEGs.list = NULL
threshold.vec <- seq(1000, 5000, 1000)
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)

duplicate_rate_cDEGs_per_variable <- function(var.name = "BMI", method.name = "DESeq2") {
  for (threshold.i in threshold.vec) {
    for (seed.i in seed.vec) {
      if (method.name %in% c("DESeq2", "edgeR")) {
        load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sUni%s.RData", method.name, seed.i))
        cDEGs <- c(intersect(get(sprintf("%s.eFDR.GeneName.train", method.name))[[var.name]][1:threshold.i], 
                             get(sprintf("%s.eFDR.GeneName.test", method.name))[[var.name]][1:threshold.i]))
      } else if (method.name == "eBayes") {
        load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sUni%s.RData", method.name, seed.i))
        cDEGs <- c(intersect(names(get(sprintf("%s.eFDR.train2", method.name))[[var.name]][1:threshold.i]), 
                             names(get(sprintf("%s.eFDR.test2", method.name))[[var.name]][1:threshold.i])))
      } else {
        load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sUni%s.RData", method.name, seed.i))
        method.name2 <- "voom"
        cDEGs <- c(intersect(names(get(sprintf("%s.eFDR.train2", method.name2))[[var.name]][1:threshold.i]), 
                             names(get(sprintf("%s.eFDR.test2", method.name2))[[var.name]][1:threshold.i])))
      }
      cDEGs.list <- append(cDEGs.list, list(cDEGs))
    }
    cDEGs.unique <- unique(unlist(cDEGs.list))
    
    # Create a matrix that shows the occurrence of each unique gene among the 10 random seed trials
    cDEGs.matrix <- matrix(0, nrow = length(cDEGs.unique), ncol = 10)
    rownames(cDEGs.matrix) <- cDEGs.unique
    colnames(cDEGs.matrix) <- seed.vec   
    for (i in 1:10) {
      cDEGs.i.seed <- cDEGs.list[[i]]
      for (j in 1:length(cDEGs.unique)) {
        if (cDEGs.unique[j] %in% cDEGs.i.seed) {
          cDEGs.matrix[j, i] <- 1
        }
      }
    }
    Sum <- apply(cDEGs.matrix, 1, sum)
    Percent <- Sum/10
    cDEGs.matrix <- cbind(cDEGs.matrix, Sum, Percent) # Add two new columns to the matrix
    saveRDS(cDEGs.matrix, sprintf("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/%sUni%s_%s.RDS", method.name, var.name, threshold.i))
    
    print(paste("Complete the duplicated rate matrix of cDEGs associated with", var.name, "identified by", method.name, "among 10 random seeds for the threshold", threshold.i))
    cDEGs = cDEGs.list =  NULL
  }
}

mapply(duplicate_rate_cDEGs_per_variable,
       rep(c("BMI", "AGE", "SEX", "MHABNWBC"), each = 4), 
       rep(c("DESeq2", "edgeR", "eBayes", "VoomLimma"), times = 4))


# For illustration purpose, only consider the phenotype BMI and 2000 ranking threshold
method.vec <- c("DESeq2", "edgeR", "eBayes", "VoomLimma")
var.name <- "BMI"
threshold.i <- 2000
for (method.name in method.vec) {
  data <- readRDS(sprintf("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/%sUni%s_%s.RDS", method.name, var.name, threshold.i))
  write.csv(data, sprintf("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/%sUni%s_%s.csv", method.name, var.name, threshold.i))
}
```


## Prepare and organize the data for boxplots
```{r}
boxplot_duplicate_rate_data <- function (var.name = NULL, method.name = NULL, threshold = NULL) {
  load.data <- readRDS(sprintf("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/%sUni%s_%s.RDS", method.name, var.name, threshold))
  temp.data <- data.table(threshold=threshold, 
                          method=method.name,
                          variable=var.name,
                          cDEGs.duplicate.rate.median=median(load.data[,12]),
                          cDEGs.duplicate.rate.mean=mean(load.data[,12]))
  return(temp.data)
}

organized_duplicate_rate_data <- mapply(boxplot_duplicate_rate_data,
                                        var.name = rep(c("BMI", "AGE", "SEX", "MHABNWBC"), each = 20), 
                                        method.name = rep(rep(c("DESeq2", "edgeR", "eBayes", "VoomLimma"), each = 5), times = 4),
                                        threshold = rep(rep(seq(1000, 5000, 1000), times = 4), times = 4), SIMPLIFY = F) |> rbindlist()

# data.frame(var.name, method.name, threshold)
```


## Make and save boxplots
```{r}
organized_duplicate_rate_data[,variable:=factor(variable,levels=c("BMI","AGE","SEX","MHABNWBC"))]
organized_duplicate_rate_data[,method:=factor(method,levels=c("DESeq2","edgeR","eBayes","VoomLimma"))]
organized_duplicate_rate_data[,cDEGs.duplicate.rate.median:=as.numeric(cDEGs.duplicate.rate.median)]
organized_duplicate_rate_data[,cDEGs.duplicate.rate.mean:=as.numeric(cDEGs.duplicate.rate.mean)]
date.analysis <- format(Sys.Date(), "%Y%b%d")
g1 <- organized_duplicate_rate_data |>
  ggplot(mapping = aes(y = cDEGs.duplicate.rate.median, x = method, group = method)) + 
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(aes(color = method), size = 0.8) + 
  facet_wrap(vars(variable)) + 
  scale_y_continuous(labels = scales::label_number(accuracy = 0.01),
                     breaks = seq(0, 1, 0.1)) + 
  labs(y = "Median duplicate rate of cDEGs", x = "Method") + 
  theme_bw() + 
  theme(legend.position = "none",
        axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
ggsave(filename = sprintf("../ApplicationResult/Uni/RandomSeed/DuplicatedRateMatrix/Boxplot/%s_median.png", date.analysis),
       plot = g1, device = "png", width = 9, height = 6, units = "in")

g2 <- organized_duplicate_rate_data |>
  ggplot(mapping = aes(y = cDEGs.duplicate.rate.mean, x = method, group = method)) + 
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(aes(color = method), size = 0.8) + 
  facet_wrap(vars(variable)) + 
  scale_y_continuous(labels = scales::label_number(accuracy = 0.01),
                     breaks = seq(0, 1, 0.1)) + 
  labs(y = "Mean duplicate rate of cDEGs", x = "Method") + 
  theme_bw() + 
  theme(legend.position = "none",
        axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
ggsave(filename = sprintf("../ApplicationResult/Uni/RandomSeed/DuplicatedRateMatrix/Boxplot/%s_mean.png", date.analysis),
       plot = g2, device = "png", width = 9, height = 6, units = "in")
```


## Make Venn digrams of cDEGs
```{r}
threshold <- 2000
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC")
date.analysis <- format(Sys.Date(), "%Y%b%d")

make_vennplot <-function(var.name = NULL, threshold = NULL) {

  BMAseq.cDEGs <- intersect(names(BMAseq.eFDR.Main.train[[var.name]][1:threshold]), names(BMAseq.eFDR.Main.test[[var.name]][1:threshold]))
  DESeq2.cDEGs <- intersect(DESeq2.eFDR.GeneName.train[[var.name]][1:threshold], DESeq2.eFDR.GeneName.test[[var.name]][1:threshold])
  edgeR.cDEGs <- intersect(edgeR.eFDR.GeneName.train[[var.name]][1:threshold], edgeR.eFDR.GeneName.test[[var.name]][1:threshold])
  eBayes.cDEGs <- intersect(names(eBayes.eFDR.train2[[var.name]][1:threshold]), names(eBayes.eFDR.test2[[var.name]][1:threshold]))
  voom.limma.cDEGs <- intersect(names(voom.eFDR.train2[[var.name]][1:threshold]), names(voom.eFDR.test2[[var.name]][1:threshold]))
  
  g <- ggVennDiagram(list(BMAseq = BMAseq.cDEGs, 
                          DESeq2 = DESeq2.cDEGs,
                          edgeR = edgeR.cDEGs,
                          eBayes = eBayes.cDEGs,
                          voom.limma = voom.limma.cDEGs),
                     label_alpha = 0, label_color = "white") +
    theme(legend.position = "none", 
          plot.title = element_text(hjust = 0.5, colour = "red", size = 18)) + 
    ggtitle(paste("The intersection of cDEGs associated with the", var.name, "\nunder the seed", seed.i, "and threshold", threshold)) + 
    scale_fill_distiller(palette = "Dark2", direction = -1)
  
  ggsave(filename = paste0("../ApplicationResult/Uni/RandomSeed/TMM_Top2000/Vennplot/", date.analysis, "_", var.name, "_", seed.i, ".png"),
         plot = g,
         device = "png",
         width = 10,
         height = 7,
         units = "in")
}

for (seed.i in seed.vec) {
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/BMAseqMulti%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/DESeq2Uni%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/edgeRUni%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/eBayesUni%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/VoomLimmaUni%s.RData", seed.i))
  for (var.i in var.vec) {
    make_vennplot(var.name = var.i, threshold = 2000)
  }
}
```


## Extract unique cDEGs identified by BMAseq per seed and all unique cDEGs identified by BMAseq across 10 seeds
```{r}
file.name <- "../ApplicationResult/Uni/RandomSeed/TMM_Top2000/Vennplot/uniq_cDEGs_BMI_BMAseq_per_seed.txt"
unique.cDEGs.list <- NULL
extract_unique_cDEGs_per_seed <-function(var.name = NULL, threshold = NULL) {
  BMAseq.cDEGs <- intersect(names(BMAseq.eFDR.Main.train[[var.name]][1:threshold]), names(BMAseq.eFDR.Main.test[[var.name]][1:threshold]))
  DESeq2.cDEGs <- intersect(DESeq2.eFDR.GeneName.train[[var.name]][1:threshold], DESeq2.eFDR.GeneName.test[[var.name]][1:threshold])
  edgeR.cDEGs <- intersect(edgeR.eFDR.GeneName.train[[var.name]][1:threshold], edgeR.eFDR.GeneName.test[[var.name]][1:threshold])
  eBayes.cDEGs <- intersect(names(eBayes.eFDR.train2[[var.name]][1:threshold]), names(eBayes.eFDR.test2[[var.name]][1:threshold]))
  voom.limma.cDEGs <- intersect(names(voom.eFDR.train2[[var.name]][1:threshold]), names(voom.eFDR.test2[[var.name]][1:threshold]))
  unique.cDEGs <- setdiff(BMAseq.cDEGs, union(DESeq2.cDEGs, edgeR.cDEGs) |> union(eBayes.cDEGs) |> union(voom.limma.cDEGs))
  return(unique.cDEGs)
}

for (seed.i in seed.vec) {
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/BMAseqMulti%s.RData", seed.i)) # Multivariable BMAseq model without interaction terms
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/DESeq2Uni%s.RData", seed.i)) # Univariable DESeq2 model
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/edgeRUni%s.RData", seed.i)) # Univariable edgeR model
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/eBayesUni%s.RData", seed.i)) # Univariable eBayes model
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/VoomLimmaUni%s.RData", seed.i)) # Univariable VoomLimma model
  unique.cDEGs <- extract_unique_cDEGs_per_seed(var.name = "BMI", threshold = 2000)
  unique.cDEGs.list <- append(unique.cDEGs.list, list(unique.cDEGs))
  write(sprintf("\nThe following gene IDs are unique gene IDs associated with the main effect of %s identified by BMAseq under the seed %s and threshold %s", var.name, seed.i, threshold), file = file.name, append = T)
  write.table(unique.cDEGs, file = file.name, quote = F, col.names = F, append = T)
}
names(unique.cDEGs.list) <- seed.vec

file.name <- "../ApplicationResult/Uni/RandomSeed/TMM_Top2000/Vennplot/uniq_cDEGs_BMI_BMAseq_all_seeds.txt"
write.table(unlist(unique.cDEGs.list) |> unique(), file = file.name, quote = F, col.names = F)
```


# Multi-variables without interaction terms analysis
## Create functions of building multivariable models without interaction terms
```{r}
multi_TMM_top <- function(seed.num = 999, threshold = 2000) {
  ##------Set variables of interest------
  vars.pool <- c("BMI", "AGE", "SEX", "MHABNWBC")  
  
  
  ##------BMAseq------
  output.multi.train.s1 <- BMAseq.multi.postprob.norm(dat.expr.counts = dat.expr.train,  
                                                      dat.pheno = dat.pheno.train, 
                                                      var.pool = vars.pool, 
                                                      max.nvar = 4, 
                                                      cut.BF = 1)
  
  output.multi.train.s2 <- BMAseq.multi.DEG(dat.pheno = output.multi.train.s1$dat.pheno.new, 
                                            model.space = output.multi.train.s1$model.space, 
                                            post.modelprob = output.multi.train.s1$post.modelprob, 
                                            var.pool = vars.pool, 
                                            cut.FDR = 0.05) 
  # BMAseq.multi.DEG() from {BMAseq} (2022Dec5 version) is used here
  
  BMAseq.eFDR.Main.train <- mclapply(1:length(vars.pool),
                                     function(i) output.multi.train.s2$eFDR.Main[, i][order(output.multi.train.s2$eFDR.Main[, i])[1:threshold]],
                                     mc.cores = 10)
  
  output.multi.test.s1 <- BMAseq.multi.postprob.norm(dat.expr.counts = dat.expr.test, 
                                                     dat.pheno = dat.pheno.test, 
                                                     var.pool = vars.pool, 
                                                     max.nvar = 4, 
                                                     cut.BF = 1)
  
  output.multi.test.s2 <- BMAseq.multi.DEG(dat.pheno = output.multi.test.s1$dat.pheno.new, 
                                           model.space = output.multi.test.s1$model.space, 
                                           post.modelprob = output.multi.test.s1$post.modelprob, 
                                           var.pool = vars.pool, 
                                           cut.FDR = 0.05)
  
  BMAseq.eFDR.Main.test <- mclapply(1:length(vars.pool),
                                    function(i) output.multi.test.s2$eFDR.Main[, i][order(output.multi.test.s2$eFDR.Main[, i])[1:threshold]],
                                    mc.cores = 10)

  names(BMAseq.eFDR.Main.train) = names(BMAseq.eFDR.Main.test) = vars.pool
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, vars.pool,
       output.multi.train.s1, output.multi.train.s2, output.multi.test.s1, output.multi.test.s2, 
       BMAseq.eFDR.Main.train, BMAseq.eFDR.Main.test, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/BMAseqMulti%s.RData", threshold, seed.num)) 
  
  
  ##------voom + limma------
  design.train <- model.matrix(~BMI + AGE + SEX + MHABNWBC, data = dat.pheno.train)
  design.test <- model.matrix(~BMI + AGE + SEX + MHABNWBC, data = dat.pheno.test)
    
  voom.train <- voom(counts = dat.expr.train,
                     design = design.train,
                     lib.size = colSums(dat.expr.train)*calcNormFactors(dat.expr.train)) 
  
  voom.fit.train <- lmFit(voom.train[["E"]],
                          design = design.train,
                          weights = voom.train[["weights"]])
  
  voom.eFDR.train <- mclapply(1:length(vars.pool),
                              function(i) {
                                t <- voom.fit.train[["coefficients"]][, i + 1]/voom.fit.train[["stdev.unscaled"]][, i + 1]/voom.fit.train[["sigma"]]
                                p <- 2*pt(-abs(t), df = voom.fit.train[["df.residual"]])
                                return(qvalue(p)) },
                              mc.cores = 10)
  
  voom.eFDR.train2 <- mclapply(1:length(vars.pool),
                               function(i) {
                                 q.val <- voom.eFDR.train[[i]][["qvalues"]]
                                 return(q.val[order(q.val)[1:threshold]])
                               },
                               mc.cores = 10)
  
  names(voom.eFDR.train) = names(voom.eFDR.train2) = vars.pool 
  
  voom.test <- voom(counts = dat.expr.test,
                    design = design.test,
                    lib.size = colSums(dat.expr.test)*calcNormFactors(dat.expr.test))
  
  voom.fit.test <- lmFit(voom.test[["E"]],
                         design = design.test,
                         weights = voom.test[["weights"]])
  
  voom.eFDR.test <- mclapply(1:length(vars.pool),
                             function(i) {
                               t <- voom.fit.test[["coefficients"]][, i + 1]/voom.fit.test[["stdev.unscaled"]][, i + 1]/voom.fit.test[["sigma"]]
                               p <- 2*pt(-abs(t), df = voom.fit.test[["df.residual"]])
                               return(qvalue(p)) },
                             mc.cores = 10)
  
  voom.eFDR.test2 <- mclapply(1:length(vars.pool),
                              function(i) {
                                q.val <- voom.eFDR.test[[i]][["qvalues"]]
                                return(q.val[order(q.val)[1:threshold]])
                              },
                              mc.cores = 10)
  
  names(voom.eFDR.test) = names(voom.eFDR.test2) = vars.pool 

  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, vars.pool,
       design.train, design.test, voom.train, voom.fit.train, voom.eFDR.train, voom.eFDR.train2, 
       voom.test, voom.fit.test, voom.eFDR.test, voom.eFDR.test2, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/VoomLimmaMulti%s.RData", threshold, seed.num))
  
  
  ##------voom + limma + eBayes------
  voom.train <- voom(counts = dat.expr.train,
                     design = design.train,
                     lib.size = colSums(dat.expr.train)*calcNormFactors(dat.expr.train))
  
  eBayes.fit.train <- lmFit(voom.train[["E"]],
                            design = design.train,
                            weights = voom.train[["weights"]]) %>% 
                        eBayes()
  
  eBayes.eFDR.train <- mclapply(1:length(vars.pool),
                                function(i) eBayes.fit.train[["p.value"]][, i + 1] %>% 
                                  qvalue(),
                                mc.cores = 10)
  
  eBayes.eFDR.train2 <- mclapply(1:length(vars.pool),
                                 function(i) {
                                   q.val <- eBayes.eFDR.train[[i]][["qvalues"]]
                                   return(q.val[order(q.val)[1:threshold]])
                                 },
                                 mc.cores = 10)
  
  names(eBayes.eFDR.train) = names(eBayes.eFDR.train2) = vars.pool 
  
  voom.test <- voom(counts = dat.expr.test,
                    design = design.test,
                    lib.size = colSums(dat.expr.test)*calcNormFactors(dat.expr.test))
  

  eBayes.fit.test <- lmFit(voom.test[["E"]],
                           design = design.test,
                           weights = voom.test[["weights"]]) %>% 
                       eBayes()
  
  eBayes.eFDR.test <- mclapply(1:length(vars.pool),
                               function(i) eBayes.fit.test[["p.value"]][, i + 1] %>% 
                                 qvalue(),
                               mc.cores = 10)
  
  eBayes.eFDR.test2 <- mclapply(1:length(vars.pool),
                                function(i) {
                                  q.val <- eBayes.eFDR.test[[i]][["qvalues"]]
                                  return(q.val[order(q.val)[1:threshold]])
                                },
                                mc.cores = 10)
  
  names(eBayes.eFDR.test) = names(eBayes.eFDR.test2) = vars.pool 
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, vars.pool,
       design.train, design.test, voom.train, eBayes.fit.train, eBayes.eFDR.train, eBayes.eFDR.train2,
       voom.test, eBayes.fit.test, eBayes.eFDR.test, eBayes.eFDR.test2, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/eBayesMulti%s.RData", threshold, seed.num))

  
  ##------edgeR------
  y.train <- DGEList(counts = dat.expr.train, 
                     lib.size = colSums(dat.expr.train)) %>% 
               calcNormFactors() %>% 
               estimateGLMTrendedDisp(design.train)
  
  edgeR.fit.train <- glmQLFit(y.train, design.train)
  
  qlf.train <- mclapply(1:length(vars.pool),
                        function(i) glmQLFTest(edgeR.fit.train, coef = i + 1),
                        mc.cores = 10)
  
  edgeR.eFDR.train <- mclapply(1:length(vars.pool),
                               function(i) qlf.train[[i]][["table"]][["PValue"]] %>%
                                 qvalue(),
                               mc.cores = 10)
  
  edgeR.eFDR.train2 <- mclapply(1:length(vars.pool),
                                function(i) {
                                  q.val <- edgeR.eFDR.train[[i]][["qvalues"]]
                                  return(q.val[order(q.val)[1:threshold]]) 
                                },
                                mc.cores = 10)
  
  edgeR.eFDR.GeneName.train <- mclapply(1:length(vars.pool),
                                        function(i) {
                                          q.val <- edgeR.eFDR.train[[i]][["qvalues"]]
                                          return(rownames(dat.expr.train)[order(q.val)[1:threshold]])
                                        },
                                        mc.cores = 10)
  
  names(qlf.train) = names(edgeR.eFDR.train) = names(edgeR.eFDR.train2) = names(edgeR.eFDR.GeneName.train) = vars.pool 
  
  y.test <- DGEList(counts = dat.expr.test, 
                    lib.size = colSums(dat.expr.test)) %>% 
              calcNormFactors() %>% 
              estimateGLMTrendedDisp(design.test)
  
  edgeR.fit.test <- glmQLFit(y.test, design.test)
  
  qlf.test <- mclapply(1:length(vars.pool),
                       function(i) glmQLFTest(edgeR.fit.test, coef = i + 1),
                       mc.cores = 10)
  
  edgeR.eFDR.test <- mclapply(1:length(vars.pool),
                              function(i) qlf.test[[i]][["table"]][["PValue"]] %>%
                                qvalue(),
                              mc.cores = 10)
  
  edgeR.eFDR.test2 <- mclapply(1:length(vars.pool),
                               function(i) {
                                 q.val <- edgeR.eFDR.test[[i]][["qvalues"]]
                                 return(q.val[order(q.val)[1:threshold]]) 
                               },
                               mc.cores = 10)
  
  edgeR.eFDR.GeneName.test <- mclapply(1:length(vars.pool),
                                       function(i) {
                                         q.val <- edgeR.eFDR.test[[i]][["qvalues"]]
                                         return(rownames(dat.expr.test)[order(q.val)[1:threshold]])
                                       },
                                      mc.cores = 10)
  
  names(qlf.test) = names(edgeR.eFDR.test) = names(edgeR.eFDR.test2) = names(edgeR.eFDR.GeneName.test) = vars.pool 
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, vars.pool, design.train, design.test, 
       y.train, edgeR.fit.train, edgeR.eFDR.train, edgeR.eFDR.train2, edgeR.eFDR.GeneName.train, 
       y.test, edgeR.fit.test, edgeR.eFDR.test, edgeR.eFDR.test2, edgeR.eFDR.GeneName.test, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/edgeRMulti%s.RData", threshold, seed.num))
  
  ##------DESeq2------
  cts.train <- as.matrix(dat.expr.train)
  cts.test <- as.matrix(dat.expr.test)
  coldata.train <- dat.pheno.train
  coldata.test <- dat.pheno.test
  
  # Check the sample order of the count matrix and column data in the training and test sets 
  all(rownames(coldata.test) == colnames(cts.test))
  all(rownames(coldata.train) == colnames(cts.train))
  
  name.formula <- c("BMI_high_vs_low", "AGE_old_vs_young", "SEX_male_vs_female", "MHABNWBC_yes_vs_no") 
  
  # Transform the TMM normalization factors to be used in DESeq2 (training set)
  lib.size <- colSums(cts.train)
  norm.factor <- calcNormFactors(cts.train, method = "TMM")
  size.factor <- lib.size*norm.factor/exp(mean(log(lib.size*norm.factor))) 
  
  dds <- DESeqDataSetFromMatrix(countData = cts.train, 
                                colData = coldata.train, 
                                design = ~BMI + AGE + SEX + MHABNWBC)
  sizeFactors(dds) <- size.factor

  res.train <- mclapply(1:length(vars.pool), 
                        function(i) {
                          return(results(DESeq(dds), name = name.formula[i])) },
                        mc.cores = 10)
  
  DESeq2.eFDR.train <- mclapply(1:length(vars.pool),
                                function(i) qvalue(res.train[[i]][["pvalue"]])$qvalues,
                                mc.cores = 10)
  
  DESeq2.eFDR.train2 <- mclapply(1:length(vars.pool),
                                 function(i) {
                                   q.val <- DESeq2.eFDR.train[[i]]
                                   return(q.val[order(q.val)[1:threshold]])
                                 },
                                 mc.cores = 10)
  
  DESeq2.eFDR.GeneName.train <- mclapply(1:length(vars.pool),
                                         function(i) {
                                           q.val <- DESeq2.eFDR.train[[i]]
                                           return(rownames(cts.train)[order(q.val)[1:threshold]])
                                         },
                                         mc.cores = 10)
  
  names(res.train) = names(DESeq2.eFDR.train) = names(DESeq2.eFDR.train2) = names(DESeq2.eFDR.GeneName.train) = vars.pool 
  
  # Transform the TMM normalization factors to be used in DESeq2 (test set)
  lib.size <- colSums(cts.test)
  norm.factor <- calcNormFactors(cts.test, method = "TMM")
  size.factor <- lib.size*norm.factor/exp(mean(log(lib.size*norm.factor))) 
  
  dds <- DESeqDataSetFromMatrix(countData = cts.test, 
                                colData = coldata.test, 
                                design = ~BMI + AGE + SEX + MHABNWBC)
  sizeFactors(dds) <- size.factor
  
  res.test <- mclapply(1:length(vars.pool), 
                       function(i) {
                         return(results(DESeq(dds), name = name.formula[i])) }, 
                       mc.cores = 10)
  
  DESeq2.eFDR.test <- mclapply(1:length(vars.pool),
                               function(i) qvalue(res.test[[i]][["pvalue"]])$qvalues,
                               mc.cores = 10)
  
  DESeq2.eFDR.test2 <- mclapply(1:length(vars.pool),
                                function(i) {
                                  q.val <- DESeq2.eFDR.test[[i]]
                                  return(q.val[order(q.val)[1:threshold]])
                                },
                                mc.cores = 10)
  
  DESeq2.eFDR.GeneName.test <- mclapply(1:length(vars.pool),
                                        function(i) {
                                          q.val <- DESeq2.eFDR.test[[i]]
                                          return(rownames(cts.test)[order(q.val)[1:threshold]])
                                        },
                                        mc.cores = 10)
  
  names(res.test) = names(DESeq2.eFDR.test) = names(DESeq2.eFDR.test2) = names(DESeq2.eFDR.GeneName.test) = vars.pool
  
  save(cts.train, cts.test, coldata.train, coldata.test, vars.pool,
       res.train, DESeq2.eFDR.train, DESeq2.eFDR.train2, DESeq2.eFDR.GeneName.train, 
       res.test, DESeq2.eFDR.test, DESeq2.eFDR.test2, DESeq2.eFDR.GeneName.test, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/DESeq2Multi%s.RData", threshold, seed.num))
}
```


## Run and save modeling results
```{r}
threshold.i <- 5000
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
for (seed.i in seed.vec) {
  dat.expr.train <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/dat.expr.train%s.RDS", seed.i))
  dat.expr.test <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/dat.expr.test%s.RDS", seed.i))
  dat.pheno.train <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/dat.pheno.train%s.RDS", seed.i))
  dat.pheno.test <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/dat.pheno.test%s.RDS", seed.i))
  multi_TMM_top(seed.num = seed.i, threshold = threshold.i)
}
```


## Create functions to generate data for making boxplots of number of cDEGs and rate of cDEGs, respectively
```{r}
##------Number of cDEGs------
boxplot_data <- function (threshold = 1000, random.seed = 8809678) {
  data.table(threshold=rep(threshold,5), 
             random.seed=random.seed,
             method=c("BMAseq","DESeq2","edgeR","eBayes","voom.limma"),
             variable=rep(c("BMI","AGE","SEX","MHABNWBC"),each=5),
             num.cDEGs=c(
               ##------BMI------
               intersect(names(BMAseq.eFDR.Main.train$BMI[1:threshold]), names(BMAseq.eFDR.Main.test$BMI[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$BMI[1:threshold], DESeq2.eFDR.GeneName.test$BMI[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$BMI[1:threshold], edgeR.eFDR.GeneName.test$BMI[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$BMI[1:threshold]), names(eBayes.eFDR.test2$BMI[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$BMI[1:threshold]), names(voom.eFDR.test2$BMI[1:threshold])) |> length(),
               
               ##------AGE------
               intersect(names(BMAseq.eFDR.Main.train$AGE[1:threshold]), names(BMAseq.eFDR.Main.test$AGE[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$AGE[1:threshold], DESeq2.eFDR.GeneName.test$AGE[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$AGE[1:threshold], edgeR.eFDR.GeneName.test$AGE[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$AGE[1:threshold]), names(eBayes.eFDR.test2$AGE[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$AGE[1:threshold]), names(voom.eFDR.test2$AGE[1:threshold])) |> length(),
               
               ##------SEX------
               intersect(names(BMAseq.eFDR.Main.train$SEX[1:threshold]), names(BMAseq.eFDR.Main.test$SEX[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$SEX[1:threshold], DESeq2.eFDR.GeneName.test$SEX[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$SEX[1:threshold], edgeR.eFDR.GeneName.test$SEX[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$SEX[1:threshold]), names(eBayes.eFDR.test2$SEX[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$SEX[1:threshold]), names(voom.eFDR.test2$SEX[1:threshold])) |> length(),
               
               ##------MHABNWBC------
               intersect(names(BMAseq.eFDR.Main.train$MHABNWBC[1:threshold]), names(BMAseq.eFDR.Main.test$MHABNWBC[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$MHABNWBC[1:threshold], DESeq2.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$MHABNWBC[1:threshold], edgeR.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$MHABNWBC[1:threshold]), names(eBayes.eFDR.test2$MHABNWBC[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$MHABNWBC[1:threshold]), names(voom.eFDR.test2$MHABNWBC[1:threshold])) |> length()))
}

##------Rate of cDEGs------
boxplot_data2 <- function (threshold = 1000, random.seed = 8809678) {
  data.table(threshold=rep(threshold,5), 
             random.seed=random.seed,
             method=c("BMAseq","DESeq2","edgeR","eBayes","voom.limma"),
             variable=rep(c("BMI","AGE","SEX","MHABNWBC"),each=5),
             rate.cDEGs=c(
               ##------BMI------
               intersect(names(BMAseq.eFDR.Main.train$BMI[1:threshold]), names(BMAseq.eFDR.Main.test$BMI[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$BMI[1:threshold], DESeq2.eFDR.GeneName.test$BMI[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$BMI[1:threshold], edgeR.eFDR.GeneName.test$BMI[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$BMI[1:threshold]), names(eBayes.eFDR.test2$BMI[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$BMI[1:threshold]), names(voom.eFDR.test2$BMI[1:threshold])) |> length() / threshold,
               
               ##------AGE------
               intersect(names(BMAseq.eFDR.Main.train$AGE[1:threshold]), names(BMAseq.eFDR.Main.test$AGE[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$AGE[1:threshold], DESeq2.eFDR.GeneName.test$AGE[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$AGE[1:threshold], edgeR.eFDR.GeneName.test$AGE[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$AGE[1:threshold]), names(eBayes.eFDR.test2$AGE[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$AGE[1:threshold]), names(voom.eFDR.test2$AGE[1:threshold])) |> length() / threshold,
               
               ##------SEX------
               intersect(names(BMAseq.eFDR.Main.train$SEX[1:threshold]), names(BMAseq.eFDR.Main.test$SEX[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$SEX[1:threshold], DESeq2.eFDR.GeneName.test$SEX[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$SEX[1:threshold], edgeR.eFDR.GeneName.test$SEX[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$SEX[1:threshold]), names(eBayes.eFDR.test2$SEX[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$SEX[1:threshold]), names(voom.eFDR.test2$SEX[1:threshold])) |> length() / threshold,
               
               ##------MHABNWBC------
               intersect(names(BMAseq.eFDR.Main.train$MHABNWBC[1:threshold]), names(BMAseq.eFDR.Main.test$MHABNWBC[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$MHABNWBC[1:threshold], DESeq2.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$MHABNWBC[1:threshold], edgeR.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$MHABNWBC[1:threshold]), names(eBayes.eFDR.test2$MHABNWBC[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$MHABNWBC[1:threshold]), names(voom.eFDR.test2$MHABNWBC[1:threshold])) |> length() / threshold))
}
```


## Organize data for boxplotting
```{r}
##------Number of cDEGs------
threshold.vec <- seq(1000, 5000, 1000)
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
boxplot.data.num <- NULL
for (threshold.i in threshold.vec) {
  for (seed.i in seed.vec) {
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/BMAseqMulti%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/DESeq2Multi%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/edgeRMulti%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/eBayesMulti%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/VoomLimmaMulti%s.RData", seed.i))
    boxplot.data.num <- rbind(boxplot.data.num, boxplot_data(threshold.i, seed.i))
  }
}
saveRDS(boxplot.data.num, "../ApplicationResult/Multi/RandomSeed/numcDEGs/Boxplot/boxplot.data.num.RDS")


##------Rate of cDEGs------
threshold.vec <- seq(1000, 5000, 1000)
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
boxplot.data.rate <- NULL
for (threshold.i in threshold.vec) {
  for (seed.i in seed.vec) {
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/BMAseqMulti%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/DESeq2Multi%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/edgeRMulti%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/eBayesMulti%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/VoomLimmaMulti%s.RData", seed.i))
    boxplot.data.rate <- rbind(boxplot.data.rate, boxplot_data2(threshold.i, seed.i))
  }
}
saveRDS(boxplot.data.rate, "../ApplicationResult/Multi/RandomSeed/ratecDEGs/Boxplot/boxplot.data.rate.RDS")
```


## Make and save boxplots
```{r}
##------Number of cDEGs------
date.analysis <- format(Sys.Date(), "%Y%b%d")
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC")

for (var.i in var.vec) {
  g <- boxplot.data.num[variable==var.i] |>
    ggplot(aes(y = num.cDEGs, x = method, group = method)) + 
    geom_boxplot(outlier.shape = NA) + 
    geom_jitter(aes(color = method), size = 0.8) + 
    stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
    facet_wrap(vars(threshold)) + 
    labs(y = "Number of common differentially expressed genes", x = "Method") + 
    theme_bw() + 
    theme(legend.position = "none",
          axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
    ggsave(filename = sprintf("../ApplicationResult/Multi/RandomSeed/numcDEGs/Boxplot/%s_%s.png", date.analysis, var.i),
           plot = g, device = "png", width = 9, height = 6, units = "in")
}


##------Rate of cDEGs------
for (var.i in var.vec) {
  g <- boxplot.data.rate[variable==var.i] |>
    ggplot(aes(y = rate.cDEGs, x = method, group = method)) + 
    geom_boxplot(outlier.shape = NA) + 
    geom_jitter(aes(color = method), size = 0.8) + 
    stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
    facet_wrap(vars(threshold)) + 
    labs(y = "Rate of common differentially expressed genes", x = "Method") + 
    theme_bw() + 
    theme(legend.position = "none",
          axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
    ggsave(filename = sprintf("../ApplicationResult/Multi/RandomSeed/ratecDEGs/Boxplot/%s_%s.png", date.analysis, var.i),
           plot = g, device = "png", width = 9, height = 6, units = "in")
}
```


## Estimate the duplicate rate of a common differentially expressed gene associated with a variable
```{r}
cDEGs = cDEGs.list = NULL
threshold.vec <- seq(1000, 5000, 1000)
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)

duplicate_rate_cDEGs_per_variable <- function(var.name = "BMI", method.name = "BMAseq") {
  for (threshold.i in threshold.vec) {
    for (seed.i in seed.vec) {
      if (method.name == "BMAseq") {
        load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sMulti%s.RData", method.name, seed.i))
        cDEGs <- c(intersect(names(get(sprintf("%s.eFDR.Main.train", method.name))[[var.name]][1:threshold.i]), 
                             names(get(sprintf("%s.eFDR.Main.test", method.name))[[var.name]][1:threshold.i])))
      } else if (method.name %in% c("DESeq2", "edgeR")) {
        load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sMulti%s.RData", method.name, seed.i))
        cDEGs <- c(intersect(get(sprintf("%s.eFDR.GeneName.train", method.name))[[var.name]][1:threshold.i], 
                             get(sprintf("%s.eFDR.GeneName.test", method.name))[[var.name]][1:threshold.i]))
      } else if (method.name == "eBayes") {
        load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sMulti%s.RData", method.name, seed.i))
        cDEGs <- c(intersect(names(get(sprintf("%s.eFDR.train2", method.name))[[var.name]][1:threshold.i]), 
                             names(get(sprintf("%s.eFDR.test2", method.name))[[var.name]][1:threshold.i])))
      } else {
        load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sMulti%s.RData", method.name, seed.i))
        method.name2 <- "voom"
        cDEGs <- c(intersect(names(get(sprintf("%s.eFDR.train2", method.name2))[[var.name]][1:threshold.i]), 
                             names(get(sprintf("%s.eFDR.test2", method.name2))[[var.name]][1:threshold.i])))
      }
      cDEGs.list <- append(cDEGs.list, list(cDEGs))
    }
    cDEGs.unique <- unique(unlist(cDEGs.list))
    
    # Create a matrix that shows the occurrence of each unique gene among the 10 random seed trials
    cDEGs.matrix <- matrix(0, nrow = length(cDEGs.unique), ncol = 10)
    rownames(cDEGs.matrix) <- cDEGs.unique
    colnames(cDEGs.matrix) <- seed.vec   
    for (i in 1:10) {
      cDEGs.i.seed <- cDEGs.list[[i]]
      for (j in 1:length(cDEGs.unique)) {
        if (cDEGs.unique[j] %in% cDEGs.i.seed) {
          cDEGs.matrix[j, i] <- 1
        }
      }
    }
    Sum <- apply(cDEGs.matrix, 1, sum)
    Percent <- Sum/10
    cDEGs.matrix <- cbind(cDEGs.matrix, Sum, Percent) # Add two new columns to the matrix
    saveRDS(cDEGs.matrix, sprintf("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/%sMulti%s_%s.RDS", method.name, var.name, threshold.i))
    
    print(paste("Complete the duplicated rate matrix of cDEGs associated with", var.name, "identified by", method.name, "among 10 random seeds for the threshold", threshold.i))
    cDEGs = cDEGs.list =  NULL
  }
}

mapply(duplicate_rate_cDEGs_per_variable,
       rep(c("BMI", "AGE", "SEX", "MHABNWBC"), each = 5), 
       rep(c("BMAseq", "DESeq2", "edgeR", "eBayes", "VoomLimma"), times = 4))

method.vec <- c("BMAseq", "DESeq2", "edgeR", "eBayes", "VoomLimma")
var.name <- "BMI"
threshold.i <- 2000
for (method.name in method.vec) {
  data <- readRDS(sprintf("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/%sMulti%s_%s.RDS", method.name, var.name, threshold.i))
  write.csv(data, sprintf("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/%sMulti%s_%s.csv", method.name, var.name, threshold.i))
}
```


## Prepare and organize the data for boxplots
```{r}
boxplot_duplicate_rate_data <- function (var.name = NULL, method.name = NULL, threshold = NULL) {
  load.data <- readRDS(sprintf("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/%sMulti%s_%s.RDS", method.name, var.name, threshold))
  temp.data <- data.table(threshold=threshold, 
                          method=method.name,
                          variable=var.name,
                          cDEGs.duplicate.rate.median=median(load.data[,12]),
                          cDEGs.duplicate.rate.mean=mean(load.data[,12]))
  return(temp.data)
}

organized_duplicate_rate_data <- mapply(boxplot_duplicate_rate_data,
                                        var.name = rep(c("BMI", "AGE","SEX", "MHABNWBC"), each = 25), 
                                        method.name = rep(rep(c("BMAseq", "DESeq2", "edgeR", "eBayes", "VoomLimma"), each = 5), times = 4),
                                        threshold = rep(rep(seq(1000, 5000, 1000), times = 5), times = 4), SIMPLIFY = F) |> rbindlist()

# data.frame(var.name, method.name, threshold)
```


## Make and save boxplots
```{r}
organized_duplicate_rate_data[,variable:=factor(variable,levels=c("BMI","AGE","SEX","MHABNWBC"))]
organized_duplicate_rate_data[,method:=factor(method,levels=c("BMAseq","DESeq2","edgeR","eBayes","VoomLimma"))]
organized_duplicate_rate_data[,cDEGs.duplicate.rate.median:=as.numeric(cDEGs.duplicate.rate.median)]
organized_duplicate_rate_data[,cDEGs.duplicate.rate.mean:=as.numeric(cDEGs.duplicate.rate.mean)]
date.analysis <- format(Sys.Date(), "%Y%b%d")
g1 <- organized_duplicate_rate_data |>
  ggplot(mapping = aes(y = cDEGs.duplicate.rate.median, x = method, group = method)) + 
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(aes(color = method), size = 0.8) + 
  facet_wrap(vars(variable)) + 
  scale_y_continuous(labels = scales::label_number(accuracy = 0.01),
                     breaks = seq(0, 1, 0.1)) + 
  labs(y = "Median duplicate rate of cDEGs", x = "Method") + 
  theme_bw() + 
  theme(legend.position = "none",
        axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
ggsave(filename = sprintf("../ApplicationResult/Multi/RandomSeed/DuplicatedRateMatrix/Boxplot/%s_median.png", date.analysis),
       plot = g1, device = "png", width = 9, height = 6, units = "in")

g2 <- organized_duplicate_rate_data |>
  ggplot(mapping = aes(y = cDEGs.duplicate.rate.mean, x = method, group = method)) + 
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(aes(color = method), size = 0.8) + 
  facet_wrap(vars(variable)) + 
  scale_y_continuous(labels = scales::label_number(accuracy = 0.01),
                     breaks = seq(0, 1, 0.1)) + 
  labs(y = "Mean duplicate rate of cDEGs", x = "Method") + 
  theme_bw() + 
  theme(legend.position = "none",
        axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
ggsave(filename = sprintf("../ApplicationResult/Multi/RandomSeed/DuplicatedRateMatrix/Boxplot/%s_mean.png", date.analysis),
       plot = g2, device = "png", width = 9, height = 6, units = "in")
```


## Make Venn digrams of cDEGs
```{r}
threshold <- 2000
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC")
date.analysis <- format(Sys.Date(), "%Y%b%d")

make_vennplot <-function(var.name = NULL, threshold = NULL) {

  BMAseq.cDEGs <- intersect(names(BMAseq.eFDR.Main.train[[var.name]][1:threshold]), names(BMAseq.eFDR.Main.test[[var.name]][1:threshold]))
  DESeq2.cDEGs <- intersect(DESeq2.eFDR.GeneName.train[[var.name]][1:threshold], DESeq2.eFDR.GeneName.test[[var.name]][1:threshold])
  edgeR.cDEGs <- intersect(edgeR.eFDR.GeneName.train[[var.name]][1:threshold], edgeR.eFDR.GeneName.test[[var.name]][1:threshold])
  eBayes.cDEGs <- intersect(names(eBayes.eFDR.train2[[var.name]][1:threshold]), names(eBayes.eFDR.test2[[var.name]][1:threshold]))
  voom.limma.cDEGs <- intersect(names(voom.eFDR.train2[[var.name]][1:threshold]), names(voom.eFDR.test2[[var.name]][1:threshold]))
  
  g <- ggVennDiagram(list(BMAseq = BMAseq.cDEGs, 
                          DESeq2 = DESeq2.cDEGs,
                          edgeR = edgeR.cDEGs,
                          eBayes = eBayes.cDEGs,
                          voom.limma = voom.limma.cDEGs),
                     label_alpha = 0, label_color = "white") +
    theme(legend.position = "none", 
          plot.title = element_text(hjust = 0.5, colour = "red", size = 18)) + 
    ggtitle(paste("The intersection of cDEGs associated with the", var.name, "\nunder the seed", seed.i, "and threshold", threshold)) + 
    scale_fill_distiller(palette = "Dark2", direction = -1)
  
  ggsave(filename = paste0("../ApplicationResult/Multi/RandomSeed/TMM_Top2000/Vennplot/", date.analysis, "_", var.name, "_", seed.i, ".png"),
         plot = g,
         device = "png",
         width = 10,
         height = 7,
         units = "in")
}

for (seed.i in seed.vec) {
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/BMAseqMulti%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/DESeq2Multi%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/edgeRMulti%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/eBayesMulti%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/VoomLimmaMulti%s.RData", seed.i))
  for (var.i in var.vec) {
    make_vennplot(var.name = var.i, threshold = 2000)
  }
}
```


## Extract unique cDEGs identified by BMAseq per seed and all unique cDEGs identified by BMAseq across 10 seeds
```{r}
file.name <- "../ApplicationResult/Multi/RandomSeed/TMM_Top2000/Vennplot/uniq_cDEGs_BMI_BMAseq_per_seed.txt"
unique.cDEGs.list <- NULL
extract_unique_cDEGs_per_seed <-function(var.name = NULL, threshold = NULL) {
  BMAseq.cDEGs <- intersect(names(BMAseq.eFDR.Main.train[[var.name]][1:threshold]), names(BMAseq.eFDR.Main.test[[var.name]][1:threshold]))
  DESeq2.cDEGs <- intersect(DESeq2.eFDR.GeneName.train[[var.name]][1:threshold], DESeq2.eFDR.GeneName.test[[var.name]][1:threshold])
  edgeR.cDEGs <- intersect(edgeR.eFDR.GeneName.train[[var.name]][1:threshold], edgeR.eFDR.GeneName.test[[var.name]][1:threshold])
  eBayes.cDEGs <- intersect(names(eBayes.eFDR.train2[[var.name]][1:threshold]), names(eBayes.eFDR.test2[[var.name]][1:threshold]))
  voom.limma.cDEGs <- intersect(names(voom.eFDR.train2[[var.name]][1:threshold]), names(voom.eFDR.test2[[var.name]][1:threshold]))
  unique.cDEGs <- setdiff(BMAseq.cDEGs, union(DESeq2.cDEGs, edgeR.cDEGs) |> union(eBayes.cDEGs) |> union(voom.limma.cDEGs))
  return(unique.cDEGs)
}

for (seed.i in seed.vec) {
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/BMAseqMulti%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/DESeq2Multi%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/edgeRMulti%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/eBayesMulti%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiModel/VoomLimmaMulti%s.RData", seed.i))
  unique.cDEGs <- extract_unique_cDEGs_per_seed(var.name = "BMI", threshold = 2000)
  unique.cDEGs.list <- append(unique.cDEGs.list, list(unique.cDEGs))
  write(sprintf("\nThe following gene IDs are unique gene IDs associated with the main effect of %s identified by BMAseq under the seed %s and threshold %s", var.name, seed.i, threshold), file = file.name, append = T)
  write.table(unique.cDEGs, file = file.name, quote = F, col.names = F, append = T)
}
names(unique.cDEGs.list) <- seed.vec

file.name <- "../ApplicationResult/Multi/RandomSeed/TMM_Top2000/Vennplot/uniq_cDEGs_BMI_BMAseq_all_seeds.txt"
write.table(unlist(unique.cDEGs.list) |> unique(), file = file.name, quote = F, col.names = F)
```


# Interaction analysis
## Create functions of building multivariable models with interaction terms
```{r}
multiInt_TMM_top <- function(seed.num = 999, threshold = 2000) {
  
  ##------set variables of interest------
  var.pool <- c("BMI", "AGE", "SEX", "MHABNWBC")  
  var.pool.add <- c(var.pool, "BMIxSEX")
  
  
  ##------BMAseq------
  output.multi.int1.train <- BMAseq.multi.postprob.norm(
    dat.expr.counts = dat.expr.train,  
    dat.pheno = dat.pheno.train, 
    var.pool = var.pool, 
    max.nvar = 4, 
    interaction = "BMI&SEX", 
    cut.BF = 1)
  
  output.multi.int2.train <- BMAseq.multi.DEG(
    dat.pheno = output.multi.int1.train$dat.pheno.new,
    model.space = output.multi.int1.train$model.space, 
    post.modelprob = output.multi.int1.train$post.modelprob, 
    var.pool = var.pool,
    interact = T, 
    cut.FDR = 0.05)
  
  output.multi.int1.test <- BMAseq.multi.postprob.norm(
    dat.expr.counts = dat.expr.test, 
    dat.pheno = dat.pheno.test, 
    var.pool = var.pool, 
    max.nvar = 4, 
    interaction = "BMI&SEX", 
    cut.BF = 1)
  
  output.multi.int2.test <- BMAseq.multi.DEG(
    dat.pheno = output.multi.int1.test$dat.pheno.new,
    model.space = output.multi.int1.test$model.space, 
    post.modelprob = output.multi.int1.test$post.modelprob, 
    var.pool = var.pool,
    interact = T, 
    cut.FDR = 0.05)
  
  BMAseq.eFDR.Main.train <- mclapply(1:length(var.pool),
                                     function(i) output.multi.int2.train$eFDR.Main[, i][order(output.multi.int2.train$eFDR.Main[, i])[1:threshold]],
                                     mc.cores = 4)
  
  BMAseq.eFDR.Main.test <- mclapply(1:length(var.pool),
                                         function(i) output.multi.int2.test$eFDR.Main[, i][order(output.multi.int2.test$eFDR.Main[, i])[1:threshold]],
                                         mc.cores = 4)
  
  names(BMAseq.eFDR.Main.train) = names(BMAseq.eFDR.Main.test) = var.pool
  
  BMAseq.eFDR.Interaction.train <- mclapply(1:length(var.pool),
                                            function(i) output.multi.int2.train$eFDR.Interaction[, i][order(output.multi.int2.train$eFDR.Interaction[, i])[1:threshold]],
                                            mc.cores = 4)
  
  BMAseq.eFDR.Interaction.test <- mclapply(1:length(var.pool),
                                           function(i) output.multi.int2.test$eFDR.Interaction[, i][order(output.multi.int2.test$eFDR.Interaction[, i])[1:threshold]],
                                           mc.cores = 4)
  
  names(BMAseq.eFDR.Interaction.train) = names(BMAseq.eFDR.Interaction.test) = var.pool
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, var.pool, var.pool.add, 
       output.multi.int1.train, output.multi.int2.train, output.multi.int1.test, output.multi.int2.test, 
       BMAseq.eFDR.Main.train, BMAseq.eFDR.Main.test, BMAseq.eFDR.Interaction.train, BMAseq.eFDR.Interaction.test, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/BMAseqMultiInt%s.RData", threshold, seed.num)) 
  
  
  ##------voom + limma------
  design.train <- model.matrix(~BMI + AGE + SEX + MHABNWBC + BMI*SEX, data = dat.pheno.train)
  design.test <- model.matrix(~BMI + AGE + SEX + MHABNWBC + BMI*SEX, data = dat.pheno.test)
  
  voom.train <- voom(counts = dat.expr.train,
                     design = design.train,
                     lib.size = colSums(dat.expr.train)*calcNormFactors(dat.expr.train)) 
  
  voom.fit.train <- lmFit(voom.train[["E"]],
                          design = design.train,
                          weights = voom.train[["weights"]])
  
  voom.eFDR.train <- mclapply(1:length(var.pool.add),
                              function(i) {
                                t <- voom.fit.train[["coefficients"]][, i + 1]/voom.fit.train[["stdev.unscaled"]][, i + 1]/voom.fit.train[["sigma"]]
                                p <- 2*pt(-abs(t), df = voom.fit.train[["df.residual"]])
                                return(qvalue(p)) },
                              mc.cores = 4)
  
  voom.eFDR.train2 <- mclapply(1:length(var.pool.add),
                               function(i) {
                                 q.val <- voom.eFDR.train[[i]][["qvalues"]]
                                 return(q.val[order(q.val)[1:threshold]])
                               },
                               mc.cores = 4)
  
  names(voom.eFDR.train) = names(voom.eFDR.train2) = var.pool.add 
  
  voom.test <- voom(counts = dat.expr.test,
                    design = design.test,
                    lib.size = colSums(dat.expr.test)*calcNormFactors(dat.expr.test))
  
  voom.fit.test <- lmFit(voom.test[["E"]],
                         design = design.test,
                         weights = voom.test[["weights"]])
  
  voom.eFDR.test <- mclapply(1:length(var.pool.add),
                             function(i) {
                               t <- voom.fit.test[["coefficients"]][, i + 1]/voom.fit.test[["stdev.unscaled"]][, i + 1]/voom.fit.test[["sigma"]]
                               p <- 2*pt(-abs(t), df = voom.fit.test[["df.residual"]])
                               return(qvalue(p)) },
                             mc.cores = 4)
  
  voom.eFDR.test2 <- mclapply(1:length(var.pool.add),
                              function(i) {
                                q.val <- voom.eFDR.test[[i]][["qvalues"]]
                                return(q.val[order(q.val)[1:threshold]])
                              },
                                mc.cores = 4)
  
  names(voom.eFDR.test) = names(voom.eFDR.test2) = var.pool.add 
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, var.pool, var.pool.add, 
       design.train, design.test, voom.train, voom.fit.train, voom.eFDR.train, voom.eFDR.train2, 
       voom.test, voom.fit.test, voom.eFDR.test, voom.eFDR.test2, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/VoomLimmaMultiInt%s.RData", threshold, seed.num))
  
  
  ##------voom + limma + eBayes------
  voom.train <- voom(counts = dat.expr.train,
                     design = design.train,
                     lib.size = colSums(dat.expr.train)*calcNormFactors(dat.expr.train))
  
  eBayes.fit.train <- lmFit(voom.train[["E"]],
                            design = design.train,
                            weights = voom.train[["weights"]]) %>% 
    eBayes()
  
  eBayes.eFDR.train <- mclapply(1:length(var.pool.add),
                                function(i) eBayes.fit.train[["p.value"]][, i + 1] %>% 
                                  qvalue(),
                                mc.cores = 4)
  
  eBayes.eFDR.train2 <- mclapply(1:length(var.pool.add),
                                 function(i) {
                                   q.val <- eBayes.eFDR.train[[i]][["qvalues"]]
                                   return(q.val[order(q.val)[1:threshold]])
                                 },
                                 mc.cores = 4)
  
  names(eBayes.eFDR.train) = names(eBayes.eFDR.train2) = var.pool.add 
  
  voom.test <- voom(counts = dat.expr.test,
                    design = design.test,
                    lib.size = colSums(dat.expr.test)*calcNormFactors(dat.expr.test))
  
  eBayes.fit.test <- lmFit(voom.test[["E"]],
                           design = design.test,
                           weights = voom.test[["weights"]]) %>% 
    eBayes()
  
  eBayes.eFDR.test <- mclapply(1:length(var.pool.add),
                               function(i) eBayes.fit.test[["p.value"]][, i + 1] %>% 
                                 qvalue(),
                               mc.cores = 4)
  
  eBayes.eFDR.test2 <- mclapply(1:length(var.pool.add),
                                function(i) {
                                  q.val <- eBayes.eFDR.test[[i]][["qvalues"]]
                                  return(q.val[order(q.val)[1:threshold]]) 
                                },
                                mc.cores = 4)
  
  names(eBayes.eFDR.test) = names(eBayes.eFDR.test2) = var.pool.add 
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, var.pool, var.pool.add, 
       design.train, design.test, voom.train, eBayes.fit.train, eBayes.eFDR.train, eBayes.eFDR.train2, 
       voom.test, eBayes.fit.test, eBayes.eFDR.test, eBayes.eFDR.test2, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/eBayesMultiInt%s.RData", threshold, seed.num))
  
  
  ##------edgeR------
  y.train <- DGEList(counts = dat.expr.train, 
                     lib.size = colSums(dat.expr.train)) %>% 
    calcNormFactors() %>% 
    estimateGLMTrendedDisp(design.train)
  
  edgeR.fit.train <- glmQLFit(y.train, design.train)
  
  qlf.train <- mclapply(1:length(var.pool.add),
                        function(i) glmQLFTest(edgeR.fit.train, coef = i + 1),
                        mc.cores = 4)
  
  edgeR.eFDR.train <- mclapply(1:length(var.pool.add),
                               function(i) qlf.train[[i]][["table"]][["PValue"]] %>%
                                 qvalue(),
                               mc.cores = 4)
  
  edgeR.eFDR.train2 <- mclapply(1:length(var.pool.add),
                                function(i) {
                                  q.val <- edgeR.eFDR.train[[i]][["qvalues"]]
                                  return(q.val[order(q.val)[1:threshold]]) 
                                },
                                mc.cores = 4)
  
  edgeR.eFDR.GeneName.train <- mclapply(1:length(var.pool.add),
                                        function(i) {
                                          q.val <- edgeR.eFDR.train[[i]][["qvalues"]]
                                          return(rownames(dat.expr.train)[order(q.val)[1:threshold]])
                                        },
                                        mc.cores = 4)
  
  names(qlf.train) = names(edgeR.eFDR.train) = names(edgeR.eFDR.train2) = names(edgeR.eFDR.GeneName.train) = var.pool.add 
  
  y.test <- DGEList(counts = dat.expr.test, 
                    lib.size = colSums(dat.expr.test)) %>% 
    calcNormFactors() %>% 
    estimateGLMTrendedDisp(design.test)
  
  edgeR.fit.test <- glmQLFit(y.test, design.test)
  
  qlf.test <- mclapply(1:length(var.pool.add),
                       function(i) glmQLFTest(edgeR.fit.test, coef = i + 1),
                       mc.cores = 4)
  
  edgeR.eFDR.test <- mclapply(1:length(var.pool.add),
                              function(i) qlf.test[[i]][["table"]][["PValue"]] %>% 
                                qvalue(),
                              mc.cores = 4)
  
  edgeR.eFDR.test2 <- mclapply(1:length(var.pool.add),
                               function(i) {
                                 q.val <- edgeR.eFDR.test[[i]][["qvalues"]]
                                 return(q.val[order(q.val)[1:threshold]]) 
                               },
                               mc.cores = 4)
  
  edgeR.eFDR.GeneName.test <- mclapply(1:length(var.pool.add),
                                       function(i) {
                                         q.val <- edgeR.eFDR.test[[i]][["qvalues"]]
                                         return(rownames(dat.expr.test)[order(q.val)[1:threshold]])
                                       },
                                       mc.cores = 4)
  
  names(qlf.test) = names(edgeR.eFDR.test) = names(edgeR.eFDR.test2) = names(edgeR.eFDR.GeneName.test) = var.pool.add 
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, var.pool, var.pool.add, design.train, design.test, 
       y.train, edgeR.fit.train, edgeR.eFDR.train, edgeR.eFDR.train2, edgeR.eFDR.GeneName.train, 
       y.test, edgeR.fit.test, edgeR.eFDR.test, edgeR.eFDR.test2, edgeR.eFDR.GeneName.test, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/edgeRMultiInt%s.RData", threshold, seed.num))
  
  
  ##------DESeq2------
  cts.train <- as.matrix(dat.expr.train)
  cts.test <- as.matrix(dat.expr.test)
  coldata.train <- dat.pheno.train
  coldata.test <- dat.pheno.test
  
  name.formula <- c("BMI_high_vs_low", "AGE_old_vs_young", "SEX_male_vs_female", 
                    "MHABNWBC_yes_vs_no", "BMIhigh.SEXmale")  
  
  lib.size <- colSums(cts.train)
  norm.factor <- calcNormFactors(cts.train, method = "TMM")
  size.factor <- lib.size*norm.factor/exp(mean(log(lib.size*norm.factor))) 
  
  dds <- DESeqDataSetFromMatrix(countData = cts.train, 
                                colData = coldata.train, 
                                design = ~BMI + AGE + SEX + MHABNWBC + BMI*SEX)
  sizeFactors(dds) <- size.factor
  
  res.train <- mclapply(1:length(var.pool.add), 
                        function(i) {
                          return(results(DESeq(dds), name = name.formula[i])) },
                        mc.cores = 4)
  
  DESeq2.eFDR.train <- mclapply(1:length(var.pool.add),
                                function(i) qvalue(res.train[[i]][["pvalue"]])$qvalues,
                                mc.cores = 4)
  
  DESeq2.eFDR.train2 <- mclapply(1:length(var.pool.add),
                                 function(i) {
                                   q.val <- DESeq2.eFDR.train[[i]]
                                   return(q.val[order(q.val)[1:threshold]])
                                 },
                                 mc.cores = 4)
  
  DESeq2.eFDR.GeneName.train <- mclapply(1:length(var.pool.add),
                                         function(i) {
                                           q.val <- DESeq2.eFDR.train[[i]]
                                           return(rownames(cts.train)[order(q.val)[1:threshold]])
                                         },
                                         mc.cores = 4)
  
  names(res.train) = names(DESeq2.eFDR.train) = names(DESeq2.eFDR.train2) = names(DESeq2.eFDR.GeneName.train) = var.pool.add 
  
  lib.size <- colSums(cts.test)
  norm.factor <- calcNormFactors(cts.test, method = "TMM")
  size.factor <- lib.size*norm.factor/exp(mean(log(lib.size*norm.factor))) 
  
  dds <- DESeqDataSetFromMatrix(countData = cts.test, 
                                colData = coldata.test, 
                                design = ~BMI + AGE + SEX + MHABNWBC + BMI*SEX)
  sizeFactors(dds) <- size.factor
  
  res.test <- mclapply(1:length(var.pool.add), 
                       function(i) {
                         return(results(DESeq(dds), name = name.formula[i])) }, 
                       mc.cores = 4)
  
  DESeq2.eFDR.test <- mclapply(1:length(var.pool.add),
                               function(i) qvalue(res.test[[i]][["pvalue"]])$qvalues,
                               mc.cores = 4)
  
  DESeq2.eFDR.test2 <- mclapply(1:length(var.pool.add),
                                function(i) {
                                  q.val <- DESeq2.eFDR.test[[i]]
                                  return(q.val[order(q.val)[1:threshold]])
                                },
                                mc.cores = 4)
  
  DESeq2.eFDR.GeneName.test <- mclapply(1:length(var.pool.add),
                                        function(i) {
                                          q.val <- DESeq2.eFDR.test[[i]]
                                          return(rownames(cts.test)[order(q.val)[1:threshold]])
                                        },
                                        mc.cores = 4)
  
  names(res.test) = names(DESeq2.eFDR.test) = names(DESeq2.eFDR.test2) = names(DESeq2.eFDR.GeneName.test) = var.pool.add
  
  save(cts.train, cts.test, coldata.train, coldata.test, var.pool, var.pool.add, 
       res.train, DESeq2.eFDR.train, DESeq2.eFDR.train2, DESeq2.eFDR.GeneName.train, 
       res.test, DESeq2.eFDR.test, DESeq2.eFDR.test2, DESeq2.eFDR.GeneName.test, 
       file = sprintf("../ApplicationData/derived/RandomSeed/Top%s/DESeq2MultiInt%s.RData", threshold, seed.num))
}
```


## Run and save modeling results
```{r}
threshold.i <- 5000
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
for (seed.i in seed.vec) {
  dat.expr.train <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/dat.expr.train%s.RDS", seed.i))
  dat.expr.test <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/dat.expr.test%s.RDS", seed.i))
  dat.pheno.train <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/dat.pheno.train%s.RDS", seed.i))
  dat.pheno.test <- readRDS(file = sprintf("../ApplicationData/derived/RandomSeed/dat.pheno.test%s.RDS", seed.i))
  multiInt_TMM_top(seed.num = seed.i, threshold = threshold.i)
}
```


## Create functions to generate data for making boxplots of number of cDEGs and rate of cDEGs, respectively
```{r}
##------Number of cDEGs------
boxplot_data <- function (threshold = 1000, random.seed = 8809678) {
  data.table(threshold=rep(threshold,5), 
             random.seed=random.seed,
             method=c("BMAseq","DESeq2","edgeR","eBayes","voom.limma"),
             variable=rep(c("BMI","AGE","SEX","WBC","BMIxSEX"),each=5),
             num.cDEGs=c(
               intersect(names(BMAseq.eFDR.Main.train$BMI[1:threshold]), names(BMAseq.eFDR.Main.test$BMI[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$BMI[1:threshold], DESeq2.eFDR.GeneName.test$BMI[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$BMI[1:threshold], edgeR.eFDR.GeneName.test$BMI[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$BMI[1:threshold]), names(eBayes.eFDR.test2$BMI[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$BMI[1:threshold]), names(voom.eFDR.test2$BMI[1:threshold])) |> length(),
               
               intersect(names(BMAseq.eFDR.Main.train$AGE[1:threshold]), names(BMAseq.eFDR.Main.test$AGE[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$AGE[1:threshold], DESeq2.eFDR.GeneName.test$AGE[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$AGE[1:threshold], edgeR.eFDR.GeneName.test$AGE[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$AGE[1:threshold]), names(eBayes.eFDR.test2$AGE[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$AGE[1:threshold]), names(voom.eFDR.test2$AGE[1:threshold])) |> length(),
               
               intersect(names(BMAseq.eFDR.Main.train$SEX[1:threshold]), names(BMAseq.eFDR.Main.test$SEX[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$SEX[1:threshold], DESeq2.eFDR.GeneName.test$SEX[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$SEX[1:threshold], edgeR.eFDR.GeneName.test$SEX[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$SEX[1:threshold]), names(eBayes.eFDR.test2$SEX[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$SEX[1:threshold]), names(voom.eFDR.test2$SEX[1:threshold])) |> length(),
               
               intersect(names(BMAseq.eFDR.Main.train$MHABNWBC[1:threshold]), names(BMAseq.eFDR.Main.test$MHABNWBC[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$MHABNWBC[1:threshold], DESeq2.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$MHABNWBC[1:threshold], edgeR.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$MHABNWBC[1:threshold]), names(eBayes.eFDR.test2$MHABNWBC[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$MHABNWBC[1:threshold]), names(voom.eFDR.test2$MHABNWBC[1:threshold])) |> length(),
               
               intersect(names(BMAseq.eFDR.Interaction.train$BMI[1:threshold]), names(BMAseq.eFDR.Interaction.test$BMI[1:threshold])) |> length(),
               intersect(DESeq2.eFDR.GeneName.train$BMIxSEX[1:threshold], DESeq2.eFDR.GeneName.test$BMIxSEX[1:threshold]) |> length(),
               intersect(edgeR.eFDR.GeneName.train$BMIxSEX[1:threshold], edgeR.eFDR.GeneName.test$BMIxSEX[1:threshold]) |> length(),
               intersect(names(eBayes.eFDR.train2$BMIxSEX[1:threshold]), names(eBayes.eFDR.test2$BMIxSEX[1:threshold])) |> length(),
               intersect(names(voom.eFDR.train2$BMIxSEX[1:threshold]), names(voom.eFDR.test2$BMIxSEX[1:threshold])) |> length()))
}


##------Rate of cDEGs------
boxplot_data2 <- function (threshold = 1000, random.seed = 8809678) {
  data.table(threshold=rep(threshold,5), 
             random.seed=random.seed,
             method=c("BMAseq","DESeq2","edgeR","eBayes","voom.limma"),
             variable=rep(c("BMI","AGE","SEX","WBC","BMIxSEX"),each=5),
             rate.cDEGs=c(
               intersect(names(BMAseq.eFDR.Main.train$BMI[1:threshold]), names(BMAseq.eFDR.Main.test$BMI[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$BMI[1:threshold], DESeq2.eFDR.GeneName.test$BMI[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$BMI[1:threshold], edgeR.eFDR.GeneName.test$BMI[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$BMI[1:threshold]), names(eBayes.eFDR.test2$BMI[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$BMI[1:threshold]), names(voom.eFDR.test2$BMI[1:threshold])) |> length() / threshold,
               
               intersect(names(BMAseq.eFDR.Main.train$AGE[1:threshold]), names(BMAseq.eFDR.Main.test$AGE[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$AGE[1:threshold], DESeq2.eFDR.GeneName.test$AGE[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$AGE[1:threshold], edgeR.eFDR.GeneName.test$AGE[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$AGE[1:threshold]), names(eBayes.eFDR.test2$AGE[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$AGE[1:threshold]), names(voom.eFDR.test2$AGE[1:threshold])) |> length() / threshold,
               
               intersect(names(BMAseq.eFDR.Main.train$SEX[1:threshold]), names(BMAseq.eFDR.Main.test$SEX[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$SEX[1:threshold], DESeq2.eFDR.GeneName.test$SEX[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$SEX[1:threshold], edgeR.eFDR.GeneName.test$SEX[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$SEX[1:threshold]), names(eBayes.eFDR.test2$SEX[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$SEX[1:threshold]), names(voom.eFDR.test2$SEX[1:threshold])) |> length() / threshold,
               
               intersect(names(BMAseq.eFDR.Main.train$MHABNWBC[1:threshold]), names(BMAseq.eFDR.Main.test$MHABNWBC[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$MHABNWBC[1:threshold], DESeq2.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$MHABNWBC[1:threshold], edgeR.eFDR.GeneName.test$MHABNWBC[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$MHABNWBC[1:threshold]), names(eBayes.eFDR.test2$MHABNWBC[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$MHABNWBC[1:threshold]), names(voom.eFDR.test2$MHABNWBC[1:threshold])) |> length() / threshold,
               
               intersect(names(BMAseq.eFDR.Interaction.train$BMI[1:threshold]), names(BMAseq.eFDR.Interaction.test$BMI[1:threshold])) |> length() / threshold,
               intersect(DESeq2.eFDR.GeneName.train$BMIxSEX[1:threshold], DESeq2.eFDR.GeneName.test$BMIxSEX[1:threshold]) |> length() / threshold,
               intersect(edgeR.eFDR.GeneName.train$BMIxSEX[1:threshold], edgeR.eFDR.GeneName.test$BMIxSEX[1:threshold]) |> length() / threshold,
               intersect(names(eBayes.eFDR.train2$BMIxSEX[1:threshold]), names(eBayes.eFDR.test2$BMIxSEX[1:threshold])) |> length() / threshold,
               intersect(names(voom.eFDR.train2$BMIxSEX[1:threshold]), names(voom.eFDR.test2$BMIxSEX[1:threshold])) |> length() / threshold))
}
```


## Organize data for boxplotting
```{r}
##------Number of cDEGs------
threshold.vec <- seq(1000, 5000, 1000)
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
boxplot.data.new <- NULL
for (threshold.i in threshold.vec) {
  for (seed.i in seed.vec) {
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/BMAseqMultiInt%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/DESeq2MultiInt%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/edgeRMultiInt%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/eBayesMultiInt%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/VoomLimmaMultiInt%s.RData", seed.i))
    boxplot.data.new <- rbind(boxplot.data.new, boxplot_data(threshold.i, seed.i))
  }
}


##------Rate of cDEGs------
threshold.vec <- seq(1000, 5000, 1000)
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
boxplot.data.new2 <- NULL
for (threshold.i in threshold.vec) {
  for (seed.i in seed.vec) {
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/BMAseqMultiInt%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/DESeq2MultiInt%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/edgeRMultiInt%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/eBayesMultiInt%s.RData", seed.i))
    load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/VoomLimmaMultiInt%s.RData", seed.i))
    boxplot.data.new2 <- rbind(boxplot.data.new2, boxplot_data2(threshold.i, seed.i))
  }
}
```


## Make and save boxplots
```{r}
##------Number of cDEGs------
date.analysis <- format(Sys.Date(), "%Y%b%d")
var.vec <- c("BMI", "AGE", "SEX", "WBC", "BMIxSEX")

for (var.i in var.vec) {
  g <- boxplot.data.new[variable==var.i] |>
    ggplot(aes(y = num.cDEGs, x = method, group = method, fill = method)) + 
    geom_boxplot() + 
    stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
    facet_wrap(vars(threshold)) + 
    labs(y = "Number of common differentially expressed genes", x = "Method") + 
    theme_bw() + 
    theme(legend.position = "none",
          axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
    ggsave(filename = sprintf("../ApplicationResult/Multi_Interaction/RandomSeed/numcDEGs/Boxplot/%s_%s.png", date.analysis, var.i),
           plot = g, device = "png", width = 9, height = 6, units = "in")
}


##------Rate of cDEGs------
for (var.i in var.vec) {
  g <- boxplot.data.new2[variable==var.i] |>
    ggplot(aes(y = rate.cDEGs, x = method, group = method, fill = method)) + 
    geom_boxplot() + 
    stat_summary(fun = median, geom = "line", aes(group = 1), color = "#0073C2FF") +  
    facet_wrap(vars(threshold)) + 
    labs(y = "Rate of common differentially expressed genes", x = "Method") + 
    theme_bw() + 
    theme(legend.position = "none",
          axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
    ggsave(filename = sprintf("../ApplicationResult/Multi_Interaction/RandomSeed/ratecDEGs/Boxplot/%s_%s.png", date.analysis, var.i),
           plot = g, device = "png", width = 9, height = 6, units = "in")
}
```


## Estimate the duplicate rate of a common differentially expressed gene associated with a variable
The duplicate rate of a common differentially expressed gene associated with a variable is the occurrences of this common differentially expressed gene associated with a variable, which is identified by a particular model (e.g., multivariable BMAseq with interaction terms), among 10 random seed trials, divided by 10, within the particular ranking threshold (e.g., top 2000)
```{r}
cDEGs = cDEGs.list = NULL
threshold.vec <- seq(1000, 5000, 1000)
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)

duplicate_rate_cDEGs_per_variable <- function(var.name = "BMI", method.name = "BMAseq") {
  for (threshold.i in threshold.vec) {
    for (seed.i in seed.vec) {
      if (method.name == "BMAseq" & var.name != "BMIxSEX") {
        load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sMultiInt%s.RData", method.name, seed.i))
        cDEGs <- c(intersect(names(get(sprintf("%s.eFDR.Main.train", method.name))[[var.name]][1:threshold.i]), 
                             names(get(sprintf("%s.eFDR.Main.test", method.name))[[var.name]][1:threshold.i])))
        } else if (method.name == "BMAseq" & var.name == "BMIxSEX") {
        load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sMultiInt%s.RData", method.name, seed.i))
        var.name2 <- "BMI"
        cDEGs <- c(intersect(names(get(sprintf("%s.eFDR.Interaction.train", method.name))[[var.name2]][1:threshold.i]), 
                             names(get(sprintf("%s.eFDR.Interaction.test", method.name))[[var.name2]][1:threshold.i])))
      } else if (method.name %in% c("DESeq2", "edgeR")) {
        load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sMultiInt%s.RData", method.name, seed.i))
        cDEGs <- c(intersect(get(sprintf("%s.eFDR.GeneName.train", method.name))[[var.name]][1:threshold.i], 
                             get(sprintf("%s.eFDR.GeneName.test", method.name))[[var.name]][1:threshold.i]))
      } else if (method.name == "VoomLimma") {
        load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sMultiInt%s.RData", method.name, seed.i))
        method.name2 <- "voom"
        cDEGs <- c(intersect(names(get(sprintf("%s.eFDR.train2", method.name2))[[var.name]][1:threshold.i]), 
                             names(get(sprintf("%s.eFDR.test2", method.name2))[[var.name]][1:threshold.i])))
      } else {
        load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/%sMultiInt%s.RData", method.name, seed.i))
        cDEGs <- c(intersect(names(get(sprintf("%s.eFDR.train2", method.name))[[var.name]][1:threshold.i]), 
                             names(get(sprintf("%s.eFDR.test2", method.name))[[var.name]][1:threshold.i])))
      }
      cDEGs.list <- append(cDEGs.list, list(cDEGs))
    }
    cDEGs.unique <- unique(unlist(cDEGs.list))
    
    # Create a matrix that shows the occurrence of each unique gene among the 10 random seed trials
    cDEGs.matrix <- matrix(0, nrow = length(cDEGs.unique), ncol = 10)
    rownames(cDEGs.matrix) <- cDEGs.unique
    colnames(cDEGs.matrix) <- seed.vec   
    for (i in 1:10) {
      cDEGs.i.seed <- cDEGs.list[[i]]
      for (j in 1:length(cDEGs.unique)) {
        if (cDEGs.unique[j] %in% cDEGs.i.seed) {
          cDEGs.matrix[j, i] <- 1
        }
      }
    }
    Sum <- apply(cDEGs.matrix, 1, sum)
    Percent <- Sum/10
    cDEGs.matrix <- cbind(cDEGs.matrix, Sum, Percent) # Add two new columns to the matrix
    saveRDS(cDEGs.matrix, sprintf("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/%sMultiInt%s_%s.RDS", method.name, var.name, threshold.i))
    
    print(paste("Complete the duplicated rate matrix of cDEGs associated with", var.name, "identified by", method.name, "among 10 random seeds for the threshold", threshold.i))
    cDEGs = cDEGs.list =  NULL
  }
}

mapply(duplicate_rate_cDEGs_per_variable,
       rep(c("BMI", "AGE","SEX", "MHABNWBC", "BMIxSEX"), each = 5), 
       rep(c("BMAseq", "DESeq2", "edgeR", "eBayes", "VoomLimma"), times = 5))
```


## Prepare and organize the data for boxplots
```{r}
boxplot_duplicate_rate_data <- function (var.name = NULL, method.name = NULL, threshold = NULL) {
  load.data <- readRDS(sprintf("../ApplicationData/derived/RandomSeed/DuplicatedRateMatrix/%sMultiInt%s_%s.RDS", method.name, var.name, threshold))
  temp.data <- data.table(threshold=threshold, 
                          method=method.name,
                          variable=var.name,
                          cDEGs.duplicate.rate.median=median(load.data[,12]),
                          cDEGs.duplicate.rate.mean=mean(load.data[,12]))
  return(temp.data)
}

organized_duplicate_rate_data <- mapply(boxplot_duplicate_rate_data,
                                        var.name = rep(c("BMI", "AGE","SEX", "MHABNWBC", "BMIxSEX"), each = 25), 
                                        method.name = rep(rep(c("BMAseq", "DESeq2", "edgeR", "eBayes", "VoomLimma"), each = 5), times = 5),
                                        threshold = rep(rep(seq(1000, 5000, 1000), times = 5), times = 5), SIMPLIFY = F) |> rbindlist()
```


## Make and save boxplots
```{r}
organized_duplicate_rate_data[,variable:=factor(variable,levels=c("BMI", "AGE","SEX", "MHABNWBC", "BMIxSEX"))]
organized_duplicate_rate_data[,cDEGs.duplicate.rate.median:=as.numeric(cDEGs.duplicate.rate.median)]
organized_duplicate_rate_data[,cDEGs.duplicate.rate.mean:=as.numeric(cDEGs.duplicate.rate.mean)]
date.analysis <- format(Sys.Date(), "%Y%b%d")
g1 <- organized_duplicate_rate_data |>
  ggplot(mapping = aes(y = cDEGs.duplicate.rate.median, x = method, group = method)) + 
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(aes(color = method)) + 
  facet_wrap(vars(variable)) + 
  scale_y_continuous(labels = scales::label_number(accuracy = 0.01),
                     breaks = seq(0, 1, 0.1)) + 
  labs(y = "Median duplicate rate of cDEGs", x = "Method") + 
  theme_bw() + 
  theme(legend.position = "none",
        axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
ggsave(filename = sprintf("../ApplicationResult/Multi_Interaction/RandomSeed/DuplicatedRateMatrix/Boxplot/%s_median.png", date.analysis),
       plot = g1, device = "png", width = 9, height = 6, units = "in")

g2 <- organized_duplicate_rate_data |>
  ggplot(mapping = aes(y = cDEGs.duplicate.rate.mean, x = method, group = method)) + 
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(aes(color = method)) + 
  facet_wrap(vars(variable)) + 
  scale_y_continuous(labels = scales::label_number(accuracy = 0.01),
                     breaks = seq(0, 1, 0.1)) + 
  labs(y = "Mean duplicate rate of cDEGs", x = "Method") + 
  theme_bw() + 
  theme(legend.position = "none",
        axis.text.x = element_text(size = 9, angle = 90, hjust = 1, color = "black", vjust = 0.5))
ggsave(filename = sprintf("../ApplicationResult/Multi_Interaction/RandomSeed/DuplicatedRateMatrix/Boxplot/%s_mean.png", date.analysis),
       plot = g2, device = "png", width = 9, height = 6, units = "in")
```


## Make Venn digrams of cDEGs
```{r}
threshold <- 2000
seed.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
var.vec <- c("BMI", "AGE", "SEX", "MHABNWBC", "BMIxSEX")

date.analysis <- format(Sys.Date(), "%Y%b%d")
make_vennplot <-function(var.name = NULL, threshold = NULL) {
  if (var.name == "BMIxSEX") {
    var.name2 <- "BMI"
    BMAseq.cDEGs <- intersect(names(BMAseq.eFDR.Interaction.train[[var.name2]][1:threshold]), names(BMAseq.eFDR.Interaction.test[[var.name2]][1:threshold]))
  } else {
    BMAseq.cDEGs <- intersect(names(BMAseq.eFDR.Main.train[[var.name]][1:threshold]), names(BMAseq.eFDR.Main.test[[var.name]][1:threshold]))
  }
  DESeq2.cDEGs <- intersect(DESeq2.eFDR.GeneName.train[[var.name]][1:threshold], DESeq2.eFDR.GeneName.test[[var.name]][1:threshold])
  edgeR.cDEGs <- intersect(edgeR.eFDR.GeneName.train[[var.name]][1:threshold], edgeR.eFDR.GeneName.test[[var.name]][1:threshold])
  eBayes.cDEGs <- intersect(names(eBayes.eFDR.train2[[var.name]][1:threshold]), names(eBayes.eFDR.test2[[var.name]][1:threshold]))
  voom.limma.cDEGs <- intersect(names(voom.eFDR.train2[[var.name]][1:threshold]), names(voom.eFDR.test2[[var.name]][1:threshold]))
  g <- ggVennDiagram(list(BMAseq = BMAseq.cDEGs, 
                          DESeq2 = DESeq2.cDEGs,
                          edgeR = edgeR.cDEGs,
                          eBayes = eBayes.cDEGs,
                          voom.limma = voom.limma.cDEGs),
                     label_alpha = 0, label_color = "white") +
    theme(legend.position = "none", 
          plot.title = element_text(hjust = 0.5, colour = "red", size = 18)) + 
    ggtitle(paste("The intersection of common differentially expressed genes \nassociated with the", var.name)) + 
    scale_fill_distiller(palette = "Dark2", direction = -1)
  
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Vennplot/", date.analysis, "_", var.name, "_", seed.i, ".png"),
         plot = g,
         device = "png",
         width = 10,
         height = 7,
         units = "in")
}

for (seed.i in seed.vec) {
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/BMAseqMultiInt%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/DESeq2MultiInt%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/edgeRMultiInt%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/eBayesMultiInt%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/VoomLimmaMultiInt%s.RData", seed.i))
  for (var.i in var.vec) {
    make_vennplot(var.name = var.i, threshold = 2000)
  }
}
```


## Extract unique BMIxSEX - related cDEGs identified by BMAseq per seed and all unique cDEGs identified by BMAseq across 10 seeds
```{r}
file.name <- "../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Vennplot/uniq_cDEGs_BMIxSEX_BMAseq_per_seed.txt"
unique.cDEGs.list <- NULL
extract_unique_cDEGs_per_seed <-function(var.name = NULL, var.name.BMA = NULL, threshold = NULL) {
  BMAseq.cDEGs <- intersect(names(BMAseq.eFDR.Interaction.train[[var.name.BMA]][1:threshold]), names(BMAseq.eFDR.Interaction.test[[var.name.BMA]][1:threshold]))
  DESeq2.cDEGs <- intersect(DESeq2.eFDR.GeneName.train[[var.name]][1:threshold], DESeq2.eFDR.GeneName.test[[var.name]][1:threshold])
  edgeR.cDEGs <- intersect(edgeR.eFDR.GeneName.train[[var.name]][1:threshold], edgeR.eFDR.GeneName.test[[var.name]][1:threshold])
  eBayes.cDEGs <- intersect(names(eBayes.eFDR.train2[[var.name]][1:threshold]), names(eBayes.eFDR.test2[[var.name]][1:threshold]))
  voom.limma.cDEGs <- intersect(names(voom.eFDR.train2[[var.name]][1:threshold]), names(voom.eFDR.test2[[var.name]][1:threshold]))
  unique.cDEGs <- setdiff(BMAseq.cDEGs, union(DESeq2.cDEGs, edgeR.cDEGs) |> union(eBayes.cDEGs) |> union(voom.limma.cDEGs))
  return(unique.cDEGs)
}

for (seed.i in seed.vec) {
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiIntModel/BMAseqMultiInt%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiIntModel/DESeq2MultiInt%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiIntModel/edgeRMultiInt%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiIntModel/eBayesMultiInt%s.RData", seed.i))
  load(sprintf("../ApplicationData/derived/RandomSeed/Top5000/MultiIntModel/VoomLimmaMultiInt%s.RData", seed.i))
  unique.cDEGs <- extract_unique_cDEGs_per_seed(var.name = "BMIxSEX", var.name.BMA = "BMI", threshold = 2000)
  unique.cDEGs.list <- append(unique.cDEGs.list, list(unique.cDEGs))
  write(sprintf("\nThe following gene IDs are unique gene IDs associated with the interaction effect of %s identified by BMAseq under the seed %s and threshold %s", var.name, seed.i, threshold), file = file.name, append = T)
  write.table(unique.cDEGs, file = file.name, quote = F, col.names = F, append = T)
}
names(unique.cDEGs.list) <- seed.vec

file.name <- "../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Vennplot/uniq_cDEGs_BMIxSEX_BMAseq_all_seeds.txt"
write.table(unlist(unique.cDEGs.list) |> unique(), file = file.name, quote = F, col.names = F)
```
