---
title: "Generate and check the RNA-seq expression levels and phenotypes in the training and test datasets"
author: "Anni Liu"
date: "`r Sys.Date()`"
output: html_document

---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r}
## RStudio keyboard shortcut
# Jump to line: Alt+Shift+G (Windows) or Option+Shift+Cmd+G (Mac)
# Cursor at the beginning of a command line: Ctrl+A
# Cursor at the end of a command line: Ctrl+E
# Clear all the code from your console: Ctrl+L
# Create a pipe operator %>%: Ctrl+Shift+M (Windows) or Cmd+Shift+M (Mac)
# Create an assignment operator <-: Alt+- (Windows) or Option+-(Mac) 
# Knit a document (knitr): Ctrl+Shift+K (Windows) or Cmd+Shift+K (Mac)
# Comment or uncomment current selection: Ctrl+Shift+C (Windows) or Cmd+Shift+C (Mac)
```

# Save and load images
```{r saveimage, eval=FALSE}
image.date <- format(Sys.Date(), "%Y%b%d")
save.image(file = paste0("../ApplicationData/derived/", image.date, "_checkdata_image_ALiu.RData"))
```

```{r loadimage}
load("../ApplicationData/derived/2023Jan23_checkdata_image_ALiu.RData")
```


# Attach libraries and functions
```{r attach_lib_func}
suppressPackageStartupMessages(easypackages::libraries("microbenchmark", "fst", "parallel"))
```

# Load original data
```{r read.data}
load("../ApplicationData/derived/rownames.RData")
dat.expr <- read_fst(path = "../ApplicationData/derived/dat.expr.Subcutaneous.RDS")
rownames(dat.expr) <- dat.expr.rownames
dat.pheno <- read_fst(path = "../ApplicationData/derived/dat.pheno.Subcutaneous.RDS")
rownames(dat.pheno) <- dat.pheno.rownames

sapply(c("dat.expr", "dat.pheno"), function(x) get(x) |> class())
#     dat.expr    dat.pheno
# "data.frame" "data.frame"

dim(dat.expr) # 24660 genes and 404 subjects
dim(dat.pheno) # 404 subjects and 13 phenotypes
dat.pheno[1:5, ]
dat.expr[1:5, 1:5]
cat(paste0("The column names of gene expression data", ifelse(all(colnames(dat.expr) == rownames(dat.pheno)), " MATCH ", " NOT MATCH"), "the row names of phenotype data."))
```


# Preprocess expression data
```{r filter.gene}
# Pre-filter the genes
# Here we perform the median absolute deviation with the threshold of 0.8 to select genes that are most likely to distinguish the samples
dat.expr.new <- dat.expr[apply(dat.expr, 1, function(x) mad(x) > 0.8), ] # We have 24455 genes
```


# One function of checking the training and test data
```{r data_func}
data_func <- function(seed.num = 999) {
  set.seed(seed.num)
  date.analysis <- format(Sys.Date(), "%Y%b%d")
  test.ind <- sample(1:nrow(dat.pheno), ceiling(0.5*nrow(dat.pheno)))
  dat.pheno.train <- dat.pheno[-test.ind, ]
  dat.pheno.test <- dat.pheno[test.ind, ]
  dat.expr.train <- dat.expr.new[, rownames(dat.pheno.train)]
  dat.expr.test <- dat.expr.new[, rownames(dat.pheno.test)]
  
  saveRDS(dat.pheno.train, file = paste0("../ApplicationData/derived/RandomSeed/dat.pheno.train", seed.num, ".RDS"))
  saveRDS(dat.pheno.test, file = paste0("../ApplicationData/derived/RandomSeed/dat.pheno.test", seed.num, ".RDS"))
  saveRDS(dat.expr.train, file = paste0("../ApplicationData/derived/RandomSeed/dat.expr.train", seed.num, ".RDS"))
  saveRDS(dat.expr.test, file = paste0("../ApplicationData/derived/RandomSeed/dat.expr.test", seed.num, ".RDS"))
  
  cat("~~~Start for one random seed~~~")
  cat(paste0("The column names of training gene expression data", ifelse(all(colnames(dat.expr.train) == rownames(dat.pheno.train)), " MATCH ", "NOT MATCH"), "the row names of training phenotype data."))
  cat("\n")
  cat(paste0("The column names of test gene expression data", ifelse(all(colnames(dat.expr.test) == rownames(dat.pheno.test)), " MATCH ", "NOT MATCH"), "the row names of test phenotype data."))
  cat("~~~End for one random seed~~~")
  cat("\n")
}
```  


# Ten trials of generating the training and test datasets
```{r}
seed.num.vec <- c(8809678, 98907, 233, 556, 7890, 120, 2390, 778, 666, 99999)
for (i in 1:10) {data_func(seed.num = seed.num.vec[i])}
```


# Check the datasets under the seeds 120, 233, and 556
```{r}
dat.expr.train120 <- readRDS("../ApplicationData/derived/RandomSeed/dat.expr.train120.RDS")
dat.expr.test120 <- readRDS("../ApplicationData/derived/RandomSeed/dat.expr.test120.RDS")
dat.pheno.train120 <- readRDS("../ApplicationData/derived/RandomSeed/dat.pheno.train120.RDS")
dat.pheno.test120 <- readRDS("../ApplicationData/derived/RandomSeed/dat.pheno.test120.RDS")

dat.expr.train233 <- readRDS("../ApplicationData/derived/RandomSeed/dat.expr.train233.RDS")
dat.expr.test233 <- readRDS("../ApplicationData/derived/RandomSeed/dat.expr.test233.RDS")
dat.pheno.train233 <- readRDS("../ApplicationData/derived/RandomSeed/dat.pheno.train233.RDS")
dat.pheno.test233 <- readRDS("../ApplicationData/derived/RandomSeed/dat.pheno.test233.RDS")

dat.expr.train556 <- readRDS("../ApplicationData/derived/RandomSeed/dat.expr.train556.RDS")
dat.expr.test556 <- readRDS("../ApplicationData/derived/RandomSeed/dat.expr.test556.RDS")
dat.pheno.train556 <- readRDS("../ApplicationData/derived/RandomSeed/dat.pheno.train556.RDS")
dat.pheno.test556 <- readRDS("../ApplicationData/derived/RandomSeed/dat.pheno.test556.RDS")
```


## Check the sample names
```{r}
length(colnames(dat.expr.train120)) # 202 samples
intersect(colnames(dat.expr.train120), colnames(dat.expr.test120)) # Intersection of sample names between the training and the test set under the seed 120
length(intersect(colnames(dat.expr.train120), colnames(dat.expr.train233))) # 95 common samples
length(intersect(colnames(dat.expr.train120), colnames(dat.expr.train556))) # 101 common samples
length(intersect(colnames(dat.expr.train233), colnames(dat.expr.train556))) # 102 common samples
```


## Check the gene expression distributions across samples - heatmap of the count matrix
```{r}
# http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#heatmap-of-the-count-matrix
library(pheatmap)
# Select the top 500 genes with the high mean counts across samples 
select.500 <- order(rowMeans(dat.expr.train120), decreasing = T)[1:500]
pheatmap(dat.expr.train120[select.500, ], 
         show_rownames = F,
         show_colnames = F,
         cluster_rows = F,
         cluster_cols = F)

select.500 <- order(rowMeans(dat.expr.train233), decreasing = T)[1:500]
pheatmap(dat.expr.train233[select.500, ], 
         show_rownames = F,
         show_colnames = F,
         cluster_rows = F,
         cluster_cols = F)

select.500 <- order(rowMeans(dat.expr.train556), decreasing = T)[1:500]
pheatmap(dat.expr.train556[select.500, ], 
         show_rownames = F,
         show_colnames = F,
         cluster_rows = F,
         cluster_cols = F)
```


## Check the gene expression distributions between two BMI levels - heatmap of the count matrix
```{r}
select.500 <- order(rowMeans(dat.expr.train120), decreasing = T)[1:500]
col.order <- c(rownames(dat.pheno.train120[dat.pheno.train120$BMI == "low", ]),
               rownames(dat.pheno.train120[dat.pheno.train120$BMI == "high", ]))
bmi <- dat.pheno.train120["BMI"]
pheatmap(dat.expr.train120[select.500, col.order], 
         show_rownames = F,
         show_colnames = F,
         cluster_rows = F,
         cluster_cols = F,
         annotation_col = bmi)

select.500 <- order(rowMeans(dat.expr.train233), decreasing = T)[1:500]
col.order <- c(rownames(dat.pheno.train233[dat.pheno.train233$BMI == "low", ]),
               rownames(dat.pheno.train233[dat.pheno.train233$BMI == "high", ]))
bmi <- dat.pheno.train233["BMI"]
pheatmap(dat.expr.train233[select.500, col.order], 
         show_rownames = F,
         show_colnames = F,
         cluster_rows = F,
         cluster_cols = F,
         annotation_col = bmi)

select.500 <- order(rowMeans(dat.expr.train556), decreasing = T)[1:500]
col.order <- c(rownames(dat.pheno.train556[dat.pheno.train556$BMI == "low", ]),
               rownames(dat.pheno.train556[dat.pheno.train556$BMI == "high", ]))
bmi <- dat.pheno.train556["BMI"]
pheatmap(dat.expr.train556[select.500, col.order], 
         show_rownames = F,
         show_colnames = F,
         cluster_rows = F,
         cluster_cols = F,
         annotation_col = bmi)
```


# Extract eFDR/q-value of the top ranked 2000 genes
## Seed: 2390
```{r}
dat.expr.train <- readRDS("../ApplicationData/derived/RandomSeed/dat.expr.train2390.RDS")
dat.expr.test <- readRDS("../ApplicationData/derived/RandomSeed/dat.expr.test2390.RDS")
dat.pheno.train <- readRDS("../ApplicationData/derived/RandomSeed/dat.pheno.train2390.RDS")
dat.pheno.test <- readRDS("../ApplicationData/derived/RandomSeed/dat.pheno.test2390.RDS")
```

```{r}
  var.pool <- c("BMI", "AGE", "SEX", "MHABNWBC")  
  var.pool.add <- c(var.pool, "BMIxSEX")
  
  ##------BMAseq------
  ##################################################################
  ##                 BMAseq multivariate analysis                 ##
  ##################################################################
  output.multi.int1.train <- BMAseq.multi.postprob.norm(
    dat.expr.counts = dat.expr.train,  
    dat.pheno = dat.pheno.train, 
    var.pool = var.pool, 
    max.nvar = 4, 
    interaction = "BMI&SEX", 
    cut.BF = 1)
  
  output.multi.int2.train <- BMAseq.multi.DEG(
    dat.pheno = output.multi.int1.train$dat.pheno.new,
    model.space = output.multi.int1.train$model.space, 
    post.modelprob = output.multi.int1.train$post.modelprob, 
    var.pool = var.pool,
    interact = T, 
    cut.FDR = 0.05)
  
  output.multi.int1.test <- BMAseq.multi.postprob.norm(
    dat.expr.counts = dat.expr.test, 
    dat.pheno = dat.pheno.test, 
    var.pool = var.pool, 
    max.nvar = 4, 
    interaction = "BMI&SEX", 
    cut.BF = 1)
  
  output.multi.int2.test <- BMAseq.multi.DEG(
    dat.pheno = output.multi.int1.test$dat.pheno.new,
    model.space = output.multi.int1.test$model.space, 
    post.modelprob = output.multi.int1.test$post.modelprob, 
    var.pool = var.pool,
    interact = T, 
    cut.FDR = 0.05)
  
  BMAseq.eFDR.Main.2000.train <- mclapply(1:length(var.pool),
                                   function(i) output.multi.int2.train$eFDR.Main[, i][order(output.multi.int2.train$eFDR.Main[, i])[1:2000]],
                                   mc.cores = 4)
  
  BMAseq.eFDR.Main.2000.test <- mclapply(1:length(var.pool),
                                  function(i) output.multi.int2.test$eFDR.Main[, i][order(output.multi.int2.test$eFDR.Main[, i])[1:2000]],
                                  mc.cores = 4)

  names(BMAseq.eFDR.Main.2000.train) = names(BMAseq.eFDR.Main.2000.test) = var.pool
  
  BMAseq.eFDR.Interaction.2000.train <- mclapply(1:length(var.pool),
                                          function(i) output.multi.int2.train$eFDR.Interaction[, i][order(output.multi.int2.train$eFDR.Interaction[, i])[1:2000]],
                                          mc.cores = 4)
  
  BMAseq.eFDR.Interaction.2000.test <- mclapply(1:length(var.pool),
                                         function(i) output.multi.int2.test$eFDR.Interaction[, i][order(output.multi.int2.test$eFDR.Interaction[, i])[1:2000]],
                                         mc.cores = 4)
  
  names(BMAseq.eFDR.Interaction.2000.train) = names(BMAseq.eFDR.Interaction.2000.test) = var.pool
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, var.pool, var.pool.add, output.multi.int1.train, output.multi.int2.train, output.multi.int1.test, output.multi.int2.test, BMAseq.eFDR.Main.2000.train, BMAseq.eFDR.Main.2000.test, BMAseq.eFDR.Interaction.2000.train, BMAseq.eFDR.Interaction.2000.test, file = "../ApplicationData/derived/RandomSeed/BMAseqMultiInt2390.RData")
  
  
  ##------voom + limma------
  ##################################################################
  ##              voom + limma multivariate analysis              ##
  ##################################################################
  design.train <- model.matrix(~BMI + AGE + SEX + MHABNWBC + BMI*SEX, data = dat.pheno.train)
  design.test <- model.matrix(~BMI + AGE + SEX + MHABNWBC + BMI*SEX, data = dat.pheno.test)
    
  voom.train <- voom(counts = dat.expr.train,
                     design = design.train,
                     lib.size = colSums(dat.expr.train)*calcNormFactors(dat.expr.train)) 
  
  voom.fit.train <- lmFit(voom.train[["E"]],
                          design = design.train,
                          weights = voom.train[["weights"]])
  
  voom.eFDR.train <- mclapply(1:length(var.pool.add),
                         function(i) {
                           t <- voom.fit.train[["coefficients"]][, i + 1]/voom.fit.train[["stdev.unscaled"]][, i+1]/voom.fit.train[["sigma"]]
                           p <- 2*pt(-abs(t), df = voom.fit.train[["df.residual"]])
                           return(qvalue(p)) },
                         mc.cores = 4L)
  
  voom.eFDR.2000.train <- mclapply(1:length(var.pool.add),
                             function(i) {
                               q.val <- voom.eFDR.train[[i]][["qvalues"]]
                               return(q.val[order(q.val)[1:2000]])
                             },
                             mc.cores = 4L)
  
  names(voom.eFDR.train) = names(voom.eFDR.2000.train) = var.pool.add 
  
  voom.test <- voom(counts = dat.expr.test,
                    design = design.test,
                    lib.size = colSums(dat.expr.test)*calcNormFactors(dat.expr.test))
  
  voom.fit.test <- lmFit(voom.test[["E"]],
                         design = design.test,
                         weights = voom.test[["weights"]])
  
  voom.eFDR.test <- mclapply(1:length(var.pool.add),
                        function(i) {
                          t <- voom.fit.test[["coefficients"]][, i + 1]/voom.fit.test[["stdev.unscaled"]][, i+1]/voom.fit.test[["sigma"]]
                          p <- 2*pt(-abs(t), df = voom.fit.test[["df.residual"]])
                          return(qvalue(p)) },
                        mc.cores = 4L)
  
  voom.eFDR.2000.test <- mclapply(1:length(var.pool.add),
                            function(i) {
                              q.val <- voom.eFDR.test[[i]][["qvalues"]]
                              return(q.val[order(q.val)[1:2000]])
                            },
                            mc.cores = 4L)
  
  names(voom.eFDR.test) = names(voom.eFDR.2000.test) = var.pool.add 
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, var.pool, var.pool.add, design.train, design.test, voom.train, voom.fit.train, voom.eFDR.train, voom.eFDR.2000.train, voom.test, voom.fit.test, voom.eFDR.test, voom.eFDR.2000.test, file = "../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt2390.RData")
  
  
  ##------voom + limma + eBayes------
  #################################################################
  ##         voom + limma + eBayes multivariate analysis         ##
  #################################################################
  voom.train <- voom(counts = dat.expr.train,
                     design = design.train,
                     lib.size = colSums(dat.expr.train)*calcNormFactors(dat.expr.train))
  
  eBayes.fit.train <- lmFit(voom.train[["E"]],
                          design = design.train,
                          weights = voom.train[["weights"]]) %>% 
                      eBayes()
  
  eBayes.eFDR.train <- mclapply(1:length(var.pool.add),
                         function(i) eBayes.fit.train[["p.value"]][, i + 1] %>% 
                           qvalue(),
                         mc.cores = 4L)
  
  eBayes.eFDR.2000.train <- mclapply(1:length(var.pool.add),
                             function(i) {
                               q.val <- eBayes.eFDR.train[[i]][["qvalues"]]
                               return(q.val[order(q.val)[1:2000]])
                               },
                             mc.cores = 4L)
  
  names(eBayes.eFDR.train) = names(eBayes.eFDR.2000.train) = var.pool.add 
  
  voom.test <- voom(counts = dat.expr.test,
                    design = design.test,
                    lib.size = colSums(dat.expr.test)*calcNormFactors(dat.expr.test))
  
  eBayes.fit.test <- lmFit(voom.test[["E"]],
                         design = design.test,
                         weights = voom.test[["weights"]]) %>% 
                     eBayes()
  
  eBayes.eFDR.test <- mclapply(1:length(var.pool.add),
                        function(i) eBayes.fit.test[["p.value"]][, i + 1] %>% 
                          qvalue(),
                        mc.cores = 4L)
  
  eBayes.eFDR.2000.test <- mclapply(1:length(var.pool.add),
                             function(i) {
                               q.val <- eBayes.eFDR.test[[i]][["qvalues"]]
                               return(q.val[order(q.val)[1:2000]]) 
                               },
                             mc.cores = 4L)
  
  names(eBayes.eFDR.test) = names(eBayes.eFDR.2000.test) = var.pool.add 
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, var.pool, var.pool.add, design.train, design.test, voom.train, eBayes.fit.train, eBayes.eFDR.train, eBayes.eFDR.2000.train, voom.test, eBayes.fit.test, eBayes.eFDR.test, eBayes.eFDR.2000.test, file = "../ApplicationData/derived/RandomSeed/eBayesMultiInt2390.RData")
  
  
  ##------edgeR------
  #################################################################
  ##                 edgeR multivariate analysis                 ##
  #################################################################
  y.train <- DGEList(counts = dat.expr.train, 
                     lib.size = colSums(dat.expr.train)) %>% 
               calcNormFactors() %>% 
               estimateGLMTrendedDisp(design.train)
  
  edgeR.fit.train <- glmQLFit(y.train, design.train)
  
  qlf.train <- mclapply(1:length(var.pool.add),
                        function(i) glmQLFTest(edgeR.fit.train, coef = i + 1),
                        mc.cores = 4L)
  
  edgeR.eFDR.train <- mclapply(1:length(var.pool.add),
                         function(i) qlf.train[[i]][["table"]][["PValue"]] %>%
                           qvalue(),
                         mc.cores = 4L)
  
  edgeR.eFDR.2000.train <- mclapply(1:length(var.pool.add),
                              function(i) {
                                q.val <- edgeR.eFDR.train[[i]][["qvalues"]]
                                return(q.val[order(q.val)[1:2000]]) 
                              },
                              mc.cores = 4L)
  
  edgeR.eFDR.2000.GeneName.train <- mclapply(1:length(var.pool.add),
                              function(i) {
                                q.val <- edgeR.eFDR.train[[i]][["qvalues"]]
                                return(rownames(dat.expr.train)[order(q.val)[1:2000]])
                              },
                              mc.cores = 4L)
  
  names(qlf.train) = names(edgeR.eFDR.train) = names(edgeR.eFDR.2000.train) = names(edgeR.eFDR.2000.GeneName.train) = var.pool.add 
  
  y.test <- DGEList(counts = dat.expr.test, 
                    lib.size = colSums(dat.expr.test)) %>% 
              calcNormFactors() %>% 
              estimateGLMTrendedDisp(design.test)
  
  edgeR.fit.test <- glmQLFit(y.test, design.test)
  
  qlf.test <- mclapply(1:length(var.pool.add),
                      function(i) glmQLFTest(edgeR.fit.test, coef = i + 1),
                      mc.cores = 4L)
  
  edgeR.eFDR.test <- mclapply(1:length(var.pool.add),
                        function(i) qlf.test[[i]][["table"]][["PValue"]] %>% 
                          qvalue(),
                        mc.cores = 4L)
  
  edgeR.eFDR.2000.test <- mclapply(1:length(var.pool.add),
                             function(i) {
                               q.val <- edgeR.eFDR.test[[i]][["qvalues"]]
                               return(q.val[order(q.val)[1:2000]]) 
                             },
                             mc.cores = 4L)
  
  edgeR.eFDR.2000.GeneName.test <- mclapply(1:length(var.pool.add),
                              function(i) {
                                q.val <- edgeR.eFDR.test[[i]][["qvalues"]]
                                return(rownames(dat.expr.test)[order(q.val)[1:2000]])
                              },
                              mc.cores = 4L)
  
  names(qlf.test) = names(edgeR.eFDR.test) = names(edgeR.eFDR.2000.test) = names(edgeR.eFDR.2000.GeneName.test) = var.pool.add 
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, var.pool, var.pool.add, design.train, design.test, y.train, edgeR.fit.train, edgeR.eFDR.train, edgeR.eFDR.2000.train, edgeR.eFDR.2000.GeneName.train, y.test, edgeR.fit.test, edgeR.eFDR.test, edgeR.eFDR.2000.test, edgeR.eFDR.2000.GeneName.test, file = "../ApplicationData/derived/RandomSeed/edgeRMultiInt2390.RData")
  
    
  ##------DESeq2------
  ##################################################################
  ##                 DESeq2 multivariate analysis                 ##
  ##################################################################
  cts.train <- as.matrix(dat.expr.train)
  cts.test <- as.matrix(dat.expr.test)
  coldata.train <- dat.pheno.train
  coldata.test <- dat.pheno.test
  
  name.formula <- c("BMI_high_vs_low", 
                    "AGE_old_vs_young", 
                    "SEX_male_vs_female", 
                    "MHABNWBC_yes_vs_no",
                    "BMIhigh.SEXmale")  
  
  lib.size <- colSums(cts.train)
  norm.factor <- calcNormFactors(cts.train, method = "TMM")
  size.factor <- lib.size*norm.factor/exp(mean(log(lib.size*norm.factor))) 
  
  dds <- DESeqDataSetFromMatrix(countData = cts.train, 
                                colData = coldata.train, 
                                design = ~BMI + AGE + SEX + MHABNWBC + BMI*SEX)
  sizeFactors(dds) <- size.factor

  res.train <- mclapply(1:length(var.pool.add), 
                        function(i) {
                          return(results(DESeq(dds), name = name.formula[i])) },
                        mc.cores = 4L)
  
  DESeq2.eFDR.train <- mclapply(1:length(var.pool.add),
                         function(i) qvalue(res.train[[i]][["pvalue"]])$qvalues,
                         mc.cores = 4L)
  
  DESeq2.eFDR.2000.train <- mclapply(1:length(var.pool.add),
                               function(i) {
                                 q.val <- DESeq2.eFDR.train[[i]]
                                 return(q.val[order(q.val)[1:2000]])
                               },
                               mc.cores = 4L)
  
  DESeq2.eFDR.2000.GeneName.train <- mclapply(1:length(var.pool.add),
                                function(i) {
                                  q.val <- DESeq2.eFDR.train[[i]]
                                  return(rownames(cts.train)[order(q.val)[1:2000]])},
                                mc.cores = 4L)
  
  names(res.train) = names(DESeq2.eFDR.train) = names(DESeq2.eFDR.2000.train) = names(DESeq2.eFDR.2000.GeneName.train) = var.pool.add 
  
  lib.size <- colSums(cts.test)
  norm.factor <- calcNormFactors(cts.test, method = "TMM")
  size.factor <- lib.size*norm.factor/exp(mean(log(lib.size*norm.factor))) 
  
  dds <- DESeqDataSetFromMatrix(countData = cts.test, 
                                colData = coldata.test, 
                                design = ~BMI + AGE + SEX + MHABNWBC + BMI*SEX)
  sizeFactors(dds) <- size.factor
  
  res.test <- mclapply(1:length(var.pool.add), 
                       function(i) {
                         return(results(DESeq(dds), name = name.formula[i])) }, 
                       mc.cores = 4L)
  
  DESeq2.eFDR.test <- mclapply(1:length(var.pool.add),
                        function(i) qvalue(res.test[[i]][["pvalue"]])$qvalues,
                        mc.cores = 4L)
  
  DESeq2.eFDR.2000.test <- mclapply(1:length(var.pool.add),
                              function(i) {
                                q.val <- DESeq2.eFDR.test[[i]]
                                return(q.val[order(q.val)[1:2000]])
                              },
                              mc.cores = 4L)
  
  DESeq2.eFDR.2000.GeneName.test <- mclapply(1:length(var.pool.add),
                                function(i) {
                                  q.val <- DESeq2.eFDR.test[[i]]
                                  return(rownames(cts.test)[order(q.val)[1:2000]])},
                                mc.cores = 4L)
  
  names(res.test) = names(DESeq2.eFDR.test) = names(DESeq2.eFDR.2000.test) = names(DESeq2.eFDR.2000.GeneName.test) = var.pool.add
  
  save(cts.train, cts.test, coldata.train, coldata.test, var.pool, var.pool.add, res.train, DESeq2.eFDR.train, DESeq2.eFDR.2000.train, DESeq2.eFDR.2000.GeneName.train, res.test, DESeq2.eFDR.test, DESeq2.eFDR.2000.test, DESeq2.eFDR.2000.GeneName.test, file = "../ApplicationData/derived/RandomSeed/DESeq2MultiInt2390.RData")
```


## Seed: 98907
```{r}
dat.expr.train <- readRDS("../ApplicationData/derived/RandomSeed/dat.expr.train98907.RDS")
dat.expr.test <- readRDS("../ApplicationData/derived/RandomSeed/dat.expr.test98907.RDS")
dat.pheno.train <- readRDS("../ApplicationData/derived/RandomSeed/dat.pheno.train98907.RDS")
dat.pheno.test <- readRDS("../ApplicationData/derived/RandomSeed/dat.pheno.test98907.RDS")
```

```{r}
  var.pool <- c("BMI", "AGE", "SEX", "MHABNWBC")  
  var.pool.add <- c(var.pool, "BMIxSEX")
  
  ##------BMAseq------
  ##################################################################
  ##                 BMAseq multivariate analysis                 ##
  ##################################################################
  output.multi.int1.train <- BMAseq.multi.postprob.norm(
    dat.expr.counts = dat.expr.train,  
    dat.pheno = dat.pheno.train, 
    var.pool = var.pool, 
    max.nvar = 4, 
    interaction = "BMI&SEX", 
    cut.BF = 1)
  
  output.multi.int2.train <- BMAseq.multi.DEG(
    dat.pheno = output.multi.int1.train$dat.pheno.new,
    model.space = output.multi.int1.train$model.space, 
    post.modelprob = output.multi.int1.train$post.modelprob, 
    var.pool = var.pool,
    interact = T, 
    cut.FDR = 0.05)
  
  output.multi.int1.test <- BMAseq.multi.postprob.norm(
    dat.expr.counts = dat.expr.test, 
    dat.pheno = dat.pheno.test, 
    var.pool = var.pool, 
    max.nvar = 4, 
    interaction = "BMI&SEX", 
    cut.BF = 1)
  
  output.multi.int2.test <- BMAseq.multi.DEG(
    dat.pheno = output.multi.int1.test$dat.pheno.new,
    model.space = output.multi.int1.test$model.space, 
    post.modelprob = output.multi.int1.test$post.modelprob, 
    var.pool = var.pool,
    interact = T, 
    cut.FDR = 0.05)
  
  BMAseq.eFDR.Main.2000.train <- mclapply(1:length(var.pool),
                                   function(i) output.multi.int2.train$eFDR.Main[, i][order(output.multi.int2.train$eFDR.Main[, i])[1:2000]],
                                   mc.cores = 4)
  
  BMAseq.eFDR.Main.2000.test <- mclapply(1:length(var.pool),
                                  function(i) output.multi.int2.test$eFDR.Main[, i][order(output.multi.int2.test$eFDR.Main[, i])[1:2000]],
                                  mc.cores = 4)

  names(BMAseq.eFDR.Main.2000.train) = names(BMAseq.eFDR.Main.2000.test) = var.pool
  
  BMAseq.eFDR.Interaction.2000.train <- mclapply(1:length(var.pool),
                                          function(i) output.multi.int2.train$eFDR.Interaction[, i][order(output.multi.int2.train$eFDR.Interaction[, i])[1:2000]],
                                          mc.cores = 4)
  
  BMAseq.eFDR.Interaction.2000.test <- mclapply(1:length(var.pool),
                                         function(i) output.multi.int2.test$eFDR.Interaction[, i][order(output.multi.int2.test$eFDR.Interaction[, i])[1:2000]],
                                         mc.cores = 4)
  
  names(BMAseq.eFDR.Interaction.2000.train) = names(BMAseq.eFDR.Interaction.2000.test) = var.pool
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, var.pool, var.pool.add, output.multi.int1.train, output.multi.int2.train, output.multi.int1.test, output.multi.int2.test, BMAseq.eFDR.Main.2000.train, BMAseq.eFDR.Main.2000.test, BMAseq.eFDR.Interaction.2000.train, BMAseq.eFDR.Interaction.2000.test, file = "../ApplicationData/derived/RandomSeed/BMAseqMultiInt98907.RData")
  
  
  ##------voom + limma------
  ##################################################################
  ##              voom + limma multivariate analysis              ##
  ##################################################################
  design.train <- model.matrix(~BMI + AGE + SEX + MHABNWBC + BMI*SEX, data = dat.pheno.train)
  design.test <- model.matrix(~BMI + AGE + SEX + MHABNWBC + BMI*SEX, data = dat.pheno.test)
    
  voom.train <- voom(counts = dat.expr.train,
                     design = design.train,
                     lib.size = colSums(dat.expr.train)*calcNormFactors(dat.expr.train)) 
  
  voom.fit.train <- lmFit(voom.train[["E"]],
                          design = design.train,
                          weights = voom.train[["weights"]])
  
  voom.eFDR.train <- mclapply(1:length(var.pool.add),
                         function(i) {
                           t <- voom.fit.train[["coefficients"]][, i + 1]/voom.fit.train[["stdev.unscaled"]][, i+1]/voom.fit.train[["sigma"]]
                           p <- 2*pt(-abs(t), df = voom.fit.train[["df.residual"]])
                           return(qvalue(p)) },
                         mc.cores = 4L)
  
  voom.eFDR.2000.train <- mclapply(1:length(var.pool.add),
                             function(i) {
                               q.val <- voom.eFDR.train[[i]][["qvalues"]]
                               return(q.val[order(q.val)[1:2000]])
                             },
                             mc.cores = 4L)
  
  names(voom.eFDR.train) = names(voom.eFDR.2000.train) = var.pool.add 
  
  voom.test <- voom(counts = dat.expr.test,
                    design = design.test,
                    lib.size = colSums(dat.expr.test)*calcNormFactors(dat.expr.test))
  
  voom.fit.test <- lmFit(voom.test[["E"]],
                         design = design.test,
                         weights = voom.test[["weights"]])
  
  voom.eFDR.test <- mclapply(1:length(var.pool.add),
                        function(i) {
                          t <- voom.fit.test[["coefficients"]][, i + 1]/voom.fit.test[["stdev.unscaled"]][, i+1]/voom.fit.test[["sigma"]]
                          p <- 2*pt(-abs(t), df = voom.fit.test[["df.residual"]])
                          return(qvalue(p)) },
                        mc.cores = 4L)
  
  voom.eFDR.2000.test <- mclapply(1:length(var.pool.add),
                            function(i) {
                              q.val <- voom.eFDR.test[[i]][["qvalues"]]
                              return(q.val[order(q.val)[1:2000]])
                            },
                            mc.cores = 4L)
  
  names(voom.eFDR.test) = names(voom.eFDR.2000.test) = var.pool.add 
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, var.pool, var.pool.add, design.train, design.test, voom.train, voom.fit.train, voom.eFDR.train, voom.eFDR.2000.train, voom.test, voom.fit.test, voom.eFDR.test, voom.eFDR.2000.test, file = "../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt98907.RData")
  
  
  ##------voom + limma + eBayes------
  #################################################################
  ##         voom + limma + eBayes multivariate analysis         ##
  #################################################################
  voom.train <- voom(counts = dat.expr.train,
                     design = design.train,
                     lib.size = colSums(dat.expr.train)*calcNormFactors(dat.expr.train))
  
  eBayes.fit.train <- lmFit(voom.train[["E"]],
                          design = design.train,
                          weights = voom.train[["weights"]]) %>% 
                      eBayes()
  
  eBayes.eFDR.train <- mclapply(1:length(var.pool.add),
                         function(i) eBayes.fit.train[["p.value"]][, i + 1] %>% 
                           qvalue(),
                         mc.cores = 4L)
  
  eBayes.eFDR.2000.train <- mclapply(1:length(var.pool.add),
                             function(i) {
                               q.val <- eBayes.eFDR.train[[i]][["qvalues"]]
                               return(q.val[order(q.val)[1:2000]])
                               },
                             mc.cores = 4L)
  
  names(eBayes.eFDR.train) = names(eBayes.eFDR.2000.train) = var.pool.add 
  
  voom.test <- voom(counts = dat.expr.test,
                    design = design.test,
                    lib.size = colSums(dat.expr.test)*calcNormFactors(dat.expr.test))
  
  eBayes.fit.test <- lmFit(voom.test[["E"]],
                         design = design.test,
                         weights = voom.test[["weights"]]) %>% 
                     eBayes()
  
  eBayes.eFDR.test <- mclapply(1:length(var.pool.add),
                        function(i) eBayes.fit.test[["p.value"]][, i + 1] %>% 
                          qvalue(),
                        mc.cores = 4L)
  
  eBayes.eFDR.2000.test <- mclapply(1:length(var.pool.add),
                             function(i) {
                               q.val <- eBayes.eFDR.test[[i]][["qvalues"]]
                               return(q.val[order(q.val)[1:2000]]) 
                               },
                             mc.cores = 4L)
  
  names(eBayes.eFDR.test) = names(eBayes.eFDR.2000.test) = var.pool.add 
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, var.pool, var.pool.add, design.train, design.test, voom.train, eBayes.fit.train, eBayes.eFDR.train, eBayes.eFDR.2000.train, voom.test, eBayes.fit.test, eBayes.eFDR.test, eBayes.eFDR.2000.test, file = "../ApplicationData/derived/RandomSeed/eBayesMultiInt98907.RData")
  
  
  ##------edgeR------
  #################################################################
  ##                 edgeR multivariate analysis                 ##
  #################################################################
  y.train <- DGEList(counts = dat.expr.train, 
                     lib.size = colSums(dat.expr.train)) %>% 
               calcNormFactors() %>% 
               estimateGLMTrendedDisp(design.train)
  
  edgeR.fit.train <- glmQLFit(y.train, design.train)
  
  qlf.train <- mclapply(1:length(var.pool.add),
                        function(i) glmQLFTest(edgeR.fit.train, coef = i + 1),
                        mc.cores = 4L)
  
  edgeR.eFDR.train <- mclapply(1:length(var.pool.add),
                         function(i) qlf.train[[i]][["table"]][["PValue"]] %>%
                           qvalue(),
                         mc.cores = 4L)
  
  edgeR.eFDR.2000.train <- mclapply(1:length(var.pool.add),
                              function(i) {
                                q.val <- edgeR.eFDR.train[[i]][["qvalues"]]
                                return(q.val[order(q.val)[1:2000]]) 
                              },
                              mc.cores = 4L)
  
  edgeR.eFDR.2000.GeneName.train <- mclapply(1:length(var.pool.add),
                              function(i) {
                                q.val <- edgeR.eFDR.train[[i]][["qvalues"]]
                                return(rownames(dat.expr.train)[order(q.val)[1:2000]])
                              },
                              mc.cores = 4L)
  
  names(qlf.train) = names(edgeR.eFDR.train) = names(edgeR.eFDR.2000.train) = names(edgeR.eFDR.2000.GeneName.train) = var.pool.add 
  
  y.test <- DGEList(counts = dat.expr.test, 
                    lib.size = colSums(dat.expr.test)) %>% 
              calcNormFactors() %>% 
              estimateGLMTrendedDisp(design.test)
  
  edgeR.fit.test <- glmQLFit(y.test, design.test)
  
  qlf.test <- mclapply(1:length(var.pool.add),
                      function(i) glmQLFTest(edgeR.fit.test, coef = i + 1),
                      mc.cores = 4L)
  
  edgeR.eFDR.test <- mclapply(1:length(var.pool.add),
                        function(i) qlf.test[[i]][["table"]][["PValue"]] %>% 
                          qvalue(),
                        mc.cores = 4L)
  
  edgeR.eFDR.2000.test <- mclapply(1:length(var.pool.add),
                             function(i) {
                               q.val <- edgeR.eFDR.test[[i]][["qvalues"]]
                               return(q.val[order(q.val)[1:2000]]) 
                             },
                             mc.cores = 4L)
  
  edgeR.eFDR.2000.GeneName.test <- mclapply(1:length(var.pool.add),
                              function(i) {
                                q.val <- edgeR.eFDR.test[[i]][["qvalues"]]
                                return(rownames(dat.expr.test)[order(q.val)[1:2000]])
                              },
                              mc.cores = 4L)
  
  names(qlf.test) = names(edgeR.eFDR.test) = names(edgeR.eFDR.2000.test) = names(edgeR.eFDR.2000.GeneName.test) = var.pool.add 
  
  save(dat.expr.train, dat.expr.test, dat.pheno.train, dat.pheno.test, var.pool, var.pool.add, design.train, design.test, y.train, edgeR.fit.train, edgeR.eFDR.train, edgeR.eFDR.2000.train, edgeR.eFDR.2000.GeneName.train, y.test, edgeR.fit.test, edgeR.eFDR.test, edgeR.eFDR.2000.test, edgeR.eFDR.2000.GeneName.test, file = "../ApplicationData/derived/RandomSeed/edgeRMultiInt98907.RData")
  
    
  ##------DESeq2------
  ##################################################################
  ##                 DESeq2 multivariate analysis                 ##
  ##################################################################
  cts.train <- as.matrix(dat.expr.train)
  cts.test <- as.matrix(dat.expr.test)
  coldata.train <- dat.pheno.train
  coldata.test <- dat.pheno.test
  
  name.formula <- c("BMI_high_vs_low", 
                    "AGE_old_vs_young", 
                    "SEX_male_vs_female", 
                    "MHABNWBC_yes_vs_no",
                    "BMIhigh.SEXmale")  
  
  lib.size <- colSums(cts.train)
  norm.factor <- calcNormFactors(cts.train, method = "TMM")
  size.factor <- lib.size*norm.factor/exp(mean(log(lib.size*norm.factor))) 
  
  dds <- DESeqDataSetFromMatrix(countData = cts.train, 
                                colData = coldata.train, 
                                design = ~BMI + AGE + SEX + MHABNWBC + BMI*SEX)
  sizeFactors(dds) <- size.factor

  res.train <- mclapply(1:length(var.pool.add), 
                        function(i) {
                          return(results(DESeq(dds), name = name.formula[i])) },
                        mc.cores = 4L)
  
  DESeq2.eFDR.train <- mclapply(1:length(var.pool.add),
                         function(i) qvalue(res.train[[i]][["pvalue"]])$qvalues,
                         mc.cores = 4L)
  
  DESeq2.eFDR.2000.train <- mclapply(1:length(var.pool.add),
                               function(i) {
                                 q.val <- DESeq2.eFDR.train[[i]]
                                 return(q.val[order(q.val)[1:2000]])
                               },
                               mc.cores = 4L)
  
  DESeq2.eFDR.2000.GeneName.train <- mclapply(1:length(var.pool.add),
                                function(i) {
                                  q.val <- DESeq2.eFDR.train[[i]]
                                  return(rownames(cts.train)[order(q.val)[1:2000]])},
                                mc.cores = 4L)
  
  names(res.train) = names(DESeq2.eFDR.train) = names(DESeq2.eFDR.2000.train) = names(DESeq2.eFDR.2000.GeneName.train) = var.pool.add 
  
  lib.size <- colSums(cts.test)
  norm.factor <- calcNormFactors(cts.test, method = "TMM")
  size.factor <- lib.size*norm.factor/exp(mean(log(lib.size*norm.factor))) 
  
  dds <- DESeqDataSetFromMatrix(countData = cts.test, 
                                colData = coldata.test, 
                                design = ~BMI + AGE + SEX + MHABNWBC + BMI*SEX)
  sizeFactors(dds) <- size.factor
  
  res.test <- mclapply(1:length(var.pool.add), 
                       function(i) {
                         return(results(DESeq(dds), name = name.formula[i])) }, 
                       mc.cores = 4L)
  
  DESeq2.eFDR.test <- mclapply(1:length(var.pool.add),
                        function(i) qvalue(res.test[[i]][["pvalue"]])$qvalues,
                        mc.cores = 4L)
  
  DESeq2.eFDR.2000.test <- mclapply(1:length(var.pool.add),
                              function(i) {
                                q.val <- DESeq2.eFDR.test[[i]]
                                return(q.val[order(q.val)[1:2000]])
                              },
                              mc.cores = 4L)
  
  DESeq2.eFDR.2000.GeneName.test <- mclapply(1:length(var.pool.add),
                                function(i) {
                                  q.val <- DESeq2.eFDR.test[[i]]
                                  return(rownames(cts.test)[order(q.val)[1:2000]])},
                                mc.cores = 4L)
  
  names(res.test) = names(DESeq2.eFDR.test) = names(DESeq2.eFDR.2000.test) = names(DESeq2.eFDR.2000.GeneName.test) = var.pool.add
  
  save(cts.train, cts.test, coldata.train, coldata.test, var.pool, var.pool.add, res.train, DESeq2.eFDR.train, DESeq2.eFDR.2000.train, DESeq2.eFDR.2000.GeneName.train, res.test, DESeq2.eFDR.test, DESeq2.eFDR.2000.test, DESeq2.eFDR.2000.GeneName.test, file = "../ApplicationData/derived/RandomSeed/DESeq2MultiInt98907.RData")
```


# Violin plot - distributions of eFDR/q-value of the top ranked 2000 genes
## Seed: 2390
```{r}
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt2390.RData")
```


### BMI
```{r}
BMI.eFDR.q.dat.2390 <- data.frame(Estimate = c(
  BMAseq.eFDR.Main.2000.train$BMI, BMAseq.eFDR.Main.2000.test$BMI,
  DESeq2.eFDR.2000.train$BMI, DESeq2.eFDR.2000.test$BMI, 
  edgeR.eFDR.2000.train$BMI, edgeR.eFDR.2000.test$BMI,
  eBayes.eFDR.2000.train$BMI, eBayes.eFDR.2000.test$BMI,
  voom.eFDR.2000.train$BMI, voom.eFDR.2000.test$BMI),
                                  Data.type = rep(c("Train", "Test"), each = 2000, times = 5),
                                  DEG.method = rep(c("BMAseq", "DESeq2", "edgeR", "voom+limma+eBayes", "voom+limma"), each = 4000)
                                  )
g <- BMI.eFDR.q.dat.2390 %>% 
  ggplot(aes(x = Data.type, 
             y = Estimate,
             fill = Data.type)) +
  geom_violin() +
  geom_boxplot(width = 0.1) + 
  facet_wrap(~ DEG.method, nrow = 1) + 
  scale_x_discrete(limits = c("Train", "Test")) + 
  scale_fill_brewer(palette = "Set1",
                    direction = -1) +
  labs(x = "Data Type",
       y = "Estimate",
       title = "The distributions of top 2000 lowest eFDRs and q-values \nassociated with the main effect of BMI in training and test sets per approach") +
  theme_bw() + 
  theme(plot.title = element_text(hjust = 0.5),
        legend.position = "none") 

date.analysis <- format(Sys.Date(), "%Y%b%d")
seed.num <- 2390
ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/", date.analysis, "_", "BMIMain_eFDR_qvalue", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 8,
         height = 4,
         units = "in")

```


### AGE
```{r}
AGE.eFDR.q.dat.2390 <- data.frame(Estimate = c(
  BMAseq.eFDR.Main.2000.train$AGE, BMAseq.eFDR.Main.2000.test$AGE,
  DESeq2.eFDR.2000.train$AGE, DESeq2.eFDR.2000.test$AGE, 
  edgeR.eFDR.2000.train$AGE, edgeR.eFDR.2000.test$AGE,
  eBayes.eFDR.2000.train$AGE, eBayes.eFDR.2000.test$AGE,
  voom.eFDR.2000.train$AGE, voom.eFDR.2000.test$AGE),
                                  Data.type = rep(c("Train", "Test"), each = 2000, times = 5),
                                  DEG.method = rep(c("BMAseq", "DESeq2", "edgeR", "voom+limma+eBayes", "voom+limma"), each = 4000)
                                  )
g <- AGE.eFDR.q.dat.2390 %>% 
  ggplot(aes(x = Data.type, 
             y = Estimate,
             fill = Data.type)) +
  geom_violin() +
  geom_boxplot(width = 0.1) + 
  facet_wrap(~ DEG.method, nrow = 1) + 
  scale_x_discrete(limits = c("Train", "Test")) + 
  scale_fill_brewer(palette = "Set1",
                    direction = -1) +
  labs(x = "Data Type",
       y = "Estimate",
       title = "The distributions of top 2000 lowest eFDRs and q-values \nassociated with the main effect of AGE in training and test sets per approach") +
  theme_bw() + 
  theme(plot.title = element_text(hjust = 0.5),
        legend.position = "none") 

date.analysis <- format(Sys.Date(), "%Y%b%d")
seed.num <- 2390
ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/", date.analysis, "_", "AGEMain_eFDR_qvalue", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 8,
         height = 4,
         units = "in")
```


### SEX
```{r}
SEX.eFDR.q.dat.2390 <- data.frame(Estimate = c(
  BMAseq.eFDR.Main.2000.train$SEX, BMAseq.eFDR.Main.2000.test$SEX,
  DESeq2.eFDR.2000.train$SEX, DESeq2.eFDR.2000.test$SEX, 
  edgeR.eFDR.2000.train$SEX, edgeR.eFDR.2000.test$SEX,
  eBayes.eFDR.2000.train$SEX, eBayes.eFDR.2000.test$SEX,
  voom.eFDR.2000.train$SEX, voom.eFDR.2000.test$SEX),
                                  Data.type = rep(c("Train", "Test"), each = 2000, times = 5),
                                  DEG.method = rep(c("BMAseq", "DESeq2", "edgeR", "voom+limma+eBayes", "voom+limma"), each = 4000)
                                  )
g <- SEX.eFDR.q.dat.2390 %>% 
  ggplot(aes(x = Data.type, 
             y = Estimate,
             fill = Data.type)) +
  geom_violin() +
  geom_boxplot(width = 0.1) + 
  facet_wrap(~ DEG.method, nrow = 1) + 
  scale_x_discrete(limits = c("Train", "Test")) + 
  scale_fill_brewer(palette = "Set1",
                    direction = -1) +
  labs(x = "Data Type",
       y = "Estimate",
       title = "The distributions of top 2000 lowest eFDRs and q-values \nassociated with the main effect of SEX in training and test sets per approach") +
  theme_bw() + 
  theme(plot.title = element_text(hjust = 0.5),
        legend.position = "none") 

date.analysis <- format(Sys.Date(), "%Y%b%d")
seed.num <- 2390
ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/", date.analysis, "_", "SEXMain_eFDR_qvalue", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 8,
         height = 4,
         units = "in")
```


### MHABNWBC
```{r}
MHABNWBC.eFDR.q.dat.2390 <- data.frame(Estimate = c(
  BMAseq.eFDR.Main.2000.train$MHABNWBC, BMAseq.eFDR.Main.2000.test$MHABNWBC,
  DESeq2.eFDR.2000.train$MHABNWBC, DESeq2.eFDR.2000.test$MHABNWBC, 
  edgeR.eFDR.2000.train$MHABNWBC, edgeR.eFDR.2000.test$MHABNWBC,
  eBayes.eFDR.2000.train$MHABNWBC, eBayes.eFDR.2000.test$MHABNWBC,
  voom.eFDR.2000.train$MHABNWBC, voom.eFDR.2000.test$MHABNWBC),
                                  Data.type = rep(c("Train", "Test"), each = 2000, times = 5),
                                  DEG.method = rep(c("BMAseq", "DESeq2", "edgeR", "voom+limma+eBayes", "voom+limma"), each = 4000)
                                  )
g <- MHABNWBC.eFDR.q.dat.2390 %>% 
  ggplot(aes(x = Data.type, 
             y = Estimate,
             fill = Data.type)) +
  geom_violin() +
  geom_boxplot(width = 0.1) + 
  facet_wrap(~ DEG.method, nrow = 1) + 
  scale_x_discrete(limits = c("Train", "Test")) + 
  scale_fill_brewer(palette = "Set1",
                    direction = -1) +
  labs(x = "Data Type",
       y = "Estimate",
       title = "The distributions of top 2000 lowest eFDRs and q-values \nassociated with the main effect of MHABNWBC in training and test sets per approach") +
  theme_bw() + 
  theme(plot.title = element_text(hjust = 0.5),
        legend.position = "none") 

date.analysis <- format(Sys.Date(), "%Y%b%d")
seed.num <- 2390
ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/", date.analysis, "_", "MHABNWBCMain_eFDR_qvalue", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 8,
         height = 4,
         units = "in")
```


### BMI X SEX
```{r}
BMI.Int.eFDR.q.dat.2390 <- data.frame(
  Estimate = c(
    BMAseq.eFDR.Interaction.2000.train$BMI, BMAseq.eFDR.Interaction.2000.test$BMI, # same if use SEX
    DESeq2.eFDR.2000.train$BMIxSEX, DESeq2.eFDR.2000.test$BMIxSEX, 
    edgeR.eFDR.2000.train$BMIxSEX, edgeR.eFDR.2000.test$BMIxSEX,
    eBayes.eFDR.2000.train$BMIxSEX, eBayes.eFDR.2000.test$BMIxSEX,
    voom.eFDR.2000.train$BMIxSEX, voom.eFDR.2000.test$BMIxSEX),
  Data.type = rep(c("Train", "Test"), each = 2000, times = 5),
  DEG.method = rep(c("BMAseq", "DESeq2", "edgeR", "voom+limma+eBayes", "voom+limma"), each = 4000)
                                  )
g <- BMI.Int.eFDR.q.dat.2390 %>% 
  ggplot(aes(x = Data.type, 
             y = Estimate,
             fill = Data.type)) +
  geom_violin() +
  geom_boxplot(width = 0.1) + 
  facet_wrap(~ DEG.method, nrow = 1) + 
  scale_x_discrete(limits = c("Train", "Test")) + 
  scale_fill_brewer(palette = "Set1",
                    direction = -1) +
  labs(x = "Data Type",
       y = "Estimate",
       title = "The distributions of top 2000 lowest eFDRs and q-values \nassociated with the interaction effect of BMI X AGE in training and test sets per approach") +
  theme_bw() + 
  theme(plot.title = element_text(hjust = 0.5),
        legend.position = "none") 

date.analysis <- format(Sys.Date(), "%Y%b%d")
seed.num <- 2390
ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/", date.analysis, "_", "BMIInt_eFDR_qvalue", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 8,
         height = 4,
         units = "in")
```


## Seed: 98907
```{r}
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt98907.RData")
```


### BMI
```{r}
BMI.eFDR.q.dat.98907 <- data.frame(Estimate = c(
  BMAseq.eFDR.Main.2000.train$BMI, BMAseq.eFDR.Main.2000.test$BMI,
  DESeq2.eFDR.2000.train$BMI, DESeq2.eFDR.2000.test$BMI, 
  edgeR.eFDR.2000.train$BMI, edgeR.eFDR.2000.test$BMI,
  eBayes.eFDR.2000.train$BMI, eBayes.eFDR.2000.test$BMI,
  voom.eFDR.2000.train$BMI, voom.eFDR.2000.test$BMI),
                                  Data.type = rep(c("Train", "Test"), each = 2000, times = 5),
                                  DEG.method = rep(c("BMAseq", "DESeq2", "edgeR", "voom+limma+eBayes", "voom+limma"), each = 4000)
                                  )
g <- BMI.eFDR.q.dat.98907 %>% 
  ggplot(aes(x = Data.type, 
             y = Estimate,
             fill = Data.type)) +
  geom_violin() +
  geom_boxplot(width = 0.1) + 
  facet_wrap(~ DEG.method, nrow = 1) + 
  scale_x_discrete(limits = c("Train", "Test")) + 
  scale_fill_brewer(palette = "Set1",
                    direction = -1) +
  labs(x = "Data Type",
       y = "Estimate",
       title = "The distributions of top 2000 lowest eFDRs and q-values \nassociated with the main effect of BMI in training and test sets per approach") +
  theme_bw() + 
  theme(plot.title = element_text(hjust = 0.5),
        legend.position = "none") 

date.analysis <- format(Sys.Date(), "%Y%b%d")
seed.num <- 98907
ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/", date.analysis, "_", "BMIMain_eFDR_qvalue", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 8,
         height = 4,
         units = "in")
```


### AGE
```{r}
AGE.eFDR.q.dat.98907 <- data.frame(Estimate = c(
  BMAseq.eFDR.Main.2000.train$AGE, BMAseq.eFDR.Main.2000.test$AGE,
  DESeq2.eFDR.2000.train$AGE, DESeq2.eFDR.2000.test$AGE, 
  edgeR.eFDR.2000.train$AGE, edgeR.eFDR.2000.test$AGE,
  eBayes.eFDR.2000.train$AGE, eBayes.eFDR.2000.test$AGE,
  voom.eFDR.2000.train$AGE, voom.eFDR.2000.test$AGE),
                                  Data.type = rep(c("Train", "Test"), each = 2000, times = 5),
                                  DEG.method = rep(c("BMAseq", "DESeq2", "edgeR", "voom+limma+eBayes", "voom+limma"), each = 4000)
                                  )
g <- AGE.eFDR.q.dat.98907 %>% 
  ggplot(aes(x = Data.type, 
             y = Estimate,
             fill = Data.type)) +
  geom_violin() +
  geom_boxplot(width = 0.1) + 
  facet_wrap(~ DEG.method, nrow = 1) + 
  scale_x_discrete(limits = c("Train", "Test")) + 
  scale_fill_brewer(palette = "Set1",
                    direction = -1) +
  labs(x = "Data Type",
       y = "Estimate",
       title = "The distributions of top 2000 lowest eFDRs and q-values \nassociated with the main effect of AGE in training and test sets per approach") +
  theme_bw() + 
  theme(plot.title = element_text(hjust = 0.5),
        legend.position = "none") 

date.analysis <- format(Sys.Date(), "%Y%b%d")
seed.num <- 98907
ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/", date.analysis, "_", "AGEMain_eFDR_qvalue", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 8,
         height = 4,
         units = "in")
```


### SEX
```{r}
SEX.eFDR.q.dat.98907 <- data.frame(Estimate = c(
  BMAseq.eFDR.Main.2000.train$SEX, BMAseq.eFDR.Main.2000.test$SEX,
  DESeq2.eFDR.2000.train$SEX, DESeq2.eFDR.2000.test$SEX, 
  edgeR.eFDR.2000.train$SEX, edgeR.eFDR.2000.test$SEX,
  eBayes.eFDR.2000.train$SEX, eBayes.eFDR.2000.test$SEX,
  voom.eFDR.2000.train$SEX, voom.eFDR.2000.test$SEX),
                                  Data.type = rep(c("Train", "Test"), each = 2000, times = 5),
                                  DEG.method = rep(c("BMAseq", "DESeq2", "edgeR", "voom+limma+eBayes", "voom+limma"), each = 4000)
                                  )
g <- SEX.eFDR.q.dat.98907 %>% 
  ggplot(aes(x = Data.type, 
             y = Estimate,
             fill = Data.type)) +
  geom_violin() +
  geom_boxplot(width = 0.1) + 
  facet_wrap(~ DEG.method, nrow = 1) + 
  scale_x_discrete(limits = c("Train", "Test")) + 
  scale_fill_brewer(palette = "Set1",
                    direction = -1) +
  labs(x = "Data Type",
       y = "Estimate",
       title = "The distributions of top 2000 lowest eFDRs and q-values \nassociated with the main effect of SEX in training and test sets per approach") +
  theme_bw() + 
  theme(plot.title = element_text(hjust = 0.5),
        legend.position = "none") 

date.analysis <- format(Sys.Date(), "%Y%b%d")
seed.num <- 98907
ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/", date.analysis, "_", "SEXMain_eFDR_qvalue", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 8,
         height = 4,
         units = "in")
```


### MHABNWBC
```{r}
MHABNWBC.eFDR.q.dat.98907 <- data.frame(Estimate = c(
  BMAseq.eFDR.Main.2000.train$MHABNWBC, BMAseq.eFDR.Main.2000.test$MHABNWBC,
  DESeq2.eFDR.2000.train$MHABNWBC, DESeq2.eFDR.2000.test$MHABNWBC, 
  edgeR.eFDR.2000.train$MHABNWBC, edgeR.eFDR.2000.test$MHABNWBC,
  eBayes.eFDR.2000.train$MHABNWBC, eBayes.eFDR.2000.test$MHABNWBC,
  voom.eFDR.2000.train$MHABNWBC, voom.eFDR.2000.test$MHABNWBC),
                                  Data.type = rep(c("Train", "Test"), each = 2000, times = 5),
                                  DEG.method = rep(c("BMAseq", "DESeq2", "edgeR", "voom+limma+eBayes", "voom+limma"), each = 4000)
                                  )
g <- MHABNWBC.eFDR.q.dat.98907 %>% 
  ggplot(aes(x = Data.type, 
             y = Estimate,
             fill = Data.type)) +
  geom_violin() +
  geom_boxplot(width = 0.1) + 
  facet_wrap(~ DEG.method, nrow = 1) + 
  scale_x_discrete(limits = c("Train", "Test")) + 
  scale_fill_brewer(palette = "Set1",
                    direction = -1) +
  labs(x = "Data Type",
       y = "Estimate",
       title = "The distributions of top 2000 lowest eFDRs and q-values \nassociated with the main effect of MHABNWBC in training and test sets per approach") +
  theme_bw() + 
  theme(plot.title = element_text(hjust = 0.5),
        legend.position = "none") 

date.analysis <- format(Sys.Date(), "%Y%b%d")
seed.num <- 98907
ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/", date.analysis, "_", "MHABNWBCMain_eFDR_qvalue", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 8,
         height = 4,
         units = "in")
```


### BMI X SEX
```{r}
BMI.Int.eFDR.q.dat.98907 <- data.frame(
  Estimate = c(
    BMAseq.eFDR.Interaction.2000.train$BMI, BMAseq.eFDR.Interaction.2000.test$BMI, # same if use SEX
    DESeq2.eFDR.2000.train$BMIxSEX, DESeq2.eFDR.2000.test$BMIxSEX, 
    edgeR.eFDR.2000.train$BMIxSEX, edgeR.eFDR.2000.test$BMIxSEX,
    eBayes.eFDR.2000.train$BMIxSEX, eBayes.eFDR.2000.test$BMIxSEX,
    voom.eFDR.2000.train$BMIxSEX, voom.eFDR.2000.test$BMIxSEX),
  Data.type = rep(c("Train", "Test"), each = 2000, times = 5),
  DEG.method = rep(c("BMAseq", "DESeq2", "edgeR", "voom+limma+eBayes", "voom+limma"), each = 4000)
                                  )
g <- BMI.Int.eFDR.q.dat.98907 %>% 
  ggplot(aes(x = Data.type, 
             y = Estimate,
             fill = Data.type)) +
  geom_violin() +
  geom_boxplot(width = 0.1) + 
  facet_wrap(~ DEG.method, nrow = 1) + 
  scale_x_discrete(limits = c("Train", "Test")) + 
  scale_fill_brewer(palette = "Set1",
                    direction = -1) +
  labs(x = "Data Type",
       y = "Estimate",
       title = "The distributions of top 2000 lowest eFDRs and q-values \nassociated with the interaction effect of BMI X AGE in training and test sets per approach") +
  theme_bw() + 
  theme(plot.title = element_text(hjust = 0.5),
        legend.position = "none") 

date.analysis <- format(Sys.Date(), "%Y%b%d")
seed.num <- 98907
ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/", date.analysis, "_", "BMIInt_eFDR_qvalue", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 8,
         height = 4,
         units = "in")
```


# GGally plot
## Seed: 2390
```{r}
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt2390.RData")
g <- GGally::ggpairs(data = dat.pheno.train[c("BMI", "AGE", "SEX", "MHABNWBC")],
                xlab = "Variables",
                ylab = "Variables",
                title = "Relationships between variables in the training set") + 
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.text.x = element_text(angle = 90, hjust = 1))

date.analysis <- format(Sys.Date(), "%Y%b%d")
seed.num <- 2390
ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/", date.analysis, "_", "GGallyTrain", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 8,
         height = 4,
         units = "in")

g <- GGally::ggpairs(data = dat.pheno.test[c("BMI", "AGE", "SEX", "MHABNWBC")],
                xlab = "Variables",
                ylab = "Variables",
                title = "Relationships between variables in the test set") + 
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.text.x = element_text(angle = 90, hjust = 1))

date.analysis <- format(Sys.Date(), "%Y%b%d")
ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/", date.analysis, "_", "GGallyTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 8,
         height = 4,
         units = "in")
```


## Seed: 98907
```{r}
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt98907.RData")
g <- GGally::ggpairs(data = dat.pheno.train[c("BMI", "AGE", "SEX", "MHABNWBC")],
                xlab = "Variables",
                ylab = "Variables",
                title = "Relationships between variables in the training set") + 
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.text.x = element_text(angle = 90, hjust = 1))

date.analysis <- format(Sys.Date(), "%Y%b%d")
seed.num <- 90907
ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/", date.analysis, "_", "GGallyTrain", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 8,
         height = 4,
         units = "in")

g <- GGally::ggpairs(data = dat.pheno.test[c("BMI", "AGE", "SEX", "MHABNWBC")],
                xlab = "Variables",
                ylab = "Variables",
                title = "Relationships between variables in the test set") + 
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.text.x = element_text(angle = 90, hjust = 1))

date.analysis <- format(Sys.Date(), "%Y%b%d")
ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/", date.analysis, "_", "GGallyTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 8,
         height = 4,
         units = "in")
```


# BMI x SEX
## Seed: 2390
```{r}
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt2390.RData")
table(dat.pheno.train$BMI, dat.pheno.train$SEX)
with(dat.pheno.train, chisq.test(x = BMI, y = SEX)$expected)
with(dat.pheno.train, chisq.test(x = BMI, y = SEX, correct = F))

table(dat.pheno.test$BMI, dat.pheno.test$SEX)
with(dat.pheno.test, chisq.test(x = BMI, y = SEX)$expected)
with(dat.pheno.test, chisq.test(x = BMI, y = SEX, correct = F))
```


## Seed: 98907
```{r}
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt98907.RData")
table(dat.pheno.train$BMI, dat.pheno.train$SEX)
with(dat.pheno.train, chisq.test(x = BMI, y = SEX)$expected)
with(dat.pheno.train, chisq.test(x = BMI, y = SEX, correct = F))

table(dat.pheno.test$BMI, dat.pheno.test$SEX)
with(dat.pheno.test, chisq.test(x = BMI, y = SEX)$expected)
with(dat.pheno.test, chisq.test(x = BMI, y = SEX, correct = F))
```


# Boxplot
## Seed: 2390 
```{r}
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt2390.RData")
library(ggpubr)
dat.expr.train.t <- t(dat.expr.train)
all(order(rownames(dat.expr.train.t)) == order(rownames(dat.pheno.train))) # TRUE
dat.bmi.sex.train <- cbind(dat.pheno.train[c("BMI", "SEX")], dat.expr.train.t)

boxplot.fun <- function(dataset, var.x, var.y, set.status) {
  dataset %>% ggplot(aes(y = get(var.y), x = get(var.x))) + 
    geom_boxplot(mapping = aes(group = get(var.x), color = get(var.x))) +
    geom_jitter(mapping = aes(color = get(var.x))) + 
    facet_wrap(~SEX) + 
    labs(title = paste0("Gene expression across different states of BMI and SEX \nin the ", set.status, " set"), y = paste0(var.y, " expression"), x = "BMI") +
    theme_bw(base_size = 14) + 
    scale_color_manual(values = c("#0073C2FF", "#EFC000FF")) + 
    guides(color = "none")
  }

boxplot.bmi.sex.train <- names(dat.bmi.sex.train)[9] %>% 
  map(.f = boxplot.fun, dataset = dat.bmi.sex.train, var.x = "BMI", set.status = "training")

seed.num <- 2390
date.analysis <- format(Sys.Date(), "%Y%b%d")
ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/", date.analysis, "_", "BoxplotTrain", "_", seed.num, ".png"),
         plot = boxplot.bmi.sex.train[[1]],
         device = "png",
         width = 8,
         height = 4,
         units = "in")

# grid.arrange(do.call("arrangeGrob", c(boxplot.bmi.sex.train, ncol = 1)))

dat.expr.test.t <- t(dat.expr.test)
all(order(rownames(dat.expr.test.t)) == order(rownames(dat.pheno.test))) # TRUE
dat.bmi.sex.test <- cbind(dat.pheno.test[c("BMI", "SEX")], dat.expr.test.t)
boxplot.bmi.sex.test <- names(dat.bmi.sex.test)[9] %>% 
  map(.f = boxplot.fun, dataset = dat.bmi.sex.test, var.x = "BMI", set.status = "test")

ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/", date.analysis, "_", "BoxplotTest", "_", seed.num, ".png"),
         plot = boxplot.bmi.sex.test[[1]],
         device = "png",
         width = 8,
         height = 4,
         units = "in")
```


## Seed: 98907
```{r}
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt98907.RData")
dat.expr.train.t <- t(dat.expr.train)
all(order(rownames(dat.expr.train.t)) == order(rownames(dat.pheno.train))) # TRUE
dat.bmi.sex.train <- cbind(dat.pheno.train[c("BMI", "SEX")], dat.expr.train.t)

boxplot.bmi.sex.train <- names(dat.bmi.sex.train)[12] %>% 
  map(.f = boxplot.fun, dataset = dat.bmi.sex.train, var.x = "BMI", set.status = "training")

seed.num <- 98907
date.analysis <- format(Sys.Date(), "%Y%b%d")
ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/", date.analysis, "_", "BoxplotTrain", "_", seed.num, ".png"),
         plot = boxplot.bmi.sex.train[[1]],
         device = "png",
         width = 8,
         height = 4,
         units = "in")

# grid.arrange(do.call("arrangeGrob", c(boxplot.bmi.sex.train, ncol = 1)))

dat.expr.test.t <- t(dat.expr.test)
all(order(rownames(dat.expr.test.t)) == order(rownames(dat.pheno.test))) # TRUE
dat.bmi.sex.test <- cbind(dat.pheno.test[c("BMI", "SEX")], dat.expr.test.t)
boxplot.bmi.sex.test <- names(dat.bmi.sex.test)[9] %>% 
  map(.f = boxplot.fun, dataset = dat.bmi.sex.test, var.x = "BMI", set.status = "test")

ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/", date.analysis, "_", "BoxplotTest", "_", seed.num, ".png"),
         plot = boxplot.bmi.sex.test[[1]],
         device = "png",
         width = 8,
         height = 4,
         units = "in")
```


# Venn digram
## Seed: 2390 
```{r}
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt2390.RData")

# names(BMAseq.eFDR.Main.2000.train$BMI)

g <- ggVennDiagram(list(BMAseq = intersect(names(BMAseq.eFDR.Main.2000.train$BMI), names(BMAseq.eFDR.Main.2000.test$BMI)), 
                        DESeq2 = intersect(DESeq2.eFDR.2000.GeneName.train$BMI, DESeq2.eFDR.2000.GeneName.test$BMI),
                        edgeR = intersect(edgeR.eFDR.2000.GeneName.train$BMI, edgeR.eFDR.2000.GeneName.test$BMI),
                        eBayes = intersect(names(eBayes.eFDR.2000.train$BMI), names(eBayes.eFDR.2000.test$BMI)),
                        voom.limma = intersect(names(voom.eFDR.2000.train$BMI), names(voom.eFDR.2000.test$BMI))),
                   label_alpha = 0, label_color = "white") +
  theme(legend.position = "none", 
        plot.title = element_text(hjust = 0.5, colour = "red", size = 18)) + 
  ggtitle("The intersection of common differentially expressed genes \nassociated with the main effect of BMI") + 
  scale_fill_distiller(palette = "Dark2", direction = -1)
                  
seed.num <- 2390
date.analysis <- format(Sys.Date(), "%Y%b%d")
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/", date.analysis, "_", "BMI.DEGIntersectionTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 10,
         height = 7,
         units = "in")
```


## Seed: 98907 
```{r}
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt98907.RData")

# names(BMAseq.eFDR.Main.2000.train$BMI)

g <- ggVennDiagram(list(BMAseq = intersect(names(BMAseq.eFDR.Main.2000.train$BMI), names(BMAseq.eFDR.Main.2000.test$BMI)), 
                        DESeq2 = intersect(DESeq2.eFDR.2000.GeneName.train$BMI, DESeq2.eFDR.2000.GeneName.test$BMI),
                        edgeR = intersect(edgeR.eFDR.2000.GeneName.train$BMI, edgeR.eFDR.2000.GeneName.test$BMI),
                        eBayes = intersect(names(eBayes.eFDR.2000.train$BMI), names(eBayes.eFDR.2000.test$BMI)),
                        voom.limma = intersect(names(voom.eFDR.2000.train$BMI), names(voom.eFDR.2000.test$BMI))),
                   label_alpha = 0, label_color = "white") +
  theme(legend.position = "none", 
        plot.title = element_text(hjust = 0.5, colour = "red", size = 18)) + 
  ggtitle("The intersection of common differentially expressed genes \nassociated with the main effect of BMI") + 
  scale_fill_distiller(palette = "Dark2", direction = -1)
                  
seed.num <- 98970
date.analysis <- format(Sys.Date(), "%Y%b%d")
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/", date.analysis, "_", "BMI.DEGIntersectionTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 10,
         height = 7,
         units = "in")
```


# Boxplot - expression distributions of top ranked genes between the training and test set (take the BMAseq as an example)
## Seed: 2390
### Top 2000 genes (associated with BMI) list from the training set
```{r}
##------Prepare data------
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt2390.RData")
seed.num <- 2390
date.analysis <- format(Sys.Date(), "%Y%b%d")

# names(BMAseq.eFDR.Main.2000.train[["BMI"]][1:10]) # ENSG IDs of top ranked 10 genes based on the eFDR from the smallest to the largest in the training set

library(ggpubr)
library(purrr)
library(gridExtra)

##------Prepare plotting function------
boxplot.fun <- function(dataset, var.x, var.y) {
  dataset |> ggplot(mapping = aes(y = get(var.y), x = get(var.x))) + 
    geom_boxplot(mapping = aes(group = get(var.x), color = get(var.x))) +
    geom_jitter(mapping = aes(color = get(var.x))) + 
    stat_compare_means(method = "wilcox.test", 
                       label.x.npc = "right",
                       label.y = 2,
                       mapping = aes(label = ifelse(p < 1.e-3, "p < 0.001", sprintf("p = %4.3f", as.numeric(..p.format..))))) +
    labs(y = var.y, 
         x = "") +
    theme_bw(base_size = 14) + 
    scale_color_manual(values = c("#0073C2FF", "#EFC000FF")) + 
    guides(color = "none")
}

for(i in seq(1, 1998, 9)) {
  ##------Process data------
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8)
  dat.expr.train2390.top9 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.train[["BMI"]])[gene.id], ]
  dat.expr.train2390.top9.t <- t(dat.expr.train2390.top9) |> as.data.frame()
  dat.expr.train2390.top9.t$Data <- "Train"
  dat.expr.test2390.traintop9 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.train[["BMI"]])[gene.id], ]
  dat.expr.test2390.traintop9.t <- t(dat.expr.test2390.traintop9) |> as.data.frame()
  dat.expr.test2390.traintop9.t$Data <- "Test"
  rownames(dat.expr.train2390.top9.t) = rownames(dat.expr.test2390.traintop9.t) = NULL
  dat.expr.full2390.traintop9 <- rbind(dat.expr.train2390.top9.t,
                                       dat.expr.test2390.traintop9.t)
  p.vec <- NULL
  for(j in 1:9) {
    p.val <- wilcox.test(dat.expr.full2390.traintop9[[j]] ~ dat.expr.full2390.traintop9[["Data"]], data = dat.expr.full2390.traintop9)$p.value
    p.vec <- c(p.vec, p.val)
  }
    
  ##------Make boxplots------
  if (any(!is.na(p.vec) & p.vec < 0.05)) {
    boxplot.full2390.traintop9 <- names(dat.expr.full2390.traintop9)[-match("Data", names(dat.expr.full2390.traintop9))] |> 
    map(.f = boxplot.fun, dataset = dat.expr.full2390.traintop9, var.x = "Data")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.traintop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/TrainTop2000/Significant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
           plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  } else {
    boxplot.full2390.traintop9 <- names(dat.expr.full2390.traintop9)[-match("Data", names(dat.expr.full2390.traintop9))] |> 
    map(.f = boxplot.fun, dataset = dat.expr.full2390.traintop9, var.x = "Data")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.traintop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/TrainTop2000/Insignificant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
           plot = g, device = "png", width = 16, height = 8, units = "in")
    print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  }
}

gene.id <- c(1999, 2000)
dat.expr.train2390.top2 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.train[["BMI"]])[gene.id], ]
dat.expr.train2390.top2.t <- t(dat.expr.train2390.top2) |> as.data.frame()
dat.expr.train2390.top2.t$Data <- "Train"
dat.expr.test2390.traintop2 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.train[["BMI"]])[gene.id], ]
dat.expr.test2390.traintop2.t <- t(dat.expr.test2390.traintop2) |> as.data.frame()
dat.expr.test2390.traintop2.t$Data <- "Test"
rownames(dat.expr.train2390.top2.t) = rownames(dat.expr.test2390.traintop2.t) = NULL
dat.expr.full2390.traintop2 <- rbind(dat.expr.train2390.top2.t,
                                     dat.expr.test2390.traintop2.t)

p.vec <- NULL
for(j in 1:2) {
  p.val <- wilcox.test(dat.expr.full2390.traintop2[[j]] ~ dat.expr.full2390.traintop2[["Data"]], data = dat.expr.full2390.traintop2)$p.value
  p.vec <- c(p.vec, p.val)
}
    
if (any(!is.na(p.vec) & p.vec < 0.05)) {
  boxplot.full2390.traintop2 <- names(dat.expr.full2390.traintop2)[-match("Data", names(dat.expr.full2390.traintop2))] |> 
  map(.f = boxplot.fun, dataset = dat.expr.full2390.traintop2, var.x = "Data")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.traintop2, ncol = 2)))
  print(paste0("[", i, "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/TrainTop2000/Significant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
} else {
  boxplot.full2390.traintop2 <- names(dat.expr.full2390.traintop2)[-match("Data", names(dat.expr.full2390.traintop2))] |> 
  map(.f = boxplot.fun, dataset = dat.expr.full2390.traintop2, var.x = "Data")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.traintop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/TrainTop2000/Insignificant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
}
```


### Top 2000 genes list (associated with BMI) from the test set
```{r}
##------Prepare data------
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt2390.RData")
seed.num <- 2390
date.analysis <- format(Sys.Date(), "%Y%b%d")

for(i in seq(1, 1998, 9)) {
  ##------Process data------
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8)
  dat.expr.train2390.testtop9 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.test[["BMI"]])[gene.id], ]
  dat.expr.train2390.testtop9.t <- t(dat.expr.train2390.testtop9) |> as.data.frame()
  dat.expr.train2390.testtop9.t$Data <- "Train"
  dat.expr.test2390.top9 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.test[["BMI"]])[gene.id], ]
  dat.expr.test2390.top9.t <- t(dat.expr.test2390.top9) |> as.data.frame()
  dat.expr.test2390.top9.t$Data <- "Test"
  rownames(dat.expr.train2390.testtop9.t) = rownames(dat.expr.test2390.top9.t) = NULL
  dat.expr.full2390.testtop9 <- rbind(dat.expr.train2390.testtop9.t,
                                      dat.expr.test2390.top9.t)
  p.vec <- NULL
  for(j in 1:9) {
    p.val <- wilcox.test(dat.expr.full2390.testtop9[[j]] ~ dat.expr.full2390.testtop9[["Data"]], data = dat.expr.full2390.testtop9)$p.value
    p.vec <- c(p.vec, p.val)
  }
    
  ##------Make boxplots------
  if (any(!is.na(p.vec) & p.vec < 0.05)) {
    boxplot.full2390.testtop9 <- names(dat.expr.full2390.testtop9)[-match("Data", names(dat.expr.full2390.testtop9))] |> 
    map(.f = boxplot.fun, dataset = dat.expr.full2390.testtop9, var.x = "Data")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.testtop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/TestTop2000/Significant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  } else {
    boxplot.full2390.testtop9 <- names(dat.expr.full2390.testtop9)[-match("Data", names(dat.expr.full2390.testtop9))] |> 
    map(.f = boxplot.fun, dataset = dat.expr.full2390.testtop9, var.x = "Data")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.testtop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/TestTop2000/Insignificant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
    print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  }
}

gene.id <- c(1999, 2000)
dat.expr.train2390.testtop2 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.test[["BMI"]])[gene.id], ]
dat.expr.train2390.testtop2.t <- t(dat.expr.train2390.testtop2) |> as.data.frame()
dat.expr.train2390.testtop2.t$Data <- "Train"
dat.expr.test2390.top2 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.test[["BMI"]])[gene.id], ]
dat.expr.test2390.top2.t <- t(dat.expr.test2390.top2) |> as.data.frame()
dat.expr.test2390.top2.t$Data <- "Test"
rownames(dat.expr.train2390.testtop2.t) = rownames(dat.expr.test2390.top2.t) = NULL
dat.expr.full2390.testtop2 <- rbind(dat.expr.train2390.testtop2.t,
                                    dat.expr.test2390.top2.t)

p.vec <- NULL
for(j in 1:2) {
  p.val <- wilcox.test(dat.expr.full2390.testtop2[[j]] ~ dat.expr.full2390.testtop2[["Data"]], data = dat.expr.full2390.testtop2)$p.value
  p.vec <- c(p.vec, p.val)
}
    
if (any(!is.na(p.vec) & p.vec < 0.05)) {
  boxplot.full2390.testtop2 <- names(dat.expr.full2390.testtop2)[-match("Data", names(dat.expr.full2390.testtop2))] |> 
  map(.f = boxplot.fun, dataset = dat.expr.full2390.testtop2, var.x = "Data")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.testtop2, ncol = 2)))
  print(paste0("[", i, "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/TestTop2000/Significant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
} else {
  boxplot.full2390.testtop2 <- names(dat.expr.full2390.testtop2)[-match("Data", names(dat.expr.full2390.testtop2))] |> 
  map(.f = boxplot.fun, dataset = dat.expr.full2390.testtop2, var.x = "Data")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.testtop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/TestTop2000/Insignificant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
}
```


### BMI
#### Top 2000 genes list from the training set
```{r}
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt2390.RData")
BMI <- c(dat.pheno.train[["BMI"]], dat.pheno.test[["BMI"]])

seed.num <- 2390
date.analysis <- format(Sys.Date(), "%Y%b%d")

##------Prepare plotting function------
boxplot.fun.2 <- function(dataset, var.x, var.y) {
  dataset |> ggplot(mapping = aes(y = get(var.y), x = get(var.x))) + 
    geom_boxplot(mapping = aes(color = get(var.x))) +
    geom_jitter(mapping = aes(color = get(var.x))) + 
    stat_compare_means(method = "wilcox.test", 
                       label.x.npc = "right",
                       label.y = 2,
                       mapping = aes(label = ifelse(p < 1.e-3, "p < 0.001", sprintf("p = %4.3f", as.numeric(..p.format..))))) +
    facet_wrap(~factor(Data, levels = c("Train", "Test"))) + 
    labs(y = var.y, 
         x = "") +
    theme_bw(base_size = 14) + 
    scale_color_manual(values = c("#0073C2FF", "#EFC000FF")) + 
    scale_x_discrete(limits = c("low", "high")) + 
    guides(color = "none")
}

for(i in seq(1, 1998, 9)) {
  ##------Process data------
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8)
  dat.expr.train2390.top9 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.train[["BMI"]])[gene.id], ]
  dat.expr.train2390.top9.t <- t(dat.expr.train2390.top9) |> as.data.frame()
dat.expr.train2390.top9.t$Data <- "Train"
  dat.expr.test2390.traintop9 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.train[["BMI"]])[gene.id], ]
  dat.expr.test2390.traintop9.t <- t(dat.expr.test2390.traintop9) |> as.data.frame()
  dat.expr.test2390.traintop9.t$Data <- "Test"
  rownames(dat.expr.train2390.top9.t) = rownames(dat.expr.test2390.traintop9.t) = NULL
  dat.expr.full2390.traintop9 <- rbind(dat.expr.train2390.top9.t,
                                       dat.expr.test2390.traintop9.t)
  
  dat.expr.full2390.traintop9 <- data.frame(dat.expr.full2390.traintop9, BMI)
  sumDis <- 0
  for(j in 1:9) {
    p.val.train <- wilcox.test(dat.expr.full2390.traintop9[dat.expr.full2390.traintop9$Data == "Train", ][[j]] ~ dat.expr.full2390.traintop9[dat.expr.full2390.traintop9$Data == "Train", ][["BMI"]], data = dat.expr.full2390.traintop9)$p.value
    p.val.test <- wilcox.test(dat.expr.full2390.traintop9[dat.expr.full2390.traintop9$Data == "Test", ][[j]] ~ dat.expr.full2390.traintop9[dat.expr.full2390.traintop9$Data == "Test", ][["BMI"]], data = dat.expr.full2390.traintop9)$p.value
    
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1 }
  }
    
  ##------Make boxplots------
  if ((!is.na(sumDis)) & (sumDis > 0)) {
    boxplot.full2390.traintop9 <- names(dat.expr.full2390.traintop9)[-match(c("Data", "BMI"), names(dat.expr.full2390.traintop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full2390.traintop9, var.x = "BMI")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.traintop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/BMI/TrainTop2000/Discordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"), plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  } else {
    boxplot.full2390.traintop9 <- names(dat.expr.full2390.traintop9)[-match(c("Data", "BMI"), names(dat.expr.full2390.traintop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full2390.traintop9, var.x = "BMI")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.traintop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/BMI/TrainTop2000/Concordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"), 
           plot = g, device = "png", width = 16, height = 8, units = "in")
    print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  }
}

gene.id <- c(1999, 2000)
dat.expr.train2390.top2 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.train[["BMI"]])[gene.id], ]
dat.expr.train2390.top2.t <- t(dat.expr.train2390.top2) |> as.data.frame()
dat.expr.train2390.top2.t$Data <- "Train"
dat.expr.test2390.traintop2 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.train[["BMI"]])[gene.id], ]
dat.expr.test2390.traintop2.t <- t(dat.expr.test2390.traintop2) |> as.data.frame()
dat.expr.test2390.traintop2.t$Data <- "Test"
rownames(dat.expr.train2390.top2.t) = rownames(dat.expr.test2390.traintop2.t) = NULL
dat.expr.full2390.traintop2 <- rbind(dat.expr.train2390.top2.t,
                                     dat.expr.test2390.traintop2.t)
dat.expr.full2390.traintop2 <- data.frame(dat.expr.full2390.traintop2, BMI)

sumDis <- 0
for(j in 1:2) {
    p.val.train <- wilcox.test(dat.expr.full2390.traintop2[dat.expr.full2390.traintop2$Data == "Train", ][[j]] ~ dat.expr.full2390.traintop2[dat.expr.full2390.traintop2$Data == "Train", ][["BMI"]], data = dat.expr.full2390.traintop2)$p.value
    p.val.test <- wilcox.test(dat.expr.full2390.traintop2[dat.expr.full2390.traintop2$Data == "Test", ][[j]] ~ dat.expr.full2390.traintop2[dat.expr.full2390.traintop2$Data == "Test", ][["BMI"]], data = dat.expr.full2390.traintop2)$p.value
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1}
}
    
if ((!is.na(sumDis)) & (sumDis > 0)) {
  boxplot.full2390.traintop2 <- names(dat.expr.full2390.traintop2)[-match(c("Data", "BMI"), names(dat.expr.full2390.traintop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full2390.traintop2, var.x = "BMI")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.traintop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/BMI/TrainTop2000/Discordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
} else {
  boxplot.full2390.traintop2 <- names(dat.expr.full2390.traintop2)[-match(c("Data", "BMI"), names(dat.expr.full2390.traintop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full2390.traintop2, var.x = "BMI")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.traintop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/BMI/TrainTop2000/Concordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
}
```


#### Top 2000 genes list from the test set
```{r}
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt2390.RData")
BMI <- c(dat.pheno.train[["BMI"]], dat.pheno.test[["BMI"]])

seed.num <- 2390
date.analysis <- format(Sys.Date(), "%Y%b%d")

##------Prepare plotting function------
boxplot.fun.2 <- function(dataset, var.x, var.y) {
  dataset |> ggplot(mapping = aes(y = get(var.y), x = get(var.x))) + 
    geom_boxplot(mapping = aes(color = get(var.x))) +
    geom_jitter(mapping = aes(color = get(var.x))) + 
    stat_compare_means(method = "wilcox.test", 
                       label.x.npc = "right",
                       label.y = 2,
                       mapping = aes(label = ifelse(p < 1.e-3, "p < 0.001", sprintf("p = %4.3f", as.numeric(..p.format..))))) +
    facet_wrap(~factor(Data, levels = c("Train", "Test"))) + 
    labs(y = var.y, 
         x = "") +
    theme_bw(base_size = 14) + 
    scale_color_manual(values = c("#0073C2FF", "#EFC000FF")) + 
    scale_x_discrete(limits = c("low", "high")) + 
    guides(color = "none")
}

for(i in seq(1, 1998, 9)) {
  ##------Process data------
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8)
  dat.expr.train2390.testtop9 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.test[["BMI"]])[gene.id], ]
  dat.expr.train2390.testtop9.t <- t(dat.expr.train2390.testtop9) |> as.data.frame()
dat.expr.train2390.testtop9.t$Data <- "Train"
  dat.expr.test2390.top9 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.test[["BMI"]])[gene.id], ]
  dat.expr.test2390.top9.t <- t(dat.expr.test2390.top9) |> as.data.frame()
  dat.expr.test2390.top9.t$Data <- "Test"
  rownames(dat.expr.train2390.testtop9.t) = rownames(dat.expr.test2390.top9.t) = NULL
  dat.expr.full2390.testtop9 <- rbind(dat.expr.train2390.testtop9.t,
                                      dat.expr.test2390.top9.t)
  
  dat.expr.full2390.testtop9 <- data.frame(dat.expr.full2390.testtop9, BMI)
  sumDis <- 0
  for(j in 1:9) {
    p.val.train <- wilcox.test(dat.expr.full2390.testtop9[dat.expr.full2390.testtop9$Data == "Train", ][[j]] ~ dat.expr.full2390.testtop9[dat.expr.full2390.testtop9$Data == "Train", ][["BMI"]], data = dat.expr.full2390.testtop9)$p.value
    p.val.test <- wilcox.test(dat.expr.full2390.testtop9[dat.expr.full2390.testtop9$Data == "Test", ][[j]] ~ dat.expr.full2390.testtop9[dat.expr.full2390.testtop9$Data == "Test", ][["BMI"]], data = dat.expr.full2390.testtop9)$p.value
    
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1 }
  }
    
  ##------Make boxplots------
  if ((!is.na(sumDis)) & (sumDis > 0)) {
    boxplot.full2390.testtop9 <- names(dat.expr.full2390.testtop9)[-match(c("Data", "BMI"), names(dat.expr.full2390.testtop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full2390.testtop9, var.x = "BMI")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.testtop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/BMI/TestTop2000/Discordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  } else {
    boxplot.full2390.testtop9 <- names(dat.expr.full2390.testtop9)[-match(c("Data", "BMI"), names(dat.expr.full2390.testtop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full2390.testtop9, var.x = "BMI")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.testtop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/BMI/TestTop2000/Concordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
    print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  }
}

gene.id <- c(1999, 2000)
dat.expr.train2390.testtop2 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.test[["BMI"]])[gene.id], ]
dat.expr.train2390.testtop2.t <- t(dat.expr.train2390.testtop2) |> as.data.frame()
dat.expr.train2390.testtop2.t$Data <- "Train"
dat.expr.test2390.top2 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.test[["BMI"]])[gene.id], ]
dat.expr.test2390.top2.t <- t(dat.expr.test2390.top2) |> as.data.frame()
dat.expr.test2390.top2.t$Data <- "Test"
rownames(dat.expr.train2390.testtop2.t) = rownames(dat.expr.test2390.top2.t) = NULL
dat.expr.full2390.testtop2 <- rbind(dat.expr.train2390.testtop2.t,
                                     dat.expr.test2390.top2.t)
dat.expr.full2390.testtop2 <- data.frame(dat.expr.full2390.testtop2, BMI)

sumDis <- 0
for(j in 1:2) {
    p.val.train <- wilcox.test(dat.expr.full2390.testtop2[dat.expr.full2390.testtop2$Data == "Train", ][[j]] ~ dat.expr.full2390.testtop2[dat.expr.full2390.testtop2$Data == "Train", ][["BMI"]], data = dat.expr.full2390.testtop2)$p.value
    p.val.test <- wilcox.test(dat.expr.full2390.testtop2[dat.expr.full2390.testtop2$Data == "Test", ][[j]] ~ dat.expr.full2390.testtop2[dat.expr.full2390.testtop2$Data == "Test", ][["BMI"]], data = dat.expr.full2390.testtop2)$p.value
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1}
}
    
if ((!is.na(sumDis)) & (sumDis > 0)) {
  boxplot.full2390.testtop2 <- names(dat.expr.full2390.testtop2)[-match(c("Data", "BMI"), names(dat.expr.full2390.testtop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full2390.testtop2, var.x = "BMI")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.testtop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/BMI/TestTop2000/Discordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
} else {
  boxplot.full2390.testtop2 <- names(dat.expr.full2390.testtop2)[-match(c("Data", "BMI"), names(dat.expr.full2390.testtop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full2390.testtop2, var.x = "BMI")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.testtop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/BMI/TestTop2000/Concordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
}
```




### AGE
#### Top 2000 genes list from the training set
```{r}
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt2390.RData")
AGE <- c(dat.pheno.train[["AGE"]], dat.pheno.test[["AGE"]])

seed.num <- 2390
date.analysis <- format(Sys.Date(), "%Y%b%d")

##------Prepare plotting function------
boxplot.fun.2 <- function(dataset, var.x, var.y) {
  dataset |> ggplot(mapping = aes(y = get(var.y), x = get(var.x))) + 
    geom_boxplot(mapping = aes(color = get(var.x))) +
    geom_jitter(mapping = aes(color = get(var.x))) + 
    stat_compare_means(method = "wilcox.test", 
                       label.x.npc = "right",
                       label.y = 2,
                       mapping = aes(label = ifelse(p < 1.e-3, "p < 0.001", sprintf("p = %4.3f", as.numeric(..p.format..))))) +
    facet_wrap(~factor(Data, levels = c("Train", "Test"))) + 
    labs(y = var.y, 
         x = "") +
    theme_bw(base_size = 14) + 
    scale_color_manual(values = c("#0073C2FF", "#EFC000FF")) + 
    scale_x_discrete(limits = c("young", "old")) + 
    guides(color = "none")
}

for(i in seq(1, 1998, 9)) {
  ##------Process data------
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8)
  dat.expr.train2390.top9 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.train[["AGE"]])[gene.id], ]
  dat.expr.train2390.top9.t <- t(dat.expr.train2390.top9) |> as.data.frame()
dat.expr.train2390.top9.t$Data <- "Train"
  dat.expr.test2390.traintop9 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.train[["AGE"]])[gene.id], ]
  dat.expr.test2390.traintop9.t <- t(dat.expr.test2390.traintop9) |> as.data.frame()
  dat.expr.test2390.traintop9.t$Data <- "Test"
  rownames(dat.expr.train2390.top9.t) = rownames(dat.expr.test2390.traintop9.t) = NULL
  dat.expr.full2390.traintop9 <- rbind(dat.expr.train2390.top9.t,
                                       dat.expr.test2390.traintop9.t)
  
  dat.expr.full2390.traintop9 <- data.frame(dat.expr.full2390.traintop9, AGE)
  sumDis <- 0
  for(j in 1:9) {
    p.val.train <- wilcox.test(dat.expr.full2390.traintop9[dat.expr.full2390.traintop9$Data == "Train", ][[j]] ~ dat.expr.full2390.traintop9[dat.expr.full2390.traintop9$Data == "Train", ][["AGE"]], data = dat.expr.full2390.traintop9)$p.value
    p.val.test <- wilcox.test(dat.expr.full2390.traintop9[dat.expr.full2390.traintop9$Data == "Test", ][[j]] ~ dat.expr.full2390.traintop9[dat.expr.full2390.traintop9$Data == "Test", ][["AGE"]], data = dat.expr.full2390.traintop9)$p.value
    
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1 }
  }
    
  ##------Make boxplots------
  if ((!is.na(sumDis)) & (sumDis > 0)) {
    boxplot.full2390.traintop9 <- names(dat.expr.full2390.traintop9)[-match(c("Data", "AGE"), names(dat.expr.full2390.traintop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full2390.traintop9, var.x = "AGE")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.traintop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/AGE/TrainTop2000/Discordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"), 
           plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  } else {
    boxplot.full2390.traintop9 <- names(dat.expr.full2390.traintop9)[-match(c("Data", "AGE"), names(dat.expr.full2390.traintop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full2390.traintop9, var.x = "AGE")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.traintop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/AGE/TrainTop2000/Concordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
           plot = g, device = "png", width = 16, height = 8, units = "in")
    print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  }
}

gene.id <- c(1999, 2000)
dat.expr.train2390.top2 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.train[["AGE"]])[gene.id], ]
dat.expr.train2390.top2.t <- t(dat.expr.train2390.top2) |> as.data.frame()
dat.expr.train2390.top2.t$Data <- "Train"
dat.expr.test2390.traintop2 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.train[["AGE"]])[gene.id], ]
dat.expr.test2390.traintop2.t <- t(dat.expr.test2390.traintop2) |> as.data.frame()
dat.expr.test2390.traintop2.t$Data <- "Test"
rownames(dat.expr.train2390.top2.t) = rownames(dat.expr.test2390.traintop2.t) = NULL
dat.expr.full2390.traintop2 <- rbind(dat.expr.train2390.top2.t,
                                     dat.expr.test2390.traintop2.t)
dat.expr.full2390.traintop2 <- data.frame(dat.expr.full2390.traintop2, AGE)

sumDis <- 0
for(j in 1:2) {
    p.val.train <- wilcox.test(dat.expr.full2390.traintop2[dat.expr.full2390.traintop2$Data == "Train", ][[j]] ~ dat.expr.full2390.traintop2[dat.expr.full2390.traintop2$Data == "Train", ][["AGE"]], data = dat.expr.full2390.traintop2)$p.value
    p.val.test <- wilcox.test(dat.expr.full2390.traintop2[dat.expr.full2390.traintop2$Data == "Test", ][[j]] ~ dat.expr.full2390.traintop2[dat.expr.full2390.traintop2$Data == "Test", ][["AGE"]], data = dat.expr.full2390.traintop2)$p.value
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1}
}
    
if ((!is.na(sumDis)) & (sumDis > 0)) {
  boxplot.full2390.traintop2 <- names(dat.expr.full2390.traintop2)[-match(c("Data", "AGE"), names(dat.expr.full2390.traintop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full2390.traintop2, var.x = "AGE")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.traintop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/AGE/TrainTop2000/Discordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
} else {
  boxplot.full2390.traintop2 <- names(dat.expr.full2390.traintop2)[-match(c("Data", "AGE"), names(dat.expr.full2390.traintop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full2390.traintop2, var.x = "AGE")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.traintop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/AGE/TrainTop2000/Concordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
}
```


#### Top 2000 genes list from the test set
```{r}
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt2390.RData")
AGE <- c(dat.pheno.train[["AGE"]], dat.pheno.test[["AGE"]])

seed.num <- 2390
date.analysis <- format(Sys.Date(), "%Y%b%d")

##------Prepare plotting function------
boxplot.fun.2 <- function(dataset, var.x, var.y) {
  dataset |> ggplot(mapping = aes(y = get(var.y), x = get(var.x))) + 
    geom_boxplot(mapping = aes(color = get(var.x))) +
    geom_jitter(mapping = aes(color = get(var.x))) + 
    stat_compare_means(method = "wilcox.test", 
                       label.x.npc = "right",
                       label.y = 2,
                       mapping = aes(label = ifelse(p < 1.e-3, "p < 0.001", sprintf("p = %4.3f", as.numeric(..p.format..))))) +
    facet_wrap(~factor(Data, levels = c("Train", "Test"))) + 
    labs(y = var.y, 
         x = "") +
    theme_bw(base_size = 14) + 
    scale_color_manual(values = c("#0073C2FF", "#EFC000FF")) + 
    scale_x_discrete(limits = c("young", "old")) + 
    guides(color = "none")
}

for(i in seq(1, 1998, 9)) {
  ##------Process data------
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8)
  dat.expr.train2390.testtop9 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.test[["AGE"]])[gene.id], ]
  dat.expr.train2390.testtop9.t <- t(dat.expr.train2390.testtop9) |> as.data.frame()
dat.expr.train2390.testtop9.t$Data <- "Train"
  dat.expr.test2390.top9 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.test[["AGE"]])[gene.id], ]
  dat.expr.test2390.top9.t <- t(dat.expr.test2390.top9) |> as.data.frame()
  dat.expr.test2390.top9.t$Data <- "Test"
  rownames(dat.expr.train2390.testtop9.t) = rownames(dat.expr.test2390.top9.t) = NULL
  dat.expr.full2390.testtop9 <- rbind(dat.expr.train2390.testtop9.t,
                                       dat.expr.test2390.top9.t)
  
  dat.expr.full2390.testtop9 <- data.frame(dat.expr.full2390.testtop9, AGE)
  sumDis <- 0
  for(j in 1:9) {
    p.val.train <- wilcox.test(dat.expr.full2390.testtop9[dat.expr.full2390.testtop9$Data == "Train", ][[j]] ~ dat.expr.full2390.testtop9[dat.expr.full2390.testtop9$Data == "Train", ][["AGE"]], data = dat.expr.full2390.testtop9)$p.value
    p.val.test <- wilcox.test(dat.expr.full2390.testtop9[dat.expr.full2390.testtop9$Data == "Test", ][[j]] ~ dat.expr.full2390.testtop9[dat.expr.full2390.testtop9$Data == "Test", ][["AGE"]], data = dat.expr.full2390.testtop9)$p.value
    
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1 }
  }
    
  ##------Make boxplots------
  if ((!is.na(sumDis)) & (sumDis > 0)) {
    boxplot.full2390.testtop9 <- names(dat.expr.full2390.testtop9)[-match(c("Data", "AGE"), names(dat.expr.full2390.testtop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full2390.testtop9, var.x = "AGE")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.testtop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/AGE/TestTop2000/Discordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
           plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  } else {
    boxplot.full2390.testtop9 <- names(dat.expr.full2390.testtop9)[-match(c("Data", "AGE"), names(dat.expr.full2390.testtop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full2390.testtop9, var.x = "AGE")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.testtop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/AGE/TestTop2000/Concordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
           plot = g, device = "png", width = 16, height = 8, units = "in")
    print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  }
}

gene.id <- c(1999, 2000)
dat.expr.train2390.testtop2 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.test[["AGE"]])[gene.id], ]
dat.expr.train2390.testtop2.t <- t(dat.expr.train2390.testtop2) |> as.data.frame()
dat.expr.train2390.testtop2.t$Data <- "Train"
dat.expr.test2390.top2 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.test[["AGE"]])[gene.id], ]
dat.expr.test2390.top2.t <- t(dat.expr.test2390.top2) |> as.data.frame()
dat.expr.test2390.top2.t$Data <- "Test"
rownames(dat.expr.train2390.testtop2.t) = rownames(dat.expr.test2390.top2.t) = NULL
dat.expr.full2390.testtop2 <- rbind(dat.expr.train2390.testtop2.t,
                                     dat.expr.test2390.top2.t)
dat.expr.full2390.testtop2 <- data.frame(dat.expr.full2390.testtop2, AGE)

sumDis <- 0
for(j in 1:2) {
    p.val.train <- wilcox.test(dat.expr.full2390.testtop2[dat.expr.full2390.testtop2$Data == "Train", ][[j]] ~ dat.expr.full2390.testtop2[dat.expr.full2390.testtop2$Data == "Train", ][["AGE"]], data = dat.expr.full2390.testtop2)$p.value
    p.val.test <- wilcox.test(dat.expr.full2390.testtop2[dat.expr.full2390.testtop2$Data == "Test", ][[j]] ~ dat.expr.full2390.testtop2[dat.expr.full2390.testtop2$Data == "Test", ][["AGE"]], data = dat.expr.full2390.testtop2)$p.value
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1}
}
    
if ((!is.na(sumDis)) & (sumDis > 0)) {
  boxplot.full2390.testtop2 <- names(dat.expr.full2390.testtop2)[-match(c("Data", "AGE"), names(dat.expr.full2390.testtop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full2390.testtop2, var.x = "AGE")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.testtop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/AGE/TestTop2000/Discordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
} else {
  boxplot.full2390.testtop2 <- names(dat.expr.full2390.testtop2)[-match(c("Data", "AGE"), names(dat.expr.full2390.testtop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full2390.testtop2, var.x = "AGE")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.testtop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/AGE/TestTop2000/Concordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
}
```


### SEX
#### Top 2000 genes list from the training set
```{r}
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt2390.RData")
SEX <- c(dat.pheno.train[["SEX"]], dat.pheno.test[["SEX"]])

seed.num <- 2390
date.analysis <- format(Sys.Date(), "%Y%b%d")

##------Prepare plotting function------
boxplot.fun.2 <- function(dataset, var.x, var.y) {
  dataset |> ggplot(mapping = aes(y = get(var.y), x = factor(get(var.x), levels = c("male", "female")))) + 
    geom_boxplot(mapping = aes(color = get(var.x))) +
    geom_jitter(mapping = aes(color = get(var.x))) + 
    stat_compare_means(method = "wilcox.test", 
                       label.x.npc = "right",
                       label.y = 2,
                       mapping = aes(label = ifelse(p < 1.e-3, "p < 0.001", sprintf("p = %4.3f", as.numeric(..p.format..))))) +
    facet_wrap(~factor(Data, levels = c("Train", "Test"))) + 
    labs(y = var.y, 
         x = "") +
    theme_bw(base_size = 14) + 
    scale_color_manual(values = c("#0073C2FF", "#EFC000FF")) + 
    scale_x_discrete(limits = c("male", "female")) + 
    guides(color = "none")
}

for(i in seq(1, 1998, 9)) {
  ##------Process data------
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8)
  dat.expr.train2390.top9 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.train[["SEX"]])[gene.id], ]
  dat.expr.train2390.top9.t <- t(dat.expr.train2390.top9) |> as.data.frame()
dat.expr.train2390.top9.t$Data <- "Train"
  dat.expr.test2390.traintop9 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.train[["SEX"]])[gene.id], ]
  dat.expr.test2390.traintop9.t <- t(dat.expr.test2390.traintop9) |> as.data.frame()
  dat.expr.test2390.traintop9.t$Data <- "Test"
  rownames(dat.expr.train2390.top9.t) = rownames(dat.expr.test2390.traintop9.t) = NULL
  dat.expr.full2390.traintop9 <- rbind(dat.expr.train2390.top9.t,
                                       dat.expr.test2390.traintop9.t)
  
  dat.expr.full2390.traintop9 <- data.frame(dat.expr.full2390.traintop9, SEX)
  sumDis <- 0
  for(j in 1:9) {
    p.val.train <- wilcox.test(dat.expr.full2390.traintop9[dat.expr.full2390.traintop9$Data == "Train", ][[j]] ~ dat.expr.full2390.traintop9[dat.expr.full2390.traintop9$Data == "Train", ][["SEX"]], data = dat.expr.full2390.traintop9)$p.value
    p.val.test <- wilcox.test(dat.expr.full2390.traintop9[dat.expr.full2390.traintop9$Data == "Test", ][[j]] ~ dat.expr.full2390.traintop9[dat.expr.full2390.traintop9$Data == "Test", ][["SEX"]], data = dat.expr.full2390.traintop9)$p.value
    
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1 }
  }
    
  ##------Make boxplots------
  if ((!is.na(sumDis)) & (sumDis > 0)) {
    boxplot.full2390.traintop9 <- names(dat.expr.full2390.traintop9)[-match(c("Data", "SEX"), names(dat.expr.full2390.traintop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full2390.traintop9, var.x = "SEX")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.traintop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/SEX/TrainTop2000/Discordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
           plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  } else {
    boxplot.full2390.traintop9 <- names(dat.expr.full2390.traintop9)[-match(c("Data", "SEX"), names(dat.expr.full2390.traintop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full2390.traintop9, var.x = "SEX")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.traintop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/SEX/TrainTop2000/Concordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
           plot = g, device = "png", width = 16, height = 8, units = "in")
    print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  }
}

gene.id <- c(1999, 2000)
dat.expr.train2390.top2 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.train[["SEX"]])[gene.id], ]
dat.expr.train2390.top2.t <- t(dat.expr.train2390.top2) |> as.data.frame()
dat.expr.train2390.top2.t$Data <- "Train"
dat.expr.test2390.traintop2 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.train[["SEX"]])[gene.id], ]
dat.expr.test2390.traintop2.t <- t(dat.expr.test2390.traintop2) |> as.data.frame()
dat.expr.test2390.traintop2.t$Data <- "Test"
rownames(dat.expr.train2390.top2.t) = rownames(dat.expr.test2390.traintop2.t) = NULL
dat.expr.full2390.traintop2 <- rbind(dat.expr.train2390.top2.t,
                                     dat.expr.test2390.traintop2.t)
dat.expr.full2390.traintop2 <- data.frame(dat.expr.full2390.traintop2, SEX)

sumDis <- 0
for(j in 1:2) {
    p.val.train <- wilcox.test(dat.expr.full2390.traintop2[dat.expr.full2390.traintop2$Data == "Train", ][[j]] ~ dat.expr.full2390.traintop2[dat.expr.full2390.traintop2$Data == "Train", ][["SEX"]], data = dat.expr.full2390.traintop2)$p.value
    p.val.test <- wilcox.test(dat.expr.full2390.traintop2[dat.expr.full2390.traintop2$Data == "Test", ][[j]] ~ dat.expr.full2390.traintop2[dat.expr.full2390.traintop2$Data == "Test", ][["SEX"]], data = dat.expr.full2390.traintop2)$p.value
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1}
}
    
if ((!is.na(sumDis)) & (sumDis > 0)) {
  boxplot.full2390.traintop2 <- names(dat.expr.full2390.traintop2)[-match(c("Data", "SEX"), names(dat.expr.full2390.traintop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full2390.traintop2, var.x = "SEX")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.traintop2, ncol = 2)))
  print(paste0("[", i, "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/SEX/TrainTop2000/Discordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
} else {
  boxplot.full2390.traintop2 <- names(dat.expr.full2390.traintop2)[-match(c("Data", "SEX"), names(dat.expr.full2390.traintop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full2390.traintop2, var.x = "SEX")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.traintop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/SEX/TrainTop2000/Concordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
}
```


#### Top 2000 genes list from the test set
```{r}
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt2390.RData")
SEX <- c(dat.pheno.train[["SEX"]], dat.pheno.test[["SEX"]])

seed.num <- 2390
date.analysis <- format(Sys.Date(), "%Y%b%d")

##------Prepare plotting function------
boxplot.fun.2 <- function(dataset, var.x, var.y) {
  dataset |> ggplot(mapping = aes(y = get(var.y), x = factor(get(var.x), levels = c("male", "female")))) + 
    geom_boxplot(mapping = aes(color = get(var.x))) +
    geom_jitter(mapping = aes(color = get(var.x))) + 
    stat_compare_means(method = "wilcox.test", 
                       label.x.npc = "right",
                       label.y = 2,
                       mapping = aes(label = ifelse(p < 1.e-3, "p < 0.001", sprintf("p = %4.3f", as.numeric(..p.format..))))) +
    facet_wrap(~factor(Data, levels = c("Train", "Test"))) + 
    labs(y = var.y, 
         x = "") +
    theme_bw(base_size = 14) + 
    scale_color_manual(values = c("#0073C2FF", "#EFC000FF")) + 
    scale_x_discrete(limits = c("male", "female")) + 
    guides(color = "none")
}

for(i in seq(1, 1998, 9)) {
  ##------Process data------
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8)
  dat.expr.train2390.testtop9 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.test[["SEX"]])[gene.id], ]
  dat.expr.train2390.testtop9.t <- t(dat.expr.train2390.testtop9) |> as.data.frame()
dat.expr.train2390.testtop9.t$Data <- "Train"
  dat.expr.test2390.top9 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.test[["SEX"]])[gene.id], ]
  dat.expr.test2390.top9.t <- t(dat.expr.test2390.top9) |> as.data.frame()
  dat.expr.test2390.top9.t$Data <- "Test"
  rownames(dat.expr.train2390.testtop9.t) = rownames(dat.expr.test2390.top9.t) = NULL
  dat.expr.full2390.testtop9 <- rbind(dat.expr.train2390.testtop9.t,
                                       dat.expr.test2390.top9.t)
  
  dat.expr.full2390.testtop9 <- data.frame(dat.expr.full2390.testtop9, SEX)
  sumDis <- 0
  for(j in 1:9) {
    p.val.train <- wilcox.test(dat.expr.full2390.testtop9[dat.expr.full2390.testtop9$Data == "Train", ][[j]] ~ dat.expr.full2390.testtop9[dat.expr.full2390.testtop9$Data == "Train", ][["SEX"]], data = dat.expr.full2390.testtop9)$p.value
    p.val.test <- wilcox.test(dat.expr.full2390.testtop9[dat.expr.full2390.testtop9$Data == "Test", ][[j]] ~ dat.expr.full2390.testtop9[dat.expr.full2390.testtop9$Data == "Test", ][["SEX"]], data = dat.expr.full2390.testtop9)$p.value
    
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1 }
  }
    
  ##------Make boxplots------
  if ((!is.na(sumDis)) & (sumDis > 0)) {
    boxplot.full2390.testtop9 <- names(dat.expr.full2390.testtop9)[-match(c("Data", "SEX"), names(dat.expr.full2390.testtop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full2390.testtop9, var.x = "SEX")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.testtop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/SEX/TestTop2000/Discordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 16,
         height = 8,
         units = "in")
  print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  } else {
    boxplot.full2390.testtop9 <- names(dat.expr.full2390.testtop9)[-match(c("Data", "SEX"), names(dat.expr.full2390.testtop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full2390.testtop9, var.x = "SEX")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.testtop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/SEX/TestTop2000/Concordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 16,
         height = 8,
         units = "in")
    print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  }
}

gene.id <- c(1999, 2000)
dat.expr.train2390.testtop2 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.test[["SEX"]])[gene.id], ]
dat.expr.train2390.testtop2.t <- t(dat.expr.train2390.testtop2) |> as.data.frame()
dat.expr.train2390.testtop2.t$Data <- "Train"
dat.expr.test2390.top2 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.test[["SEX"]])[gene.id], ]
dat.expr.test2390.top2.t <- t(dat.expr.test2390.top2) |> as.data.frame()
dat.expr.test2390.top2.t$Data <- "Test"
rownames(dat.expr.train2390.testtop2.t) = rownames(dat.expr.test2390.top2.t) = NULL
dat.expr.full2390.testtop2 <- rbind(dat.expr.train2390.testtop2.t,
                                     dat.expr.test2390.top2.t)
dat.expr.full2390.testtop2 <- data.frame(dat.expr.full2390.testtop2, SEX)

sumDis <- 0
for(j in 1:2) {
    p.val.train <- wilcox.test(dat.expr.full2390.testtop2[dat.expr.full2390.testtop2$Data == "Train", ][[j]] ~ dat.expr.full2390.testtop2[dat.expr.full2390.testtop2$Data == "Train", ][["SEX"]], data = dat.expr.full2390.testtop2)$p.value
    p.val.test <- wilcox.test(dat.expr.full2390.testtop2[dat.expr.full2390.testtop2$Data == "Test", ][[j]] ~ dat.expr.full2390.testtop2[dat.expr.full2390.testtop2$Data == "Test", ][["SEX"]], data = dat.expr.full2390.testtop2)$p.value
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1}
}
    
if ((!is.na(sumDis)) & (sumDis > 0)) {
  boxplot.full2390.testtop2 <- names(dat.expr.full2390.testtop2)[-match(c("Data", "SEX"), names(dat.expr.full2390.testtop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full2390.testtop2, var.x = "SEX")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.testtop2, ncol = 2)))
  print(paste0("[", i, "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/SEX/TestTop2000/Discordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 16,
         height = 8,
         units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
} else {
  boxplot.full2390.testtop2 <- names(dat.expr.full2390.testtop2)[-match(c("Data", "SEX"), names(dat.expr.full2390.testtop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full2390.testtop2, var.x = "SEX")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.testtop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/SEX/TestTop2000/Concordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 16,
         height = 8,
         units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
}
```


### MHABNWBC
#### Top 2000 genes list from the training set
```{r}
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt2390.RData")
MHABNWBC <- c(dat.pheno.train[["MHABNWBC"]], dat.pheno.test[["MHABNWBC"]])

seed.num <- 2390
date.analysis <- format(Sys.Date(), "%Y%b%d")

##------Prepare plotting function------
boxplot.fun.2 <- function(dataset, var.x, var.y) {
  dataset |> ggplot(mapping = aes(y = get(var.y), x = factor(get(var.x), levels = c("yes", "no")))) +
    geom_boxplot(mapping = aes(color = get(var.x))) +
    geom_jitter(mapping = aes(color = get(var.x))) + 
    stat_compare_means(method = "wilcox.test",
                       label.x.npc = "right",
                       label.y = 2,
                       mapping = aes(label = ifelse(p < 1.e-3, "p < 0.001", sprintf("p = %4.3f", as.numeric(..p.format..))))) +
    facet_wrap(~factor(Data, levels = c("Train", "Test"))) + 
    labs(y = var.y, 
         x = "") +
    theme_bw(base_size = 14) + 
    scale_color_manual(values = c("#0073C2FF", "#EFC000FF")) + 
    scale_x_discrete(limits = c("yes", "no")) + 
    guides(color = "none")
}

for(i in seq(1, 1998, 9)) {
  ##------Process data------
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8)
  dat.expr.train2390.top9 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.train[["MHABNWBC"]])[gene.id], ]
  dat.expr.train2390.top9.t <- t(dat.expr.train2390.top9) |> as.data.frame()
dat.expr.train2390.top9.t$Data <- "Train"
  dat.expr.test2390.traintop9 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.train[["MHABNWBC"]])[gene.id], ]
  dat.expr.test2390.traintop9.t <- t(dat.expr.test2390.traintop9) |> as.data.frame()
  dat.expr.test2390.traintop9.t$Data <- "Test"
  rownames(dat.expr.train2390.top9.t) = rownames(dat.expr.test2390.traintop9.t) = NULL
  dat.expr.full2390.traintop9 <- rbind(dat.expr.train2390.top9.t,
                                       dat.expr.test2390.traintop9.t)
  
  dat.expr.full2390.traintop9 <- data.frame(dat.expr.full2390.traintop9, MHABNWBC)
  sumDis <- 0
  for(j in 1:9) {
    p.val.train <- wilcox.test(dat.expr.full2390.traintop9[dat.expr.full2390.traintop9$Data == "Train", ][[j]] ~ dat.expr.full2390.traintop9[dat.expr.full2390.traintop9$Data == "Train", ][["MHABNWBC"]], data = dat.expr.full2390.traintop9)$p.value
    p.val.test <- wilcox.test(dat.expr.full2390.traintop9[dat.expr.full2390.traintop9$Data == "Test", ][[j]] ~ dat.expr.full2390.traintop9[dat.expr.full2390.traintop9$Data == "Test", ][["MHABNWBC"]], data = dat.expr.full2390.traintop9)$p.value
    
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1 }
  }
    
  ##------Make boxplots------
  if ((!is.na(sumDis)) & (sumDis > 0)) {
    boxplot.full2390.traintop9 <- names(dat.expr.full2390.traintop9)[-match(c("Data", "MHABNWBC"), names(dat.expr.full2390.traintop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full2390.traintop9, var.x = "MHABNWBC")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.traintop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/MHABNWBC/TrainTop2000/Discordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 16,
         height = 8,
         units = "in")
  print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  } else {
    boxplot.full2390.traintop9 <- names(dat.expr.full2390.traintop9)[-match(c("Data", "MHABNWBC"), names(dat.expr.full2390.traintop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full2390.traintop9, var.x = "MHABNWBC")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.traintop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/MHABNWBC/TrainTop2000/Concordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 16,
         height = 8,
         units = "in")
    print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  }
}

gene.id <- c(1999, 2000)
dat.expr.train2390.top2 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.train[["MHABNWBC"]])[gene.id], ]
dat.expr.train2390.top2.t <- t(dat.expr.train2390.top2) |> as.data.frame()
dat.expr.train2390.top2.t$Data <- "Train"
dat.expr.test2390.traintop2 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.train[["MHABNWBC"]])[gene.id], ]
dat.expr.test2390.traintop2.t <- t(dat.expr.test2390.traintop2) |> as.data.frame()
dat.expr.test2390.traintop2.t$Data <- "Test"
rownames(dat.expr.train2390.top2.t) = rownames(dat.expr.test2390.traintop2.t) = NULL
dat.expr.full2390.traintop2 <- rbind(dat.expr.train2390.top2.t,
                                     dat.expr.test2390.traintop2.t)
dat.expr.full2390.traintop2 <- data.frame(dat.expr.full2390.traintop2, MHABNWBC)

sumDis <- 0
for(j in 1:2) {
    p.val.train <- wilcox.test(dat.expr.full2390.traintop2[dat.expr.full2390.traintop2$Data == "Train", ][[j]] ~ dat.expr.full2390.traintop2[dat.expr.full2390.traintop2$Data == "Train", ][["MHABNWBC"]], data = dat.expr.full2390.traintop2)$p.value
    p.val.test <- wilcox.test(dat.expr.full2390.traintop2[dat.expr.full2390.traintop2$Data == "Test", ][[j]] ~ dat.expr.full2390.traintop2[dat.expr.full2390.traintop2$Data == "Test", ][["MHABNWBC"]], data = dat.expr.full2390.traintop2)$p.value
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1}
}
    
if ((!is.na(sumDis)) & (sumDis > 0)) {
  boxplot.full2390.traintop2 <- names(dat.expr.full2390.traintop2)[-match(c("Data", "MHABNWBC"), names(dat.expr.full2390.traintop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full2390.traintop2, var.x = "MHABNWBC")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.traintop2, ncol = 2)))
  print(paste0("[", i, "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/MHABNWBC/TrainTop2000/Discordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 16,
         height = 8,
         units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
} else {
  boxplot.full2390.traintop2 <- names(dat.expr.full2390.traintop2)[-match(c("Data", "MHABNWBC"), names(dat.expr.full2390.traintop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full2390.traintop2, var.x = "MHABNWBC")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.traintop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/MHABNWBC/TrainTop2000/Concordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 16,
         height = 8,
         units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
}
```


#### Top 2000 genes list from the test set
```{r}
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt2390.RData")
MHABNWBC <- c(dat.pheno.train[["MHABNWBC"]], dat.pheno.test[["MHABNWBC"]])

seed.num <- 2390
date.analysis <- format(Sys.Date(), "%Y%b%d")

##------Prepare plotting function------
boxplot.fun.2 <- function(dataset, var.x, var.y) {
  dataset |> ggplot(mapping = aes(y = get(var.y), x = factor(get(var.x), levels = c("yes", "no")))) + 
    geom_boxplot(mapping = aes(color = get(var.x))) +
    geom_jitter(mapping = aes(color = get(var.x))) + 
    stat_compare_means(method = "wilcox.test", 
                       label.x.npc = "right",
                       label.y = 2,
                       mapping = aes(label = ifelse(p < 1.e-3, "p < 0.001", sprintf("p = %4.3f", as.numeric(..p.format..))))) +
    facet_wrap(~factor(Data, levels = c("Train", "Test"))) + 
    labs(y = var.y, 
         x = "") +
    theme_bw(base_size = 14) + 
    scale_color_manual(values = c("#0073C2FF", "#EFC000FF")) + 
    scale_x_discrete(limits = c("yes", "no")) + 
    guides(color = "none")
}

for(i in seq(1, 1998, 9)) {
  ##------Process data------
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8)
  dat.expr.train2390.testtop9 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.test[["MHABNWBC"]])[gene.id], ]
  dat.expr.train2390.testtop9.t <- t(dat.expr.train2390.testtop9) |> as.data.frame()
dat.expr.train2390.testtop9.t$Data <- "Train"
  dat.expr.test2390.top9 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.test[["MHABNWBC"]])[gene.id], ]
  dat.expr.test2390.top9.t <- t(dat.expr.test2390.top9) |> as.data.frame()
  dat.expr.test2390.top9.t$Data <- "Test"
  rownames(dat.expr.train2390.testtop9.t) = rownames(dat.expr.test2390.top9.t) = NULL
  dat.expr.full2390.testtop9 <- rbind(dat.expr.train2390.testtop9.t,
                                      dat.expr.test2390.top9.t)
  
  dat.expr.full2390.testtop9 <- data.frame(dat.expr.full2390.testtop9, MHABNWBC)
  sumDis <- 0
  for(j in 1:9) {
    p.val.train <- wilcox.test(dat.expr.full2390.testtop9[dat.expr.full2390.testtop9$Data == "Train", ][[j]] ~ dat.expr.full2390.testtop9[dat.expr.full2390.testtop9$Data == "Train", ][["MHABNWBC"]], data = dat.expr.full2390.testtop9)$p.value
    p.val.test <- wilcox.test(dat.expr.full2390.testtop9[dat.expr.full2390.testtop9$Data == "Test", ][[j]] ~ dat.expr.full2390.testtop9[dat.expr.full2390.testtop9$Data == "Test", ][["MHABNWBC"]], data = dat.expr.full2390.testtop9)$p.value
    
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1 }
  }
    
  ##------Make boxplots------
  if ((!is.na(sumDis)) & (sumDis > 0)) {
    boxplot.full2390.testtop9 <- names(dat.expr.full2390.testtop9)[-match(c("Data", "MHABNWBC"), names(dat.expr.full2390.testtop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full2390.testtop9, var.x = "MHABNWBC")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.testtop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/MHABNWBC/TestTop2000/Discordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
           plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  } else {
    boxplot.full2390.testtop9 <- names(dat.expr.full2390.testtop9)[-match(c("Data", "MHABNWBC"), names(dat.expr.full2390.testtop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full2390.testtop9, var.x = "MHABNWBC")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.testtop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/MHABNWBC/TestTop2000/Concordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
           plot = g, device = "png", width = 16, height = 8, units = "in")
    print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  }
}

gene.id <- c(1999, 2000)
dat.expr.train2390.testtop2 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.test[["MHABNWBC"]])[gene.id], ]
dat.expr.train2390.testtop2.t <- t(dat.expr.train2390.testtop2) |> as.data.frame()
dat.expr.train2390.testtop2.t$Data <- "Train"
dat.expr.test2390.top2 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.test[["MHABNWBC"]])[gene.id], ]
dat.expr.test2390.top2.t <- t(dat.expr.test2390.top2) |> as.data.frame()
dat.expr.test2390.top2.t$Data <- "Test"
rownames(dat.expr.train2390.testtop2.t) = rownames(dat.expr.test2390.top2.t) = NULL
dat.expr.full2390.testtop2 <- rbind(dat.expr.train2390.testtop2.t,
                                     dat.expr.test2390.top2.t)
dat.expr.full2390.testtop2 <- data.frame(dat.expr.full2390.testtop2, MHABNWBC)

sumDis <- 0
for(j in 1:2) {
    p.val.train <- wilcox.test(dat.expr.full2390.testtop2[dat.expr.full2390.testtop2$Data == "Train", ][[j]] ~ dat.expr.full2390.testtop2[dat.expr.full2390.testtop2$Data == "Train", ][["MHABNWBC"]], data = dat.expr.full2390.testtop2)$p.value
    p.val.test <- wilcox.test(dat.expr.full2390.testtop2[dat.expr.full2390.testtop2$Data == "Test", ][[j]] ~ dat.expr.full2390.testtop2[dat.expr.full2390.testtop2$Data == "Test", ][["MHABNWBC"]], data = dat.expr.full2390.testtop2)$p.value
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1}
}
    
if ((!is.na(sumDis)) & (sumDis > 0)) {
  boxplot.full2390.testtop2 <- names(dat.expr.full2390.testtop2)[-match(c("Data", "MHABNWBC"), names(dat.expr.full2390.testtop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full2390.testtop2, var.x = "MHABNWBC")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.testtop2, ncol = 2)))
  print(paste0("[", i, "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/MHABNWBC/TestTop2000/Discordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
} else {
  boxplot.full2390.testtop2 <- names(dat.expr.full2390.testtop2)[-match(c("Data", "MHABNWBC"), names(dat.expr.full2390.testtop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full2390.testtop2, var.x = "MHABNWBC")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.testtop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/MHABNWBC/TestTop2000/Concordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
}
```


### BMI X SEX
#### Top 2000 genes list from the training set
```{r}
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt2390.RData")
BMI.0 <- c(dat.pheno.train[["BMI"]], dat.pheno.test[["BMI"]])
SEX.0 <- c(dat.pheno.train[["SEX"]], dat.pheno.test[["SEX"]])
# expand.grid(BMI = c("low", "high"), SEX = c("male", "female"))
BMI <- dplyr::case_when(BMI.0 == "low" & SEX.0 == "male" ~ "LowMale", 
                        BMI.0 == "high" & SEX.0 == "male" ~ "HighMale",
                        BMI.0 == "low" & SEX.0 == "female" ~ "LowFemale", 
                        BMI.0 == "high" & SEX.0 == "female" ~ "HighFemale")

seed.num <- 2390
date.analysis <- format(Sys.Date(), "%Y%b%d")


##------Prepare plotting function------
boxplot.fun.2 <- function(dataset, var.x, var.y) {
  dataset |> ggplot(mapping = aes(y = get(var.y), x = factor(get(var.x), levels = c("LowMale", "LowFemale", "HighMale", "HighFemale")))) + 
    geom_boxplot(mapping = aes(color = get(var.x))) +
    geom_jitter(mapping = aes(color = get(var.x))) + 
    stat_compare_means(method = "kruskal.test", 
                       label.x.npc = "center",
                       label.y = 2,
                       mapping = aes(label = ifelse(p < 1.e-3, "p < 0.001", sprintf("p = %4.3f", as.numeric(..p.format..))))) +
    facet_wrap(~factor(Data, levels = c("Train", "Test"))) + 
    labs(y = var.y, 
         x = "") +
    theme_bw(base_size = 14) + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1, color = "black", vjust = 0.5)) + 
    scale_color_manual(values = c("#0073C2FF", "#EFC000FF", "#868686FF", "#CD534CFF")) + 
    scale_x_discrete(limits = c("LowMale", "LowFemale", "HighMale", "HighFemale")) + 
    guides(color = "none")
}


for(i in seq(1, 1998, 9)) {
  ##------Process data------
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8)
  dat.expr.train2390.top9 <- dat.expr.train[names(BMAseq.eFDR.Interaction.2000.train[["BMI"]])[gene.id], ]
  dat.expr.train2390.top9.t <- t(dat.expr.train2390.top9) |> as.data.frame()
dat.expr.train2390.top9.t$Data <- "Train"
  dat.expr.test2390.traintop9 <- dat.expr.test[names(BMAseq.eFDR.Interaction.2000.train[["BMI"]])[gene.id], ]
  dat.expr.test2390.traintop9.t <- t(dat.expr.test2390.traintop9) |> as.data.frame()
  dat.expr.test2390.traintop9.t$Data <- "Test"
  rownames(dat.expr.train2390.top9.t) = rownames(dat.expr.test2390.traintop9.t) = NULL
  dat.expr.full2390.traintop9 <- rbind(dat.expr.train2390.top9.t,
                                       dat.expr.test2390.traintop9.t)
  
  dat.expr.full2390.traintop9 <- data.frame(dat.expr.full2390.traintop9, BMI)
  sumDis <- 0
  for(j in 1:9) {
    p.val.train <- kruskal.test(dat.expr.full2390.traintop9[dat.expr.full2390.traintop9$Data == "Train", ][[j]] ~ dat.expr.full2390.traintop9[dat.expr.full2390.traintop9$Data == "Train", ][["BMI"]], data = dat.expr.full2390.traintop9)$p.value
    p.val.test <- kruskal.test(dat.expr.full2390.traintop9[dat.expr.full2390.traintop9$Data == "Test", ][[j]] ~ dat.expr.full2390.traintop9[dat.expr.full2390.traintop9$Data == "Test", ][["BMI"]], data = dat.expr.full2390.traintop9)$p.value
    
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1 }
  }
    
  ##------Make boxplots------
  if ((!is.na(sumDis)) & (sumDis > 0)) {
    boxplot.full2390.traintop9 <- names(dat.expr.full2390.traintop9)[-match(c("Data", "BMI"), names(dat.expr.full2390.traintop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full2390.traintop9, var.x = "BMI")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.traintop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/BMISEX/TrainTop2000/Discordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
           plot = g, device = "png", width = 16, height = 10, units = "in")
  print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  } else {
    boxplot.full2390.traintop9 <- names(dat.expr.full2390.traintop9)[-match(c("Data", "BMI"), names(dat.expr.full2390.traintop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full2390.traintop9, var.x = "BMI")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.traintop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/BMISEX/TrainTop2000/Concordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
           plot = g, device = "png", width = 16, height = 10, units = "in")
    print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  }
}

gene.id <- c(1999, 2000)
dat.expr.train2390.top2 <- dat.expr.train[names(BMAseq.eFDR.Interaction.2000.train[["BMI"]])[gene.id], ]
dat.expr.train2390.top2.t <- t(dat.expr.train2390.top2) |> as.data.frame()
dat.expr.train2390.top2.t$Data <- "Train"
dat.expr.test2390.traintop2 <- dat.expr.test[names(BMAseq.eFDR.Interaction.2000.train[["BMI"]])[gene.id], ]
dat.expr.test2390.traintop2.t <- t(dat.expr.test2390.traintop2) |> as.data.frame()
dat.expr.test2390.traintop2.t$Data <- "Test"
rownames(dat.expr.train2390.top2.t) = rownames(dat.expr.test2390.traintop2.t) = NULL
dat.expr.full2390.traintop2 <- rbind(dat.expr.train2390.top2.t,
                                     dat.expr.test2390.traintop2.t)
dat.expr.full2390.traintop2 <- data.frame(dat.expr.full2390.traintop2, BMI)

sumDis <- 0
for(j in 1:2) {
    p.val.train <- kruskal.test(dat.expr.full2390.traintop2[dat.expr.full2390.traintop2$Data == "Train", ][[j]] ~ dat.expr.full2390.traintop2[dat.expr.full2390.traintop2$Data == "Train", ][["BMI"]], data = dat.expr.full2390.traintop2)$p.value
    p.val.test <- kruskal.test(dat.expr.full2390.traintop2[dat.expr.full2390.traintop2$Data == "Test", ][[j]] ~ dat.expr.full2390.traintop2[dat.expr.full2390.traintop2$Data == "Test", ][["BMI"]], data = dat.expr.full2390.traintop2)$p.value
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1}
}
    
if ((!is.na(sumDis)) & (sumDis > 0)) {
  boxplot.full2390.traintop2 <- names(dat.expr.full2390.traintop2)[-match(c("Data", "BMI"), names(dat.expr.full2390.traintop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full2390.traintop2, var.x = "BMI")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.traintop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/BMISEX/TrainTop2000/Discordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 10, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
} else {
  boxplot.full2390.traintop2 <- names(dat.expr.full2390.traintop2)[-match(c("Data", "BMI"), names(dat.expr.full2390.traintop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full2390.traintop2, var.x = "BMI")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.traintop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/BMISEX/TrainTop2000/Concordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 10, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
}
```


#### Top 2000 genes list from the test set
```{r}
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt2390.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt2390.RData")
BMI.0 <- c(dat.pheno.train[["BMI"]], dat.pheno.test[["BMI"]])
SEX.0 <- c(dat.pheno.train[["SEX"]], dat.pheno.test[["SEX"]])
BMI <- dplyr::case_when(BMI.0 == "low" & SEX.0 == "male" ~ "LowMale", 
                        BMI.0 == "high" & SEX.0 == "male" ~ "HighMale",
                        BMI.0 == "low" & SEX.0 == "female" ~ "LowFemale", 
                        BMI.0 == "high" & SEX.0 == "female" ~ "HighFemale")

seed.num <- 2390
date.analysis <- format(Sys.Date(), "%Y%b%d")

##------Prepare plotting function------
boxplot.fun.2 <- function(dataset, var.x, var.y) {
  dataset |> ggplot(mapping = aes(y = get(var.y), x = factor(get(var.x), levels = c("LowMale", "LowFemale", "HighMale", "HighFemale")))) + 
    geom_boxplot(mapping = aes(color = get(var.x))) +
    geom_jitter(mapping = aes(color = get(var.x))) + 
    stat_compare_means(method = "kruskal.test", 
                       label.x.npc = "center",
                       label.y = 2,
                       mapping = aes(label = ifelse(p < 1.e-3, "p < 0.001", sprintf("p = %4.3f", as.numeric(..p.format..))))) +
    facet_wrap(~factor(Data, levels = c("Train", "Test"))) + 
    labs(y = var.y, 
         x = "") +
    theme_bw(base_size = 14) + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1, color = "black", vjust = 0.5)) + 
    scale_color_manual(values = c("#0073C2FF", "#EFC000FF", "#868686FF", "#CD534CFF")) + 
    scale_x_discrete(limits = c("LowMale", "LowFemale", "HighMale", "HighFemale")) + 
    guides(color = "none")
}

for(i in seq(1, 1998, 9)) {
  ##------Process data------
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8)
  dat.expr.train2390.testtop9 <- dat.expr.train[names(BMAseq.eFDR.Interaction.2000.test[["BMI"]])[gene.id], ]
  dat.expr.train2390.testtop9.t <- t(dat.expr.train2390.testtop9) |> as.data.frame()
dat.expr.train2390.testtop9.t$Data <- "Train"
  dat.expr.test2390.top9 <- dat.expr.test[names(BMAseq.eFDR.Interaction.2000.test[["BMI"]])[gene.id], ]
  dat.expr.test2390.top9.t <- t(dat.expr.test2390.top9) |> as.data.frame()
  dat.expr.test2390.top9.t$Data <- "Test"
  rownames(dat.expr.train2390.testtop9.t) = rownames(dat.expr.test2390.top9.t) = NULL
  dat.expr.full2390.testtop9 <- rbind(dat.expr.train2390.testtop9.t,
                                      dat.expr.test2390.top9.t)
  
  dat.expr.full2390.testtop9 <- data.frame(dat.expr.full2390.testtop9, BMI)
  sumDis <- 0
  for(j in 1:9) {
    p.val.train <- kruskal.test(dat.expr.full2390.testtop9[dat.expr.full2390.testtop9$Data == "Train", ][[j]] ~ dat.expr.full2390.testtop9[dat.expr.full2390.testtop9$Data == "Train", ][["BMI"]], data = dat.expr.full2390.testtop9)$p.value
    p.val.test <- kruskal.test(dat.expr.full2390.testtop9[dat.expr.full2390.testtop9$Data == "Test", ][[j]] ~ dat.expr.full2390.testtop9[dat.expr.full2390.testtop9$Data == "Test", ][["BMI"]], data = dat.expr.full2390.testtop9)$p.value
    
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1 }
  }
    
  ##------Make boxplots------
  if ((!is.na(sumDis)) & (sumDis > 0)) {
    boxplot.full2390.testtop9 <- names(dat.expr.full2390.testtop9)[-match(c("Data", "BMI"), names(dat.expr.full2390.testtop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full2390.testtop9, var.x = "BMI")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.testtop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/BMISEX/TestTop2000/Discordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 10, units = "in")
  print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  } else {
    boxplot.full2390.testtop9 <- names(dat.expr.full2390.testtop9)[-match(c("Data", "BMI"), names(dat.expr.full2390.testtop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full2390.testtop9, var.x = "BMI")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.testtop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/BMISEX/TestTop2000/Concordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 10, units = "in")
    print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  }
}

gene.id <- c(1999, 2000)
dat.expr.train2390.testtop2 <- dat.expr.train[names(BMAseq.eFDR.Interaction.2000.test[["BMI"]])[gene.id], ]
dat.expr.train2390.testtop2.t <- t(dat.expr.train2390.testtop2) |> as.data.frame()
dat.expr.train2390.testtop2.t$Data <- "Train"
dat.expr.test2390.top2 <- dat.expr.test[names(BMAseq.eFDR.Interaction.2000.test[["BMI"]])[gene.id], ]
dat.expr.test2390.top2.t <- t(dat.expr.test2390.top2) |> as.data.frame()
dat.expr.test2390.top2.t$Data <- "Test"
rownames(dat.expr.train2390.testtop2.t) = rownames(dat.expr.test2390.top2.t) = NULL
dat.expr.full2390.testtop2 <- rbind(dat.expr.train2390.testtop2.t,
                                     dat.expr.test2390.top2.t)
dat.expr.full2390.testtop2 <- data.frame(dat.expr.full2390.testtop2, BMI)

sumDis <- 0
for(j in 1:2) {
    p.val.train <- kruskal.test(dat.expr.full2390.testtop2[dat.expr.full2390.testtop2$Data == "Train", ][[j]] ~ dat.expr.full2390.testtop2[dat.expr.full2390.testtop2$Data == "Train", ][["BMI"]], data = dat.expr.full2390.testtop2)$p.value
    p.val.test <- kruskal.test(dat.expr.full2390.testtop2[dat.expr.full2390.testtop2$Data == "Test", ][[j]] ~ dat.expr.full2390.testtop2[dat.expr.full2390.testtop2$Data == "Test", ][["BMI"]], data = dat.expr.full2390.testtop2)$p.value
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1}
}
    
if ((!is.na(sumDis)) & (sumDis > 0)) {
  boxplot.full2390.testtop2 <- names(dat.expr.full2390.testtop2)[-match(c("Data", "BMI"), names(dat.expr.full2390.testtop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full2390.testtop2, var.x = "BMI")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.testtop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/BMISEX/TestTop2000/Discordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 10, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
} else {
  boxplot.full2390.testtop2 <- names(dat.expr.full2390.testtop2)[-match(c("Data", "BMI"), names(dat.expr.full2390.testtop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full2390.testtop2, var.x = "BMI")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full2390.testtop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/2390/BMISEX/TestTop2000/Concordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 10, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
}
```


## Seed: 98907
### Top 2000 genes list (associated with BMI) from the training set
```{r}
##------Prepare data------
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt98907.RData")
seed.num <- 98907
date.analysis <- format(Sys.Date(), "%Y%b%d")

# names(BMAseq.eFDR.Main.2000.train[["BMI"]][1:10]) # ENSG IDs of top ranked 10 genes based on the eFDR from the smallest to the largest in the training set


for(i in seq(1, 1998, 9)) {
  ##------Process data------
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8)
  dat.expr.train98907.top9 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.train[["BMI"]])[gene.id], ]
  dat.expr.train98907.top9.t <- t(dat.expr.train98907.top9) |> as.data.frame()
dat.expr.train98907.top9.t$Data <- "Train"
  dat.expr.test98907.traintop9 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.train[["BMI"]])[gene.id], ]
  dat.expr.test98907.traintop9.t <- t(dat.expr.test98907.traintop9) |> as.data.frame()
  dat.expr.test98907.traintop9.t$Data <- "Test"
  rownames(dat.expr.train98907.top9.t) = rownames(dat.expr.test98907.traintop9.t) = NULL
  dat.expr.full98907.traintop9 <- rbind(dat.expr.train98907.top9.t,
                                      dat.expr.test98907.traintop9.t)
  p.vec <- NULL
  for(j in 1:9) {
    p.val <- wilcox.test(dat.expr.full98907.traintop9[[j]] ~ dat.expr.full98907.traintop9[["Data"]], data = dat.expr.full98907.traintop9)$p.value
    p.vec <- c(p.vec, p.val)
  }
    
  ##------Make boxplots------
  if (any(!is.na(p.vec) & p.vec < 0.05)) {
    boxplot.full98907.traintop9 <- names(dat.expr.full98907.traintop9)[-match("Data", names(dat.expr.full98907.traintop9))] |> 
    map(.f = boxplot.fun, dataset = dat.expr.full98907.traintop9, var.x = "Data")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.traintop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/Significant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 16,
         height = 8,
         units = "in")
  print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  } else {
    boxplot.full98907.traintop9 <- names(dat.expr.full98907.traintop9)[-match("Data", names(dat.expr.full98907.traintop9))] |> 
    map(.f = boxplot.fun, dataset = dat.expr.full98907.traintop9, var.x = "Data")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.traintop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/Insignificant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 16,
         height = 8,
         units = "in")
    print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  }
}

gene.id <- c(1999, 2000)
dat.expr.train98907.top2 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.train[["BMI"]])[gene.id], ]
dat.expr.train98907.top2.t <- t(dat.expr.train98907.top2) |> as.data.frame()
dat.expr.train98907.top2.t$Data <- "Train"
dat.expr.test98907.traintop2 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.train[["BMI"]])[gene.id], ]
dat.expr.test98907.traintop2.t <- t(dat.expr.test98907.traintop2) |> as.data.frame()
dat.expr.test98907.traintop2.t$Data <- "Test"
rownames(dat.expr.train98907.top2.t) = rownames(dat.expr.test98907.traintop2.t) = NULL
dat.expr.full98907.traintop2 <- rbind(dat.expr.train98907.top2.t,
                                     dat.expr.test98907.traintop2.t)

p.vec <- NULL
for(j in 1:2) {
  p.val <- wilcox.test(dat.expr.full98907.traintop2[[j]] ~ dat.expr.full98907.traintop2[["Data"]], data = dat.expr.full98907.traintop2)$p.value
  p.vec <- c(p.vec, p.val)
}
    
if (any(!is.na(p.vec) & p.vec < 0.05)) {
  boxplot.full98907.traintop2 <- names(dat.expr.full98907.traintop2)[-match("Data", names(dat.expr.full98907.traintop2))] |> 
  map(.f = boxplot.fun, dataset = dat.expr.full98907.traintop2, var.x = "Data")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.traintop2, ncol = 2)))
  print(paste0("[", i, "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/Significant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 16,
         height = 8,
         units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
} else {
  boxplot.full98907.traintop2 <- names(dat.expr.full98907.traintop2)[-match("Data", names(dat.expr.full98907.traintop2))] |> 
  map(.f = boxplot.fun, dataset = dat.expr.full98907.traintop2, var.x = "Data")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.traintop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/Insignificant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 16,
         height = 8,
         units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
}
```


### Top 2000 genes list (associated with BMI) from the test set
```{r}
##------Prepare data------
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt98907.RData")
seed.num <- 98907
date.analysis <- format(Sys.Date(), "%Y%b%d")

for(i in seq(1, 1998, 9)) {
  ##------Process data------
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8)
  dat.expr.train98907.testtop9 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.test[["BMI"]])[gene.id], ]
  dat.expr.train98907.testtop9.t <- t(dat.expr.train98907.testtop9) |> as.data.frame()
  dat.expr.train98907.testtop9.t$Data <- "Train"
  dat.expr.test98907.top9 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.test[["BMI"]])[gene.id], ]
  dat.expr.test98907.top9.t <- t(dat.expr.test98907.top9) |> as.data.frame()
  dat.expr.test98907.top9.t$Data <- "Test"
  rownames(dat.expr.train98907.testtop9.t) = rownames(dat.expr.test98907.top9.t) = NULL
  dat.expr.full98907.testtop9 <- rbind(dat.expr.train98907.testtop9.t,
                                      dat.expr.test98907.top9.t)
  p.vec <- NULL
  for(j in 1:9) {
    p.val <- wilcox.test(dat.expr.full98907.testtop9[[j]] ~ dat.expr.full98907.testtop9[["Data"]], data = dat.expr.full98907.testtop9)$p.value
    p.vec <- c(p.vec, p.val)
  }
    
  ##------Make boxplots------
  if (any(!is.na(p.vec) & p.vec < 0.05)) {
    boxplot.full98907.testtop9 <- names(dat.expr.full98907.testtop9)[-match("Data", names(dat.expr.full98907.testtop9))] |> 
    map(.f = boxplot.fun, dataset = dat.expr.full98907.testtop9, var.x = "Data")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.testtop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/TestTop2000/Significant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 16,
         height = 8,
         units = "in")
  print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  } else {
    boxplot.full98907.testtop9 <- names(dat.expr.full98907.testtop9)[-match("Data", names(dat.expr.full98907.testtop9))] |> 
    map(.f = boxplot.fun, dataset = dat.expr.full98907.testtop9, var.x = "Data")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.testtop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/TestTop2000/Insignificant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 16,
         height = 8,
         units = "in")
    print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  }
}

gene.id <- c(1999, 2000)
dat.expr.train98907.testtop2 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.test[["BMI"]])[gene.id], ]
dat.expr.train98907.testtop2.t <- t(dat.expr.train98907.testtop2) |> as.data.frame()
dat.expr.train98907.testtop2.t$Data <- "Train"
dat.expr.test98907.top2 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.test[["BMI"]])[gene.id], ]
dat.expr.test98907.top2.t <- t(dat.expr.test98907.top2) |> as.data.frame()
dat.expr.test98907.top2.t$Data <- "Test"
rownames(dat.expr.train98907.testtop2.t) = rownames(dat.expr.test98907.top2.t) = NULL
dat.expr.full98907.testtop2 <- rbind(dat.expr.train98907.testtop2.t,
                                    dat.expr.test98907.top2.t)

p.vec <- NULL
for(j in 1:2) {
  p.val <- wilcox.test(dat.expr.full98907.testtop2[[j]] ~ dat.expr.full98907.testtop2[["Data"]], data = dat.expr.full98907.testtop2)$p.value
  p.vec <- c(p.vec, p.val)
}
    
if (any(!is.na(p.vec) & p.vec < 0.05)) {
  boxplot.full98907.testtop2 <- names(dat.expr.full98907.testtop2)[-match("Data", names(dat.expr.full98907.testtop2))] |> 
  map(.f = boxplot.fun, dataset = dat.expr.full98907.testtop2, var.x = "Data")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.testtop2, ncol = 2)))
  print(paste0("[", i, "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/TestTop2000/Significant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 16,
         height = 8,
         units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
} else {
  boxplot.full98907.testtop2 <- names(dat.expr.full98907.testtop2)[-match("Data", names(dat.expr.full98907.testtop2))] |> 
  map(.f = boxplot.fun, dataset = dat.expr.full98907.testtop2, var.x = "Data")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.testtop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/TestTop2000/Insignificant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 16,
         height = 8,
         units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
}
```


### BMI
#### Top 2000 genes list from the training set
```{r}
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt98907.RData")
BMI <- c(dat.pheno.train[["BMI"]], dat.pheno.test[["BMI"]])

seed.num <- 98907
date.analysis <- format(Sys.Date(), "%Y%b%d")

##------Prepare plotting function------
boxplot.fun.2 <- function(dataset, var.x, var.y) {
  dataset |> ggplot(mapping = aes(y = get(var.y), x = get(var.x))) + 
    geom_boxplot(mapping = aes(color = get(var.x))) +
    geom_jitter(mapping = aes(color = get(var.x))) + 
    stat_compare_means(method = "wilcox.test", 
                       label.x.npc = "right",
                       label.y = 2,
                       mapping = aes(label = ifelse(p < 1.e-3, "p < 0.001", sprintf("p = %4.3f", as.numeric(..p.format..))))) +
    facet_wrap(~factor(Data, levels = c("Train", "Test"))) + 
    labs(y = var.y, 
         x = "") +
    theme_bw(base_size = 14) + 
    scale_color_manual(values = c("#0073C2FF", "#EFC000FF")) + 
    scale_x_discrete(limits = c("low", "high")) + 
    guides(color = "none")
}

for(i in seq(1, 1998, 9)) {
  ##------Process data------
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8)
  dat.expr.train98907.top9 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.train[["BMI"]])[gene.id], ]
  dat.expr.train98907.top9.t <- t(dat.expr.train98907.top9) |> as.data.frame()
dat.expr.train98907.top9.t$Data <- "Train"
  dat.expr.test98907.traintop9 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.train[["BMI"]])[gene.id], ]
  dat.expr.test98907.traintop9.t <- t(dat.expr.test98907.traintop9) |> as.data.frame()
  dat.expr.test98907.traintop9.t$Data <- "Test"
  rownames(dat.expr.train98907.top9.t) = rownames(dat.expr.test98907.traintop9.t) = NULL
  dat.expr.full98907.traintop9 <- rbind(dat.expr.train98907.top9.t,
                                       dat.expr.test98907.traintop9.t)
  
  dat.expr.full98907.traintop9 <- data.frame(dat.expr.full98907.traintop9, BMI)
  sumDis <- 0
  for(j in 1:9) {
    p.val.train <- wilcox.test(dat.expr.full98907.traintop9[dat.expr.full98907.traintop9$Data == "Train", ][[j]] ~ dat.expr.full98907.traintop9[dat.expr.full98907.traintop9$Data == "Train", ][["BMI"]], data = dat.expr.full98907.traintop9)$p.value
    p.val.test <- wilcox.test(dat.expr.full98907.traintop9[dat.expr.full98907.traintop9$Data == "Test", ][[j]] ~ dat.expr.full98907.traintop9[dat.expr.full98907.traintop9$Data == "Test", ][["BMI"]], data = dat.expr.full98907.traintop9)$p.value
    
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1 }
  }
    
  ##------Make boxplots------
  if ((!is.na(sumDis)) & (sumDis > 0)) {
    boxplot.full98907.traintop9 <- names(dat.expr.full98907.traintop9)[-match(c("Data", "BMI"), names(dat.expr.full98907.traintop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full98907.traintop9, var.x = "BMI")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.traintop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/BMI/TrainTop2000/Discordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
           plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  } else {
    boxplot.full98907.traintop9 <- names(dat.expr.full98907.traintop9)[-match(c("Data", "BMI"), names(dat.expr.full98907.traintop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full98907.traintop9, var.x = "BMI")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.traintop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/BMI/TrainTop2000/Concordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"), 
           plot = g, device = "png", width = 16, height = 8, units = "in")
    print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  }
}

gene.id <- c(1999, 2000)
dat.expr.train98907.top2 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.train[["BMI"]])[gene.id], ]
dat.expr.train98907.top2.t <- t(dat.expr.train98907.top2) |> as.data.frame()
dat.expr.train98907.top2.t$Data <- "Train"
dat.expr.test98907.traintop2 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.train[["BMI"]])[gene.id], ]
dat.expr.test98907.traintop2.t <- t(dat.expr.test98907.traintop2) |> as.data.frame()
dat.expr.test98907.traintop2.t$Data <- "Test"
rownames(dat.expr.train98907.top2.t) = rownames(dat.expr.test98907.traintop2.t) = NULL
dat.expr.full98907.traintop2 <- rbind(dat.expr.train98907.top2.t,
                                     dat.expr.test98907.traintop2.t)
dat.expr.full98907.traintop2 <- data.frame(dat.expr.full98907.traintop2, BMI)

sumDis <- 0
for(j in 1:2) {
    p.val.train <- wilcox.test(dat.expr.full98907.traintop2[dat.expr.full98907.traintop2$Data == "Train", ][[j]] ~ dat.expr.full98907.traintop2[dat.expr.full98907.traintop2$Data == "Train", ][["BMI"]], data = dat.expr.full98907.traintop2)$p.value
    p.val.test <- wilcox.test(dat.expr.full98907.traintop2[dat.expr.full98907.traintop2$Data == "Test", ][[j]] ~ dat.expr.full98907.traintop2[dat.expr.full98907.traintop2$Data == "Test", ][["BMI"]], data = dat.expr.full98907.traintop2)$p.value
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1}
}
    
if ((!is.na(sumDis)) & (sumDis > 0)) {
  boxplot.full98907.traintop2 <- names(dat.expr.full98907.traintop2)[-match(c("Data", "BMI"), names(dat.expr.full98907.traintop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full98907.traintop2, var.x = "BMI")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.traintop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/BMI/TrainTop2000/Discordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
} else {
  boxplot.full98907.traintop2 <- names(dat.expr.full98907.traintop2)[-match(c("Data", "BMI"), names(dat.expr.full98907.traintop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full98907.traintop2, var.x = "BMI")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.traintop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/BMI/TrainTop2000/Concordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
}
```


#### Top 2000 genes list from the test set
```{r}
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt98907.RData")
BMI <- c(dat.pheno.train[["BMI"]], dat.pheno.test[["BMI"]])

seed.num <- 98907
date.analysis <- format(Sys.Date(), "%Y%b%d")

##------Prepare plotting function------
boxplot.fun.2 <- function(dataset, var.x, var.y) {
  dataset |> ggplot(mapping = aes(y = get(var.y), x = get(var.x))) + 
    geom_boxplot(mapping = aes(color = get(var.x))) +
    geom_jitter(mapping = aes(color = get(var.x))) + 
    stat_compare_means(method = "wilcox.test", 
                       label.x.npc = "right",
                       label.y = 2,
                       mapping = aes(label = ifelse(p < 1.e-3, "p < 0.001", sprintf("p = %4.3f", as.numeric(..p.format..))))) +
    facet_wrap(~factor(Data, levels = c("Train", "Test"))) + 
    labs(y = var.y, 
         x = "") +
    theme_bw(base_size = 14) + 
    scale_color_manual(values = c("#0073C2FF", "#EFC000FF")) + 
    scale_x_discrete(limits = c("low", "high")) + 
    guides(color = "none")
}

for(i in seq(1, 1998, 9)) {
  ##------Process data------
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8)
  dat.expr.train98907.testtop9 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.test[["BMI"]])[gene.id], ]
  dat.expr.train98907.testtop9.t <- t(dat.expr.train98907.testtop9) |> as.data.frame()
dat.expr.train98907.testtop9.t$Data <- "Train"
  dat.expr.test98907.top9 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.test[["BMI"]])[gene.id], ]
  dat.expr.test98907.top9.t <- t(dat.expr.test98907.top9) |> as.data.frame()
  dat.expr.test98907.top9.t$Data <- "Test"
  rownames(dat.expr.train98907.testtop9.t) = rownames(dat.expr.test98907.top9.t) = NULL
  dat.expr.full98907.testtop9 <- rbind(dat.expr.train98907.testtop9.t,
                                       dat.expr.test98907.top9.t)
  
  dat.expr.full98907.testtop9 <- data.frame(dat.expr.full98907.testtop9, BMI)
  sumDis <- 0
  for(j in 1:9) {
    p.val.train <- wilcox.test(dat.expr.full98907.testtop9[dat.expr.full98907.testtop9$Data == "Train", ][[j]] ~ dat.expr.full98907.testtop9[dat.expr.full98907.testtop9$Data == "Train", ][["BMI"]], data = dat.expr.full98907.testtop9)$p.value
    p.val.test <- wilcox.test(dat.expr.full98907.testtop9[dat.expr.full98907.testtop9$Data == "Test", ][[j]] ~ dat.expr.full98907.testtop9[dat.expr.full98907.testtop9$Data == "Test", ][["BMI"]], data = dat.expr.full98907.testtop9)$p.value
    
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1 }
  }
    
  ##------Make boxplots------
  if ((!is.na(sumDis)) & (sumDis > 0)) {
    boxplot.full98907.testtop9 <- names(dat.expr.full98907.testtop9)[-match(c("Data", "BMI"), names(dat.expr.full98907.testtop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full98907.testtop9, var.x = "BMI")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.testtop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/BMI/TestTop2000/Discordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
           plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  } else {
    boxplot.full98907.testtop9 <- names(dat.expr.full98907.testtop9)[-match(c("Data", "BMI"), names(dat.expr.full98907.testtop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full98907.testtop9, var.x = "BMI")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.testtop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/BMI/TestTop2000/Concordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"), 
           plot = g, device = "png", width = 16, height = 8, units = "in")
    print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  }
}

gene.id <- c(1999, 2000)
dat.expr.train98907.testtop2 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.test[["BMI"]])[gene.id], ]
dat.expr.train98907.testtop2.t <- t(dat.expr.train98907.testtop2) |> as.data.frame()
dat.expr.train98907.testtop2.t$Data <- "Train"
dat.expr.test98907.top2 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.test[["BMI"]])[gene.id], ]
dat.expr.test98907.top2.t <- t(dat.expr.test98907.top2) |> as.data.frame()
dat.expr.test98907.top2.t$Data <- "Test"
rownames(dat.expr.train98907.testtop2.t) = rownames(dat.expr.test98907.top2.t) = NULL
dat.expr.full98907.testtop2 <- rbind(dat.expr.train98907.testtop2.t,
                                     dat.expr.test98907.top2.t)
dat.expr.full98907.testtop2 <- data.frame(dat.expr.full98907.testtop2, BMI)

sumDis <- 0
for(j in 1:2) {
    p.val.train <- wilcox.test(dat.expr.full98907.testtop2[dat.expr.full98907.testtop2$Data == "Train", ][[j]] ~ dat.expr.full98907.testtop2[dat.expr.full98907.testtop2$Data == "Train", ][["BMI"]], data = dat.expr.full98907.testtop2)$p.value
    p.val.test <- wilcox.test(dat.expr.full98907.testtop2[dat.expr.full98907.testtop2$Data == "Test", ][[j]] ~ dat.expr.full98907.testtop2[dat.expr.full98907.testtop2$Data == "Test", ][["BMI"]], data = dat.expr.full98907.testtop2)$p.value
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1}
}
    
if ((!is.na(sumDis)) & (sumDis > 0)) {
  boxplot.full98907.testtop2 <- names(dat.expr.full98907.testtop2)[-match(c("Data", "BMI"), names(dat.expr.full98907.testtop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full98907.testtop2, var.x = "BMI")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.testtop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/BMI/TestTop2000/Discordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
} else {
  boxplot.full98907.testtop2 <- names(dat.expr.full98907.testtop2)[-match(c("Data", "BMI"), names(dat.expr.full98907.testtop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full98907.testtop2, var.x = "BMI")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.testtop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/BMI/TestTop2000/Concordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
}
```


### AGE
#### Top 2000 genes list from the training set
```{r}
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt98907.RData")
AGE <- c(dat.pheno.train[["AGE"]], dat.pheno.test[["AGE"]])

seed.num <- 98907
date.analysis <- format(Sys.Date(), "%Y%b%d")

##------Prepare plotting function------
boxplot.fun.2 <- function(dataset, var.x, var.y) {
  dataset |> ggplot(mapping = aes(y = get(var.y), x = get(var.x))) + 
    geom_boxplot(mapping = aes(color = get(var.x))) +
    geom_jitter(mapping = aes(color = get(var.x))) + 
    stat_compare_means(method = "wilcox.test", 
                       label.x.npc = "right",
                       label.y = 2,
                       mapping = aes(label = ifelse(p < 1.e-3, "p < 0.001", sprintf("p = %4.3f", as.numeric(..p.format..))))) +
    facet_wrap(~factor(Data, levels = c("Train", "Test"))) + 
    labs(y = var.y, 
         x = "") +
    theme_bw(base_size = 14) + 
    scale_color_manual(values = c("#0073C2FF", "#EFC000FF")) + 
    scale_x_discrete(limits = c("young", "old")) + 
    guides(color = "none")
}

for(i in seq(1, 1998, 9)) {
  ##------Process data------
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8)
  dat.expr.train98907.top9 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.train[["AGE"]])[gene.id], ]
  dat.expr.train98907.top9.t <- t(dat.expr.train98907.top9) |> as.data.frame()
dat.expr.train98907.top9.t$Data <- "Train"
  dat.expr.test98907.traintop9 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.train[["AGE"]])[gene.id], ]
  dat.expr.test98907.traintop9.t <- t(dat.expr.test98907.traintop9) |> as.data.frame()
  dat.expr.test98907.traintop9.t$Data <- "Test"
  rownames(dat.expr.train98907.top9.t) = rownames(dat.expr.test98907.traintop9.t) = NULL
  dat.expr.full98907.traintop9 <- rbind(dat.expr.train98907.top9.t,
                                       dat.expr.test98907.traintop9.t)
  
  dat.expr.full98907.traintop9 <- data.frame(dat.expr.full98907.traintop9, AGE)
  sumDis <- 0
  for(j in 1:9) {
    p.val.train <- wilcox.test(dat.expr.full98907.traintop9[dat.expr.full98907.traintop9$Data == "Train", ][[j]] ~ dat.expr.full98907.traintop9[dat.expr.full98907.traintop9$Data == "Train", ][["AGE"]], data = dat.expr.full98907.traintop9)$p.value
    p.val.test <- wilcox.test(dat.expr.full98907.traintop9[dat.expr.full98907.traintop9$Data == "Test", ][[j]] ~ dat.expr.full98907.traintop9[dat.expr.full98907.traintop9$Data == "Test", ][["AGE"]], data = dat.expr.full98907.traintop9)$p.value
    
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1 }
  }
    
  ##------Make boxplots------
  if ((!is.na(sumDis)) & (sumDis > 0)) {
    boxplot.full98907.traintop9 <- names(dat.expr.full98907.traintop9)[-match(c("Data", "AGE"), names(dat.expr.full98907.traintop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full98907.traintop9, var.x = "AGE")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.traintop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/AGE/TrainTop2000/Discordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"), 
           plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  } else {
    boxplot.full98907.traintop9 <- names(dat.expr.full98907.traintop9)[-match(c("Data", "AGE"), names(dat.expr.full98907.traintop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full98907.traintop9, var.x = "AGE")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.traintop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/AGE/TrainTop2000/Concordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
           plot = g, device = "png", width = 16, height = 8, units = "in")
    print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  }
}

gene.id <- c(1999, 2000)
dat.expr.train98907.top2 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.train[["AGE"]])[gene.id], ]
dat.expr.train98907.top2.t <- t(dat.expr.train98907.top2) |> as.data.frame()
dat.expr.train98907.top2.t$Data <- "Train"
dat.expr.test98907.traintop2 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.train[["AGE"]])[gene.id], ]
dat.expr.test98907.traintop2.t <- t(dat.expr.test98907.traintop2) |> as.data.frame()
dat.expr.test98907.traintop2.t$Data <- "Test"
rownames(dat.expr.train98907.top2.t) = rownames(dat.expr.test98907.traintop2.t) = NULL
dat.expr.full98907.traintop2 <- rbind(dat.expr.train98907.top2.t,
                                     dat.expr.test98907.traintop2.t)
dat.expr.full98907.traintop2 <- data.frame(dat.expr.full98907.traintop2, AGE)

sumDis <- 0
for(j in 1:2) {
    p.val.train <- wilcox.test(dat.expr.full98907.traintop2[dat.expr.full98907.traintop2$Data == "Train", ][[j]] ~ dat.expr.full98907.traintop2[dat.expr.full98907.traintop2$Data == "Train", ][["AGE"]], data = dat.expr.full98907.traintop2)$p.value
    p.val.test <- wilcox.test(dat.expr.full98907.traintop2[dat.expr.full98907.traintop2$Data == "Test", ][[j]] ~ dat.expr.full98907.traintop2[dat.expr.full98907.traintop2$Data == "Test", ][["AGE"]], data = dat.expr.full98907.traintop2)$p.value
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1}
}
    
if ((!is.na(sumDis)) & (sumDis > 0)) {
  boxplot.full98907.traintop2 <- names(dat.expr.full98907.traintop2)[-match(c("Data", "AGE"), names(dat.expr.full98907.traintop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full98907.traintop2, var.x = "AGE")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.traintop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/AGE/TrainTop2000/Discordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
} else {
  boxplot.full98907.traintop2 <- names(dat.expr.full98907.traintop2)[-match(c("Data", "AGE"), names(dat.expr.full98907.traintop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full98907.traintop2, var.x = "AGE")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.traintop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/AGE/TrainTop2000/Concordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
}
```


#### Top 2000 genes list from the test set
```{r}
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt98907.RData")
AGE <- c(dat.pheno.train[["AGE"]], dat.pheno.test[["AGE"]])

seed.num <- 98907
date.analysis <- format(Sys.Date(), "%Y%b%d")

##------Prepare plotting function------
boxplot.fun.2 <- function(dataset, var.x, var.y) {
  dataset |> ggplot(mapping = aes(y = get(var.y), x = get(var.x))) + 
    geom_boxplot(mapping = aes(color = get(var.x))) +
    geom_jitter(mapping = aes(color = get(var.x))) + 
    stat_compare_means(method = "wilcox.test", 
                       label.x.npc = "right",
                       label.y = 2,
                       mapping = aes(label = ifelse(p < 1.e-3, "p < 0.001", sprintf("p = %4.3f", as.numeric(..p.format..))))) +
    facet_wrap(~factor(Data, levels = c("Train", "Test"))) + 
    labs(y = var.y, 
         x = "") +
    theme_bw(base_size = 14) + 
    scale_color_manual(values = c("#0073C2FF", "#EFC000FF")) + 
    scale_x_discrete(limits = c("young", "old")) + 
    guides(color = "none")
}

for(i in seq(1, 1998, 9)) {
  ##------Process data------
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8)
  dat.expr.train98907.testtop9 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.test[["AGE"]])[gene.id], ]
  dat.expr.train98907.testtop9.t <- t(dat.expr.train98907.testtop9) |> as.data.frame()
dat.expr.train98907.testtop9.t$Data <- "Train"
  dat.expr.test98907.top9 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.test[["AGE"]])[gene.id], ]
  dat.expr.test98907.top9.t <- t(dat.expr.test98907.top9) |> as.data.frame()
  dat.expr.test98907.top9.t$Data <- "Test"
  rownames(dat.expr.train98907.testtop9.t) = rownames(dat.expr.test98907.top9.t) = NULL
  dat.expr.full98907.testtop9 <- rbind(dat.expr.train98907.testtop9.t,
                                       dat.expr.test98907.top9.t)
  
  dat.expr.full98907.testtop9 <- data.frame(dat.expr.full98907.testtop9, AGE)
  sumDis <- 0
  for(j in 1:9) {
    p.val.train <- wilcox.test(dat.expr.full98907.testtop9[dat.expr.full98907.testtop9$Data == "Train", ][[j]] ~ dat.expr.full98907.testtop9[dat.expr.full98907.testtop9$Data == "Train", ][["AGE"]], data = dat.expr.full98907.testtop9)$p.value
    p.val.test <- wilcox.test(dat.expr.full98907.testtop9[dat.expr.full98907.testtop9$Data == "Test", ][[j]] ~ dat.expr.full98907.testtop9[dat.expr.full98907.testtop9$Data == "Test", ][["AGE"]], data = dat.expr.full98907.testtop9)$p.value
    
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1 }
  }
    
  ##------Make boxplots------
  if ((!is.na(sumDis)) & (sumDis > 0)) {
    boxplot.full98907.testtop9 <- names(dat.expr.full98907.testtop9)[-match(c("Data", "AGE"), names(dat.expr.full98907.testtop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full98907.testtop9, var.x = "AGE")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.testtop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/AGE/TestTop2000/Discordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
           plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  } else {
    boxplot.full98907.testtop9 <- names(dat.expr.full98907.testtop9)[-match(c("Data", "AGE"), names(dat.expr.full98907.testtop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full98907.testtop9, var.x = "AGE")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.testtop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/AGE/TestTop2000/Concordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
           plot = g, device = "png", width = 16, height = 8, units = "in")
    print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  }
}

gene.id <- c(1999, 2000)
dat.expr.train98907.testtop2 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.test[["AGE"]])[gene.id], ]
dat.expr.train98907.testtop2.t <- t(dat.expr.train98907.testtop2) |> as.data.frame()
dat.expr.train98907.testtop2.t$Data <- "Train"
dat.expr.test98907.top2 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.test[["AGE"]])[gene.id], ]
dat.expr.test98907.top2.t <- t(dat.expr.test98907.top2) |> as.data.frame()
dat.expr.test98907.top2.t$Data <- "Test"
rownames(dat.expr.train98907.testtop2.t) = rownames(dat.expr.test98907.top2.t) = NULL
dat.expr.full98907.testtop2 <- rbind(dat.expr.train98907.testtop2.t,
                                     dat.expr.test98907.top2.t)
dat.expr.full98907.testtop2 <- data.frame(dat.expr.full98907.testtop2, AGE)

sumDis <- 0
for(j in 1:2) {
    p.val.train <- wilcox.test(dat.expr.full98907.testtop2[dat.expr.full98907.testtop2$Data == "Train", ][[j]] ~ dat.expr.full98907.testtop2[dat.expr.full98907.testtop2$Data == "Train", ][["AGE"]], data = dat.expr.full98907.testtop2)$p.value
    p.val.test <- wilcox.test(dat.expr.full98907.testtop2[dat.expr.full98907.testtop2$Data == "Test", ][[j]] ~ dat.expr.full98907.testtop2[dat.expr.full98907.testtop2$Data == "Test", ][["AGE"]], data = dat.expr.full98907.testtop2)$p.value
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1}
}
    
if ((!is.na(sumDis)) & (sumDis > 0)) {
  boxplot.full98907.testtop2 <- names(dat.expr.full98907.testtop2)[-match(c("Data", "AGE"), names(dat.expr.full98907.testtop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full98907.testtop2, var.x = "AGE")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.testtop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/AGE/TestTop2000/Discordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
} else {
  boxplot.full98907.testtop2 <- names(dat.expr.full98907.testtop2)[-match(c("Data", "AGE"), names(dat.expr.full98907.testtop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full98907.testtop2, var.x = "AGE")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.testtop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/AGE/TestTop2000/Concordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
}
```


### SEX
#### Top 2000 genes list from the training set
```{r}
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt98907.RData")
SEX <- c(dat.pheno.train[["SEX"]], dat.pheno.test[["SEX"]])

seed.num <- 98907
date.analysis <- format(Sys.Date(), "%Y%b%d")

##------Prepare plotting function------
boxplot.fun.2 <- function(dataset, var.x, var.y) {
  dataset |> ggplot(mapping = aes(y = get(var.y), x = factor(get(var.x), levels = c("male", "female")))) + 
    geom_boxplot(mapping = aes(color = get(var.x))) +
    geom_jitter(mapping = aes(color = get(var.x))) + 
    stat_compare_means(method = "wilcox.test", 
                       label.x.npc = "right",
                       label.y = 2,
                       mapping = aes(label = ifelse(p < 1.e-3, "p < 0.001", sprintf("p = %4.3f", as.numeric(..p.format..))))) +
    facet_wrap(~factor(Data, levels = c("Train", "Test"))) + 
    labs(y = var.y, 
         x = "") +
    theme_bw(base_size = 14) + 
    scale_color_manual(values = c("#0073C2FF", "#EFC000FF")) + 
    scale_x_discrete(limits = c("male", "female")) + 
    guides(color = "none")
}

for(i in seq(1, 1998, 9)) {
  ##------Process data------
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8)
  dat.expr.train98907.top9 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.train[["SEX"]])[gene.id], ]
  dat.expr.train98907.top9.t <- t(dat.expr.train98907.top9) |> as.data.frame()
dat.expr.train98907.top9.t$Data <- "Train"
  dat.expr.test98907.traintop9 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.train[["SEX"]])[gene.id], ]
  dat.expr.test98907.traintop9.t <- t(dat.expr.test98907.traintop9) |> as.data.frame()
  dat.expr.test98907.traintop9.t$Data <- "Test"
  rownames(dat.expr.train98907.top9.t) = rownames(dat.expr.test98907.traintop9.t) = NULL
  dat.expr.full98907.traintop9 <- rbind(dat.expr.train98907.top9.t,
                                       dat.expr.test98907.traintop9.t)
  
  dat.expr.full98907.traintop9 <- data.frame(dat.expr.full98907.traintop9, SEX)
  sumDis <- 0
  for(j in 1:9) {
    p.val.train <- wilcox.test(dat.expr.full98907.traintop9[dat.expr.full98907.traintop9$Data == "Train", ][[j]] ~ dat.expr.full98907.traintop9[dat.expr.full98907.traintop9$Data == "Train", ][["SEX"]], data = dat.expr.full98907.traintop9)$p.value
    p.val.test <- wilcox.test(dat.expr.full98907.traintop9[dat.expr.full98907.traintop9$Data == "Test", ][[j]] ~ dat.expr.full98907.traintop9[dat.expr.full98907.traintop9$Data == "Test", ][["SEX"]], data = dat.expr.full98907.traintop9)$p.value
    
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1 }
  }
    
  ##------Make boxplots------
  if ((!is.na(sumDis)) & (sumDis > 0)) {
    boxplot.full98907.traintop9 <- names(dat.expr.full98907.traintop9)[-match(c("Data", "SEX"), names(dat.expr.full98907.traintop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full98907.traintop9, var.x = "SEX")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.traintop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/SEX/TrainTop2000/Discordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
           plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  } else {
    boxplot.full98907.traintop9 <- names(dat.expr.full98907.traintop9)[-match(c("Data", "SEX"), names(dat.expr.full98907.traintop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full98907.traintop9, var.x = "SEX")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.traintop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/SEX/TrainTop2000/Concordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
           plot = g, device = "png", width = 16, height = 8, units = "in")
    print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  }
}

gene.id <- c(1999, 2000)
dat.expr.train98907.top2 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.train[["SEX"]])[gene.id], ]
dat.expr.train98907.top2.t <- t(dat.expr.train98907.top2) |> as.data.frame()
dat.expr.train98907.top2.t$Data <- "Train"
dat.expr.test98907.traintop2 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.train[["SEX"]])[gene.id], ]
dat.expr.test98907.traintop2.t <- t(dat.expr.test98907.traintop2) |> as.data.frame()
dat.expr.test98907.traintop2.t$Data <- "Test"
rownames(dat.expr.train98907.top2.t) = rownames(dat.expr.test98907.traintop2.t) = NULL
dat.expr.full98907.traintop2 <- rbind(dat.expr.train98907.top2.t,
                                     dat.expr.test98907.traintop2.t)
dat.expr.full98907.traintop2 <- data.frame(dat.expr.full98907.traintop2, SEX)

sumDis <- 0
for(j in 1:2) {
    p.val.train <- wilcox.test(dat.expr.full98907.traintop2[dat.expr.full98907.traintop2$Data == "Train", ][[j]] ~ dat.expr.full98907.traintop2[dat.expr.full98907.traintop2$Data == "Train", ][["SEX"]], data = dat.expr.full98907.traintop2)$p.value
    p.val.test <- wilcox.test(dat.expr.full98907.traintop2[dat.expr.full98907.traintop2$Data == "Test", ][[j]] ~ dat.expr.full98907.traintop2[dat.expr.full98907.traintop2$Data == "Test", ][["SEX"]], data = dat.expr.full98907.traintop2)$p.value
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1}
}
    
if ((!is.na(sumDis)) & (sumDis > 0)) {
  boxplot.full98907.traintop2 <- names(dat.expr.full98907.traintop2)[-match(c("Data", "SEX"), names(dat.expr.full98907.traintop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full98907.traintop2, var.x = "SEX")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.traintop2, ncol = 2)))
  print(paste0("[", i, "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/SEX/TrainTop2000/Discordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
} else {
  boxplot.full98907.traintop2 <- names(dat.expr.full98907.traintop2)[-match(c("Data", "SEX"), names(dat.expr.full98907.traintop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full98907.traintop2, var.x = "SEX")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.traintop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/SEX/TrainTop2000/Concordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
}
```


#### Top 2000 genes list from the test set
```{r}
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt98907.RData")
SEX <- c(dat.pheno.train[["SEX"]], dat.pheno.test[["SEX"]])

seed.num <- 98907
date.analysis <- format(Sys.Date(), "%Y%b%d")

##------Prepare plotting function------
boxplot.fun.2 <- function(dataset, var.x, var.y) {
  dataset |> ggplot(mapping = aes(y = get(var.y), x = factor(get(var.x), levels = c("male", "female")))) + 
    geom_boxplot(mapping = aes(color = get(var.x))) +
    geom_jitter(mapping = aes(color = get(var.x))) + 
    stat_compare_means(method = "wilcox.test", 
                       label.x.npc = "right",
                       label.y = 2,
                       mapping = aes(label = ifelse(p < 1.e-3, "p < 0.001", sprintf("p = %4.3f", as.numeric(..p.format..))))) +
    facet_wrap(~factor(Data, levels = c("Train", "Test"))) + 
    labs(y = var.y, 
         x = "") +
    theme_bw(base_size = 14) + 
    scale_color_manual(values = c("#0073C2FF", "#EFC000FF")) + 
    scale_x_discrete(limits = c("male", "female")) + 
    guides(color = "none")
}

for(i in seq(1, 1998, 9)) {
  ##------Process data------
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8)
  dat.expr.train98907.testtop9 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.test[["SEX"]])[gene.id], ]
  dat.expr.train98907.testtop9.t <- t(dat.expr.train98907.testtop9) |> as.data.frame()
dat.expr.train98907.testtop9.t$Data <- "Train"
  dat.expr.test98907.top9 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.test[["SEX"]])[gene.id], ]
  dat.expr.test98907.top9.t <- t(dat.expr.test98907.top9) |> as.data.frame()
  dat.expr.test98907.top9.t$Data <- "Test"
  rownames(dat.expr.train98907.testtop9.t) = rownames(dat.expr.test98907.top9.t) = NULL
  dat.expr.full98907.testtop9 <- rbind(dat.expr.train98907.testtop9.t,
                                       dat.expr.test98907.top9.t)
  
  dat.expr.full98907.testtop9 <- data.frame(dat.expr.full98907.testtop9, SEX)
  sumDis <- 0
  for(j in 1:9) {
    p.val.train <- wilcox.test(dat.expr.full98907.testtop9[dat.expr.full98907.testtop9$Data == "Train", ][[j]] ~ dat.expr.full98907.testtop9[dat.expr.full98907.testtop9$Data == "Train", ][["SEX"]], data = dat.expr.full98907.testtop9)$p.value
    p.val.test <- wilcox.test(dat.expr.full98907.testtop9[dat.expr.full98907.testtop9$Data == "Test", ][[j]] ~ dat.expr.full98907.testtop9[dat.expr.full98907.testtop9$Data == "Test", ][["SEX"]], data = dat.expr.full98907.testtop9)$p.value
    
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1 }
  }
    
  ##------Make boxplots------
  if ((!is.na(sumDis)) & (sumDis > 0)) {
    boxplot.full98907.testtop9 <- names(dat.expr.full98907.testtop9)[-match(c("Data", "SEX"), names(dat.expr.full98907.testtop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full98907.testtop9, var.x = "SEX")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.testtop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/SEX/TestTop2000/Discordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 16,
         height = 8,
         units = "in")
  print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  } else {
    boxplot.full98907.testtop9 <- names(dat.expr.full98907.testtop9)[-match(c("Data", "SEX"), names(dat.expr.full98907.testtop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full98907.testtop9, var.x = "SEX")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.testtop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/SEX/TestTop2000/Concordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 16,
         height = 8,
         units = "in")
    print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  }
}

gene.id <- c(1999, 2000)
dat.expr.train98907.testtop2 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.test[["SEX"]])[gene.id], ]
dat.expr.train98907.testtop2.t <- t(dat.expr.train98907.testtop2) |> as.data.frame()
dat.expr.train98907.testtop2.t$Data <- "Train"
dat.expr.test98907.top2 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.test[["SEX"]])[gene.id], ]
dat.expr.test98907.top2.t <- t(dat.expr.test98907.top2) |> as.data.frame()
dat.expr.test98907.top2.t$Data <- "Test"
rownames(dat.expr.train98907.testtop2.t) = rownames(dat.expr.test98907.top2.t) = NULL
dat.expr.full98907.testtop2 <- rbind(dat.expr.train98907.testtop2.t,
                                     dat.expr.test98907.top2.t)
dat.expr.full98907.testtop2 <- data.frame(dat.expr.full98907.testtop2, SEX)

sumDis <- 0
for(j in 1:2) {
    p.val.train <- wilcox.test(dat.expr.full98907.testtop2[dat.expr.full98907.testtop2$Data == "Train", ][[j]] ~ dat.expr.full98907.testtop2[dat.expr.full98907.testtop2$Data == "Train", ][["SEX"]], data = dat.expr.full98907.testtop2)$p.value
    p.val.test <- wilcox.test(dat.expr.full98907.testtop2[dat.expr.full98907.testtop2$Data == "Test", ][[j]] ~ dat.expr.full98907.testtop2[dat.expr.full98907.testtop2$Data == "Test", ][["SEX"]], data = dat.expr.full98907.testtop2)$p.value
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1}
}
    
if ((!is.na(sumDis)) & (sumDis > 0)) {
  boxplot.full98907.testtop2 <- names(dat.expr.full98907.testtop2)[-match(c("Data", "SEX"), names(dat.expr.full98907.testtop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full98907.testtop2, var.x = "SEX")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.testtop2, ncol = 2)))
  print(paste0("[", i, "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/SEX/TestTop2000/Discordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 16,
         height = 8,
         units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
} else {
  boxplot.full98907.testtop2 <- names(dat.expr.full98907.testtop2)[-match(c("Data", "SEX"), names(dat.expr.full98907.testtop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full98907.testtop2, var.x = "SEX")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.testtop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/SEX/TestTop2000/Concordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 16,
         height = 8,
         units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
}
```


### MHABNWBC
#### Top 2000 genes list from the training set
```{r}
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt98907.RData")
MHABNWBC <- c(dat.pheno.train[["MHABNWBC"]], dat.pheno.test[["MHABNWBC"]])

seed.num <- 98907
date.analysis <- format(Sys.Date(), "%Y%b%d")

##------Prepare plotting function------
boxplot.fun.2 <- function(dataset, var.x, var.y) {
  dataset |> ggplot(mapping = aes(y = get(var.y), x = factor(get(var.x), levels = c("yes", "no")))) +
    geom_boxplot(mapping = aes(color = get(var.x))) +
    geom_jitter(mapping = aes(color = get(var.x))) + 
    stat_compare_means(method = "wilcox.test",
                       label.x.npc = "right",
                       label.y = 2,
                       mapping = aes(label = ifelse(p < 1.e-3, "p < 0.001", sprintf("p = %4.3f", as.numeric(..p.format..))))) +
    facet_wrap(~factor(Data, levels = c("Train", "Test"))) + 
    labs(y = var.y, 
         x = "") +
    theme_bw(base_size = 14) + 
    scale_color_manual(values = c("#0073C2FF", "#EFC000FF")) + 
    scale_x_discrete(limits = c("yes", "no")) + 
    guides(color = "none")
}

for(i in seq(1, 1998, 9)) {
  ##------Process data------
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8)
  dat.expr.train98907.top9 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.train[["MHABNWBC"]])[gene.id], ]
  dat.expr.train98907.top9.t <- t(dat.expr.train98907.top9) |> as.data.frame()
dat.expr.train98907.top9.t$Data <- "Train"
  dat.expr.test98907.traintop9 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.train[["MHABNWBC"]])[gene.id], ]
  dat.expr.test98907.traintop9.t <- t(dat.expr.test98907.traintop9) |> as.data.frame()
  dat.expr.test98907.traintop9.t$Data <- "Test"
  rownames(dat.expr.train98907.top9.t) = rownames(dat.expr.test98907.traintop9.t) = NULL
  dat.expr.full98907.traintop9 <- rbind(dat.expr.train98907.top9.t,
                                       dat.expr.test98907.traintop9.t)
  
  dat.expr.full98907.traintop9 <- data.frame(dat.expr.full98907.traintop9, MHABNWBC)
  sumDis <- 0
  for(j in 1:9) {
    p.val.train <- wilcox.test(dat.expr.full98907.traintop9[dat.expr.full98907.traintop9$Data == "Train", ][[j]] ~ dat.expr.full98907.traintop9[dat.expr.full98907.traintop9$Data == "Train", ][["MHABNWBC"]], data = dat.expr.full98907.traintop9)$p.value
    p.val.test <- wilcox.test(dat.expr.full98907.traintop9[dat.expr.full98907.traintop9$Data == "Test", ][[j]] ~ dat.expr.full98907.traintop9[dat.expr.full98907.traintop9$Data == "Test", ][["MHABNWBC"]], data = dat.expr.full98907.traintop9)$p.value
    
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1 }
  }
    
  ##------Make boxplots------
  if ((!is.na(sumDis)) & (sumDis > 0)) {
    boxplot.full98907.traintop9 <- names(dat.expr.full98907.traintop9)[-match(c("Data", "MHABNWBC"), names(dat.expr.full98907.traintop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full98907.traintop9, var.x = "MHABNWBC")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.traintop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/MHABNWBC/TrainTop2000/Discordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 16,
         height = 8,
         units = "in")
  print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  } else {
    boxplot.full98907.traintop9 <- names(dat.expr.full98907.traintop9)[-match(c("Data", "MHABNWBC"), names(dat.expr.full98907.traintop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full98907.traintop9, var.x = "MHABNWBC")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.traintop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/MHABNWBC/TrainTop2000/Concordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 16,
         height = 8,
         units = "in")
    print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  }
}

gene.id <- c(1999, 2000)
dat.expr.train98907.top2 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.train[["MHABNWBC"]])[gene.id], ]
dat.expr.train98907.top2.t <- t(dat.expr.train98907.top2) |> as.data.frame()
dat.expr.train98907.top2.t$Data <- "Train"
dat.expr.test98907.traintop2 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.train[["MHABNWBC"]])[gene.id], ]
dat.expr.test98907.traintop2.t <- t(dat.expr.test98907.traintop2) |> as.data.frame()
dat.expr.test98907.traintop2.t$Data <- "Test"
rownames(dat.expr.train98907.top2.t) = rownames(dat.expr.test98907.traintop2.t) = NULL
dat.expr.full98907.traintop2 <- rbind(dat.expr.train98907.top2.t,
                                     dat.expr.test98907.traintop2.t)
dat.expr.full98907.traintop2 <- data.frame(dat.expr.full98907.traintop2, MHABNWBC)

sumDis <- 0
for(j in 1:2) {
    p.val.train <- wilcox.test(dat.expr.full98907.traintop2[dat.expr.full98907.traintop2$Data == "Train", ][[j]] ~ dat.expr.full98907.traintop2[dat.expr.full98907.traintop2$Data == "Train", ][["MHABNWBC"]], data = dat.expr.full98907.traintop2)$p.value
    p.val.test <- wilcox.test(dat.expr.full98907.traintop2[dat.expr.full98907.traintop2$Data == "Test", ][[j]] ~ dat.expr.full98907.traintop2[dat.expr.full98907.traintop2$Data == "Test", ][["MHABNWBC"]], data = dat.expr.full98907.traintop2)$p.value
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1}
}
    
if ((!is.na(sumDis)) & (sumDis > 0)) {
  boxplot.full98907.traintop2 <- names(dat.expr.full98907.traintop2)[-match(c("Data", "MHABNWBC"), names(dat.expr.full98907.traintop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full98907.traintop2, var.x = "MHABNWBC")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.traintop2, ncol = 2)))
  print(paste0("[", i, "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/MHABNWBC/TrainTop2000/Discordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 16,
         height = 8,
         units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
} else {
  boxplot.full98907.traintop2 <- names(dat.expr.full98907.traintop2)[-match(c("Data", "MHABNWBC"), names(dat.expr.full98907.traintop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full98907.traintop2, var.x = "MHABNWBC")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.traintop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/MHABNWBC/TrainTop2000/Concordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g,
         device = "png",
         width = 16,
         height = 8,
         units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
}
```


#### Top 2000 genes list from the test set
```{r}
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt98907.RData")
MHABNWBC <- c(dat.pheno.train[["MHABNWBC"]], dat.pheno.test[["MHABNWBC"]])

seed.num <- 98907
date.analysis <- format(Sys.Date(), "%Y%b%d")

##------Prepare plotting function------
boxplot.fun.2 <- function(dataset, var.x, var.y) {
  dataset |> ggplot(mapping = aes(y = get(var.y), x = factor(get(var.x), levels = c("yes", "no")))) + 
    geom_boxplot(mapping = aes(color = get(var.x))) +
    geom_jitter(mapping = aes(color = get(var.x))) + 
    stat_compare_means(method = "wilcox.test", 
                       label.x.npc = "right",
                       label.y = 2,
                       mapping = aes(label = ifelse(p < 1.e-3, "p < 0.001", sprintf("p = %4.3f", as.numeric(..p.format..))))) +
    facet_wrap(~factor(Data, levels = c("Train", "Test"))) + 
    labs(y = var.y, 
         x = "") +
    theme_bw(base_size = 14) + 
    scale_color_manual(values = c("#0073C2FF", "#EFC000FF")) + 
    scale_x_discrete(limits = c("yes", "no")) + 
    guides(color = "none")
}

for(i in seq(1, 1998, 9)) {
  ##------Process data------
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8)
  dat.expr.train98907.testtop9 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.test[["MHABNWBC"]])[gene.id], ]
  dat.expr.train98907.testtop9.t <- t(dat.expr.train98907.testtop9) |> as.data.frame()
dat.expr.train98907.testtop9.t$Data <- "Train"
  dat.expr.test98907.top9 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.test[["MHABNWBC"]])[gene.id], ]
  dat.expr.test98907.top9.t <- t(dat.expr.test98907.top9) |> as.data.frame()
  dat.expr.test98907.top9.t$Data <- "Test"
  rownames(dat.expr.train98907.testtop9.t) = rownames(dat.expr.test98907.top9.t) = NULL
  dat.expr.full98907.testtop9 <- rbind(dat.expr.train98907.testtop9.t,
                                      dat.expr.test98907.top9.t)
  
  dat.expr.full98907.testtop9 <- data.frame(dat.expr.full98907.testtop9, MHABNWBC)
  sumDis <- 0
  for(j in 1:9) {
    p.val.train <- wilcox.test(dat.expr.full98907.testtop9[dat.expr.full98907.testtop9$Data == "Train", ][[j]] ~ dat.expr.full98907.testtop9[dat.expr.full98907.testtop9$Data == "Train", ][["MHABNWBC"]], data = dat.expr.full98907.testtop9)$p.value
    p.val.test <- wilcox.test(dat.expr.full98907.testtop9[dat.expr.full98907.testtop9$Data == "Test", ][[j]] ~ dat.expr.full98907.testtop9[dat.expr.full98907.testtop9$Data == "Test", ][["MHABNWBC"]], data = dat.expr.full98907.testtop9)$p.value
    
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1 }
  }
    
  ##------Make boxplots------
  if ((!is.na(sumDis)) & (sumDis > 0)) {
    boxplot.full98907.testtop9 <- names(dat.expr.full98907.testtop9)[-match(c("Data", "MHABNWBC"), names(dat.expr.full98907.testtop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full98907.testtop9, var.x = "MHABNWBC")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.testtop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/MHABNWBC/TestTop2000/Discordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
           plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  } else {
    boxplot.full98907.testtop9 <- names(dat.expr.full98907.testtop9)[-match(c("Data", "MHABNWBC"), names(dat.expr.full98907.testtop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full98907.testtop9, var.x = "MHABNWBC")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.testtop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/MHABNWBC/TestTop2000/Concordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
           plot = g, device = "png", width = 16, height = 8, units = "in")
    print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  }
}

gene.id <- c(1999, 2000)
dat.expr.train98907.testtop2 <- dat.expr.train[names(BMAseq.eFDR.Main.2000.test[["MHABNWBC"]])[gene.id], ]
dat.expr.train98907.testtop2.t <- t(dat.expr.train98907.testtop2) |> as.data.frame()
dat.expr.train98907.testtop2.t$Data <- "Train"
dat.expr.test98907.top2 <- dat.expr.test[names(BMAseq.eFDR.Main.2000.test[["MHABNWBC"]])[gene.id], ]
dat.expr.test98907.top2.t <- t(dat.expr.test98907.top2) |> as.data.frame()
dat.expr.test98907.top2.t$Data <- "Test"
rownames(dat.expr.train98907.testtop2.t) = rownames(dat.expr.test98907.top2.t) = NULL
dat.expr.full98907.testtop2 <- rbind(dat.expr.train98907.testtop2.t,
                                     dat.expr.test98907.top2.t)
dat.expr.full98907.testtop2 <- data.frame(dat.expr.full98907.testtop2, MHABNWBC)

sumDis <- 0
for(j in 1:2) {
    p.val.train <- wilcox.test(dat.expr.full98907.testtop2[dat.expr.full98907.testtop2$Data == "Train", ][[j]] ~ dat.expr.full98907.testtop2[dat.expr.full98907.testtop2$Data == "Train", ][["MHABNWBC"]], data = dat.expr.full98907.testtop2)$p.value
    p.val.test <- wilcox.test(dat.expr.full98907.testtop2[dat.expr.full98907.testtop2$Data == "Test", ][[j]] ~ dat.expr.full98907.testtop2[dat.expr.full98907.testtop2$Data == "Test", ][["MHABNWBC"]], data = dat.expr.full98907.testtop2)$p.value
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1}
}
    
if ((!is.na(sumDis)) & (sumDis > 0)) {
  boxplot.full98907.testtop2 <- names(dat.expr.full98907.testtop2)[-match(c("Data", "MHABNWBC"), names(dat.expr.full98907.testtop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full98907.testtop2, var.x = "MHABNWBC")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.testtop2, ncol = 2)))
  print(paste0("[", i, "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/MHABNWBC/TestTop2000/Discordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
} else {
  boxplot.full98907.testtop2 <- names(dat.expr.full98907.testtop2)[-match(c("Data", "MHABNWBC"), names(dat.expr.full98907.testtop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full98907.testtop2, var.x = "MHABNWBC")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.testtop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/MHABNWBC/TestTop2000/Concordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 8, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
}
```


### BMI X SEX
#### Top 2000 genes list from the training set
```{r}
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt98907.RData")
BMI.0 <- c(dat.pheno.train[["BMI"]], dat.pheno.test[["BMI"]])
SEX.0 <- c(dat.pheno.train[["SEX"]], dat.pheno.test[["SEX"]])
# expand.grid(BMI = c("low", "high"), SEX = c("male", "female"))
BMI <- dplyr::case_when(BMI.0 == "low" & SEX.0 == "male" ~ "LowMale", 
                        BMI.0 == "high" & SEX.0 == "male" ~ "HighMale",
                        BMI.0 == "low" & SEX.0 == "female" ~ "LowFemale", 
                        BMI.0 == "high" & SEX.0 == "female" ~ "HighFemale")

seed.num <- 98907
date.analysis <- format(Sys.Date(), "%Y%b%d")


##------Prepare plotting function------
boxplot.fun.2 <- function(dataset, var.x, var.y) {
  dataset |> ggplot(mapping = aes(y = get(var.y), x = factor(get(var.x), levels = c("LowMale", "LowFemale", "HighMale", "HighFemale")))) + 
    geom_boxplot(mapping = aes(color = get(var.x))) +
    geom_jitter(mapping = aes(color = get(var.x))) + 
    stat_compare_means(method = "kruskal.test", 
                       label.x.npc = "center",
                       label.y = 2,
                       mapping = aes(label = ifelse(p < 1.e-3, "p < 0.001", sprintf("p = %4.3f", as.numeric(..p.format..))))) +
    facet_wrap(~factor(Data, levels = c("Train", "Test"))) + 
    labs(y = var.y, 
         x = "") +
    theme_bw(base_size = 14) + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1, color = "black", vjust = 0.5)) + 
    scale_color_manual(values = c("#0073C2FF", "#EFC000FF", "#868686FF", "#CD534CFF")) + 
    scale_x_discrete(limits = c("LowMale", "LowFemale", "HighMale", "HighFemale")) + 
    guides(color = "none")
}


for(i in seq(1, 1998, 9)) {
  ##------Process data------
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8)
  dat.expr.train98907.top9 <- dat.expr.train[names(BMAseq.eFDR.Interaction.2000.train[["BMI"]])[gene.id], ]
  dat.expr.train98907.top9.t <- t(dat.expr.train98907.top9) |> as.data.frame()
dat.expr.train98907.top9.t$Data <- "Train"
  dat.expr.test98907.traintop9 <- dat.expr.test[names(BMAseq.eFDR.Interaction.2000.train[["BMI"]])[gene.id], ]
  dat.expr.test98907.traintop9.t <- t(dat.expr.test98907.traintop9) |> as.data.frame()
  dat.expr.test98907.traintop9.t$Data <- "Test"
  rownames(dat.expr.train98907.top9.t) = rownames(dat.expr.test98907.traintop9.t) = NULL
  dat.expr.full98907.traintop9 <- rbind(dat.expr.train98907.top9.t,
                                       dat.expr.test98907.traintop9.t)
  
  dat.expr.full98907.traintop9 <- data.frame(dat.expr.full98907.traintop9, BMI)
  sumDis <- 0
  for(j in 1:9) {
    p.val.train <- kruskal.test(dat.expr.full98907.traintop9[dat.expr.full98907.traintop9$Data == "Train", ][[j]] ~ dat.expr.full98907.traintop9[dat.expr.full98907.traintop9$Data == "Train", ][["BMI"]], data = dat.expr.full98907.traintop9)$p.value
    p.val.test <- kruskal.test(dat.expr.full98907.traintop9[dat.expr.full98907.traintop9$Data == "Test", ][[j]] ~ dat.expr.full98907.traintop9[dat.expr.full98907.traintop9$Data == "Test", ][["BMI"]], data = dat.expr.full98907.traintop9)$p.value
    
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1 }
  }
    
  ##------Make boxplots------
  if ((!is.na(sumDis)) & (sumDis > 0)) {
    boxplot.full98907.traintop9 <- names(dat.expr.full98907.traintop9)[-match(c("Data", "BMI"), names(dat.expr.full98907.traintop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full98907.traintop9, var.x = "BMI")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.traintop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/BMISEX/TrainTop2000/Discordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
           plot = g, device = "png", width = 16, height = 10, units = "in")
  print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  } else {
    boxplot.full98907.traintop9 <- names(dat.expr.full98907.traintop9)[-match(c("Data", "BMI"), names(dat.expr.full98907.traintop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full98907.traintop9, var.x = "BMI")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.traintop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/BMISEX/TrainTop2000/Concordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
           plot = g, device = "png", width = 16, height = 10, units = "in")
    print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  }
}

gene.id <- c(1999, 2000)
dat.expr.train98907.top2 <- dat.expr.train[names(BMAseq.eFDR.Interaction.2000.train[["BMI"]])[gene.id], ]
dat.expr.train98907.top2.t <- t(dat.expr.train98907.top2) |> as.data.frame()
dat.expr.train98907.top2.t$Data <- "Train"
dat.expr.test98907.traintop2 <- dat.expr.test[names(BMAseq.eFDR.Interaction.2000.train[["BMI"]])[gene.id], ]
dat.expr.test98907.traintop2.t <- t(dat.expr.test98907.traintop2) |> as.data.frame()
dat.expr.test98907.traintop2.t$Data <- "Test"
rownames(dat.expr.train98907.top2.t) = rownames(dat.expr.test98907.traintop2.t) = NULL
dat.expr.full98907.traintop2 <- rbind(dat.expr.train98907.top2.t,
                                     dat.expr.test98907.traintop2.t)
dat.expr.full98907.traintop2 <- data.frame(dat.expr.full98907.traintop2, BMI)

sumDis <- 0
for(j in 1:2) {
    p.val.train <- kruskal.test(dat.expr.full98907.traintop2[dat.expr.full98907.traintop2$Data == "Train", ][[j]] ~ dat.expr.full98907.traintop2[dat.expr.full98907.traintop2$Data == "Train", ][["BMI"]], data = dat.expr.full98907.traintop2)$p.value
    p.val.test <- kruskal.test(dat.expr.full98907.traintop2[dat.expr.full98907.traintop2$Data == "Test", ][[j]] ~ dat.expr.full98907.traintop2[dat.expr.full98907.traintop2$Data == "Test", ][["BMI"]], data = dat.expr.full98907.traintop2)$p.value
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1}
}
    
if ((!is.na(sumDis)) & (sumDis > 0)) {
  boxplot.full98907.traintop2 <- names(dat.expr.full98907.traintop2)[-match(c("Data", "BMI"), names(dat.expr.full98907.traintop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full98907.traintop2, var.x = "BMI")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.traintop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/BMISEX/TrainTop2000/Discordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 10, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
} else {
  boxplot.full98907.traintop2 <- names(dat.expr.full98907.traintop2)[-match(c("Data", "BMI"), names(dat.expr.full98907.traintop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full98907.traintop2, var.x = "BMI")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.traintop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/BMISEX/TrainTop2000/Concordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 10, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
}
```


#### Top 2000 genes list from the test set
```{r}
load("../ApplicationData/derived/RandomSeed/BMAseqMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt98907.RData")
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt98907.RData")
BMI.0 <- c(dat.pheno.train[["BMI"]], dat.pheno.test[["BMI"]])
SEX.0 <- c(dat.pheno.train[["SEX"]], dat.pheno.test[["SEX"]])
BMI <- dplyr::case_when(BMI.0 == "low" & SEX.0 == "male" ~ "LowMale", 
                        BMI.0 == "high" & SEX.0 == "male" ~ "HighMale",
                        BMI.0 == "low" & SEX.0 == "female" ~ "LowFemale", 
                        BMI.0 == "high" & SEX.0 == "female" ~ "HighFemale")

seed.num <- 98907
date.analysis <- format(Sys.Date(), "%Y%b%d")

##------Prepare plotting function------
boxplot.fun.2 <- function(dataset, var.x, var.y) {
  dataset |> ggplot(mapping = aes(y = get(var.y), x = factor(get(var.x), levels = c("LowMale", "LowFemale", "HighMale", "HighFemale")))) + 
    geom_boxplot(mapping = aes(color = get(var.x))) +
    geom_jitter(mapping = aes(color = get(var.x))) + 
    stat_compare_means(method = "kruskal.test", 
                       label.x.npc = "center",
                       label.y = 2,
                       mapping = aes(label = ifelse(p < 1.e-3, "p < 0.001", sprintf("p = %4.3f", as.numeric(..p.format..))))) +
    facet_wrap(~factor(Data, levels = c("Train", "Test"))) + 
    labs(y = var.y, 
         x = "") +
    theme_bw(base_size = 14) + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1, color = "black", vjust = 0.5)) + 
    scale_color_manual(values = c("#0073C2FF", "#EFC000FF", "#868686FF", "#CD534CFF")) + 
    scale_x_discrete(limits = c("LowMale", "LowFemale", "HighMale", "HighFemale")) + 
    guides(color = "none")
}

for(i in seq(1, 1998, 9)) {
  ##------Process data------
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8)
  dat.expr.train98907.testtop9 <- dat.expr.train[names(BMAseq.eFDR.Interaction.2000.test[["BMI"]])[gene.id], ]
  dat.expr.train98907.testtop9.t <- t(dat.expr.train98907.testtop9) |> as.data.frame()
dat.expr.train98907.testtop9.t$Data <- "Train"
  dat.expr.test98907.top9 <- dat.expr.test[names(BMAseq.eFDR.Interaction.2000.test[["BMI"]])[gene.id], ]
  dat.expr.test98907.top9.t <- t(dat.expr.test98907.top9) |> as.data.frame()
  dat.expr.test98907.top9.t$Data <- "Test"
  rownames(dat.expr.train98907.testtop9.t) = rownames(dat.expr.test98907.top9.t) = NULL
  dat.expr.full98907.testtop9 <- rbind(dat.expr.train98907.testtop9.t,
                                      dat.expr.test98907.top9.t)
  
  dat.expr.full98907.testtop9 <- data.frame(dat.expr.full98907.testtop9, BMI)
  sumDis <- 0
  for(j in 1:9) {
    p.val.train <- kruskal.test(dat.expr.full98907.testtop9[dat.expr.full98907.testtop9$Data == "Train", ][[j]] ~ dat.expr.full98907.testtop9[dat.expr.full98907.testtop9$Data == "Train", ][["BMI"]], data = dat.expr.full98907.testtop9)$p.value
    p.val.test <- kruskal.test(dat.expr.full98907.testtop9[dat.expr.full98907.testtop9$Data == "Test", ][[j]] ~ dat.expr.full98907.testtop9[dat.expr.full98907.testtop9$Data == "Test", ][["BMI"]], data = dat.expr.full98907.testtop9)$p.value
    
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1 }
  }
    
  ##------Make boxplots------
  if ((!is.na(sumDis)) & (sumDis > 0)) {
    boxplot.full98907.testtop9 <- names(dat.expr.full98907.testtop9)[-match(c("Data", "BMI"), names(dat.expr.full98907.testtop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full98907.testtop9, var.x = "BMI")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.testtop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/BMISEX/TestTop2000/Discordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 10, units = "in")
  print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  } else {
    boxplot.full98907.testtop9 <- names(dat.expr.full98907.testtop9)[-match(c("Data", "BMI"), names(dat.expr.full98907.testtop9))] |> 
    map(.f = boxplot.fun.2, dataset = dat.expr.full98907.testtop9, var.x = "BMI")
    g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.testtop9, ncol = 3)))
    print(paste0("[", i, "] Plot a 9-boxplot matrix successfully!"))
    ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/BMISEX/TestTop2000/Concordant/", date.analysis, "_", i, "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 10, units = "in")
    print(paste0("[", i, "] Save a 9-boxplot matrix successfully!"))
  }
}

gene.id <- c(1999, 2000)
dat.expr.train98907.testtop2 <- dat.expr.train[names(BMAseq.eFDR.Interaction.2000.test[["BMI"]])[gene.id], ]
dat.expr.train98907.testtop2.t <- t(dat.expr.train98907.testtop2) |> as.data.frame()
dat.expr.train98907.testtop2.t$Data <- "Train"
dat.expr.test98907.top2 <- dat.expr.test[names(BMAseq.eFDR.Interaction.2000.test[["BMI"]])[gene.id], ]
dat.expr.test98907.top2.t <- t(dat.expr.test98907.top2) |> as.data.frame()
dat.expr.test98907.top2.t$Data <- "Test"
rownames(dat.expr.train98907.testtop2.t) = rownames(dat.expr.test98907.top2.t) = NULL
dat.expr.full98907.testtop2 <- rbind(dat.expr.train98907.testtop2.t,
                                     dat.expr.test98907.top2.t)
dat.expr.full98907.testtop2 <- data.frame(dat.expr.full98907.testtop2, BMI)

sumDis <- 0
for(j in 1:2) {
    p.val.train <- kruskal.test(dat.expr.full98907.testtop2[dat.expr.full98907.testtop2$Data == "Train", ][[j]] ~ dat.expr.full98907.testtop2[dat.expr.full98907.testtop2$Data == "Train", ][["BMI"]], data = dat.expr.full98907.testtop2)$p.value
    p.val.test <- kruskal.test(dat.expr.full98907.testtop2[dat.expr.full98907.testtop2$Data == "Test", ][[j]] ~ dat.expr.full98907.testtop2[dat.expr.full98907.testtop2$Data == "Test", ][["BMI"]], data = dat.expr.full98907.testtop2)$p.value
    if ((p.val.train < 0.05 & p.val.test >= 0.05) | (p.val.train >= 0.05 & p.val.test < 0.05)) { sumDis <- sumDis + 1}
}
    
if ((!is.na(sumDis)) & (sumDis > 0)) {
  boxplot.full98907.testtop2 <- names(dat.expr.full98907.testtop2)[-match(c("Data", "BMI"), names(dat.expr.full98907.testtop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full98907.testtop2, var.x = "BMI")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.testtop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/BMISEX/TestTop2000/Discordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 10, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
} else {
  boxplot.full98907.testtop2 <- names(dat.expr.full98907.testtop2)[-match(c("Data", "BMI"), names(dat.expr.full98907.testtop2))] |> 
  map(.f = boxplot.fun.2, dataset = dat.expr.full98907.testtop2, var.x = "BMI")
  g <- grid.arrange(do.call("arrangeGrob", c(boxplot.full98907.testtop2, ncol = 2)))
  print(paste0("[", gene.id[1], "] Plot a 2-boxplot matrix successfully!"))
  ggsave(filename = paste0("../ApplicationResult/Multi_Interaction/RandomSeed/TMM_Top2000/Boxplot/98907/BMISEX/TestTop2000/Concordant/", date.analysis, "_", gene.id[1], "GeneExpBoxplotTrainTest", "_", seed.num, ".png"),
         plot = g, device = "png", width = 16, height = 10, units = "in")
  print(paste0("[", gene.id[1], "] Save a 2-boxplot matrix successfully!"))
}
```


# Count - expression distributions of top ranked genes between the training and test set (Other approaches) 
## Seed 2390
### Top 2000 genes list (associated with BMI) from the training set
```{r}
##------DESeq2------
load("../ApplicationData/derived/RandomSeed/DESeq2MultiInt2390.RData")
sumDis <- 0
Discordant.ENSG <- NULL
for(i in seq(1, 2000, 10)) {
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8, i + 9)
  temp.ENSG <- DESeq2.eFDR.2000.GeneName.train[["BMI"]][gene.id]
  cts.train2390.top10 <- cts.train[temp.ENSG, ]
  cts.train2390.top10.t <- t(cts.train2390.top10) |> as.data.frame()
  cts.train2390.top10.t$Data <- "Train"
  cts.test2390.traintop10 <- cts.test[temp.ENSG, ]
  cts.test2390.traintop10.t <- t(cts.test2390.traintop10) |> as.data.frame()
  cts.test2390.traintop10.t$Data <- "Test"
  rownames(cts.train2390.top10.t) = rownames(cts.test2390.traintop10.t) = NULL
  dat.expr.full2390.traintop10 <- rbind(cts.train2390.top10.t,
                                        cts.test2390.traintop10.t)

  for(j in 1:10) {
    p.val <- wilcox.test(dat.expr.full2390.traintop10[[j]] ~ dat.expr.full2390.traintop10[["Data"]], data = dat.expr.full2390.traintop10)$p.value
    if (p.val < 0.05) { 
      sumDis <- sumDis + 1 
      Discordant.ENSG <- c(Discordant.ENSG, temp.ENSG[j])
    }
  }
}
##------Make the boxplot------
boxplot.fun <- function(dataset, var.x, var.y) {
  dataset |> ggplot(mapping = aes(y = get(var.y), x = factor(get(var.x), levels = c("Train", "Test")))) + 
    geom_boxplot(mapping = aes(group = get(var.x), color = get(var.x))) +
    geom_jitter(mapping = aes(color = get(var.x))) + 
    stat_compare_means(method = "wilcox.test", 
                       label.x.npc = "right",
                       label.y = 2,
                       mapping = aes(label = ifelse(p < 1.e-3, "p < 0.001", sprintf("p = %4.3f", as.numeric(..p.format..))))) +
    labs(y = var.y, 
         x = "") +
    theme_bw(base_size = 14) + 
    scale_color_manual(values = c("#0073C2FF", "#EFC000FF")) + 
    scale_x_discrete(limits = c("Train", "Test")) + 
    guides(color = "none")
}

cts.train.new <- t(cts.train) |> as.data.frame(); cts.test.new <- t(cts.test) |> as.data.frame()
rownames(cts.train.new) = rownames(cts.test.new) = NULL
cts.train.new$Data <- "Train"; cts.test.new$Data <- "Test"
cts.new <- rbind(cts.train.new, cts.test.new)

boxplot.Discordant.ENSG.DESeq2 <- Discordant.ENSG |> map(.f = boxplot.fun, dataset = cts.new, var.x = "Data")


##------edgeR------
load("../ApplicationData/derived/RandomSeed/edgeRMultiInt2390.RData")
sumDis <- 0
Discordant.ENSG <- NULL
for(i in seq(1, 2000, 10)) {
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8, i + 9)
  temp.ENSG <- edgeR.eFDR.2000.GeneName.train[["BMI"]][gene.id]
  dat.expr.train2390.top10 <- dat.expr.train[temp.ENSG, ]
  dat.expr.train2390.top10.t <- t(dat.expr.train2390.top10) |> as.data.frame()
  dat.expr.train2390.top10.t$Data <- "Train"
  dat.expr.test2390.traintop10 <- dat.expr.test[temp.ENSG, ]
  dat.expr.test2390.traintop10.t <- t(dat.expr.test2390.traintop10) |> as.data.frame()
  dat.expr.test2390.traintop10.t$Data <- "Test"
  rownames(dat.expr.train2390.top10.t) = rownames(dat.expr.test2390.traintop10.t) = NULL
  dat.expr.full2390.traintop10 <- rbind(dat.expr.train2390.top10.t,
                                        dat.expr.test2390.traintop10.t)

  for(j in 1:10) {
    p.val <- wilcox.test(dat.expr.full2390.traintop10[[j]] ~ dat.expr.full2390.traintop10[["Data"]], data = dat.expr.full2390.traintop10)$p.value
    if (p.val < 0.05) { 
      sumDis <- sumDis + 1 
      Discordant.ENSG <- c(Discordant.ENSG, temp.ENSG[j])
    }
  }
}

dat.expr.train.new <- t(dat.expr.train) |> as.data.frame(); dat.expr.test.new <- t(dat.expr.test) |> as.data.frame()
rownames(dat.expr.train.new) = rownames(dat.expr.test.new) = NULL
dat.expr.train.new$Data <- "Train"; dat.expr.test.new$Data <- "Test"
dat.expr.new <- rbind(dat.expr.train.new, dat.expr.test.new)

boxplot.Discordant.ENSG.edgeR <- Discordant.ENSG |> map(.f = boxplot.fun, dataset = dat.expr.new, var.x = "Data")


##-----eBayes------
load("../ApplicationData/derived/RandomSeed/eBayesMultiInt2390.RData")
sumDis <- 0
Discordant.ENSG <- NULL
for(i in seq(1, 2000, 10)) {
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8, i + 9)
  temp.ENSG <- names(eBayes.eFDR.2000.train[["BMI"]])[gene.id]
  dat.expr.train2390.top10 <- dat.expr.train[temp.ENSG, ]
  dat.expr.train2390.top10.t <- t(dat.expr.train2390.top10) |> as.data.frame()
dat.expr.train2390.top10.t$Data <- "Train"
  dat.expr.test2390.traintop10 <- dat.expr.test[temp.ENSG, ]
  dat.expr.test2390.traintop10.t <- t(dat.expr.test2390.traintop10) |> as.data.frame()
  dat.expr.test2390.traintop10.t$Data <- "Test"
  rownames(dat.expr.train2390.top10.t) = rownames(dat.expr.test2390.traintop10.t) = NULL
  dat.expr.full2390.traintop10 <- rbind(dat.expr.train2390.top10.t,
                                        dat.expr.test2390.traintop10.t)
  for(j in 1:10) {
    p.val <- wilcox.test(dat.expr.full2390.traintop10[[j]] ~ dat.expr.full2390.traintop10[["Data"]], data = dat.expr.full2390.traintop10)$p.value
    if (p.val < 0.05) { 
      sumDis <- sumDis + 1 
      Discordant.ENSG <- c(Discordant.ENSG, temp.ENSG[j])
    }
  }
}

dat.expr.train.new <- t(dat.expr.train) |> as.data.frame(); dat.expr.test.new <- t(dat.expr.test) |> as.data.frame()
rownames(dat.expr.train.new) = rownames(dat.expr.test.new) = NULL
dat.expr.train.new$Data <- "Train"; dat.expr.test.new$Data <- "Test"
dat.expr.new <- rbind(dat.expr.train.new, dat.expr.test.new)

boxplot.Discordant.ENSG.eBayes <- Discordant.ENSG |> map(.f = boxplot.fun, dataset = dat.expr.new, var.x = "Data")

##-----voom + limma------
load("../ApplicationData/derived/RandomSeed/VoomLimmaMultiInt2390.RData")
sumDis <- 0
Discordant.ENSG <- NULL
for(i in seq(1, 2000, 10)) {
  gene.id <- c(i, i + 1, i + 2, i + 3, i + 4, i + 5, i + 6, i + 7, i + 8, i + 9)
  temp.ENSG <- names(voom.eFDR.2000.train[["BMI"]])[gene.id]
  dat.expr.train2390.top10 <- dat.expr.train[temp.ENSG, ]
  dat.expr.train2390.top10.t <- t(dat.expr.train2390.top10) |> as.data.frame()
dat.expr.train2390.top10.t$Data <- "Train"
  dat.expr.test2390.traintop10 <- dat.expr.test[temp.ENSG, ]
  dat.expr.test2390.traintop10.t <- t(dat.expr.test2390.traintop10) |> as.data.frame()
  dat.expr.test2390.traintop10.t$Data <- "Test"
  rownames(dat.expr.train2390.top10.t) = rownames(dat.expr.test2390.traintop10.t) = NULL
  dat.expr.full2390.traintop10 <- rbind(dat.expr.train2390.top10.t,
                                        dat.expr.test2390.traintop10.t)
  for(j in 1:10) {
    p.val <- wilcox.test(dat.expr.full2390.traintop10[[j]] ~ dat.expr.full2390.traintop10[["Data"]], data = dat.expr.full2390.traintop10)$p.value
    if (p.val < 0.05) { 
      sumDis <- sumDis + 1 
      Discordant.ENSG <- c(Discordant.ENSG, temp.ENSG[j])
    }
  }
}

dat.expr.train.new <- t(dat.expr.train) |> as.data.frame(); dat.expr.test.new <- t(dat.expr.test) |> as.data.frame()
rownames(dat.expr.train.new) = rownames(dat.expr.test.new) = NULL
dat.expr.train.new$Data <- "Train"; dat.expr.test.new$Data <- "Test"
dat.expr.new <- rbind(dat.expr.train.new, dat.expr.test.new)

boxplot.Discordant.ENSG.voom <- Discordant.ENSG |> map(.f = boxplot.fun, dataset = dat.expr.new, var.x = "Data")
```

